---
title: "COMT integrated MAVEs"
author: "Ian Hoskins"
date: "8/01/2023"
output:
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE, cache=TRUE)
```


```{r Setup}

library(data.table)
library(ggplot2)
library(GGally)
library(cowplot)
library(ggsci)
library(viridis)
library(bnstruct)
library(pheatmap)
library(RColorBrewer)
library(limma)
library(edgeR)
library(sva)
library(ggsci)
library(ggrepel)
library(readxl)
library(ribor)
library(kohonen)
library(apcluster)
library(lawstat)
library(easyCODA)
library(ALDEx2)
library(gplots)
library(ggcoverage)
library(rtracklayer)

source("/Users/ianhoskins/satmut_utils_dev/satmut_utils/src/prototype/summarization_utils.r")
source("/Users/ianhoskins/satmut_utils_dev/satmut_utils/src/prototype/modeling_utils.r")

comt_base_dir<- "/Users/ianhoskins/data_staging/integrative_MAVEs"
comt_02dec2021_base_dir<- paste(comt_base_dir, "COMT_02DEC2021", sep="/")
comt_15aug2022_base_dir<- paste(comt_base_dir, "COMT_15AUG2022", sep="/")
comt_03nov2022_base_dir<- paste(comt_base_dir, "COMT_03NOV2022", sep="/")
comt_18feb2023_base_dir<- paste(comt_base_dir, "COMT_18FEB2023", sep="/")
comt_23mar2023_base_dir<- paste(comt_base_dir, "COMT_23MAR2023", sep="/")

comt_02dec2021_sim_dir<- paste(comt_02dec2021_base_dir, "sim", sep="/")
comt_03nov2022_sim_dir<- paste(comt_03nov2022_base_dir, "sim", sep="/")

comt_res_base_dir<- paste(comt_base_dir, "results", sep="/")

comt_clinvar_dir<- paste(comt_base_dir, "COMT_clinvar", sep="/")
kozak_dir<- paste(comt_base_dir, "kozak", sep="/")
comt_twist_qc_dir<- paste(comt_base_dir, "Twist_QC", sep="/")

comt_lysis_buffers_dir<- paste(comt_base_dir, "COMT_lysis_buffers_trial1", sep="/")
comt_dox_titration_dir<- paste(comt_base_dir, "UTR5_COMT_Dox_titration", sep="/")

comt_vec_val_dir<- paste(comt_base_dir, "COMT_endo_UTR5_vs_UTR5_del", sep="/")
comt_vec_val_tr1_dir<- paste(comt_vec_val_dir, "293T_LLP_COMT_vector_validation_trial1", sep="/")
comt_vec_val_tr2_dir<- paste(comt_vec_val_dir, "293T_LLP_COMT_vector_validation_trial2", sep="/")
# Not assayed yet; have DNaseI RNAs and lysates for WB
#comt_vec_val_tr3_dir<- paste(comt_vec_val_dir, "293T_LLP_COMT_vector_validation_trial3", sep="/")

comt_utr5_mutant_dir<- paste(comt_base_dir, "UTR5_mutants", sep="/")

tripseq_dir<- "/Users/ianhoskins/TrIP_seq"
comt_pp_dir<- paste(comt_base_dir, "COMT_polysome_profiling", sep="/")
comt_rp_dir<- paste(comt_base_dir, "COMT_ribosome_profiling", sep="/")
comt_facs_dir<- paste(comt_base_dir, "COMT_FACS", sep="/")

comt_pp_roi1_trial1_dir<- paste(comt_pp_dir, "IH_15SEP2021_COMT_ROI_1_trial1", sep="/")
comt_pp_roi1_trial2_dir<- paste(comt_pp_dir, "IH_24SEP2021_COMT_ROI_1_trial2", sep="/")
comt_pp_roi1_trial3_dir<- paste(comt_pp_dir, "IH_29SEP2021_COMT_ROI_1_trial3", sep="/")
comt_pp_roi1_trial4_dir<- paste(comt_pp_dir, "IH_07JUL2022_COMT_ROI_1_2_trial4", sep="/")
comt_pp_roi_trial5_dir<- paste(comt_pp_dir, "IH_12JUL2022_COMT_ROI_1_2_trial5", sep="/")
comt_pp_roi_trial6_dir<- paste(comt_pp_dir, "IH_22JUL2022_COMT_ROI_1_2_trial6", sep="/")

probing_data_dir<- paste(comt_base_dir, "RASP_files", "selected_data", sep="/")
codon_data_dir<- paste(comt_base_dir, "codon_data", sep="/")
demask_dir<- paste(comt_base_dir, "DeMaSk", sep="/")
gnomad_dir<- paste(comt_base_dir, "gnomAD", sep="/")

comt_clinvar_bed_dt<- fread(paste(comt_clinvar_dir, "clinvar_COMT_mutagenesis_genomic.bed", sep="/"))
names(comt_clinvar_bed_dt)<- c("contig", "start", "stop")

# Mutagenesis region was defined initially as a 64 codon region
grch38_roi1_mutagenesis_coords<- seq(19962644, 19962748)
grch38_roi2_mutagenesis_coords<- seq(19963682, 19963750)
pDEST_roi1_mutagenesis_coords<- seq(4794, 4898)
pDEST_roi2_mutagenesis_coords<- seq(5082, 5150)
ENST00000361682_roi1_mutagenesis_coords<- seq(500,605)
ENST00000361682_roi2_mutagenesis_coords<- seq(788,856)
genomic_mutagenesis_coords<- c(grch38_roi1_mutagenesis_coords, grch38_roi2_mutagenesis_coords)
pDEST_mutagenesis_coords<- c(pDEST_roi1_mutagenesis_coords, pDEST_roi2_mutagenesis_coords)
ENST00000361682_mutagenesis_coords<- c(
  ENST00000361682_roi1_mutagenesis_coords, ENST00000361682_roi2_mutagenesis_coords)

mutagenesis_coord_dt<- data.table(
  Genomic=genomic_mutagenesis_coords, 
  Expression=pDEST_mutagenesis_coords, 
  ROI=as.factor(rep(c("ROI_1", "ROI_2"), times=c(
    length(grch38_roi1_mutagenesis_coords), 
    length(grch38_roi2_mutagenesis_coords)))))

# 3BWM human S-COMT; add 50 aa for MB-COMT position
mb_comt_offset<- 50

# DNC substrate
# M40, K140, D141 (Mg2+), L198, W143, W43, P174
substrate_aa_pos<- c(40, 140, 141, 198, 143, 43, 174)

# https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3952716/
# S119, q220, E90, I91, A118, W143, V42, S72, D141, Y68
sam_aa_pos<- c(119, 120, 90, 91, 118, 143, 42, 72, 141, 68)

roi1_rgb<- rgb(110, 0, 98, maxColorValue=255)
roi2_rgb<- rgb(53, 96, 141, maxColorValue=255)
roi_cols<- c(roi1_rgb, roi2_rgb)

clr<- function(x){
  x_geommean<- exp(mean(log(x))) 
  res<- log(x) - log(x_geommean)
  return(res)
}

reclose_coda<- function(x){
  reclosed<- x/sum(x, na.rm=TRUE)
}

iqlr<- function(x){
  iqr<- quantile(x, probs=0.75, na.rm=TRUE) - quantile(x, probs=0.25, na.rm=TRUE)
  res<- log(x) - log(iqr)
  return(res)
}

```


# COMT RNA probing data

```{r COMT RNA probing}

probing_metadata_dt<- data.table(SampleName=c("DIM-2P-seq_hg38_Cell_2017_DMS_invivo_both_score_hg38_DMS_treated_diff.both.COMT", 
                                              "DMS-MaPseq_hg38_Nature_Methods_2017_DMS_invivo_score_both_score_DMSMapSeq_2017.score.COMT",
                                              "DMS-seq_hg38_Nature_2014_DMS_vivo_both_score_fbl_vivo_score.COMT", 
                                              "DMS-seq_hg38_Nature_2014_DMS_vivo_both_score_K562_vivo_score.COMT"), 
                                 Type=c("DIM-2P-seq", "DMS-MaPseq", "DMS-seq", "DMS-seq"), 
                                 Group=c(rep("HEK293T", 2), "FBL", "K562"))

comt_rna_probing_dt<- as.data.table(LoadTrackFile(track.folder=probing_data_dir, format="bedgraph", meta.info=probing_metadata_dt))

comt_rna_probing_gtf<- import.gff(paste(probing_data_dir, "gencode.v29.annotation.exons.COMT.clean.gtf", sep="/"), format="gtf")

comt_rna_probing_regions=data.frame(start=c(grch38_roi1_mutagenesis_coords[1], grch38_roi2_mutagenesis_coords[1]),
                                    end=c(grch38_roi1_mutagenesis_coords[length(grch38_roi1_mutagenesis_coords)], 
                                          grch38_roi2_mutagenesis_coords[length(grch38_roi2_mutagenesis_coords)]),
                                    label=c("ROI 1", "ROI 2"))

#comt_rna_probing_cov_gg<- ggcoverage(data=comt_rna_probing_dt)
#print(comt_rna_probing_cov_gg)

# For some reason I get errors when running the command
#ggcoverage(as.data.frame(comt_rna_probing_dt), mark.region=comt_rna_probing_regions, mark.color="thistle2", gene.name="COMT")+geom_gene(gtf.gr=comt_rna_probing_gtf)

comt_rna_probing_sum_dt<- comt_rna_probing_dt[,.(Sum_score=sum(score)), by=.(Type, Group)]

comt_rna_probing_sum_dt[,Key:=paste(Type, Group, sep=":")]
comt_rna_probing_dt[,Key:=paste(Type, Group, sep=":")]

comt_rna_probing_dt[,Sum_score:=comt_rna_probing_sum_dt[match(comt_rna_probing_dt[,Key], Key), Sum_score]]

comt_rna_probing_dt[, Proportion:=score/Sum_score]

is_in_name<- function(x, exon_list){
  exon_names<- names(exon_list)
  return(exon_names[sapply(exon_list, is_in, x)])
  }
    
# Plot proportion of the score across target positions
exon_list<- list(
  "Exon_1"=seq(19941607, 19941897),
  "Exon_2"=seq(19961199, 19961289), 
  "Exon_3"=seq(19962527, 19962815),
  "Exon_4"=seq(19963566, 19963759),
  "Exon_5"=seq(19964168, 19964299), 
  "Exon_6"=seq(19968536, 19969975))

comt_rna_probing_dt[,Exon:=sapply(comt_rna_probing_dt[,start], is_in_name, exon_list)]

comt_rna_probing_dt[,Type:=factor(Type, levels=c("DMS-seq", "DIM-2P-seq", "DMS-MaPseq"))]

probing_cols<- pal_jama()(3)

comt_rna_probing_gg<- ggplot(
  data=comt_rna_probing_dt[Group!="FBL"], 
  aes(x=start, y=Proportion, fill=Type))

comt_rna_probing_gg+
  facet_wrap(~Exon, scales="free_x")+
  geom_bar(stat="identity")+
  scale_fill_manual(values=probing_cols)+
  theme_cowplot()

# Exon 2 and 3
comt_rna_probing_gg<- ggplot(
  data=comt_rna_probing_dt[Group!="FBL" & Exon=="Exon_2"], 
  aes(x=start, y=Proportion, fill=Type))

comt_rna_probing_gg+
  facet_wrap(~Exon, scales="free_x")+
  geom_bar(stat="identity")+
  scale_fill_manual(values=probing_cols[c(1,3)])+
  theme_cowplot()

# In ROI_1 only
comt_rna_probing_gg<- ggplot(
  data=comt_rna_probing_dt[Group!="FBL" & Exon=="Exon_3"], 
  aes(x=start, y=Proportion, fill=Type))

comt_rna_probing_gg+
  facet_wrap(~Exon, scales="free_x")+
  geom_rect(aes(xmin=grch38_roi1_mutagenesis_coords[1], 
                xmax=grch38_roi1_mutagenesis_coords[length(grch38_roi1_mutagenesis_coords)], 
                ymin=-Inf, ymax=Inf), fill="thistle2")+
  geom_bar(stat="identity")+
  scale_fill_manual(values=probing_cols[c(1,3)])+
  theme_cowplot()

# In ROI_2 only
comt_rna_probing_gg<- ggplot(
  data=comt_rna_probing_dt[Group!="FBL" & Exon=="Exon_4"], 
  aes(x=start, y=Proportion, fill=Type))

comt_rna_probing_gg+
  facet_wrap(~Exon, scales="free_x")+
  geom_rect(aes(xmin=grch38_roi2_mutagenesis_coords[1], 
                xmax=grch38_roi2_mutagenesis_coords[length(grch38_roi2_mutagenesis_coords)], 
                ymin=-Inf, ymax=Inf), fill="thistle2")+
  geom_bar(stat="identity")+
  scale_fill_manual(values=probing_cols[c(1,3)])+
  theme_cowplot()

# Exon 6
comt_rna_probing_gg<- ggplot(
  data=comt_rna_probing_dt[Group!="FBL" & Exon=="Exon_6"], 
  aes(x=start, y=Proportion, fill=Type))

comt_rna_probing_gg+
  facet_wrap(~Exon, scales="free_x")+
  geom_bar(stat="identity")+
  scale_fill_manual(values=probing_cols)+
  theme_cowplot()

rm(comt_rna_probing_gg)

```


# COMT mutagenesis region mapping and uniformity by Twist readout

```{r Mapping ClinVar variants from genomic to expression clone position}

mutagenesis_coord_clinvar_dt<- mutagenesis_coord_dt[Genomic%in%comt_clinvar_bed_dt[,start]]

comt_clinvar_expr_bed_dt<- data.table(
  contig="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP", 
  start=mutagenesis_coord_clinvar_dt[,Expression], 
  stop=mutagenesis_coord_clinvar_dt[,Expression+1])

```

```{r COMT library uniformity}

comt_library_p1_dt<- as.data.table(read_xlsx(paste(comt_twist_qc_dir, "Final_Q-118856_P1_QC_Report.xlsx", sep="/"), sheet=3))

comt_library_p2_dt<- as.data.table(read_xlsx(paste(comt_twist_qc_dir, "Final_Q-118856_P2_QC_Report.xlsx", sep="/"), sheet=3))

comt_library_both_dt<- rbind(comt_library_p1_dt, comt_library_p2_dt)

comt_library_both_dt[,Plate:=as.factor(
  rep(c("P1", "P2"), times=c(nrow(comt_library_p1_dt), nrow(comt_library_p2_dt))))]

rm(comt_library_p1_dt)
rm(comt_library_p2_dt)

comt_library_both_dt[,variant_proportion:=as.numeric(
  sub(pattern="%",replacement="",variant_proportion,fixed=T))/100]

comt_library_both_dt[,ROI:=as.factor(ifelse(`Codon Position` < 400, "ROI_1", "ROI_2"))]

# This confirms the proportion is for all variants at each codon, not across codons
comt_library_both_prop_summ_dt<- comt_library_both_dt[
  ,sum(variant_proportion),by=.(Plate,ROI,`Codon Position`)]

#comt_library_both_prop_summ_dt[V1>101]

# Compute the proportion relative to all positions assuming equal weighting
comt_library_both_dt[,Global_proportion:=variant_proportion/sum(variant_proportion), by=.(ROI)]

two_cols_roi<- viridis(n=2, alpha=0.75, begin=0.2, end=0.8, option="D")

comt_library_both_gg<- ggplot(
  data=comt_library_both_dt, 
  aes(x=log10(Global_proportion), col=ROI))

comt_library_both_gg+
  geom_density()+
  scale_color_manual(values=two_cols_roi)+
  labs(x="log10 proportion in library")+
  theme_cowplot()

rm(comt_library_both_gg)

# Plot frequencies by type
comt_library_both_dt[,HD:=hamming.distance(
  unlist(strsplit(wt, "")), unlist(strsplit(variant, ""))), 
  by=seq_len(nrow(comt_library_both_dt))]

comt_library_both_dt[HD==1,TYPE:="SNP"]
comt_library_both_dt[HD==2,TYPE:="di_nt_MNP"]
comt_library_both_dt[HD==3,TYPE:="tri_nt_MNP"]
comt_library_both_dt[,TYPE:=factor(TYPE, levels=c("SNP", "di_nt_MNP", "tri_nt_MNP"))]

comt_library_both_type_gg<- ggplot(
  data=comt_library_both_dt, 
  aes(x=log10(Global_proportion), col=TYPE))

comt_library_both_type_gg+
  facet_wrap(~ROI)+
  geom_density()+
  labs(x="log10 proportion in library")+
  theme_cowplot()

# Examine position-by-position for each type
comt_library_both_type_gg<- ggplot(
  data=comt_library_both_dt[!is.na(TYPE)], 
  aes(x=as.factor(`Codon Position`), y=variant_proportion, col=TYPE))

comt_library_both_type_gg+
  facet_wrap(~ROI, scales="free_x", nrow=2)+
  geom_boxplot(position="dodge")+
  labs(y="Variant proportion of codon position")+
  theme_cowplot()+
  coord_cartesian(ylim=c(0,0.25))+
  theme(axis.text.x=element_text(angle=90))

# Examine position-by-position for each plate
comt_library_both_type_gg<- ggplot(
  data=comt_library_both_dt[!is.na(TYPE)], 
  aes(x=as.factor(`Codon Position`), y=variant_proportion, col=TYPE))

comt_library_both_type_gg+
  facet_wrap(~Plate, scales="free_x", nrow=2)+
  geom_boxplot(position="dodge")+
  labs(y="Variant proportion of codon position")+
  theme_cowplot()+
  coord_cartesian(ylim=c(0,0.5))+
  theme(axis.text.x=element_text(angle=90))

rm(comt_library_both_type_gg)

```


# COMT experimental optimization

## COMT lysis buffer testing

Purpose was to determine if there was a particular detergent that solubilizes maximal levels of COMT membrane and cytoplasmic protein isoforms.

I harvested passage 19 LLP UTR5-COMT cells at 50% confluency after 2 µg/mL Dox induction a day prior (4 µg/mL induction at days 2 and 3 pre-harvest). Before harvest cells were stimulated with fresh media for 2.5 h, washed once with warm PBS, then lysed with 325 µL the following lysis buffers, all modified from the RP lysis buffer with the exception of the RIPA buffers:

1. 1% Triton X-100
2. 1% NP-40
3. 1x RIPA (1% NP-40, 0.1% SDS, 0.25% sodium deoxycholate)
4. 1% TritonX-100, 1% DDM
5. 1% NP-40, 1% DDM
6. 1x RIPA, 1% DDM

I included the UTR5-COMT and UTR5del-COMT 1x RIPA control lysates prepared in the "COMT nativePAGE trial 1" experiment. These lysates were harvested with the same RIPA buffer but were grown to higher confluency (~70%) than the current experiment. 

```{r COMT lysis buffers}

comt_lysis_dt<- fread(paste(
  comt_lysis_buffers_dir, "COMT_lysis_buffer_trial1_tidy.csv", sep="/"), sep=",")

comt_lysis_dt[,Flag_Actin_ratio:=Flag/Actin]

# Plot Flag:actin ratio
comt_lysis_gg<- ggplot(
  data=comt_lysis_dt[
    Construct=="UTR5-COMT" & Experiment=="COMT_lysis_buffers_trial1"], 
  aes(x=Buffer, y=Flag_Actin_ratio))

comt_lysis_gg+
  geom_point()+
  scale_y_continuous(breaks=seq(0,1.5,0.25))+
  coord_cartesian(ylim=c(0,1.5))+
  labs(x="Lysis buffer", y="Flag:action signal ratio")+
  theme_cowplot()

# Ratio may be misleading; plot effect on just Flag signal
comt_lysis_gg<- ggplot(
  data=comt_lysis_dt[
    Construct=="UTR5-COMT" & Experiment=="COMT_lysis_buffers_trial1"], 
  aes(x=Buffer, y=Flag))

comt_lysis_gg+
  geom_point()+
  labs(x="Lysis buffer", y="Flag signal")+
  theme_cowplot()

rm(comt_lysis_gg)
  
# Use RIPA control lysates to estimate effect size of UTR5 del on protein abundance
# of membrane isoform
comt_construct_gg<- ggplot(
  data=comt_lysis_dt[Experiment=="COMT_nativePAGE_trial1"], 
  aes(x=Construct, y=Flag_Actin_ratio))

comt_construct_gg+
  geom_point()+
  scale_y_continuous(breaks=seq(0,2.5,0.25))+
  coord_cartesian(ylim=c(0,2.5))+
  labs(x="COMT construct", y="Flag:action signal ratio")+
  theme_cowplot()

# Flag signal only
comt_construct_gg<- ggplot(
  data=comt_lysis_dt[Experiment=="COMT_nativePAGE_trial1"], 
  aes(x=Construct, y=Flag))

comt_construct_gg+
  geom_point()+
  labs(x="COMT construct", y="Flag signal")+
  theme_cowplot()

rm(comt_construct_gg)
  
```

Results indicate that inclusion of 1% DDM may help improve capture of the membrane isoform, as Flag:actin signal was slightly higher when including DDM in three ionic detergents. However, this may be misguided as when examining specifically Flag signal, DDM results in decreased Flag signal. Thus, stronger decreases to Actin solubilization may lead to an errant conclusion that DDM improves MB-COMT solubilization.

Which ionic detergent is best is not clear. While Triton-X showed slightly higher Flag:actin ratio than RIPA, when examining just Flag signal, RIPA yields the highest signal.

Thus, it appears there is no lysis buffer formulation that outcompetes others for COMT solubilization. Stick with standard RIPA buffer without DDM.


## Dox titration for COMT induction

Purpose was to test higher levels of Dox (>2 µg/mL) for higher COMT transcript abundance.

```{r COMT Dox titration}

comt_dox_titr_dt<- fread(paste(
  comt_dox_titration_dir, "IH_09JUL2021_COMT_Dox_titration_qPCR.csv", sep="/"), sep=",", stringsAsFactors=T)

comt_dox_titr_dt[,Dox:=as.factor(Dox)]

three_cols<- c("darkgreen", "brown", "tan")

comt_dox_titr_gg<- ggplot(
  data=comt_dox_titr_dt, 
  aes(x=Dox, y=FC, col=Assay))

comt_dox_titr_gg+
  geom_jitter(size=2.5, width=0.1)+
  coord_cartesian(ylim=c(0,1.2))+
  scale_y_continuous(breaks=seq(0,1.2,.2))+
  scale_color_manual(values=three_cols)+
  labs(x="Doxycyline concentration (µg/mL)", col="qPCR assay",
       y="Vector mRNA fold change \nrelative to lowest Dox")+
  theme_cowplot()
  
```

No effect of increased Dox was observed, which was similarly communicated by other researchers (Kamyab Javanmardi, Atine Cote). While increased Dox may affect the proportion of cells with fluorescence as read by IRES-mCherry, for each individual cell, expression is likely driven by stochastic bursting transcription.


# COMT 5' UTR deletion affect on mRNA abundance

Purpose was to test for a difference in mRNA abundance, polysome association, and protein abundance for the 5' UTR del (53 nt) covering the exon 2 uORFs versus the endogenous COMT 5' UTR. Here, endogenous may be a slight misnomer, as the first COMT exon is not included in the construct (it is high-GC and probably difficult to clone).

Three experiments have been conducted so far:
COMT vector validation trial 1, 2, 3

```{r COMT RT-qPCR for endo UTR5 and UTR5 del}

# First qPCR on total lysate RNA from polysome profiling trial 1
# IH_14JUL2021_LLP_COMT_vec_val_trial2_total_RTqPCR_tidy.csv

# IH_09AUG2021_293TLLP_vec_val_tr2_pp1_2_mCherry_qPCR_cntrl_tidy.csv

comt_rna_exp1_dt<- fread(paste(
  comt_vec_val_tr2_dir, "IH_14JUL2021_LLP_COMT_vec_val_trial2_total_RTqPCR.csv", sep="/"), sep=",")

comt_rna_exp2_dt<- fread(paste(
  comt_vec_val_tr2_dir, "IH_04AUG2021_293T_LLP_vec_val_tr2_pp_tr1_2_rtqPCR.csv", sep="/"), sep=",")

comt_rna_exp3_dt<- fread(paste(comt_vec_val_tr2_dir, "IH_09AUG2021_293TLLP_vec_val_tr2_pp1_2_mCherry_qPCR_cntrl.csv",sep="/"),sep=",")

# Repeated for control qPCR primers and complete cDNA input for COMT_vec_RNA assay
# (See exp 3 explanation)
comt_rna_exp4_dt<- fread(paste(comt_vec_val_tr2_dir, "IH_10AUG2021_293T_LLP_vec_val_tr2_pp_tr1_2_mCherry_cntrl_repeat.csv",sep="/"),sep=",")

comt_rna_all_dt<- rbind(
  comt_rna_exp1_dt, comt_rna_exp2_dt, 
  comt_rna_exp3_dt[Assay=="mCherry_qPCR"], 
  comt_rna_exp4_dt)

comt_rna_all_summ_dt<- comt_rna_all_dt[
  ,.(Median_FC=median(FC), Mean_FC=mean(FC)), 
  by=.(Sample, `Sample Name`, Assay)]

comt_rna_exp1_dt[,Sample:=as.factor(Sample)]
comt_rna_exp2_dt[,Sample:=as.factor(Sample)]
comt_rna_exp3_dt[,Sample:=as.factor(Sample)]
comt_rna_exp4_dt[,Sample:=as.factor(Sample)]

two_cols<- viridis(n=2, alpha=0.7, end=0.3)

# Experiment 1- polysome profiling trial 1 (10 µg/mL CHX) total lysate
comt_rna_exp1_gg<- ggplot(
  data=comt_rna_exp1_dt, aes(x=Sample, y=FC, col=Sample))

comt_rna_exp1_gg+
  geom_boxplot(width=0.5, outlier.shape=NA)+
  geom_jitter(size=2, width=0.2)+
  scale_color_manual(values=two_cols)+
  scale_y_continuous(breaks=seq(0, 1.2, 0.2))+
  coord_cartesian(ylim=c(0,1.2))+
  labs(x="COMT construct", col="Construct",
       y="Trasgene mRNA fold change \nUTR5 del to endo UTR")+
  theme_cowplot()

comt_rna_exp1_ttest_res<- comt_rna_exp1_dt[, t.test(FC ~ Sample, data=.SD)]


# Experiment 2- polysome profiling trial 2 (100 µg/mL CHX) total lysate
# (plus trial 1 samples)
comt_rna_exp2_gg<- ggplot(
  data=comt_rna_exp2_dt[Assay=="COMT_vec_RNA"], 
  aes(x=Sample, y=FC, col=Sample))

comt_rna_exp2_gg+
  geom_boxplot(width=0.5, outlier.shape=NA)+
  geom_jitter(size=2, width=0.2)+
  scale_color_manual(values=two_cols)+
  scale_y_continuous(breaks=seq(0, 1.2, 0.2))+
  coord_cartesian(ylim=c(0,1.2))+
  labs(x="COMT construct", col="Construct",
       y="Trasgene mRNA fold change \nUTR5 del to endo UTR")+
  theme_cowplot()

comt_rna_exp2_ttest_res<- comt_rna_exp2_dt[Assay=="COMT_vec_RNA", t.test(FC ~ Sample, data=.SD)]
comt_roi2_exp2_ttest_res<- comt_rna_exp2_dt[Assay=="COMT_vec_RNA", t.test(FC ~ Sample, data=.SD)]


# Experiment 3- ran on Viia7 instrument C and to include mCherry assay
# used COMT_vec_RNA assay again but ran out of template so input was 
# ~ 1 µL instead of the 2 µL used for mCherry and EEF2 controls
comt_rna_exp3_gg<- ggplot(
  data=comt_rna_exp3_dt, 
  aes(x=Sample, y=FC, col=Sample))

comt_rna_exp3_gg+
  facet_wrap(~Assay)+
  geom_boxplot(width=0.5, outlier.shape=NA)+
  geom_jitter(size=2, width=0.2)+
  scale_color_manual(values=two_cols)+
  scale_y_continuous(breaks=seq(0, 1.2, 0.2))+
  coord_cartesian(ylim=c(0,1.2))+
  labs(x="COMT construct", col="Construct",
       y="Trasgene mRNA fold change \nUTR5 del to endo UTR")+
  theme_cowplot()

comt_rna_exp3_ttest_res<- comt_rna_exp3_dt[Assay=="COMT_vec_RNA", t.test(FC ~ Sample, data=.SD)]
mcherry_rna_exp3_ttest_res<- comt_rna_exp3_dt[Assay=="mCherry_qPCR", t.test(FC ~ Sample, data=.SD)]

# Experiment 4- ran on Viia7 instrument C 
# repeat all samples and mCherry assay but switching well positions for assays
comt_rna_exp4_gg<- ggplot(
  data=comt_rna_exp4_dt, 
  aes(x=Sample, y=FC, col=Sample))

comt_rna_exp4_gg+
  facet_wrap(~Assay)+
  geom_boxplot(width=0.5, outlier.shape=NA)+
  geom_jitter(size=2, width=0.2)+
  scale_color_manual(values=two_cols)+
  scale_y_continuous(breaks=seq(0, 1.2, 0.2))+
  coord_cartesian(ylim=c(0,1.2))+
  labs(x="COMT construct", col="Construct",
       y="Trasgene mRNA fold change \nUTR5 del to endo UTR")+
  theme_cowplot()

comt_rna_exp4_ttest_res<- comt_rna_exp4_dt[Assay=="COMT_vec_RNA", t.test(FC ~ Sample, data=.SD)]
mcherry_rna_exp4_ttest_res<- comt_rna_exp4_dt[Assay=="mCherry_qPCR", t.test(FC ~ Sample, data=.SD)]


# Plot mean FC across qPCR experiments
# ran on Viia7 instrument C but switching well positions for assays
# this should in theory help diminish block/well-specific effects
# on expression measurements
comt_rna_all_summ_gg<- ggplot(
  data=comt_rna_all_summ_dt, 
  aes(x=Sample, y=Mean_FC, col=Sample))

comt_rna_all_summ_gg+
  facet_wrap(~Assay)+
  geom_boxplot(width=0.5, outlier.shape=NA)+
  geom_jitter(size=2, width=0.2)+
  scale_color_manual(values=two_cols)+
  scale_y_continuous(breaks=seq(0, 1.2, 0.2))+
  coord_cartesian(ylim=c(0,1.2))+
  labs(x="COMT construct", col="Construct",
       y="Trasgene mRNA fold change \nUTR5 del to endo UTR")+
  theme_cowplot()

# Plot median FC across qPCR experiments
comt_rna_all_summ_gg<- ggplot(
  data=comt_rna_all_summ_dt, 
  aes(x=Sample, y=Median_FC, col=Sample))

comt_rna_all_summ_gg+
  facet_wrap(~Assay)+
  geom_boxplot(width=0.5, outlier.shape=NA)+
  geom_jitter(size=2, width=0.2)+
  scale_color_manual(values=two_cols)+
  scale_y_continuous(breaks=seq(0, 1.2, 0.2))+
  coord_cartesian(ylim=c(0,1.2))+
  labs(x="COMT construct", col="Construct",
       y="Trasgene mRNA fold change \nUTR5 del to endo UTR")+
  theme_cowplot()

comt_ttest_res<- comt_rna_all_summ_dt[Assay=="COMT_vec_RNA", t.test(Median_FC ~ Sample, data=.SD)]
comt_roi2_ttest_res<- comt_rna_all_summ_dt[Assay=="COMT_ROI_2", t.test(Median_FC ~ Sample, data=.SD)]
mcherry_ttest_res<- comt_rna_all_summ_dt[Assay=="mCherry_qPCR", t.test(Median_FC ~ Sample, data=.SD)]


# Finally plot just COMT vec RNA and mCherry assays
comt_rna_all_summ_final_gg<- ggplot(
  data=comt_rna_all_summ_dt[Assay!="COMT_ROI_2"], 
  aes(x=Assay, y=Median_FC, col=Sample))

comt_rna_all_summ_final_gg+
  geom_boxplot(width=0.5, outlier.shape=NA)+
  geom_point(position=position_jitterdodge(jitter.width=0.2), size=2)+
  scale_color_manual(values=two_cols)+
  scale_y_continuous(breaks=seq(0, 1.2, 0.2))+
  coord_cartesian(ylim=c(0,1.2))+
  labs(x="COMT construct", col="Construct",
       y="Trasgene mRNA fold change \nUTR5 del to endo UTR")+
  theme_cowplot()


# Polysome profiling indicated tr2_1C may have been degraded
# if we remove this sample, are results more consistent?
comt_rna_all_summ_final_gg<- ggplot(
  data=comt_rna_all_summ_dt[Assay!="COMT_ROI_2" & `Sample Name`!="tr2_1C"], 
  aes(x=Sample, y=Median_FC, col=Sample))

comt_rna_all_summ_final_gg+
  facet_wrap(~Assay)+
  geom_boxplot(width=0.5, outlier.shape=NA)+
  geom_jitter(size=2, width=0.2)+
  scale_color_manual(values=two_cols)+
  scale_y_continuous(breaks=seq(0, 1.2, 0.2))+
  coord_cartesian(ylim=c(0,1.2))+
  labs(x="COMT construct", col="Construct",
       y="Trasgene mRNA fold change \nUTR5 del to endo UTR")+
  theme_cowplot()

comt_filt_ttest_res<- comt_rna_all_summ_dt[
  Assay=="COMT_vec_RNA" & `Sample Name`!="tr2_1C", t.test(Median_FC ~ Sample, data=.SD)]

comt_filt_roi2_ttest_res<- comt_rna_all_summ_dt[
  Assay=="COMT_ROI_2" & `Sample Name`!="tr2_1C", t.test(Median_FC ~ Sample, data=.SD)]

mcherry_filt_ttest_res<- comt_rna_all_summ_dt[
  Assay=="mCherry_qPCR" & `Sample Name`!="tr2_1C", t.test(Median_FC ~ Sample, data=.SD)]

# The filtering made a big difference for the mCherry assay, as it removed the high
# outlier point for the UTR5 del construct
# However it made a small impact on the COMT assays, making them less significant

```

The 53 bp deletion clone (UTR5del) was _depleted_ relative to the endogenous 5' UTR by over 2-fold. This is consistent with polysome profiling from the same lysate, but contrary to the results obtained from flow cytometry and immunoblotting.


# COMT 5' UTR deletion affect on mRNA and protein abundance by flow cytometry

```{r COMT flow cytometry for endo UTR5 and UTR5 del}

comt_vec_val_tr1_select_dir<- paste(comt_vec_val_tr1_dir, "IH_24MAY2021_UTR5_COMT_moxGFP_validation_selected", sep="/")
comt_vec_val_tr1_select_endo_file<- paste(comt_vec_val_tr1_select_dir, "export_4_UTR5_COMT_moxGFP_Tube_001.csv", sep="/")
comt_vec_val_tr1_select_del_file<- paste(comt_vec_val_tr1_select_dir, "export_2_noUTR_COMT_moxGFP_Tube_001.csv", sep="/")

comt_vec_val_tr1_select_endo_dt<- fread(comt_vec_val_tr1_select_endo_file, sep=",")
comt_vec_val_tr1_select_del_dt<- fread(comt_vec_val_tr1_select_del_file, sep=",")

comt_vec_val_tr1_select_both_dt<- rbind(comt_vec_val_tr1_select_endo_dt, comt_vec_val_tr1_select_del_dt)

comt_vec_val_tr1_select_both_dt[,Sample:=as.factor(rep(c("endo", "del"), times=c(
  nrow(comt_vec_val_tr1_select_endo_dt), nrow(comt_vec_val_tr1_select_del_dt))))]

rm(comt_vec_val_tr1_select_endo_dt)
rm(comt_vec_val_tr1_select_del_dt)

comt_utr5_endo_vs_del_flow_gg<- ggplot(
  data=comt_vec_val_tr1_select_both_dt, 
  aes(x=log10(`GFP-A`), col=Sample))

comt_utr5_endo_vs_del_flow_gg+
  geom_density()+
  theme_cowplot()

comt_vec_val_tr1_select_both_dt[,median(`GFP-A`), by=.(Sample)]
# 8115.58/4064.82

comt_vec_val_tr1_select_both_dt[,median(log10(`GFP-A`), na.rm=TRUE), by=.(Sample)]
(10^3.912751)/(10^3.619586)

comt_vec_val_tr1_select_both_dt[,wilcox.test(`GFP-A`~Sample)]

```


# TrIP-seq polysome profiling data for COMT

Examine the distribution of COMT observed in the TrIP-seq dataset.

```{r COMT TrIP-seq polysome profiling}

tripseq_gene_dt<- fread(paste(
  tripseq_dir, "GSE69352_tripseq_gene_tpm_ensembl_v75.csv", sep="/"), sep=",")

tripseq_gene_melt_dt<- melt(
  data=tripseq_gene_dt, id.vars=c("gene_name", "gene_id", "gene_biotype"), 
  value.name="TPM", variable.name="Fraction")

tripseq_gene_melt_dt[,Fraction_base:=sapply(strsplit(as.character(Fraction), "_"), head, 1)]
tripseq_gene_melt_dt[,Fraction_base:=as.factor(Fraction_base)]

tripseq_gene_melt_comt_dt<- tripseq_gene_melt_dt[gene_name=="COMT"]

tripseq_gene_melt_comt_dt[,Ribo_num:=substr(Fraction_base,5,5)]
tripseq_gene_melt_comt_dt[Fraction_base=="cyto", Ribo_num:=0]
tripseq_gene_melt_comt_dt[Fraction_base=="80S", Ribo_num:=1]
tripseq_gene_melt_comt_dt[,Ribo_num:=as.integer(Ribo_num)]

tripseq_gene_melt_comt_gg<- ggplot(
  data=tripseq_gene_melt_comt_dt, 
  aes(x=Ribo_num, y=TPM))

tripseq_gene_melt_comt_gg+
  geom_point(size=2)+
  geom_smooth(se=FALSE)+
  scale_x_continuous(breaks=seq(0,8))+
  labs(x="Fraction", y="TPM")+
  theme_cowplot()

# Summarize replicates
tripseq_gene_melt_comt_summ_dt<- tripseq_gene_melt_comt_dt[,.(Mean_TPM=mean(TPM), SD_TPM=sd(TPM)), by=.(Fraction_base)]

tripseq_gene_melt_comt_summ_dt[,Fraction_base:=factor(
  Fraction_base, levels=c("cyto", "80S", "poly2", "poly3", "poly4", "poly5", "poly6", "poly7", "poly8"))]

tripseq_gene_melt_comt_gg<- ggplot(
  data=tripseq_gene_melt_comt_summ_dt, 
  aes(x=Fraction_base, y=Mean_TPM, 
      ymax=Mean_TPM+SD_TPM, 
      ymin=Mean_TPM-SD_TPM))

tripseq_gene_melt_comt_gg+
  geom_bar(stat="identity", fill="steelblue")+
  geom_errorbar(col="grey", width=0.2)+
  labs(x="Fraction", y="Mean TPM +/- SD")+
  theme_cowplot()

rm(tripseq_gene_melt_comt_gg)

```

It appears COMT is highly engaged with ribosomes, although variability in the monosome and mid-polysome fractions makes the interpretation less clear.


## UTR5-Flag-COMT-moxGFP and 5' UTR del polysome profiling trial 1

This experiment used 10 µg/mL CHX in the PBS wash buffers, with no pre-incubation in media. Cells were stimulated for 1 h with fresh warm media and were at 50-60% confluency at harvest.

```{r COMT UTR5 polysome profiling trial 1}

comt_pp1_1a_dt<- fread(paste(comt_pp_dir, "13JUL2021_polysome_profiling_logs", "IH_13JUL2021_COMT_vector_validation_trial2_1A.csv",sep="/"),sep=",")
comt_pp1_1b_dt<- fread(paste(comt_pp_dir, "13JUL2021_polysome_profiling_logs", "IH_13JUL2021_COMT_vector_validation_trial2_1B.csv",sep="/"),sep=",")
comt_pp1_1c_dt<- fread(paste(comt_pp_dir, "13JUL2021_polysome_profiling_logs", "IH_13JUL2021_COMT_vector_validation_trial2_1C.csv",sep="/"),sep=",")
comt_pp1_3a_dt<- fread(paste(comt_pp_dir, "13JUL2021_polysome_profiling_logs", "IH_13JUL2021_COMT_vector_validation_trial2_3A.csv",sep="/"),sep=",")
comt_pp1_3b_dt<- fread(paste(comt_pp_dir, "13JUL2021_polysome_profiling_logs", "IH_13JUL2021_COMT_vector_validation_trial2_3B.csv",sep="/"),sep=",")
comt_pp1_3c_dt<- fread(paste(comt_pp_dir, "13JUL2021_polysome_profiling_logs", "IH_13JUL2021_COMT_vector_validation_trial2_3C.csv",sep="/"),sep=",")

comt_pp1_1_nrows<- sum(nrow(comt_pp1_1a_dt), nrow(comt_pp1_1b_dt), nrow(comt_pp1_1c_dt))
comt_pp1_3_nrows<- sum(nrow(comt_pp1_3a_dt), nrow(comt_pp1_3b_dt), nrow(comt_pp1_3c_dt))
comt_pp1_all_dt<- rbind(comt_pp1_1a_dt, comt_pp1_1b_dt, comt_pp1_1c_dt, comt_pp1_3a_dt, comt_pp1_3b_dt, comt_pp1_3c_dt)

comt_pp1_all_dt[,Construct:=as.factor(rep(c("UTR5 del.", "nat. UTR5"), times=c(comt_pp1_1_nrows, comt_pp1_3_nrows)))]
comt_pp1_all_dt[,Gradient:=as.factor(rep(c("A", "B", "C", "A", "B", "C"), times=c(nrow(comt_pp1_1a_dt), nrow(comt_pp1_1b_dt), nrow(comt_pp1_1c_dt), nrow(comt_pp1_3a_dt), nrow(comt_pp1_3b_dt), nrow(comt_pp1_3c_dt))))]

# Polysome profiling trial 1 Gradient 1A
comt_pp1_1a_gg<- ggplot(
  data=comt_pp1_1a_dt, aes(x=`Position(mm)`, y=Absorbance))

comt_pp1_1a_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(0,5))+
  geom_vline(xintercept=comt_pp1_1a_dt[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

rm(comt_pp1_1a_gg)

# Polysome profiling trial 1 Gradient 1B
comt_pp1_1b_gg<- ggplot(
  data=comt_pp1_1b_dt, aes(x=`Position(mm)`, y=Absorbance))

comt_pp1_1b_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(0,5))+
  geom_vline(xintercept=comt_pp1_1b_dt[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

rm(comt_pp1_1b_gg)

# Polysome profiling trial 1 Gradient 1C
comt_pp1_1c_gg<- ggplot(
  data=comt_pp1_1c_dt, aes(x=`Position(mm)`, y=Absorbance))

comt_pp1_1c_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(0,5))+
  geom_vline(xintercept=comt_pp1_1c_dt[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

rm(comt_pp1_1c_gg)

# Polysome profiling trial 1 Gradient 3A
comt_pp1_3a_gg<- ggplot(
  data=comt_pp1_3a_dt, aes(x=`Position(mm)`, y=Absorbance))

comt_pp1_3a_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(0,5))+
  geom_vline(xintercept=comt_pp1_3a_dt[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

rm(comt_pp1_3a_gg)

# Polysome profiling trial 1 Gradient 3B
comt_pp1_3b_gg<- ggplot(
  data=comt_pp1_3b_dt, aes(x=`Position(mm)`, y=Absorbance))

comt_pp1_3b_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(0,5))+
  geom_vline(xintercept=comt_pp1_3b_dt[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

rm(comt_pp1_3b_gg)

# Polysome profiling trial 1 Gradient 3C
comt_pp1_3c_gg<- ggplot(
  data=comt_pp1_3c_dt, aes(x=`Position(mm)`, y=Absorbance))

comt_pp1_3c_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(0,5))+
  geom_vline(xintercept=comt_pp1_3c_dt[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

rm(comt_pp1_3c_gg)

```


# UTR5-Flag-COMT-moxGFP and 5' UTR del polysome profiling

Trial 2:

This experiment used 100 µg/mL CHX pre-treatment in media for 10 min at 37 ˚C, prior to wash in PBS with 100 µg/mL CHX. Cells were stimulated for 2.5 h with fresh warm media and were at 90-95% confluency at harvest. Polysomes were fractionated on Biocomp Piston Gradient Fractionator with version 2 and Triax detector with simple UV detection. Gradients were fractionated into 24 raw fractions of ~480 µL volume at 0.2 mm/s speed.

```{r COMT UTR5 polysome profiling trial 2}

comt_pp2_1a_dt<- fread(paste(comt_pp_dir, "16JUL2021_polysome_profiling_logs", "IH_16JUL2021_LLP_COMT_vec_val_trial2_pp_trial2_1A.csv",sep="/"),sep=",")
comt_pp2_1b_dt<- fread(paste(comt_pp_dir, "16JUL2021_polysome_profiling_logs", "IH_16JUL2021_LLP_COMT_vec_val_trial2_pp_trial2_1B.csv",sep="/"),sep=",")
comt_pp2_1c_dt<- fread(paste(comt_pp_dir, "16JUL2021_polysome_profiling_logs", "IH_16JUL2021_LLP_COMT_vec_val_trial2_pp_trial2_1C.csv",sep="/"),sep=",")
comt_pp2_3a_dt<- fread(paste(comt_pp_dir, "16JUL2021_polysome_profiling_logs", "IH_16JUL2021_LLP_COMT_vec_val_trial2_pp_trial2_3A.csv",sep="/"),sep=",")
comt_pp2_3b_dt<- fread(paste(comt_pp_dir, "16JUL2021_polysome_profiling_logs", "IH_16JUL2021_LLP_COMT_vec_val_trial2_pp_trial2_3B.csv",sep="/"),sep=",")
comt_pp2_3c_dt<- fread(paste(comt_pp_dir, "16JUL2021_polysome_profiling_logs", "IH_16JUL2021_LLP_COMT_vec_val_trial2_pp_trial2_3C.csv",sep="/"),sep=",")

comt_pp2_1_nrows<- sum(nrow(comt_pp2_1a_dt), nrow(comt_pp2_1b_dt), nrow(comt_pp2_1c_dt))
comt_pp2_3_nrows<- sum(nrow(comt_pp2_3a_dt), nrow(comt_pp2_3b_dt), nrow(comt_pp2_3c_dt))
comt_pp2_all_dt<- rbind(comt_pp2_1a_dt, comt_pp2_1b_dt, comt_pp2_1c_dt, comt_pp2_3a_dt, comt_pp2_3b_dt, comt_pp2_3c_dt)

comt_pp2_all_dt[,Construct:=as.factor(rep(c("UTR5 del", "endo UTR5"), 
                                          times=c(comt_pp2_1_nrows, comt_pp2_3_nrows)))]

comt_pp2_all_dt[,Gradient:=as.factor(rep(c("A", "B", "C", "A", "B", "C"), times=c(nrow(comt_pp2_1a_dt), nrow(comt_pp2_1b_dt), nrow(comt_pp2_1c_dt), nrow(comt_pp2_3a_dt), nrow(comt_pp2_3b_dt), nrow(comt_pp2_3c_dt))))]

comt_pp2_all_dt[,ID:=as.factor(paste(Construct, Gradient, sep=":"))]

# Polysome profiling trial 2 Gradient 1A
comt_pp2_1a_gg<- ggplot(
  data=comt_pp2_1a_dt, aes(x=`Position(mm)`, y=Absorbance))

comt_pp2_1a_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_pp2_1a_dt[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

rm(comt_pp2_1a_gg)

# Polysome profiling trial 2 Gradient 1B
comt_pp2_1b_gg<- ggplot(
  data=comt_pp2_1b_dt, aes(x=`Position(mm)`, y=Absorbance))

comt_pp2_1b_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_pp2_1b_dt[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

rm(comt_pp2_1b_gg)

# Polysome profiling trial 2 Gradient 1C
comt_pp2_1c_gg<- ggplot(
  data=comt_pp1_1c_dt, aes(x=`Position(mm)`, y=Absorbance))

comt_pp2_1c_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_pp2_1c_dt[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

rm(comt_pp2_1c_gg)

# Polysome profiling trial 2 Gradient 3A
comt_pp2_3a_gg<- ggplot(
  data=comt_pp2_3a_dt, aes(x=`Position(mm)`, y=Absorbance))

comt_pp2_3a_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_pp2_3a_dt[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

rm(comt_pp2_3a_gg)

# Polysome profiling trial 2 Gradient 3B
comt_pp2_3b_gg<- ggplot(
  data=comt_pp2_3b_dt, aes(x=`Position(mm)`, y=Absorbance))

comt_pp2_3b_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_pp2_3b_dt[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

rm(comt_pp2_3b_gg)

# Polysome profiling trial 2 Gradient 3C
comt_pp2_3a_gg<- ggplot(
  data=comt_pp2_3a_dt, aes(x=`Position(mm)`, y=Absorbance))

comt_pp2_3a_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_pp2_3a_dt[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

rm(comt_pp2_3a_gg)

# Plot all profiles together
comt_pp2_all_gg<- ggplot(
  data=comt_pp2_all_dt, 
  aes(x=`Position(mm)`, y=Absorbance, col=ID))

comt_pp2_all_gg+
  geom_line(lwd=0.5)+
  scale_color_npg()+
  coord_cartesian(ylim=c(-0.5,4), xlim=c(4.5,70))+
  theme_cowplot()

rm(comt_pp2_all_gg)

# Create an example
comt_pp2_3b_ex_gg<- ggplot(
  data=comt_pp2_3a_dt, aes(x=`Position(mm)`, y=Absorbance))

comt_pp2_3b_ex_gg+
  geom_line(col="steelblue", lwd=1)+
  coord_cartesian(ylim=c(-0.5,4), xlim=c(4.5,75))+
  theme_cowplot()

rm(comt_pp2_3b_ex_gg)

```

RNA yields were checked after phenol-chloroform extraction and then after DNaseI-treatment and column re-purification of RNA. 

```{r COMT UTR5 polysome profiling trial 2 RNA yields}

comt_pp2_rna_yields_dt<- as.data.table(read_xlsx(paste(comt_pp_dir, "IH_21JUL2021_LLP_COMT_vec_val_trial2_pp_trial2_RNA_extract.xlsx",sep="/")))

comt_pp2_rna_yields_dt[,Construct_short:=factor(Construct_short, levels=c("endoUTR5", "UTR5del"))]

# First plot raw RNA yields from sucrose precipitation and then phenol-chloroform extraction
comt_pp2_rna_yields_gg<- ggplot(
  data=comt_pp2_rna_yields_dt, 
  aes(x=Metafraction, y=Concentration_ng_uL, col=Construct_short))

comt_pp2_rna_yields_gg+
  expand_limits(y=0)+
  geom_point(size=2, position=position_dodge(width=0.9))+
  scale_color_manual(values=two_cols)+
  labs(y="DNaseI-treated RNA yield, ng/uL", 
       col="COMT construct")+
  theme_cowplot()+
  theme(legend.position="bottom")

rm(comt_pp2_rna_yields_gg)

# Yields after DNaseI treatment and purification with 
# Zymo RNA Clean and Concentrator, 10 µL elution in water
comt_pp2_rna_yields_gg<- ggplot(
  data=comt_pp2_rna_yields_dt, 
  aes(x=Metafraction, y=Concentration_ng_uL_DNaseI_repure, col=Construct_short))

comt_pp2_rna_yields_gg+
  expand_limits(y=0)+
  geom_point(size=2, position=position_dodge(width=0.9), alpha=0.7)+
  scale_color_manual(values=two_cols)+
  labs(y="DNaseI-treated RNA yield, ng/uL", 
       col="COMT construct")+
  theme_cowplot()+
  theme(legend.position="bottom")

rm(comt_pp2_rna_yields_gg)

# So RNA yields were similar between the two constructs
# The subunit fraction shows the lowest yield
# and the terminal fractions show highest yield
# in accordance with polysome profiles

```

1st strand cDNA was synthesized with 2.5 µM oligo dT (18) and 2.5 µM random hexamers with Superscript IV at 23 ˚C for 10 min, 55 ˚C for 1 h, then 80 ˚C for 10 min. Then RNA was degraded by addition of 1 µL NEB RNaseH and incubation at 37 ˚C for 20 min. 1st strand cDNA was diluted 1:10 in water and 2 µL was input into a 10 µL PowerUp SYBR qPCR reaction with 3-step standard cycling with anneal at 55 ˚C for 15 s and extend at 72 ˚C for 1 min.

```{r COMT UTR5 polysome profiling trial 2 RT-qPCR results}

# How well did we extract the RNA based on uniformity of IVT_B Cts?
ivt_pp2_rtqpcr_dt<- fread(paste(comt_pp_dir, "IH_22JUL2021_COMT_vec_val_tr2_pp_tr2_qPCR_IVT_tidy.csv",sep="/"),sep=",")

ivt_pp2_rtqpcr_gg<- ggplot(
  data=ivt_pp2_rtqpcr_dt, 
  aes(x=Metafraction, y=CT, col=Construct_short))

ivt_pp2_rtqpcr_gg+
  geom_point(size=2, position=position_dodge(width=0.9))+
  scale_y_continuous(breaks=seq(13,33,2))+
  coord_cartesian(ylim=c(13,33))+
  scale_color_manual(values=two_cols)+
  labs(y="IVT_B Ct, 5M copies", 
       col="COMT construct")+
  theme_cowplot()+
  theme(legend.position="bottom")

rm(ivt_pp2_rtqpcr_gg)

# I extended the limits to the range of Cts observed across COMT and IVT_B assays

# Now plot the COMT RNA abundance across polysome metafractions
comt_pp2_rtqpcr_dt<- fread(paste(
  comt_pp_dir, "IH_22JUL2021_COMT_vec_val_tr2_pp_tr2_qPCR_tidy.csv",
  sep="/"),sep=",")

comt_pp2_rtqpcr_dt[,Metafraction:=as.factor(Metafraction)]

comt_pp2_rtqpcr_summ_dt<- comt_pp2_rtqpcr_dt[
  ,.(Mean_log2_FC=mean(log2(dCt_FC))),
  by=.(Construct_short, Gradient, Metafraction)]

# Plot non-normalized log2 FC
comt_pp2_rtqpcr_gg<- ggplot(
  data=comt_pp2_rtqpcr_summ_dt, 
  aes(x=Metafraction, y=Mean_log2_FC, col=Construct_short))

comt_pp2_rtqpcr_gg+
  geom_boxplot(position=position_dodge(width=0.9))+
  geom_point(size=2, position=position_dodge(width=0.9))+
  scale_color_manual(values=two_cols)+
  labs(y="log2(FC) relative to 1A.1", 
       col="COMT construct")+
  theme_cowplot()+
  theme(legend.position="bottom")

rm(comt_pp2_rtqpcr_gg)

# Plot non-normalized log2 FC for metafractions 4 and 5
comt_pp2_4_5_rtqpcr_gg<- ggplot(
  data=comt_pp2_rtqpcr_summ_dt[Metafraction%in%c(4,5)], 
  aes(x=Metafraction, y=Mean_log2_FC, col=Construct_short))

comt_pp2_4_5_rtqpcr_gg+
  geom_boxplot(position=position_dodge(width=0.9))+
  geom_point(size=2, position=position_dodge(width=0.9))+
  scale_color_manual(values=two_cols)+
  labs(y="log2(FC) relative to 1A.1", 
       col="COMT construct")+
  theme_cowplot()+
  theme(legend.position="bottom")

rm(comt_pp2_4_5_rtqpcr_gg)

# After normalization to IVT copies
comt_pp2_rtqpcr_norm_gg<- ggplot(
  data=comt_pp2_rtqpcr_dt, 
  aes(x=Metafraction, y=log2(ddCt_FC), col=Construct_short))

comt_pp2_rtqpcr_norm_gg+
  geom_boxplot(position=position_dodge(width=0.9))+
  geom_point(size=2, position=position_dodge(width=0.9))+
  scale_color_manual(values=two_cols)+
  labs(y="IVT-normalized log2(FC)\nrelative to 1A.1", 
       col="COMT construct")+
  theme_cowplot()+
  theme(legend.position="bottom")

rm(comt_pp2_rtqpcr_norm_gg)

# Previously I observed the total mRNA abundance for the COMT UTR5 del construct
# to be ~2-fold less than the mRNA for the endo. UTR5 COMT construct
# Do we see the same in the cytoplasmic fraction?

comt_pp2_rtqpcr_cyto_gg<- ggplot(
  data=comt_pp2_rtqpcr_dt[Metafraction=="1"], 
  aes(x=Construct_short, y=ddCt_FC, col=Construct_short))

comt_pp2_rtqpcr_cyto_gg+
  geom_boxplot()+
  geom_jitter(size=2, width=0.1)+
  coord_cartesian(ylim=c(0,4))+
  scale_color_manual(values=two_cols)+
  labs(y="IVT-normalized FC\nrelative to 1A.1", 
       x="COMT construct")+
  theme_cowplot()+
  theme(legend.position="bottom")

rm(comt_pp2_rtqpcr_cyto_gg)

# Metafraction 1
t.test(comt_pp2_rtqpcr_dt[Metafraction==1 & Construct_short=="natUTR5", log2(dCt_FC)],
       comt_pp2_rtqpcr_dt[Metafraction==1 & Construct_short=="UTR5del", log2(dCt_FC)],
       alternative="two.sided")

# Metafraction 2
t.test(comt_pp2_rtqpcr_dt[Metafraction==2 & Construct_short=="natUTR5", log2(dCt_FC)],
       comt_pp2_rtqpcr_dt[Metafraction==2 & Construct_short=="UTR5del", log2(dCt_FC)],
       alternative="two.sided")

# Metafraction 3
t.test(comt_pp2_rtqpcr_dt[Metafraction==3 & Construct_short=="natUTR5", log2(dCt_FC)],
       comt_pp2_rtqpcr_dt[Metafraction==3 & Construct_short=="UTR5del", log2(dCt_FC)],
       alternative="two.sided")

# Metafraction 4
t.test(comt_pp2_rtqpcr_dt[Metafraction==4 & Construct_short=="natUTR5", log2(dCt_FC)],
       comt_pp2_rtqpcr_dt[Metafraction==4 & Construct_short=="UTR5del", log2(dCt_FC)],
       alternative="two.sided")

# Metafraction 5
t.test(comt_pp2_rtqpcr_dt[Metafraction==5 & Construct_short=="natUTR5", log2(dCt_FC)],
       comt_pp2_rtqpcr_dt[Metafraction==5 & Construct_short=="UTR5del", log2(dCt_FC)],
       alternative="two.sided")

```

The UTR5 del construct is less engaged with 80S to 7-somes, but more engaged with the heavy polysomes (without synthetic spike-in normalization: see comment below). The overall decrease in COMT RNA in metafractions 2-4 compared to metafractions 1 and 5 agrees with the lower mRNA abundance observed in RT-qPCR results total RNA. 

However, when normalizing COMT signal to synthetic spike-in RNA signal, the increase for the UTR5 del in metafraction 5 (heavy polysomes) disappears. This suggests that observed difference in this metafraction may be due to differences in extraction efficiency.

Thus, the only consistent trend is lower COMT mRNA in monosome to 7-some fractions.


# UTR5 mutants

```{r COMT UTR5 mutants}

comt_utr5_mutant_WT_file<- paste(comt_utr5_mutant_dir, "export_WT_Tube_001_Q2 DsRed-H+ , GFP-H+.csv", sep="/")
comt_utr5_mutant_uORFA_file<- paste(comt_utr5_mutant_dir, "export_uORF_A_mutant_Tube_001_Q2 DsRed-H+ , GFP-H+.csv", sep="/")
comt_utr5_mutant_uORFB_file<- paste(comt_utr5_mutant_dir, "export_uORF_B_mutant_Tube_001_Q2 DsRed-H+ , GFP-H+.csv", sep="/")
comt_utr5_mutant_uORFB_restore_file<- paste(comt_utr5_mutant_dir, "export_uORF_B_restore_mutant_Tube_001_Q2 DsRed-H+ , GFP-H+.csv", sep="/")
comt_utr5_mutant_Flagdel_file<- paste(comt_utr5_mutant_dir, "export_Flag_del_Tube_001_Q2 DsRed-H+ , GFP-H+.csv", sep="/")
comt_utr5_mutant_PCBPdel_file<- paste(comt_utr5_mutant_dir, "export_PCBP_del_Tube_001_Q2 DsRed-H+ , GFP-H+.csv", sep="/")

comt_utr5_mutant_files<- list(comt_utr5_mutant_WT_file, comt_utr5_mutant_uORFA_file, comt_utr5_mutant_uORFB_file, 
                              comt_utr5_mutant_uORFB_restore_file, comt_utr5_mutant_Flagdel_file, comt_utr5_mutant_PCBPdel_file)

comt_utr5_mutant_list<- lapply(comt_utr5_mutant_files, fread, sep=",")
comt_utr5_mutant_list_nrows<- sapply(comt_utr5_mutant_list, nrow)
comt_utr5_mutant_list_all_dt<- rbindlist(comt_utr5_mutant_list)

rm(comt_utr5_mutant_list)

comt_utr5_mutant_list_all_dt[,Sample:=as.factor(
  rep(c("WT", "uORF_A", "uORF_B", "uORF_B_restore", "Flag_del", "PCBP_del"), times=comt_utr5_mutant_list_nrows))]

comt_utr5_mutant_list_all_dt[,Is_del:=FALSE]
comt_utr5_mutant_list_all_dt[grepl("del", Sample), Is_del:=TRUE]

comt_utr5_mutants_gg<- ggplot(
  data=comt_utr5_mutant_list_all_dt, 
  aes(x=log10(`GFP-A`), col=Sample))

comt_utr5_mutants_gg+
  facet_wrap(~Is_del)+
  geom_density()+
  scale_color_brewer(palette="Set1")+
  theme_cowplot()

# Protein abundance
comt_utr5_mutant_list_all_dt[,median(log10(`GFP-A`), na.rm=TRUE), by=.(Sample)]
# uORF_A: (10^3.691297)/(10^3.694595)
# uORF_B: (10^3.998965)/(10^3.694595)
# uORF_B_restore: (10^4.100805)/(10^3.694595)
# Flag_del: (10^4.195139)/(10^3.694595)
# PCBP_del: (10^3.826845)/(10^3.694595)

comt_utr5_mutant_list_all_dt[Sample%in%c("WT", "uORF_A"), wilcox.test(log10(`GFP-A`)~Sample, na.option=na.omit)]
comt_utr5_mutant_list_all_dt[Sample%in%c("WT", "uORF_B"), wilcox.test(log10(`GFP-A`)~Sample, na.option=na.omit)]
comt_utr5_mutant_list_all_dt[Sample%in%c("WT", "uORF_B_restore"), wilcox.test(log10(`GFP-A`)~Sample, na.option=na.omit)]
comt_utr5_mutant_list_all_dt[Sample%in%c("WT", "Flag_del"), wilcox.test(log10(`GFP-A`)~Sample, na.option=na.omit)]
comt_utr5_mutant_list_all_dt[Sample%in%c("WT", "PCBP_del"), wilcox.test(log10(`GFP-A`)~Sample, na.option=na.omit)]

# All except uORF_A were significant

# RNA abundance
comt_utr5_mutant_list_all_dt[,median(log10(`DsRed-A`), na.rm=TRUE), by=.(Sample)]
# uORF_A: (10^3.846602)/(10^3.861384)
# uORF_B: (10^3.872886)/(10^3.861384)
# uORF_B_restore: (10^3.985646)/(10^3.861384)
# Flag_del: (10^3.950203)/(10^3.861384)
# PCBP_del: (10^3.953305)/(10^3.861384)

comt_utr5_mutant_list_all_dt[Sample%in%c("WT", "uORF_A"), wilcox.test(log10(`DsRed-A`)~Sample, na.option=na.omit)]
comt_utr5_mutant_list_all_dt[Sample%in%c("WT", "uORF_B"), wilcox.test(log10(`DsRed-A`)~Sample, na.option=na.omit)]
comt_utr5_mutant_list_all_dt[Sample%in%c("WT", "uORF_B_restore"), wilcox.test(log10(`DsRed-A`)~Sample, na.option=na.omit)]
comt_utr5_mutant_list_all_dt[Sample%in%c("WT", "Flag_del"), wilcox.test(log10(`DsRed-A`)~Sample, na.option=na.omit)]
comt_utr5_mutant_list_all_dt[Sample%in%c("WT", "PCBP_del"), wilcox.test(log10(`DsRed-A`)~Sample, na.option=na.omit)]





```


# UTR5-Flag-COMT-moxGFP ribosome profiling

In this experiment, cells were treated with both CHX and ANS at 100 µg/mL. Both conventional and targeted (i.e. selective) ribosome profiling against the Flag tag were performed. For targeted RP two conditions were used: 100 µL and 400 µL Agarose M2 anti-Flag beads in pulldown.

```{r COMT convential and targeted RP}

trp_trial4_ribofile<- Ribo(paste(comt_rp_dir, "riboflow_output", "ribo", "all.ribo", sep="/"))

reads_post_dedup<- experiment_info(trp_trial4_ribofile)$total.reads
names(reads_post_dedup)<- experiment_info(trp_trial4_ribofile)$experiment

reference_names<- get_reference_names(trp_trial4_ribofile)
comt_reference_name<- reference_names[grepl("COMT-202", reference_names)]

# Stop coordinate considers 24 nt Flag tag after start codon
comt_cds_start<- 252
comt_cds_stop<- 1805
comt_cds_positions<- seq(comt_cds_start,comt_cds_stop)

# Offset determined by examining stop codon metagene plot
psite_fiveprime_offset<- 13
mean_rpf_length<- 32

# about 30 codons/AAs
exit_tunnel_offset<- 90

two_cols<- viridis(n=2, alpha=0.7, end=0.3)
three_cols<- c("powderblue", "lightseagreen", "forestgreen")

min_rpf_len<- 28
max_rpf_len<- 35

```


```{r T-RP trial 4 QC}

region.counts<- get_length_distribution(
  ribo.object=trp_trial4_ribofile,
  region="CDS", range.lower=15, range.upper=40)

plot_length_distribution(region.counts, fraction=TRUE)

plot_region_counts(trp_trial4_ribofile, range.lower=min_rpf_len, range.upper=max_rpf_len)

plot_metagene(trp_trial4_ribofile,
              site="start", normalize=TRUE, title="Metagene start coverage",
              experiment=experiments(trp_trial4_ribofile))

plot_metagene(trp_trial4_ribofile,
              site="stop",  normalize=TRUE, title="Metagene stop coverage",
              experiment=experiments(trp_trial4_ribofile))

comt_coverage_dt<- as.data.table(
  get_coverage(
    ribo.object=trp_trial4_ribofile,
    range.lower=min_rpf_len, range.upper=max_rpf_len,
    name=comt_reference_name, length=TRUE, tidy=TRUE))

comt_coverage_dt[,position:=as.integer(position)]

comt_coverage_cds_nonzero_dt<- comt_coverage_dt[
  count>0 & position%in%seq(comt_cds_start, comt_cds_stop)]

comt_coverage_cds_nonzero_sum_dt<- comt_coverage_cds_nonzero_dt[
  ,.(Sum_counts=sum(count)),by=.(experiment)]

comt_coverage_cds_nonzero_sum_dt[,Reads_post_dedup:=reads_post_dedup]
comt_coverage_cds_nonzero_sum_dt[,CPM:=Sum_counts/(Reads_post_dedup/1000000)]

comt_coverage_cds_nonzero_gg<- ggplot(
  data=comt_coverage_cds_nonzero_sum_dt, 
  aes(x=experiment, y=CPM))

comt_coverage_cds_nonzero_gg+
  geom_bar(stat="identity")+
  labs(x="Sample", y="Sum CPM")+
  theme_cowplot()

# Instead of normalizing to total unique reads, normalize to CPM after read length filtering
all_region_counts_dt<- as.data.table(
  get_region_counts(
    ribo.object=trp_trial4_ribofile,
    range.lower=min_rpf_len, range.upper=max_rpf_len, region=c("CDS"),
    length=TRUE, tidy=TRUE, normalize=TRUE, transcript=FALSE))

comt_coverage_cds_sum_dt<- all_region_counts_dt[
  transcript==comt_reference_name, .(experiment, count)]

comt_coverage_cds_nonzero_gg<- ggplot(
  data=all_region_counts_dt[transcript==comt_reference_name, .(experiment, count)], 
  aes(x=experiment, y=count))

comt_coverage_cds_nonzero_gg+
  geom_bar(stat="identity")+
  scale_color_manual(values=three_cols)+
  labs(x="Sample", y="Sum CPM")+
  theme_cowplot()

# This rescues counts for the targeted RP samples relative to conventional, indicating
# a higher proportion of reads in the conventional RP are short
rm(comt_coverage_cds_nonzero_gg)

```


```{r T-RP trial 4 COMT expression}

control_ref_names<- sort(reference_names[grepl("ACTB-217|CHMP2A-206|GAPDH-205|GPI-201|OAZ1-211|PSMB2-201|VCP-201", reference_names)])

psmb2_coverage_dt<- as.data.table(
  get_coverage(
    ribo.object=trp_trial4_ribofile,
    range.lower=min_rpf_len, range.upper=max_rpf_len,
    name=control_ref_names[3], length=TRUE, tidy=TRUE))

oaz1_coverage_dt<- as.data.table(
  get_coverage(
    ribo.object=trp_trial4_ribofile,
    range.lower=min_rpf_len, range.upper=max_rpf_len,
    name=control_ref_names[6], length=TRUE, tidy=TRUE))

gpi_coverage_dt<- as.data.table(
  get_coverage(
    ribo.object=trp_trial4_ribofile,
    range.lower=min_rpf_len, range.upper=max_rpf_len,
    name=control_ref_names[1], length=TRUE, tidy=TRUE))

gapdh_coverage_dt<- as.data.table(
  get_coverage(
    ribo.object=trp_trial4_ribofile,
    range.lower=min_rpf_len, range.upper=max_rpf_len,
    name=control_ref_names[4], length=TRUE, tidy=TRUE))

actb_coverage_dt<- as.data.table(
  get_coverage(
    ribo.object=trp_trial4_ribofile,
    range.lower=min_rpf_len, range.upper=max_rpf_len,
    name=control_ref_names[7], length=TRUE, tidy=TRUE))

all_coverage_dt<- rbind(comt_coverage_dt, psmb2_coverage_dt, oaz1_coverage_dt, 
                        gpi_coverage_dt, gapdh_coverage_dt, actb_coverage_dt)

all_trx<- c("COMT", "PSMB2", "OAZ1", "GPI", "GAPDH", "ACTB")

all_coverage_dt[,Gene:=rep(all_trx, times=c(
  nrow(comt_coverage_dt), nrow(psmb2_coverage_dt), nrow(oaz1_coverage_dt), 
  nrow(gpi_coverage_dt), nrow(gapdh_coverage_dt), nrow(actb_coverage_dt)))]

rm(list=c("comt_coverage_dt", "psmb2_coverage_dt", "oaz1_coverage_dt", "gpi_coverage_dt", "gapdh_coverage_dt", "actb_coverage_dt"))

all_coverage_dt[,position:=as.integer(position)]

all_coverage_sum_dt<- all_coverage_dt[
  ,.(Sum_counts=sum(count)),by=.(experiment, Gene)]

all_coverage_sum_dt[,Reads_post_dedup:=rep(reads_post_dedup,times=length(all_trx))]
all_coverage_sum_dt[,CPM:=Sum_counts/(Reads_post_dedup/1000000)]

comt_expr_gg<- ggplot(
  data=all_coverage_sum_dt, 
  aes(x=Gene, y=CPM, fill=experiment))

comt_expr_gg+
  geom_bar(stat="identity", position="dodge", alpha=0.75)+
  labs(x="Transcript", y="CPM, sum all positions", fill="Sample")+
  scale_fill_aaas()+
  theme_cowplot()+
  theme(legend.position="bottom")

comt_cds_cov_dt<- all_coverage_dt[experiment=="20210607-FlagCOMT-400uL" & Gene=="COMT" & 
                                   position<=comt_cds_stop & position>=comt_cds_start]

comt_expr_gg<- ggplot(
  data=comt_cds_cov_dt, 
  aes(x=position, y=count))

comt_expr_gg+
  geom_point(alpha=0.5)+
  scale_color_manual(values=two_cols[1])+
  labs(x="COMT transcript position", y="RPF counts")+
  theme_cowplot()

rm(comt_expr_gg)

```


```{r COMT coverage}

comt_cov_by_pos_dt<- all_coverage_dt[Gene=="COMT",]
comt_cov_by_pos_dt[,Reads_post_dedup:=rep(reads_post_dedup, nrow(comt_cov_by_pos_dt)/3)]
comt_cov_by_pos_dt[,CPM:=count/(Reads_post_dedup/1000000)]

comt_cov_by_pos_cast_dt<- dcast(
  data=comt_cov_by_pos_dt, formula=Gene + position ~ experiment, 
  value.var="CPM")

comt_cov_by_pos_cast_dt[,CPM_diff:=`20210607-FlagCOMT-400uL` - `20210607-conventional`]

cov_spearman_cor<- comt_cov_by_pos_cast_dt[
  ,cor(`20210607-FlagCOMT-400uL`, `20210607-conventional`, method="spearman")]

cov_pearson_cor<- comt_cov_by_pos_cast_dt[
  ,cor(log10(`20210607-FlagCOMT-400uL`+1), log10(`20210607-conventional`+1))]

# comt normalized coverage in T-RP - C-RP CPM diff, nonzero positions plotted
comt_cov_by_pos_gg<- ggplot(
  data=comt_cov_by_pos_cast_dt[CPM_diff!=0],
  aes(x=log10(position), y=CPM_diff))

comt_cov_by_pos_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth(method="loess", col="firebrick")+
  labs(x="log10 transcript position",
       y="Flag-COMT T-RP:C-RP CPM diff.")+
  theme_cowplot()

# Coverage by position for T-RP only, nonzero positions
comt_cov_by_pos_gg<- ggplot(
  data=comt_cov_by_pos_cast_dt[CPM_diff!=0],
  aes(x=log10(position), y=`20210607-FlagCOMT-400uL`))

comt_cov_by_pos_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth(method="loess", col="firebrick")+
  labs(x="log10 transcript position", y="Flag-COMT T-RP CPM")+
  theme_cowplot()

rm(comt_cov_by_pos_gg)

# Show mirror plots for T-RP vs. C-RP
comt_cov_by_pos_cast_dt[,CRP_mirror:=-1*`20210607-conventional`]

# Bin positions to show position enrichment more clearly
comt_cov_by_pos_cast_dt[position%in%comt_cds_positions,
                       Pos_bin:=cut(position, dig.lab=10, 
                                    breaks=round(quantile(position, probs=seq(0,1,0.1))))]

comt_cov_by_pos_cast_melt_dt<- melt(
  data=comt_cov_by_pos_cast_dt, 
  id.vars=c("Gene", "position", "Pos_bin"), 
  measure.vars=c("20210607-FlagCOMT-400uL", "CRP_mirror"), 
  variable.name="Experiment", value.name="CPM")

comt_cov_by_pos_cast_melt_dt[grepl("FlagCOMT", Experiment), Experiment_clean:="T-RP"]
comt_cov_by_pos_cast_melt_dt[is.na(Experiment_clean), Experiment_clean:="C-RP"]
comt_cov_by_pos_cast_melt_dt[,Experiment_clean:=as.factor(Experiment_clean)]

comt_cov_by_pos_mirror_gg<- ggplot(
  data=comt_cov_by_pos_cast_melt_dt[!is.na(Pos_bin)],
  aes(x=position, y=CPM, col=Experiment_clean))

comt_cov_by_pos_mirror_gg+
  facet_wrap(~Pos_bin, ncol=5, scales="free_x")+
  geom_point(alpha=0.75)+
  scale_color_manual(values=two_cols)+
  labs(x="Transcript position", y="CPM", col="Method")+
  theme_cowplot()+
  theme(legend.position="bottom",
        axis.text.x=element_text(angle=90))

rm(comt_cov_by_pos_mirror_gg)

# Plot footprints at the H62-V63 hotspot
# 393 is start of ROI_1, 483 is end
# 459 is Flag-adjusted position of H62 in the CDS
# N71 is at pos 486
# S74 is at pos 495
# M51 is at pos 426


# 446 (459 - 13) is P-site adjusted 5' end for H62
# 473 N71
# 495 S74
# 413 M51

comt_cov_by_pos_cast_melt_dt[Experiment_clean=="C-RP", CPM:=-1*CPM]

# ROI_1
comt_cov_rna_hotspot_gg<- ggplot(
  data=comt_cov_by_pos_cast_melt_dt[Pos_bin=="(407,563]" & Experiment_clean=="C-RP"],
  aes(x=position, y=CPM))

# Convert marks to A-site adjusted locations
comt_cov_rna_hotspot_gg+
  geom_vline(xintercept=c(413,446,473,482)+3, col="grey", linetype=2)+
  geom_point(alpha=0.75, size=2)+
  coord_cartesian(xlim=c(408,495))+
  scale_x_continuous(breaks=seq(407,494,6))+
  scale_color_manual(values=two_cols)+
  labs(x="Transcript position", y="CPM", col="Method")+
  theme_cowplot()

# ROI_2
# 681 is ROI_2 start, 750 is end
# M152 is 729, V158 is 750
# Plot V158 position
comt_cov_rna_hotspot_gg<- ggplot(
  data=comt_cov_by_pos_cast_melt_dt[Pos_bin%in%c("(563,718]", "(718,873]") & Experiment_clean=="C-RP"],
  aes(x=position, y=CPM))

# Convert marks to A-site adjusted locations
comt_cov_rna_hotspot_gg+
  geom_vline(xintercept=c(716+3), col="grey", linetype=2)+
  geom_point(alpha=0.75, size=2)+
  coord_cartesian(xlim=c(681,750), ylim=c(0,10))+
  scale_x_continuous(breaks=seq(677,749,6))+
  scale_color_manual(values=two_cols)+
  labs(x="Transcript position", y="CPM", col="Method")+
  theme_cowplot()

rm(comt_cov_rna_hotspot_gg)

```


```{r COMT coverage in UTR5}

# comt_reference_name
uorf_a_pos<- 178  # GCTTTG[CTG]CCG
uorf_b_pos<- 224  # AGAGGG[CTG]GAG
comt_membrane_start<- comt_cds_start  # TTGAAG[ATG]GAC
comt_cytosolic_start<- 426  # CTGCTC[ATG]GGT
comt_downstream_start<- 543  # TGGGCC[ATG]AAC
comt_downstream_start_2<- 651  # GTGCGC[ATG]GCC
comt_oof_downstream_start<- 472 # TGCAGC[ATG]CGG
mcherry_start<- 2444

comt_uorf_starts<- c(uorf_a_pos, uorf_b_pos)
comt_orf_starts<- c(comt_membrane_start, comt_cytosolic_start)
comt_downstream_starts<- c(comt_downstream_start, comt_downstream_start_2, comt_oof_downstream_start)

comt_utr5_cds_positions<- seq(0,1805)
comt_utr5_cds_positions_zoom<- seq(0,comt_cytosolic_start+240)
mcherry_utr5_cds_positions<- seq(mcherry_start-90, mcherry_start+90)

# First look at the signal in the 5' UTR, CDS, and downstream IRES for the conventional RP sample
comt_cov_by_pos_zoom_dt<- comt_cov_by_pos_dt[
    experiment=="20210607-conventional" & position%in%comt_utr5_cds_positions_zoom,]

# 5' RPF ends plotted with start codons/offsets marked with vertical lines
comt_cov_by_pos_gg<- ggplot(
  data=comt_cov_by_pos_zoom_dt,
  aes(x=position, y=CPM))

asite_fiveprime_offset<- 16
psite_fiveprime_offset<- asite_fiveprime_offset - 3
comt_orf_5prime_adj_pos<- comt_orf_starts - psite_fiveprime_offset
comt_uorf_5prime_adj_pos<- comt_uorf_starts - psite_fiveprime_offset

comt_cov_by_pos_gg+
  geom_vline(xintercept=comt_orf_5prime_adj_pos, col="darkorange", linetype=2)+
  geom_vline(xintercept=comt_uorf_5prime_adj_pos, col="pink", linetype=2)+
  geom_point(alpha=0.75, size=2)+
  geom_smooth(span=0.2, se=FALSE)+
  scale_x_continuous(breaks=seq(0,660,30))+
  scale_color_manual(values=two_cols[1])+
  labs(x="Transcript position", y="CPM")+
  theme_cowplot()+
  theme(axis.text.x=element_text(angle=90))

# Suggests initiation at uORF_A and membrane isoform
# possibly initiation at uORF_B

# Zoom in to uORF and cORF start codons
comt_cov_by_pos_gg+
  geom_vline(xintercept=comt_orf_5prime_adj_pos, col="darkorange", linetype=2)+
  geom_vline(xintercept=comt_uorf_5prime_adj_pos, col="pink", linetype=2)+
  geom_point(alpha=0.75, size=2)+
  geom_smooth(span=0.2, se=FALSE)+
  scale_x_continuous(breaks=seq(0,700,30))+
  coord_cartesian(xlim=c(60, 450))+
  scale_color_manual(values=two_cols[1])+
  labs(x="Transcript position", y="CPM")+
  theme_cowplot()

comt_cov_by_pos_gg+
  geom_vline(xintercept=comt_orf_5prime_adj_pos, col="darkorange", linetype=2)+
  geom_vline(xintercept=comt_uorf_5prime_adj_pos, col="pink", linetype=2)+
  geom_point(alpha=0.75, size=2)+
  scale_x_continuous(breaks=seq(0,700,3))+
  coord_cartesian(xlim=c(200, 250))+
  scale_color_manual(values=two_cols[1])+
  labs(x="Transcript position", y="CPM")+
  theme_cowplot()

# -12 nt difference between expected p site site and signal
# -8-9 bp from expected cORF start and signal

# Look for disomes with leading ribosome having open A site
comt_cov_by_pos_gg<- ggplot(
  data=comt_cov_by_pos_zoom_dt,
  aes(x=position, y=CPM))

psite_fiveprime_offset<- 45
comt_orf_5prime_adj_pos<- comt_orf_starts - psite_fiveprime_offset
comt_uorf_5prime_adj_pos<- comt_uorf_starts - psite_fiveprime_offset

comt_cov_by_pos_gg+
  geom_vline(xintercept=comt_orf_5prime_adj_pos, col="darkorange", linetype=3)+
  geom_vline(xintercept=comt_uorf_5prime_adj_pos, col="pink", linetype=3)+
  geom_point(alpha=0.75, size=2)+
  geom_smooth(span=0.2, se=FALSE)+
  scale_x_continuous(breaks=seq(0,660,30))+
  scale_color_manual(values=two_cols[1])+
  labs(x="Transcript position", y="CPM", title="45 nt P-site offset")+
  theme_cowplot()

rm(comt_cov_by_pos_gg)

# This suggests uORF_A and the cytosolic isoform have disomes with
# leading ribosome awaiting Met-tRNA

# Investigate signal for the IRES-mCherry start
comt_cov_by_pos_mcherry_dt<- comt_cov_by_pos_dt[
    experiment=="20210607-conventional" & position%in%mcherry_utr5_cds_positions,]

comt_cov_by_pos_mcherry_gg<- ggplot(
  data=comt_cov_by_pos_mcherry_dt,
  aes(x=position, y=CPM))

psite_fiveprime_offset<- asite_fiveprime_offset - 3
mcherry_orf_5prime_adj_pos<- mcherry_start - psite_fiveprime_offset

comt_cov_by_pos_mcherry_gg+
  geom_vline(xintercept=mcherry_orf_5prime_adj_pos, col="darkorange", linetype=2)+
  geom_point(alpha=0.75, size=2)+
  geom_smooth(span=0.2, se=FALSE)+
  scale_x_continuous(breaks=seq(2350,2560,30))+
  scale_color_manual(values=two_cols[1])+
  labs(x="Transcript position", y="CPM", 
       title="mCherry start, 15 nt P-site offset")+
  theme_cowplot()

# No clear initiating signal for mCherry with 15 nt offset

# Now for disome offset
comt_cov_by_pos_mcherry_gg<- ggplot(
  data=comt_cov_by_pos_mcherry_dt,
  aes(x=position, y=CPM))

psite_fiveprime_offset<- 45
mcherry_orf_5prime_adj_pos<- mcherry_start - psite_fiveprime_offset

comt_cov_by_pos_mcherry_gg+
  geom_vline(lwd=1.05, xintercept=mcherry_orf_5prime_adj_pos, col="darkorange", linetype=3)+
  geom_point(alpha=0.75, size=2)+
  geom_smooth(span=0.2, se=FALSE)+
  scale_x_continuous(breaks=seq(2350,2560,30))+
  scale_color_manual(values=two_cols[1])+
  labs(x="Transcript position", y="CPM", 
       title="mCherry start, 45 nt P-site offset")+
  theme_cowplot()

rm(comt_cov_by_pos_mcherry_gg)

# IRES-mCherry, translation initiation likely starts just upstream of the start codon


# This peak corresponds to uORF_A signal
first_peak_pos<- 131

# comt_cov_by_pos_zoom_dt[position%in%seq(uorf_a_pos-8,uorf_a_pos+8),]
# This peak corresponds to uORF_B signal
second_peak_pos<- 183  # seq(177,188)

# comt_cov_by_pos_zoom_dt[position%in%seq(uorf_b_pos-8,uorf_b_pos+8),]
# This peak corresponds to memORF signal
third_peak_pos<- 220  # seq(217,223)

# comt_cov_by_pos_zoom_dt[position%in%seq(comt_cytosolic_start-8,comt_cytosolic_start+8),]
# This peak corresponds to cytoORF signal
fourth_peak_pos<- 426  # seq(424,427)

#comt_cov_by_pos_zoom_dt[position%in%seq(uorf_a_pos-8,uorf_a_pos+8),]
fifth_peak_pos<- 493  # seq(493,497)

# Where are the local peaks relative to the starts?

# uorf_a_pos - first_peak_pos
# 47 nt suggests a disome with leading ribosome having an open A site

# uorf_b_pos - second_peak_pos
# 39-42 nt length is too long for a single ribosome
# and is slightly too short for a disome with open A site

# comt_membrane_start - third_peak_pos
# 32 nt is the length of a single ribosome footprint
# this is inconsistent with a ~15 bp P-site offset
# and inconsistent with a disome with leading ribosome having open A site

# comt_cytosolic_start - fourth_peak_pos
# 0 nt indicates this peak is not reflecting the membrane start signal
# could this be due to the out-of-frame downstream ATG?

# comt_oof_downstream_start - fourth_peak_pos
# 46 nt length suggests disome with leading ribosome having open A site

# Does the first in-frame downstream ATG line up with the fifth peak?
# comt_downstream_start - fifth_peak_pos
# 50 nt 
# However, if using the max position for this peak region (pos 497)
# the length is 46 nt, suggesting the fifth peak may be due to 
# initiation at the downstream in-frame ATG

# 20210607-FlagCOMT-400uL
# 20210607-FlagCOMT-100uL
# 20210607-conventional

```


# Polysome profiling traces

```{r COMT ROI_1 polysome profiling trial 1}

comt_roi1_1<- fread(paste(comt_pp_roi1_trial1_dir, "IH_15SEP2021_COMT_ROI1_1.csv",sep="/"),sep=",")
comt_roi1_2<- fread(paste(comt_pp_roi1_trial1_dir, "IH_15SEP2021_COMT_ROI1_2.csv",sep="/"),sep=",")
comt_roi1_3<- fread(paste(comt_pp_roi1_trial1_dir, "IH_15SEP2021_COMT_ROI1_3.csv",sep="/"),sep=",")
comt_roi1_4<- fread(paste(comt_pp_roi1_trial1_dir, "IH_15SEP2021_COMT_ROI1_4.csv",sep="/"),sep=",")
comt_roi1_5<- fread(paste(comt_pp_roi1_trial1_dir, "IH_15SEP2021_COMT_ROI1_5.csv",sep="/"),sep=",")
comt_roi1_6<- fread(paste(comt_pp_roi1_trial1_dir, "IH_15SEP2021_COMT_ROI1_6.csv",sep="/"),sep=",")

# Gradient 1
comt_roi1_1_gg<- ggplot(
  data=comt_roi1_1, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_1_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_1[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 2
comt_roi1_2_gg<- ggplot(
  data=comt_roi1_2, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_2_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_2[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 3
comt_roi1_3_gg<- ggplot(
  data=comt_roi1_3, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_3_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_3[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 4
comt_roi1_4_gg<- ggplot(
  data=comt_roi1_4, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_4_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_4[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 5
comt_roi1_5_gg<- ggplot(
  data=comt_roi1_5, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_5_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_5[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 6
comt_roi1_6_gg<- ggplot(
  data=comt_roi1_6, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_6_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_6[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()



```


```{r COMT ROI_1 polysome profiling trial 2}

comt_roi1_1<- fread(paste(comt_pp_roi1_trial2_dir, "IH_24SEP2021_COMT_ROI1_trial2_pp_1.csv",sep="/"),sep=",")
comt_roi1_2<- fread(paste(comt_pp_roi1_trial2_dir, "IH_24SEP2021_COMT_ROI1_trial2_pp_2.csv",sep="/"),sep=",")
comt_roi1_3<- fread(paste(comt_pp_roi1_trial2_dir, "IH_24SEP2021_COMT_ROI1_trial2_pp_3.csv",sep="/"),sep=",")
comt_roi1_4<- fread(paste(comt_pp_roi1_trial2_dir, "IH_24SEP2021_COMT_ROI1_trial2_pp_4.csv",sep="/"),sep=",")
comt_roi1_5<- fread(paste(comt_pp_roi1_trial2_dir, "IH_24SEP2021_COMT_ROI1_trial2_pp_5.csv",sep="/"),sep=",")
comt_roi1_6<- fread(paste(comt_pp_roi1_trial2_dir, "IH_24SEP2021_COMT_ROI1_trial2_pp_6.csv",sep="/"),sep=",")

# Gradient 1
comt_roi1_1_gg<- ggplot(
  data=comt_roi1_1, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_1_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_1[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 2
comt_roi1_2_gg<- ggplot(
  data=comt_roi1_2, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_2_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_2[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 3
comt_roi1_3_gg<- ggplot(
  data=comt_roi1_3, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_3_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_3[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 4
comt_roi1_4_gg<- ggplot(
  data=comt_roi1_4, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_4_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_4[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 5
comt_roi1_5_gg<- ggplot(
  data=comt_roi1_5, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_5_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_5[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 6
comt_roi1_6_gg<- ggplot(
  data=comt_roi1_6, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_6_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_6[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()



```


```{r COMT ROI_1 polysome profiling trial 3}

comt_roi1_1<- fread(paste(comt_pp_roi1_trial3_dir, "IH_29SEP2021_COMT_ROI1_trial3_1.csv",sep="/"),sep=",")
comt_roi1_2<- fread(paste(comt_pp_roi1_trial3_dir, "IH_29SEP2021_COMT_ROI1_trial3_2.csv",sep="/"),sep=",")
comt_roi1_3<- fread(paste(comt_pp_roi1_trial3_dir, "IH_29SEP2021_COMT_ROI1_trial3_3.csv",sep="/"),sep=",")
comt_roi1_4<- fread(paste(comt_pp_roi1_trial3_dir, "IH_29SEP2021_COMT_ROI1_trial3_4.csv",sep="/"),sep=",")
comt_roi1_5<- fread(paste(comt_pp_roi1_trial3_dir, "IH_29SEP2021_COMT_ROI1_trial3_5.csv",sep="/"),sep=",")
comt_roi1_6<- fread(paste(comt_pp_roi1_trial3_dir, "IH_29SEP2021_COMT_ROI1_trial3_6.csv",sep="/"),sep=",")

# Gradient 1
comt_roi1_1_gg<- ggplot(
  data=comt_roi1_1, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_1_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_1[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 2
comt_roi1_2_gg<- ggplot(
  data=comt_roi1_2, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_2_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_2[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 3
comt_roi1_3_gg<- ggplot(
  data=comt_roi1_3, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_3_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_3[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 4
comt_roi1_4_gg<- ggplot(
  data=comt_roi1_4, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_4_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_4[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 5
comt_roi1_5_gg<- ggplot(
  data=comt_roi1_5, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_5_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_5[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 6
comt_roi1_6_gg<- ggplot(
  data=comt_roi1_6, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_6_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_6[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()



```


```{r COMT ROI_1,2 polysome profiling trial 4}

comt_roi1_1<- fread(paste(comt_pp_roi1_trial4_dir, "SR_IH_1_07JUL2022.csv",sep="/"),sep=",")
comt_roi1_2<- fread(paste(comt_pp_roi1_trial4_dir, "SR_IH_2_07JUL2022.csv",sep="/"),sep=",")
comt_roi2_1<- fread(paste(comt_pp_roi1_trial4_dir, "SR_IH_3_07JUL2022.csv",sep="/"),sep=",")
comt_roi2_2<- fread(paste(comt_pp_roi1_trial4_dir, "SR_IH_4_07JUL2022.csv",sep="/"),sep=",")

# Gradient 1
comt_roi1_1_gg<- ggplot(
  data=comt_roi1_1, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_1_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_1[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 2
comt_roi1_2_gg<- ggplot(
  data=comt_roi1_2, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_2_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_2[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 3
comt_roi2_1_gg<- ggplot(
  data=comt_roi2_1, aes(x=`Position(mm)`, y=Absorbance))

comt_roi2_1_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi2_1[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 4
comt_roi2_2_gg<- ggplot(
  data=comt_roi2_2, aes(x=`Position(mm)`, y=Absorbance))

comt_roi2_2_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi2_2[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()


```

Metafraction pooling:
M1: 1-6, 100 µL water
M2: 7-12, 100 µL water
M3: 13-18, 100 µL water
M4: 19-23, 150 µL water


```{r COMT ROI_1,2 polysome profiling trial 5}

comt_roi1_c<- fread(paste(comt_pp_roi_trial5_dir, "IH_12JUL2022_COMT_ROI_1_2_trial5_1.csv",sep="/"),sep=",")

comt_roi1_d<- fread(paste(comt_pp_roi_trial5_dir, "IH_12JUL2022_COMT_ROI_1_2_trial5_2.csv",sep="/"),sep=",")

comt_roi2_c<- fread(paste(comt_pp_roi_trial5_dir, "IH_12JUL2022_COMT_ROI_1_2_trial5_3.csv",sep="/"),sep=",")

comt_roi2_d<- fread(paste(comt_pp_roi_trial5_dir, "IH_12JUL2022_COMT_ROI_1_2_trial5_4.csv",sep="/"),sep=",")

comt_roi2_a<- fread(paste(comt_pp_roi_trial5_dir, "IH_12JUL2022_COMT_ROI_1_2_trial5_5.csv",sep="/"),sep=",")

comt_roi2_b<- fread(paste(comt_pp_roi_trial5_dir, "IH_12JUL2022_COMT_ROI_1_2_trial5_6.csv",sep="/"),sep=",")


# Gradient 1
comt_roi1_c_gg<- ggplot(
  data=comt_roi1_c, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_c_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_c[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 2
comt_roi1_d_gg<- ggplot(
  data=comt_roi1_d, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_d_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_d[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 3
comt_roi2_c_gg<- ggplot(
  data=comt_roi2_c, aes(x=`Position(mm)`, y=Absorbance))

comt_roi2_c_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi2_c[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 4
comt_roi2_d_gg<- ggplot(
  data=comt_roi2_d, aes(x=`Position(mm)`, y=Absorbance))

comt_roi2_d_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi2_d[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 5
comt_roi2_a_gg<- ggplot(
  data=comt_roi2_a, aes(x=`Position(mm)`, y=Absorbance))

comt_roi2_a_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi2_a[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 6
comt_roi2_b_gg<- ggplot(
  data=comt_roi2_b, aes(x=`Position(mm)`, y=Absorbance))

comt_roi2_b_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi2_b[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()


```

Metafraction pooling:
M1: 1-5, 150 µL water
M2: 6-12, 50 µL water
M3: 13-18, 150 µL water
M4: 19-23, 150 µL water

```{r COMT ROI_1,2 polysome profiling trial 6}

comt_roi1_c<- fread(paste(comt_pp_roi_trial6_dir, "IH_22JUL2022_COMT_ROI_1_2_trial6_1.csv",sep="/"),sep=",")

comt_roi1_d<- fread(paste(comt_pp_roi_trial6_dir, "IH_22JUL2022_COMT_ROI_1_2_trial6_2.csv",sep="/"),sep=",")

comt_roi2_c<- fread(paste(comt_pp_roi_trial6_dir, "IH_22JUL2022_COMT_ROI_1_2_trial6_3.csv",sep="/"),sep=",")

comt_roi2_d<- fread(paste(comt_pp_roi_trial6_dir, "IH_22JUL2022_COMT_ROI_1_2_trial6_4.csv",sep="/"),sep=",")

# Gradient 1
comt_roi1_c_gg<- ggplot(
  data=comt_roi1_c, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_c_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_c[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 2
comt_roi1_d_gg<- ggplot(
  data=comt_roi1_d, aes(x=`Position(mm)`, y=Absorbance))

comt_roi1_d_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi1_d[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 3
comt_roi2_c_gg<- ggplot(
  data=comt_roi2_c, aes(x=`Position(mm)`, y=Absorbance))

comt_roi2_c_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi2_c[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Gradient 4
comt_roi2_d_gg<- ggplot(
  data=comt_roi2_d, aes(x=`Position(mm)`, y=Absorbance))

comt_roi2_d_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  geom_vline(xintercept=comt_roi2_d[
    `Fraction Number`!="", `Position(mm)`], col="grey", linetype=2)+
  theme_cowplot()

# Example trace
comt_roi2_c_gg<- ggplot(
  data=comt_roi2_c, aes(x=`Position(mm)`, y=Absorbance))

comt_roi2_c_gg+
  geom_line(col="steelblue")+
  coord_cartesian(ylim=c(-1,5))+
  theme_cowplot()


```

Metafraction pooling:
M1: 1-6, 100 µL water
M2: 7-12, 100 µL water
M3: 13-18, 100 µL water
M4: 19-24, 100 µL water


# COMT multiplexed assay data pre-processing

COMT_02DEC2021: polysome profiling trial 3
COMT_15AUG2022: FACS trials 1, 2
COMT_03NOV2022: polysome profiling trial 4-6 aggregate
COMT_18FEB2023: FACS trial 3, trial 2 repeat

```{r COMT library yields}

library_yield_file<- paste(comt_res_base_dir, "COMT_library_yields.txt", sep="/")
comt_lib_yield_dt<- fread(library_yield_file, sep="\t")

comt_lib_yield_gg<- ggplot(
  data=comt_lib_yield_dt, 
  aes(x=Fraction, y=Concentration_nM, col=as.factor(ROI)))

comt_lib_yield_gg+
  geom_point(position=position_jitterdodge())+
  scale_color_aaas()+
  theme_cowplot()
  
```


```{r COMT ROI_1 and ROI_2 gDNA, total cDNA, polysomal cDNA, FACS-sorted gDNA sequencing}

log10_caf_thresh<- (-5.8)

comt_02dec2021_dt<- read_vcf_summary_files(comt_02dec2021_base_dir)
comt_03nov2022_dt<- read_vcf_summary_files(comt_03nov2022_base_dir)
comt_15aug2022_dt<- read_vcf_summary_files(comt_15aug2022_base_dir)
comt_18feb2023_dt<- read_vcf_summary_files(comt_18feb2023_base_dir)
comt_23mar2023_dt<- read_vcf_summary_files(comt_23mar2023_base_dir)
comt_all_dt<- rbind(comt_02dec2021_dt, comt_03nov2022_dt, comt_15aug2022_dt, comt_18feb2023_dt, comt_23mar2023_dt)

rm(list=c("comt_02dec2021_dt", "comt_03nov2022_dt", "comt_15aug2022_dt", "comt_18feb2023_dt", "comt_23mar2023_dt"))

# Annotate input NA type

# We included two final NC libraries for ROI_1 and ROI_2 (COMT_173, COMT_174)
# However did not include these for initial manuscript submission as analysis for RNA and protein is complete
nc_samples<- paste("COMT", c(40, 108, 109), "R", sep="_")
input_lib_samples<- paste("COMT", c(41, 110, 111), "R", sep="_")

gdna_samples<- paste("COMT", c(seq(37,39), seq(112, 119)), "R", sep="_")
gdna_bio1_samples<- paste("COMT", c(37, 112, 116), "R", sep="_")
gdna_bio2_samples<- paste("COMT", c(38, 39, 113, 117), "R", sep="_")
gdna_bio3_samples<- paste("COMT", c(114, 118) , "R", sep="_")
gdna_bio4_samples<- paste("COMT", c(115, 119), "R", sep="_")

cdna_bio1_samples<- paste("COMT", c(seq(1,3), 94, 98, 102), "R", sep="_")
cdna_bio2_samples<- paste("COMT", c(seq(4,6), 95, 99, 103), "R", sep="_")
cdna_bio3_samples<- paste("COMT", c(96, 100, 104, 106), "R", sep="_")
cdna_bio4_samples<- paste("COMT", c(97, 101, 105, 107), "R", sep="_")
cdna_samples<- c(cdna_bio1_samples, cdna_bio2_samples, cdna_bio3_samples, cdna_bio4_samples)

poly_bio1_samples<- paste("COMT", c(seq(7,21), seq(62,65), seq(70,73), seq(157,160), seq(165,168)), "R", sep="_")
poly_bio2_samples<- paste("COMT", c(seq(22,36), seq(66,69), seq(74,77), seq(161,164), seq(169,172)), "R", sep="_")
poly_bio3_samples<- paste("COMT", c(seq(78,81), seq(86,89), seq(149,152)), "R", sep="_")
poly_bio4_samples<- paste("COMT", c(seq(82,85), seq(90,93), seq(153,156)), "R", sep="_")

facs_total_samples<- paste("COMT", c(seq(42,57,5), seq(120,135,5), 145, 146), "R", sep="_")
facs_q1_samples<- paste("COMT", c(seq(43,58,5), seq(121,136,5), 141), "R", sep="_")
facs_q2_samples<- paste("COMT", c(seq(44,59,5), seq(122,137,5), 140, 142), "R", sep="_")
facs_q3_samples<- paste("COMT", c(seq(45,60,5), seq(123,143,5)), "R", sep="_")
facs_q4_samples<- paste("COMT", c(seq(46,61,5), seq(124,144,5)), "R", sep="_")

facs_bio1_samples<- paste("COMT", c(seq(42,51), 145), "R", sep="_")
facs_bio2_samples<- paste("COMT", c(seq(52,61), seq(140,144)), "R", sep="_")
facs_bio3_samples<- paste("COMT", c(seq(120,129), 146), "R", sep="_")
facs_bio4_samples<- paste("COMT", c(seq(130,139)), "R", sep="_")

comt_all_dt[Sample%in%nc_samples, Input:="Negative_control"]
comt_all_dt[Sample%in%input_lib_samples, Input:="Input_library"]

comt_all_dt[Sample%in%gdna_bio1_samples, Input:="gDNA_bio1"]
comt_all_dt[Sample%in%gdna_bio2_samples, Input:="gDNA_bio2"]
comt_all_dt[Sample%in%gdna_bio3_samples, Input:="gDNA_bio3"]
comt_all_dt[Sample%in%gdna_bio4_samples, Input:="gDNA_bio4"]

comt_all_dt[Sample%in%cdna_bio1_samples, Input:="cDNA_bio1"]
comt_all_dt[Sample%in%cdna_bio2_samples, Input:="cDNA_bio2"]
comt_all_dt[Sample%in%cdna_bio3_samples, Input:="cDNA_bio3"]
comt_all_dt[Sample%in%cdna_bio4_samples, Input:="cDNA_bio4"]

comt_all_dt[Sample%in%poly_bio1_samples, Input:="Poly_bio1"]
comt_all_dt[Sample%in%poly_bio2_samples, Input:="Poly_bio2"]
comt_all_dt[Sample%in%poly_bio3_samples, Input:="Poly_bio3"]
comt_all_dt[Sample%in%poly_bio4_samples, Input:="Poly_bio4"]

comt_all_dt[Sample%in%facs_bio1_samples, Input:="FACS_bio1"]
comt_all_dt[Sample%in%facs_bio2_samples, Input:="FACS_bio2"]
comt_all_dt[Sample%in%facs_bio3_samples, Input:="FACS_bio3"]
comt_all_dt[Sample%in%facs_bio4_samples, Input:="FACS_bio4"]
comt_all_dt[,Input:=as.factor(Input)]

# Annotate region of interest/library
roi1_samples<- paste("COMT", c(
  seq(1,46), seq(52,56), seq(62,69), seq(78,85), 
  seq(94,97), 104, 105, 108, 110, seq(112,115), 
  seq(120,124), seq(130,134), 140, 146, 147, 
  seq(149,156), 173), "R", sep="_")

roi2_samples<- paste("COMT", c(
  seq(47,51), seq(57,61), seq(70,77), seq(86,93), 
  seq(98,103), 106, 107, 109, 111, seq(116,119), 
  seq(125,129), seq(135,139), seq(141,145), 148,
  seq(157,172), 174), "R", sep="_")

comt_all_dt[Sample%in%roi1_samples & POS%in%pDEST_roi1_mutagenesis_coords, ROI:="ROI_1"]
comt_all_dt[Sample%in%roi2_samples & POS%in%pDEST_roi2_mutagenesis_coords, ROI:="ROI_2"]
comt_all_dt[,ROI:=as.factor(ROI)]

# Annotate polysome metafractions
poly_metafrac_1_set1_samples<- paste("COMT", seq(7,32,5), "R", sep="_")
poly_metafrac_2_set1_samples<- paste("COMT", seq(8,33,5), "R", sep="_")
poly_metafrac_3_set1_samples<- paste("COMT", seq(9,34,5), "R", sep="_")
poly_metafrac_4_set1_samples<- paste("COMT", seq(10,35,5), "R", sep="_")
poly_metafrac_5_set1_samples<- paste("COMT", seq(11,36,5), "R", sep="_")

poly_metafrac_1_set2_samples<- paste("COMT", c(seq(62,90,4), seq(149,169,4)), "R", sep="_")
poly_metafrac_2_set2_samples<- paste("COMT", c(seq(63,91,4), seq(150,170,4)), "R", sep="_")
poly_metafrac_3_set2_samples<- paste("COMT", c(seq(64,92,4), seq(151,171,4)), "R", sep="_")
poly_metafrac_4_set2_samples<- paste("COMT", c(seq(65,93,4), seq(152,172,4)), "R", sep="_")

poly_all_metafrac_set1_samples<- c(
  poly_metafrac_1_set1_samples, poly_metafrac_2_set1_samples, 
  poly_metafrac_3_set1_samples, poly_metafrac_4_set1_samples, 
  poly_metafrac_5_set1_samples)

poly_all_metafrac_set2_samples<- c(
  poly_metafrac_1_set2_samples, poly_metafrac_2_set2_samples, 
  poly_metafrac_3_set2_samples, poly_metafrac_4_set2_samples)

poly_all_metafrac_samples<- c(poly_all_metafrac_set1_samples, poly_all_metafrac_set2_samples)

comt_all_dt[Sample%in%poly_metafrac_1_set1_samples, Fraction:="1"]
comt_all_dt[Sample%in%poly_metafrac_2_set1_samples, Fraction:="2"]
comt_all_dt[Sample%in%poly_metafrac_3_set1_samples, Fraction:="3"]
comt_all_dt[Sample%in%poly_metafrac_4_set1_samples, Fraction:="4"]
comt_all_dt[Sample%in%poly_metafrac_5_set1_samples, Fraction:="5"]

comt_all_dt[Sample%in%poly_metafrac_1_set2_samples, Fraction:="1"]
comt_all_dt[Sample%in%poly_metafrac_2_set2_samples, Fraction:="2"]
comt_all_dt[Sample%in%poly_metafrac_3_set2_samples, Fraction:="3"]
comt_all_dt[Sample%in%poly_metafrac_4_set2_samples, Fraction:="4"]
comt_all_dt[is.na(Fraction), Fraction:=""]
comt_all_dt[,Fraction:=as.factor(Fraction)]

comt_all_dt[Sample%in%facs_total_samples, Quadrant:="Input"]
comt_all_dt[Sample%in%facs_q1_samples, Quadrant:="1"]
comt_all_dt[Sample%in%facs_q2_samples, Quadrant:="2"]
comt_all_dt[Sample%in%facs_q3_samples, Quadrant:="3"]
comt_all_dt[Sample%in%facs_q4_samples, Quadrant:="4"]
comt_all_dt[is.na(Quadrant), Quadrant:=""]
comt_all_dt[,Quadrant:=as.factor(Quadrant)]

# Need the VAR_ID for RF filtering
comt_all_dt[, VAR_ID:=paste(`#CHROM`, POS, REF, ALT, sep=":")]

# Remove libraries that were accidentally swapped between ROIs during 
# library prep
comt_failed_libraries<- paste("COMT", c(seq(70,85), 147, 148), "R", sep="_")

```


```{r COMT ROI_1, ROI_2 gDNA and total and polysomal RNA sequencing variant filtering, batch correction, and rep. summarization}

prob_cutoff<- 0.5

# Train error correction models for each ROI
comt_03nov2022_sim_roi1_dt<- fread(paste(
  comt_03nov2022_sim_dir, "COMT_108_train.var.cand.vcf.summary.txt", sep="/"), sep="\t")

comt_03nov2022_sim_roi1_dt[,Sample:=as.factor("COMT_108_train")]

comt_03nov2022_sim_roi2_dt<- fread(paste(
  comt_03nov2022_sim_dir, "COMT_109_train.var.cand.vcf.summary.txt", sep="/"), sep="\t")

comt_03nov2022_sim_roi2_dt[,Sample:=as.factor("COMT_109_train")]

comt_03nov2022_sim_roi1_truth_dt<- fread(paste(
  comt_03nov2022_sim_dir, "COMT_108_train.truth.vcf", sep="/"), sep="\t", skip=20)

comt_03nov2022_sim_roi1_truth_dt[
  ,TYPE:=get_var_type(REF, ALT), by=seq_len(nrow(comt_03nov2022_sim_roi1_truth_dt))]

comt_03nov2022_sim_roi1_truth_dt[,table(TYPE)]

comt_03nov2022_sim_roi2_truth_dt<- fread(paste(
  comt_03nov2022_sim_dir, "COMT_109_train.truth.vcf", sep="/"), sep="\t", skip=20)

comt_03nov2022_sim_roi2_truth_dt[
  ,TYPE:=get_var_type(REF, ALT), by=seq_len(nrow(comt_03nov2022_sim_roi2_truth_dt))]

comt_03nov2022_sim_roi2_truth_dt[,table(TYPE)]

# log10_CAF, SUBST, TYPE, etc. are added in prep
modeling_features<- c(
  "VAR_ID", "UP_REF_NT", "DOWN_REF_NT", 
  "TYPE", "log10_CAF",
  "R1_PLUS_MED_RP", "R2_MINUS_MED_RP", 
  "R1_PLUS_MED_BQ", "R2_MINUS_MED_BQ", 
  "R1_PLUS_MED_NM", "R2_MINUS_MED_NM")

# Because the NM and RP are collinear between R1 and R2, just select the stat for one read
modeling_features_no_collinear<- modeling_features[
  !grepl("R1_PLUS_MED_RP|R1_PLUS_MED_NM", modeling_features)]

# Prep the datasets for modeling

# For ROI_1, filter out very low frequency MNPs as these likely arose
# from barcode crosstalk
comt_03nov2022_sim_roi1_dt[,TYPE:=get_var_type(REF, ALT), by=seq_len(nrow(comt_03nov2022_sim_roi1_dt))]

comt_roi1_mnp_distr_gg<- ggplot(
  data=comt_03nov2022_sim_roi1_dt[POS%in%pDEST_roi1_mutagenesis_coords],
  aes(x=log10(CAF), col=TYPE))

comt_roi1_mnp_distr_gg+
  geom_density()+
  theme_cowplot()

rm(comt_roi1_mnp_distr_gg)

# Use a log10 frequency threshold of -4 for MNPs as those observed in negative control are likely 
# from barcode crosstalk on the sequencer
comt_03nov2022_sim_roi1_filt_dt<- 
  comt_03nov2022_sim_roi1_dt[
    POS%in%pDEST_roi1_mutagenesis_coords & (TYPE=="SNP" | log10(CAF)>-4)]

comt_03nov2022_sim_roi1_prep_dt<- prep_train_dt(
  train_dt=comt_03nov2022_sim_roi1_filt_dt, 
  truth_dt=comt_03nov2022_sim_roi1_truth_dt, 
  modeling_features=modeling_features_no_collinear)

comt_03nov2022_sim_roi2_prep_dt<- prep_train_dt(
  train_dt=comt_03nov2022_sim_roi2_dt[
    POS%in%pDEST_roi2_mutagenesis_coords], 
  truth_dt=comt_03nov2022_sim_roi2_truth_dt, 
  modeling_features=modeling_features_no_collinear)

# Generate validation sets for testing
comt_03nov2022_sim_roi1_prep_folds<- create_kfolds(comt_03nov2022_sim_roi1_prep_dt)
comt_03nov2022_sim_roi1_prep_train_dt<- comt_03nov2022_sim_roi1_prep_dt[!comt_03nov2022_sim_roi1_prep_folds[[1]]]
comt_03nov2022_sim_roi1_prep_test_dt<- comt_03nov2022_sim_roi1_prep_dt[comt_03nov2022_sim_roi1_prep_folds[[1]]]

comt_03nov2022_sim_roi2_prep_folds<- create_kfolds(comt_03nov2022_sim_roi2_prep_dt)
comt_03nov2022_sim_roi2_prep_train_dt<- comt_03nov2022_sim_roi2_prep_dt[!comt_03nov2022_sim_roi2_prep_folds[[1]]]
comt_03nov2022_sim_roi2_prep_test_dt<- comt_03nov2022_sim_roi2_prep_dt[comt_03nov2022_sim_roi2_prep_folds[[1]]]


# ROI_1 model; determine mtry using CV
set.seed(9)

comt_03nov2022_sim_roi1_prep_train_cv_res<- run_nested_cv(in_dt=comt_03nov2022_sim_roi1_prep_train_dt, hyperparam_prop=0.3, kfolds=5)

comt_03nov2022_sim_roi1_prep_train_rf<- randomForest(
  Truth~., comt_03nov2022_sim_roi1_prep_train_dt, 
  mtry=round(median(comt_03nov2022_sim_roi1_prep_train_cv_res[[2]]$mtry)))

comt_03nov2022_sim_roi1_prep_train_predict<- predict(
  comt_03nov2022_sim_roi1_prep_train_rf, comt_03nov2022_sim_roi1_prep_test_dt, type="prob")

comt_03nov2022_sim_roi1_prep_train_predict_labels<- factor(
  comt_03nov2022_sim_roi1_prep_train_predict[,2]>=prob_cutoff, levels=cont_table_factor_levels)

comt_03nov2022_sim_roi1_prep_test_contigency_table<- create_cont_table(
  comt_03nov2022_sim_roi1_prep_test_dt[,Truth], comt_03nov2022_sim_roi1_prep_train_predict_labels)

comt_03nov2022_sim_roi1_prep_test_acc<- get_accuracy(comt_03nov2022_sim_roi1_prep_test_contigency_table)
comt_03nov2022_sim_roi1_prep_test_recall<- get_sensitivity(comt_03nov2022_sim_roi1_prep_test_contigency_table)
comt_03nov2022_sim_roi1_prep_test_prec<- get_precision(comt_03nov2022_sim_roi1_prep_test_contigency_table)


# ROI_2 model
set.seed(9)

comt_03nov2022_sim_roi2_prep_train_cv_res<- run_nested_cv(in_dt=comt_03nov2022_sim_roi2_prep_train_dt, hyperparam_prop=0.3, kfolds=5)

comt_03nov2022_sim_roi2_prep_train_rf<- randomForest(
  Truth~., comt_03nov2022_sim_roi2_prep_train_dt, 
  mtry=round(median(comt_03nov2022_sim_roi2_prep_train_cv_res[[2]]$mtry)))

comt_03nov2022_sim_roi2_prep_train_predict<- predict(
  comt_03nov2022_sim_roi2_prep_train_rf, comt_03nov2022_sim_roi2_prep_test_dt, type="prob")

comt_03nov2022_sim_roi2_prep_train_predict_labels<- factor(
  comt_03nov2022_sim_roi2_prep_train_predict[,2]>=prob_cutoff, levels=cont_table_factor_levels)

comt_03nov2022_sim_roi2_prep_test_contigency_table<- create_cont_table(
  comt_03nov2022_sim_roi2_prep_test_dt[,Truth], comt_03nov2022_sim_roi2_prep_train_predict_labels)

comt_03nov2022_sim_roi2_prep_test_acc<- get_accuracy(comt_03nov2022_sim_roi2_prep_test_contigency_table)
comt_03nov2022_sim_roi2_prep_test_recall<- get_sensitivity(comt_03nov2022_sim_roi2_prep_test_contigency_table)
comt_03nov2022_sim_roi2_prep_test_prec<- get_precision(comt_03nov2022_sim_roi2_prep_test_contigency_table)


# First filter out unexpected variants
# Only analyze variants
# 1) in the target regions
# 2) nonsense variants that match TAG (only stop codon expected/mutagenized)

comt_all_filt_dt<- comt_all_dt[
  !Sample%in%comt_failed_libraries & POS%in%pDEST_mutagenesis_coords & !is.na(ROI) & !ALT_CODON%in%c("TAA", "TGA")]

comt_all_roi1_dt<- comt_all_filt_dt[ROI=="ROI_1"]
comt_all_roi2_dt<- comt_all_filt_dt[ROI=="ROI_2"]

comt_all_roi1_prep_dt<- prep_train_dt(
  train_dt=comt_all_roi1_dt, modeling_features=modeling_features_no_collinear)

# Train a model on the whole dataset
set.seed(9)

comt_03nov2022_sim_roi1_prep_rf<- randomForest(
  Truth~., comt_03nov2022_sim_roi1_prep_dt, 
  mtry=round(median(comt_03nov2022_sim_roi1_prep_train_cv_res[[2]]$mtry)))

comt_all_roi1_prep_predict<- predict(
  comt_03nov2022_sim_roi1_prep_rf, comt_all_roi1_prep_dt, type="prob")

comt_all_roi1_prep_predict_labels<- factor(
  comt_all_roi1_prep_predict[,2]>=prob_cutoff, 
  levels=cont_table_factor_levels)

comt_all_roi1_prep_dt[,Prediction:=comt_all_roi1_prep_predict_labels]

# Count false positive predictions for gDNA samples only
comt_all_roi1_prep_fp_counts_dt<- comt_all_roi1_prep_dt[
  Sample%in%gdna_samples & Prediction=="FALSE", .N, by=.(Sample, VAR_ID, TYPE)]

# Filter SNPs determined FP in more than half of gDNA samples
comt_all_roi1_prep_fp_counts_snp_dt<- comt_all_roi1_prep_fp_counts_dt[TYPE=="SNP" & N==1, .N, by=.(VAR_ID)]

comt_all_roi1_prep_fp_varids<- comt_all_roi1_prep_fp_counts_snp_dt[N>=4, VAR_ID]

rm(comt_all_roi1_dt)

# comt_all_roi1_prep_fp_counts_dt[VAR_ID%in%comt_all_roi1_prep_fp_varids, table(TYPE)]/nrow(comt_all_roi1_prep_fp_counts_dt)

# For some reason most of the predicted false positives are MNPs
# require refactor of TYPE level?

caf_by_predict_gg<- ggplot(
  data=comt_all_roi1_prep_dt, 
  aes(x=log10_CAF, col=Prediction))

caf_by_predict_gg+
  geom_density()

rm(caf_by_predict_gg)


# ROI_2
comt_all_roi2_prep_dt<- prep_train_dt(
  train_dt=comt_all_roi2_dt, modeling_features=modeling_features_no_collinear)

# Train a model on the whole dataset
set.seed(9)
comt_03nov2022_sim_roi2_prep_rf<- randomForest(
  Truth~., comt_03nov2022_sim_roi2_prep_dt, 
  mtry=round(median(comt_03nov2022_sim_roi2_prep_train_cv_res[[2]]$mtry)))

comt_all_roi2_prep_predict<- predict(
  comt_03nov2022_sim_roi2_prep_rf, comt_all_roi2_prep_dt, type="prob")

comt_all_roi2_prep_predict_labels<- factor(
  comt_all_roi2_prep_predict[,2]>=prob_cutoff, 
  levels=cont_table_factor_levels)

comt_all_roi2_prep_dt[,Prediction:=comt_all_roi2_prep_predict_labels]

# Count false positive predictions
comt_all_roi2_prep_fp_counts_dt<- comt_all_roi2_prep_dt[
  Sample%in%gdna_samples & Prediction=="FALSE", .N, by=.(Sample, VAR_ID, TYPE)]

comt_all_roi2_prep_fp_counts_snp_dt<- comt_all_roi2_prep_fp_counts_dt[TYPE=="SNP" & N==1, .N, by=.(VAR_ID)]

comt_all_roi2_prep_fp_varids<- comt_all_roi2_prep_fp_counts_snp_dt[N>=2, VAR_ID]

rm(comt_all_roi2_dt)

#comt_all_roi2_prep_fp_counts_dt[VAR_ID%in%comt_all_roi2_prep_fp_varids, table(TYPE)]/nrow(comt_all_roi2_prep_fp_counts_dt)

caf_by_predict_gg<- ggplot(
  data=comt_all_roi2_prep_dt, 
  aes(x=log10_CAF, col=Prediction))

caf_by_predict_gg+
  geom_density()

rm(caf_by_predict_gg)

comt_all_roi1_2_prep_fp_varids<- c(comt_all_roi1_prep_fp_varids, comt_all_roi2_prep_fp_varids)

# Next filter variants for final analysis
# Only analyze variants
# 1) With frequency higher than the min
# 2) For SNPs, more than half of gDNA samples with True RF predictions

comt_all_filt_dt<- comt_all_filt_dt[log10(CAF)>log10_caf_thresh & !VAR_ID%in%comt_all_roi1_2_prep_fp_varids]

#rm(comt_all_dt)

# Adjust the position of the amino acid sequence to account for the N-terminal
# Flag tag
update_aa_pos<- function(aa_pos, offset){
  aa_pos_split<- unlist(strsplit(aa_pos, ","))
  aa_pos_split_offset<- as.integer(aa_pos_split) + offset
  aa_pos_offset<- paste(aa_pos_split_offset, collapse=",")
  return(aa_pos_offset)
}

update_aa_change_multi<- function(aa_pos, ref_aa, alt_aa){
  aa_pos_split<- unlist(strsplit(aa_pos, ","))
  ref_aa_split<- unlist(strsplit(ref_aa, ","))
  alt_aa_split<- unlist(strsplit(alt_aa, ","))
  aa_change<- paste("p.", ref_aa_split, aa_pos_split, alt_aa_split, sep="", collapse=",")
  return(aa_change)
}

comt_all_filt_dt[,AA_POS:=update_aa_pos(AA_POS, offset=-8), by=seq_len(nrow(comt_all_filt_dt))]

comt_all_filt_dt[,AA_CHANGE:=update_aa_change_multi(AA_POS, REF_AA, ALT_AA), 
                 by=seq_len(nrow(comt_all_filt_dt))]

# Postprocess, annotate data
comt_all_filt_postprocess_dt<- postprocess_vcf_summary_dt(
  comt_all_filt_dt, key_features=c("Sample", "ROI", "Input", "Fraction", "Quadrant", "VAR_ID"))

#rm(comt_all_filt_dt)

comt_all_filt_postprocess_dt[,log_CAF:=log(CAF)]

# Determine counts in positive and negative control samples
comt_all_filt_postprocess_dt[Input%in%c("Negative_control", "Input_library"), .N, by=.(Input, TYPE)]

# Why is there such as high count for MNPs in the negative control sample? Contamination? Crosstalk?
# Plot frequency for each type in negative control
comt_all_nc_distr_gg<- ggplot(
  data=comt_all_filt_postprocess_dt[
    Input%in%c("Negative_control", "Input_library")],
  aes(x=log10_CAF, col=TYPE))

comt_all_nc_distr_gg+
  facet_wrap(ROI~Input, scales="free")+
  geom_density()+
  labs(x="log10 CAF")+
  theme_cowplot()

rm(comt_all_nc_distr_gg)

# Compute normalized log frequency by comparison to WT count at each position
# Must do this before background subtraction as it relies on CAO, which is affected by 
# library size
comt_all_filt_postprocess_wt_dt<- comt_all_filt_postprocess_dt[
  ,.(CAO_sum=sum(CAO), DP=unique(DP)), by=.(ROI, Input, Fraction, Quadrant, Sample, POS)]

comt_all_filt_postprocess_wt_dt[,WT_CAO:=DP-CAO_sum]

# Update the postprocess data.table with the WT counts (within each sample, position)
comt_all_filt_postprocess_dt[,WT_key:=paste(Sample, POS, sep=":")]
comt_all_filt_postprocess_wt_dt[,WT_key:=paste(Sample, POS, sep=":")]

comt_all_filt_postprocess_dt[,WT_CAO:=comt_all_filt_postprocess_wt_dt[
  match(comt_all_filt_postprocess_dt[,WT_key], WT_key), WT_CAO]]

comt_all_filt_postprocess_dt[,NORM_CAF:=(CAO+0.5) / (WT_CAO+0.5)]

# Re-define the subtraction function to subtract normalized frequency
subtract_af<- function(in_dt){
  
  neg_dt<- in_dt[Input=="Negative_control", .(
    CAO=median(CAO), CAF=median(CAF), NORM_CAO=median(NORM_CAO), NORM_CAF=median(NORM_CAF)), by=.(VAR_ID)]
  
  # Merge the two tables so we can easily subtract the negative frequencies via a vectorized operation
  merged_dt<- merge(in_dt[Input!="Negative_control"], neg_dt, by="VAR_ID", all.x=TRUE)
  
  # Set baseline to 0 for variants unique to a polysomal fraction
  merged_dt[is.na(CAF.y), CAF.y:=0.0]
  merged_dt[is.na(CAO.y), CAO.y:=0]
  merged_dt[is.na(NORM_CAO.y), NORM_CAO.y:=0.0]
  merged_dt[is.na(NORM_CAF.y), NORM_CAF.y:=0.0]

  # Finally subtract the frequencies and remove those variants with negative resultant frequencies
  merged_dt[,CAF:=CAF.x-CAF.y]
  merged_dt[,CAO:=CAO.x-CAO.y]
  merged_dt[,NORM_CAO:=NORM_CAO.x-NORM_CAO.y]
  merged_dt[,NORM_CAF:=NORM_CAF.x-NORM_CAF.y]
  filtered_dt<- merged_dt[NORM_CAF>0,]
  filtered_dt[,log10_CAF:=log10(NORM_CAF)]
  filtered_dt[,log_CAF:=log(NORM_CAF)]
  
  return(filtered_dt)
}

# Results indicate MNPs are very low frequency (< 1 in 1 M)
# Could be either ambient contamination or barcode crosstalk
comt_all_filt_postprocess_subtract_roi1_dt<- subtract_af(comt_all_filt_postprocess_dt[ROI=="ROI_1"])
comt_all_filt_postprocess_subtract_roi2_dt<- subtract_af(comt_all_filt_postprocess_dt[ROI=="ROI_2"])

comt_all_filt_postprocess_subtract_dt<- rbind(
  comt_all_filt_postprocess_subtract_roi1_dt, comt_all_filt_postprocess_subtract_roi2_dt)

rm(list=c("comt_all_filt_postprocess_subtract_roi1_dt", "comt_all_filt_postprocess_subtract_roi2_dt"))

comt_all_filt_postprocess_subtract_dt[,log_NORM_CAF:=log(NORM_CAF)]
comt_all_filt_postprocess_subtract_dt[,log10_NORM_CAF:=log10(NORM_CAF)]

# Define several keys for analysis
comt_all_filt_postprocess_subtract_dt[
  ,Input_clean:=sapply(strsplit(as.character(Input), "_"), "[", 1)]

comt_all_filt_postprocess_subtract_dt[,Input_clean:=factor(
  Input_clean, levels=c("Input", "gDNA", "cDNA", "Poly", "FACS"))]

comt_all_filt_postprocess_subtract_dt[,Key:=paste(Input_clean, Fraction, sep=".")]
comt_all_filt_postprocess_subtract_dt[Input_clean=="FACS", Key:=paste(Input_clean, Quadrant, sep=".")]

comt_all_filt_postprocess_subtract_dt[,ID_SHORT:=paste(POS, REF, ALT, AA_CHANGE, sep=":")]

comt_all_filt_postprocess_subtract_dt[,Sample_num:=sapply(
  strsplit(as.character(Sample), "_"), "[" , 2)]

comt_all_filt_postprocess_subtract_dt[,Key_verbose:=paste(Input, Fraction, Sample_num, sep=":")]

comt_all_filt_postprocess_subtract_dt[
  Input_clean=="FACS", Key_verbose:=paste(Input, Quadrant, Sample_num, sep=":")]

comt_all_filt_postprocess_subtract_dt[,Key_bio:=paste(Input, Fraction, sep="_")]
comt_all_filt_postprocess_subtract_dt[Input_clean=="FACS", Key_bio:=paste(Input, Quadrant, sep="_")]

comt_all_filt_postprocess_subtract_clean_dt<- 
  comt_all_filt_postprocess_subtract_dt[,-c(
    "#CHROM", "ID", "FILTER", "QUAL", "R1_MINUS_AO", "R2_PLUS_AO", "R1_MINUS_MED_RP", "R2_PLUS_MED_RP",
    "R1_MINUS_MED_BQ", "R2_PLUS_MED_BQ", "R1_MINUS_MED_NM", "R2_PLUS_MED_NM", "WT_key",
    "CAF.x", "CAF.y", "NORM_CAF.x", "NORM_CAF.y", "CAO.x", "CAO.y", "NORM_CAO.x", "NORM_CAO.y")]

rm(comt_all_filt_postprocess_subtract_dt)

comt_all_filt_postprocess_subtract_cast_dt<- dcast(
  data=comt_all_filt_postprocess_subtract_clean_dt, 
  formula=VAR_ID~Key_verbose, value.var="log_NORM_CAF")

write.table(comt_all_filt_postprocess_subtract_cast_dt, 
            paste(comt_res_base_dir, "COMT_log_freq.txt", sep="/"),
            quote=FALSE, row.names=FALSE, sep="\t")


# Cast for comparison between single samples
comt_rep_cast_dt<- dcast(
  data=comt_all_filt_postprocess_subtract_clean_dt, formula=ROI+VAR_ID~Sample, value.var="log10_NORM_CAF")

# Remove FACS total input libraries as many were bottlenecked; also remove WT libraries from second polysome batch
facs_total_samples_drop<- c("COMT_120_R", "COMT_130_R", "COMT_146_R", "COMT_145_R", "COMT_47_R", "COMT_57_R", "COMT_173_R", "COMT_174_R")

# Finally add in polysome libraries from set 1 (first experiment) which used a different metafraction pooling strategy
comt_libraries_drop<- c(comt_failed_libraries, facs_total_samples_drop, poly_all_metafrac_set1_samples)

comt_all_filt_postprocess_subtract_drop_dt<- comt_all_filt_postprocess_subtract_clean_dt[!Sample%in%comt_libraries_drop]

rm(comt_all_filt_postprocess_subtract_clean_dt)

# Run batch correction using ComBat
# Batch-correct ROI_1 and ROI_2 independently, as we do not want low-frequency variants in the opposite region
# to influence the estimation of variability (any ROI_2 variant found in ROI_1 libraries is likely from 
# trace contamination or barcode crosstalk on the sequencer)

# Drop variants that are not observed across all replicates
row_is_complete<- function(x){sum(is.na(x))==0}

get_batch<- function(sample_name, batch_groups){
  match_logical<- sapply(batch_groups, is_in, sample_name)
  res<- names(batch_groups)[match_logical]
  return(res)
}

adjust_batch_effects<- function(cast_dt, annot_dt, batch_groups, annot_drop_ids=annot_drop_ids, 
                                factor_levels=c("gDNA", "cDNA", "Poly", "FACS"), 
                                no_correction=FALSE, no_covariates=FALSE){
  
  # Create matrix for batch correction
  cast_mat<- as.matrix(cast_dt[,-annot_drop_ids, with=FALSE], rownames.value=cast_dt[,VAR_ID])
  
  # Filter for variants observed in all replicates/batches
  cast_mat_comp<- cast_mat[which(apply(cast_mat, 1, row_is_complete)),]
  
  if(!no_covariates){
    
    # Get names of the input matrix to generate the model matrix
    cast_mat_comp_names<- sapply(strsplit(as.character(colnames(cast_mat_comp)), "_"), "[", 1)

    # Use minimal input type factor levels
    factor_levels_minimal<- factor_levels[factor_levels%in%unique(cast_mat_comp_names)]
    cast_mat_comp_names<- factor(cast_mat_comp_names, levels=factor_levels_minimal)

    model_matrix<- model.matrix(~cast_mat_comp_names)
  }

  if(!no_correction){
    batch_labels<- sapply(colnames(cast_mat_comp), get_batch, batch_groups)
    
    if(!no_covariates){
      cast_mat_comp_combat<- ComBat(dat=cast_mat_comp, batch=as.factor(batch_labels), mod=model_matrix)
    } else {
      cast_mat_comp_combat<- ComBat(dat=cast_mat_comp, batch=as.factor(batch_labels))
    }
    
  } else {
    cast_mat_comp_combat<- cast_mat_comp
  }
  
  # Convert to dt and annotate with VAR_ID
  cast_mat_comp_combat_dt<- as.data.table(
    cast_mat_comp_combat, keep.rownames=T)
  
  names(cast_mat_comp_combat_dt)[1]<- "VAR_ID"
  
  # Convert back to long format and summarize biological replicates
  cast_mat_comp_combat_melt_dt<- melt(
    data=cast_mat_comp_combat_dt, 
    id.vars="VAR_ID", variable.name="Key_verbose", value.name="log_NORM_CAF")
  
  cast_mat_comp_combat_melt_merge_dt<- merge(
    cast_mat_comp_combat_melt_dt, annot_dt, by="VAR_ID")
  
  # Re-generate Input, Key, and Key_bio columns
  cast_mat_comp_combat_melt_merge_dt[
    ,Input:=as.factor(sapply(strsplit(as.character(Key_verbose), "_"), "[", 1))]
  
  cast_mat_comp_combat_melt_merge_dt[
    ,Key:=as.factor(sapply(strsplit(as.character(Key_verbose), ":"), "[", 1))]
  
  fraction_quadrants<- cast_mat_comp_combat_melt_merge_dt[,sapply(strsplit(as.character(Key_verbose), ":"), "[", 2)]
    
  cast_mat_comp_combat_melt_merge_dt[
    ,Key_bio:=as.factor(paste(Key, fraction_quadrants, sep="_"))]
  
  cast_mat_comp_combat_melt_merge_dt[,log10_NORM_CAF:=log10(exp(log_NORM_CAF))]
  
  # Summarize technical replicates for each biological replicate
  cast_mat_comp_combat_melt_merge_summ_dt<- 
    cast_mat_comp_combat_melt_merge_dt[, .(
        Median_log10_CAF=median(log10_NORM_CAF, na.rm=T), 
        Median_log_CAF=median(log_NORM_CAF, na.rm=T)), 
      by=.(Input, Key, Key_bio, VAR_ID, POS, REF, ALT, REF_CODON, ALT_CODON,
           AA_POS, AA_POS2, REF_AA, ALT_AA, AA_CHANGE, VAR_TYPE, TYPE, ID_SHORT)]
  
  return(cast_mat_comp_combat_melt_merge_summ_dt)
}

annot_ids<- c("ROI", "VAR_ID", "POS", "REF", "ALT", "REF_CODON", "ALT_CODON", "AA_POS", "AA_POS2", 
              "REF_AA", "ALT_AA", "AA_CHANGE", "VAR_TYPE", "TYPE", "ID_SHORT")

# Get variant annotations
annot_dt<- unique(comt_all_filt_postprocess_subtract_drop_dt[,annot_ids, with=FALSE])

# Drop these annotation fields from the expression matrix
annot_drop_ids<- c("VAR_ID")

# Cast from long to wide format for each ROI
comt_all_filt_postprocess_subtract_drop_cast_cdna_gdna_roi1_dt<- dcast(
  data=comt_all_filt_postprocess_subtract_drop_dt[
    ROI=="ROI_1" & Key%in%c("cDNA.","gDNA.")],
  formula=VAR_ID~Key_verbose, value.var="log_NORM_CAF")

comt_all_filt_postprocess_subtract_drop_cast_cdna_gdna_roi2_dt<- dcast(
  data=comt_all_filt_postprocess_subtract_drop_dt[
    ROI=="ROI_2" & Key%in%c("cDNA.","gDNA.")],
  formula=VAR_ID~Key_verbose, value.var="log_NORM_CAF")

# Define the batches for gDNA-cDNA
# Batch defined by growth and time of nucleic acid extraction
cdna_gdna_batch_groups_growth<- list("c1"=c("cDNA_bio1::1", "cDNA_bio1::2", "cDNA_bio1::3", "cDNA_bio2::4", "cDNA_bio2::5", "cDNA_bio2::6"), 
                                     "c4"=c("cDNA_bio1::94", "cDNA_bio2::95", "cDNA_bio1::98", "cDNA_bio2::99"), 
                                     "c5"=c("cDNA_bio3::96", "cDNA_bio4::97", "cDNA_bio3::100", "cDNA_bio4::101", "cDNA_bio1::102", "cDNA_bio2::103"), 
                                     "c6"=c("cDNA_bio3::104", "cDNA_bio4::105", "cDNA_bio3::106", "cDNA_bio4::107"),
                                     "g1"=c("gDNA_bio1::37", "gDNA_bio2::38", "gDNA_bio2::39", "gDNA_bio1::112", "gDNA_bio2::113"),
                                     "g2"=c("gDNA_bio3::114", "gDNA_bio4::115", "gDNA_bio1::116", "gDNA_bio2::117", "gDNA_bio3::118", "gDNA_bio4::119"))

# Library prep and sequencing batch
cdna_gdna_batch_groups_prep<- list("b1"=c("cDNA_bio1::1", "cDNA_bio1::2", "cDNA_bio1::3", "cDNA_bio2::4", "cDNA_bio2::5", 
                                          "cDNA_bio2::6", "gDNA_bio1::37", "gDNA_bio2::38", "gDNA_bio2::39"), 
                                   "b2"=c("cDNA_bio1::94", "cDNA_bio2::95", "cDNA_bio3::96", "cDNA_bio4::97", "cDNA_bio1::98", "cDNA_bio2::99", 
                                          "cDNA_bio3::100", "cDNA_bio4::101", "cDNA_bio1::102", "cDNA_bio2::103", "cDNA_bio3::104", "cDNA_bio4::105",
                                          "cDNA_bio3::106", "cDNA_bio4::107", "gDNA_bio1::112", "gDNA_bio2::113", "gDNA_bio3::114", "gDNA_bio4::115", 
                                          "gDNA_bio1::116", "gDNA_bio2::117", "gDNA_bio3::118", "gDNA_bio4::119"))

# Just use library prep batch as batches become quite small if intersecting extraction x prep batch
comt_all_filt_postprocess_subtract_drop_cast_cdna_gdna_roi1_combat_dt<- adjust_batch_effects(cast_dt=comt_all_filt_postprocess_subtract_drop_cast_cdna_gdna_roi1_dt, annot_dt=annot_dt, 
                       batch_groups=cdna_gdna_batch_groups_prep, annot_drop_ids=annot_drop_ids)

# For ROI_2, no batch correction needed as all gDNA, cDNA samples in same batch
# However, apply the same internal filtering as for ROI_1
comt_all_filt_postprocess_subtract_drop_cast_cdna_gdna_roi2_combat_dt<- adjust_batch_effects(cast_dt=comt_all_filt_postprocess_subtract_drop_cast_cdna_gdna_roi2_dt, annot_dt=annot_dt, 
                       batch_groups=cdna_gdna_batch_groups_prep, annot_drop_ids=annot_drop_ids, no_correction=TRUE)

# Finally bind ROI_1 and ROI_2 back together
comt_all_filt_postprocess_subtract_drop_cast_cdna_gdna_combat_dt<- rbind(
  comt_all_filt_postprocess_subtract_drop_cast_cdna_gdna_roi1_combat_dt,
  comt_all_filt_postprocess_subtract_drop_cast_cdna_gdna_roi2_combat_dt)

# Re-annotate with the ROI
comt_all_filt_postprocess_subtract_drop_cast_cdna_gdna_combat_dt[
  ,ROI:=as.factor(rep(c("ROI_1", "ROI_2"), times=c(
    nrow(comt_all_filt_postprocess_subtract_drop_cast_cdna_gdna_roi1_combat_dt),
      nrow(comt_all_filt_postprocess_subtract_drop_cast_cdna_gdna_roi2_combat_dt))))]

rm(list=c("comt_all_filt_postprocess_subtract_drop_cast_cdna_gdna_roi1_combat_dt",
          "comt_all_filt_postprocess_subtract_drop_cast_cdna_gdna_roi2_combat_dt",
          "comt_all_filt_postprocess_subtract_drop_cast_cdna_gdna_roi1_dt",
          "comt_all_filt_postprocess_subtract_drop_cast_cdna_gdna_roi2_dt"))

# Batch correct polysomal samples
# Here we do not provide a model matrix as the RNA samples to be compared belonged to only one batch
# pass no_covariates=TRUE to adjust_batch_effects()

# Need to batch correction fractions independently as many variants are not observed in 
# fractions 1,2; otherwise these variants drop out during check for complete observations
# across samples

comt_all_filt_postprocess_subtract_drop_cast_poly_roi1_dt<- dcast(
  data=comt_all_filt_postprocess_subtract_drop_dt[
    ROI=="ROI_1" & Input_clean=="Poly" & Fraction%in%c(3,4)],
  formula=VAR_ID~Key_verbose, value.var="log_NORM_CAF")

comt_all_filt_postprocess_subtract_drop_cast_poly_roi2_dt<- dcast(
  data=comt_all_filt_postprocess_subtract_drop_dt[
    ROI=="ROI_2" & Input_clean=="Poly" & Fraction%in%c(2,3,4)],
  formula=VAR_ID~Key_verbose, value.var="log_NORM_CAF")

# Polysome library prep and sequencing batch
poly_batch_groups_prep<- list("b1"=c("Poly_bio1:1:62", "Poly_bio1:2:63", "Poly_bio1:3:64", "Poly_bio1:4:65", "Poly_bio2:1:66", 
                                     "Poly_bio2:2:67", "Poly_bio2:3:68", "Poly_bio2:4:69", "Poly_bio3:1:86", "Poly_bio3:2:87", 
                                     "Poly_bio3:3:88", "Poly_bio3:4:89", "Poly_bio4:1:90", "Poly_bio4:2:91", "Poly_bio4:3:92", "Poly_bio4:4:93"),
                              "b2"=c("Poly_bio3:1:149", "Poly_bio3:2:150", "Poly_bio3:3:151", "Poly_bio3:4:152", "Poly_bio4:1:153",
                                     "Poly_bio4:2:154", "Poly_bio4:3:155", "Poly_bio4:4:156", "Poly_bio1:1:157", "Poly_bio1:2:158",
                                     "Poly_bio1:3:159", "Poly_bio1:4:160", "Poly_bio2:1:161", "Poly_bio2:2:162", "Poly_bio2:3:163",
                                     "Poly_bio2:4:164", "Poly_bio1:1:165", "Poly_bio1:2:166", "Poly_bio1:3:167", "Poly_bio1:4:168", 
                                     "Poly_bio2:1:169", "Poly_bio2:2:170", "Poly_bio2:3:171", "Poly_bio2:4:172"))

comt_all_filt_postprocess_subtract_drop_cast_poly_roi1_combat_dt<- 
  adjust_batch_effects(cast_dt=comt_all_filt_postprocess_subtract_drop_cast_poly_roi1_dt, annot_dt=annot_dt, 
                       batch_groups=poly_batch_groups_prep, annot_drop_ids=annot_drop_ids, no_covariates=TRUE)

comt_all_filt_postprocess_subtract_drop_cast_poly_roi2_combat_dt<- 
  adjust_batch_effects(cast_dt=comt_all_filt_postprocess_subtract_drop_cast_poly_roi2_dt, annot_dt=annot_dt, 
                       batch_groups=poly_batch_groups_prep, annot_drop_ids=annot_drop_ids, no_covariates=TRUE)

# Finally bind ROI_1 and ROI_2 back together
comt_all_filt_postprocess_subtract_drop_cast_poly_combat_dt<- rbind(
  comt_all_filt_postprocess_subtract_drop_cast_poly_roi1_combat_dt,
  comt_all_filt_postprocess_subtract_drop_cast_poly_roi2_combat_dt)

# Re-annotate with the ROI
comt_all_filt_postprocess_subtract_drop_cast_poly_combat_dt[
  ,ROI:=as.factor(rep(c("ROI_1", "ROI_2"), times=c(
    nrow(comt_all_filt_postprocess_subtract_drop_cast_poly_roi1_combat_dt),
      nrow(comt_all_filt_postprocess_subtract_drop_cast_poly_roi2_combat_dt))))]

rm(list=c("comt_all_filt_postprocess_subtract_drop_cast_poly_roi1_combat_dt", 
          "comt_all_filt_postprocess_subtract_drop_cast_poly_roi2_combat_dt"))


# Batch correct FACS samples
# Here we do not provide a model matrix as the gDNA and FACS quadrants to be compared were different batches
# pass no_covariates=TRUE to adjust_batch_effects()

# Bio replicate, growth, and sorting batch
facs_batch_groups_sort<- list("B"=c("FACS_bio1:Input:42", "FACS_bio1:1:43", "FACS_bio1:2:44", "FACS_bio1:3:45", "FACS_bio1:4:46", 
                                     "FACS_bio1:Input:47", "FACS_bio1:1:48", "FACS_bio1:2:49", "FACS_bio1:3:50", "FACS_bio1:4:51",
                                     "FACS_bio2:2:140", "FACS_bio2:1:141", "FACS_bio2:2:142", "FACS_bio2:3:143", "FACS_bio2:4:144"),
                              "A"=c("FACS_bio2:Input:52", "FACS_bio2:1:53", "FACS_bio2:2:54", "FACS_bio2:3:55", "FACS_bio2:4:56", 
                                     "FACS_bio2:Input:57", "FACS_bio2:1:58", "FACS_bio2:2:59", "FACS_bio2:3:60", "FACS_bio2:4:61",
                                    "FACS_bio1:Input:145"),
                              "C"=c("FACS_bio3:Input:120", "FACS_bio3:1:121", "FACS_bio3:2:122", "FACS_bio3:3:123", "FACS_bio3:4:124",
                                     "FACS_bio3:Input:125", "FACS_bio3:1:126", "FACS_bio3:2:127", "FACS_bio3:3:128", "FACS_bio3:4:129",
                                     "FACS_bio3:Input:146"),
                              "D"=c("FACS_bio4:Input:130", "FACS_bio4:1:131", "FACS_bio4:2:132", "FACS_bio4:3:133", "FACS_bio4:4:134", 
                                     "FACS_bio4:Input:135", "FACS_bio4:1:136", "FACS_bio4:2:137", "FACS_bio4:3:138", "FACS_bio4:4:139"))

# Library prep and sequencing batch
facs_batch_groups_prep<- list("b1"=c("FACS_bio1:1:43", "FACS_bio1:1:48", "FACS_bio1:2:44", "FACS_bio1:2:49", "FACS_bio1:3:45", 
                                     "FACS_bio1:3:50", "FACS_bio1:4:46", "FACS_bio1:4:51",  "FACS_bio1:Input:42", "FACS_bio1:Input:47", 
                                     "FACS_bio2:1:53", "FACS_bio2:1:58", "FACS_bio2:2:54", "FACS_bio2:2:59", "FACS_bio2:3:55", 
                                     "FACS_bio2:3:60", "FACS_bio2:4:56", "FACS_bio2:4:61", "FACS_bio2:Input:52", "FACS_bio2:Input:57"),
                              "b2"=c("FACS_bio3:1:121", "FACS_bio3:1:126", "FACS_bio3:2:122", "FACS_bio3:2:127", "FACS_bio3:3:123", 
                                     "FACS_bio3:3:128", "FACS_bio3:4:124", "FACS_bio3:4:129", "FACS_bio3:Input:120", "FACS_bio3:Input:125",
                                     "FACS_bio3:Input:146", "FACS_bio4:1:131", "FACS_bio4:1:136", "FACS_bio4:2:132", "FACS_bio4:2:137", 
                                     "FACS_bio4:Input:130",  "FACS_bio2:1:141", "FACS_bio2:3:143", "FACS_bio1:Input:145"),
                              "b3"=c("FACS_bio4:3:133", "FACS_bio4:4:134", "FACS_bio4:Input:135", "FACS_bio4:3:138", "FACS_bio4:4:139",
                                     "FACS_bio2:2:140", "FACS_bio2:2:142", "FACS_bio2:4:144"))

# Use the library prep/sequencing batch for correction as the sorting batches are different biological replicates
# and we want to preserve biological variation

comt_all_filt_postprocess_subtract_drop_cast_facs_roi1_q1_q2_q3_dt<- dcast(
  data=comt_all_filt_postprocess_subtract_drop_dt[
    ROI=="ROI_1" & Input_clean=="FACS" & Quadrant%in%c(1,2,3)],
  formula=VAR_ID~Key_verbose, value.var="log_NORM_CAF")

comt_all_filt_postprocess_subtract_drop_cast_facs_roi1_q4_dt<- dcast(
  data=comt_all_filt_postprocess_subtract_drop_dt[
    ROI=="ROI_1" & Input_clean=="FACS" & Quadrant==4],
  formula=VAR_ID~Key_verbose, value.var="log_NORM_CAF")

comt_all_filt_postprocess_subtract_drop_cast_facs_roi2_q1_q2_q3_dt<- dcast(
  data=comt_all_filt_postprocess_subtract_drop_dt[
    ROI=="ROI_2" & Input_clean=="FACS" & Quadrant%in%c(1,2,3)],
  formula=VAR_ID~Key_verbose, value.var="log_NORM_CAF")

comt_all_filt_postprocess_subtract_drop_cast_facs_roi2_q4_dt<- dcast(
  data=comt_all_filt_postprocess_subtract_drop_dt[
    ROI=="ROI_2" & Input_clean=="FACS" & Quadrant==4],
  formula=VAR_ID~Key_verbose, value.var="log_NORM_CAF")

comt_all_filt_postprocess_subtract_drop_cast_facs_roi1_q1_q2_q3_combat_dt<- 
  adjust_batch_effects(cast_dt=comt_all_filt_postprocess_subtract_drop_cast_facs_roi1_q1_q2_q3_dt, annot_dt=annot_dt, 
                       batch_groups=facs_batch_groups_prep, annot_drop_ids=annot_drop_ids, no_covariates=TRUE)

comt_all_filt_postprocess_subtract_drop_cast_facs_roi1_q4_combat_dt<- 
  adjust_batch_effects(cast_dt=comt_all_filt_postprocess_subtract_drop_cast_facs_roi1_q4_dt, annot_dt=annot_dt, 
                       batch_groups=facs_batch_groups_prep, annot_drop_ids=annot_drop_ids, no_covariates=TRUE)

comt_all_filt_postprocess_subtract_drop_cast_facs_roi2_q1_q2_q3_combat_dt<- 
  adjust_batch_effects(cast_dt=comt_all_filt_postprocess_subtract_drop_cast_facs_roi2_q1_q2_q3_dt, annot_dt=annot_dt, 
                       batch_groups=facs_batch_groups_prep, annot_drop_ids=annot_drop_ids, no_covariates=TRUE)

comt_all_filt_postprocess_subtract_drop_cast_facs_roi2_q4_combat_dt<- 
  adjust_batch_effects(cast_dt=comt_all_filt_postprocess_subtract_drop_cast_facs_roi2_q4_dt, annot_dt=annot_dt, 
                       batch_groups=facs_batch_groups_prep, annot_drop_ids=annot_drop_ids, no_covariates=TRUE)

comt_all_filt_postprocess_subtract_drop_cast_facs_combat_dt<- rbind(
  comt_all_filt_postprocess_subtract_drop_cast_facs_roi1_q1_q2_q3_combat_dt,
  comt_all_filt_postprocess_subtract_drop_cast_facs_roi1_q4_combat_dt,
  comt_all_filt_postprocess_subtract_drop_cast_facs_roi2_q1_q2_q3_combat_dt,
  comt_all_filt_postprocess_subtract_drop_cast_facs_roi2_q4_combat_dt)

facs_roi1_nrow<- sum(nrow(comt_all_filt_postprocess_subtract_drop_cast_facs_roi1_q1_q2_q3_combat_dt),
                     nrow(comt_all_filt_postprocess_subtract_drop_cast_facs_roi1_q4_combat_dt))

facs_roi2_nrow<- sum(nrow(comt_all_filt_postprocess_subtract_drop_cast_facs_roi2_q1_q2_q3_combat_dt),
                     nrow(comt_all_filt_postprocess_subtract_drop_cast_facs_roi2_q4_combat_dt))

comt_all_filt_postprocess_subtract_drop_cast_facs_combat_dt[
  ,ROI:=as.factor(rep(c("ROI_1", "ROI_2"), times=c(facs_roi1_nrow, facs_roi2_nrow)))]

rm(list=c("comt_all_filt_postprocess_subtract_drop_cast_facs_roi1_q1_q2_q3_combat_dt",
          "comt_all_filt_postprocess_subtract_drop_cast_facs_roi1_q4_combat_dt",
          "comt_all_filt_postprocess_subtract_drop_cast_facs_roi2_q1_q2_q3_combat_dt",
          "comt_all_filt_postprocess_subtract_drop_cast_facs_roi2_q4_combat_dt", 
          "comt_all_filt_postprocess_subtract_drop_cast_facs_roi1_q1_q2_q3_dt", 
          "comt_all_filt_postprocess_subtract_drop_cast_facs_roi1_q4_dt",
          "comt_all_filt_postprocess_subtract_drop_cast_facs_roi2_q1_q2_q3_dt",
          "comt_all_filt_postprocess_subtract_drop_cast_facs_roi2_q4_dt"))

# Add all batch-corrected data back together in a single dt
comt_all_filt_postprocess_subtract_drop_cast_combat_dt<- rbind(
  comt_all_filt_postprocess_subtract_drop_cast_cdna_gdna_combat_dt,
  comt_all_filt_postprocess_subtract_drop_cast_poly_combat_dt,
  comt_all_filt_postprocess_subtract_drop_cast_facs_combat_dt)

# Add back fraction/quadrant columns
comt_all_filt_postprocess_subtract_drop_cast_combat_dt[,Fraction:=""]
comt_all_filt_postprocess_subtract_drop_cast_combat_dt[,Quadrant:=""]

comt_all_filt_postprocess_subtract_drop_cast_combat_dt[
  grepl("Poly", Key), Fraction:=sapply(strsplit(as.character(Key_bio), "_"), "[", 3)]

comt_all_filt_postprocess_subtract_drop_cast_combat_dt[
  grepl("FACS", Key), Quadrant:=sapply(strsplit(as.character(Key_bio), "_"), "[", 3)]

# Summarize biological replicates

# On strategy is using weighted mean based on achieved depth of coverage
# Rationale is to down-weight samples with low depth
comt_all_filt_postprocess_wt_dt[,DP_key:=paste(ROI, Input, Fraction, sep="_")]
comt_all_filt_postprocess_wt_dt[grepl("FACS", Input), DP_key:=paste(ROI, Input, Quadrant, sep="_")]
comt_all_filt_postprocess_wt_bio_dt<- comt_all_filt_postprocess_wt_dt[,.(DP=median(DP)), by=.(DP_key)]

# Annotate the median DP across target regions
comt_all_filt_postprocess_subtract_drop_cast_combat_dt[,DP_key:=paste(ROI, Key_bio, sep="_")]

comt_all_filt_postprocess_subtract_drop_cast_combat_dt<- merge(
  comt_all_filt_postprocess_subtract_drop_cast_combat_dt, 
  comt_all_filt_postprocess_wt_bio_dt, by="DP_key")

# Favor median instead of weighted mean

# Another strategy is just taking the median
# Comparison of different input types suggests this slightly improves correlation
# compared to the weighted mean
comt_all_filt_postprocess_subtract_drop_cast_combat_bio_dt<- 
    comt_all_filt_postprocess_subtract_drop_cast_combat_dt[
      ,.(log_NORM_CAF=median(Median_log_CAF), 
         log10_NORM_CAF=median(Median_log10_CAF)),
      by=.(ROI, Input, Fraction, Quadrant, TYPE, VAR_TYPE, VAR_ID, POS, REF, ALT, 
           REF_CODON, ALT_CODON, AA_POS, AA_POS2, REF_AA, ALT_AA, AA_CHANGE, ID_SHORT)]

comt_all_filt_postprocess_subtract_drop_cast_combat_bio_dt[,Key:=paste(Input, Fraction, sep=".")]
comt_all_filt_postprocess_subtract_drop_cast_combat_bio_dt[Input=="FACS", Key:=paste(Input, Quadrant, sep=".")]

#rm(list=c("comt_all_filt_postprocess_wt_dt", "comt_all_filt_postprocess_wt_bio_dt"))

```


```{r Coverage of each codon and amino acid}

# Amino acids changes
comt_aa_cov_dt<- comt_all_filt_postprocess_subtract_drop_dt[
  !grepl(",", AA_POS), .(N=length(unique(AA_CHANGE))), by=.(Input, AA_POS)]

comt_aa_cov_dt[,AA_POS:=as.integer(AA_POS)]

comt_aa_cov_gg<- ggplot(
  data=comt_aa_cov_dt[Input!="Input_library"], 
  aes(x=AA_POS, y=N))

comt_aa_cov_gg+
  facet_wrap(~Input)+
  geom_point()+
  geom_line()+
  expand_limits(y=0)+
  labs(x="Amino acid position", 
       y="Number amino acid changes")+
  theme_cowplot()

# Codon changes
comt_codon_cov_dt<- comt_all_filt_postprocess_subtract_drop_dt[
  !grepl(",", AA_POS), .(N=length(unique(ALT_CODON))), by=.(Input, AA_POS)]

comt_codon_cov_dt[,AA_POS:=as.integer(AA_POS)]

comt_codon_cov_gg<- ggplot(
  data=comt_codon_cov_dt[Input!="Input_library"], 
  aes(x=AA_POS, y=N))

comt_codon_cov_gg+
  facet_wrap(~Input)+
  geom_hline(yintercept=62, col="grey", linetype=2)+
  geom_point()+
  geom_line()+
  expand_limits(y=0)+
  labs(x="Amino acid position", 
       y="Number codon changes")+
  theme_cowplot()

rm(list=c("comt_aa_cov_gg", "comt_codon_cov_gg"))

```


```{r Variant frequency distr}

# gDNA, cDNA CAF
comt_rep_summ_freq_gg<- ggplot(
  data=comt_all_filt_postprocess_subtract_drop_cast_combat_dt[grepl("gDNA", Input) | grepl("cDNA", Input)], 
  aes(x=Input, y=Median_log10_CAF))

comt_rep_summ_freq_gg+
  geom_boxplot(position="dodge", outlier.shape=NA, varwidth=T)+
  facet_wrap(ROI~TYPE)+
  theme_cowplot()+
  labs(y="Median log10 frequency")


# cDNA, poly CAF
comt_rep_summ_freq_gg<- ggplot(
  data=comt_all_filt_postprocess_subtract_drop_cast_combat_dt[grepl("Poly", Input) | grepl("cDNA", Input)], 
  aes(x=Input, y=Median_log10_CAF, fill=Fraction))

comt_rep_summ_freq_gg+
  geom_boxplot(position="dodge", outlier.shape=NA, varwidth=T)+
  facet_wrap(ROI~TYPE)+
  theme_cowplot()+
  labs(y="Median log10 frequency")

```


## Replicate reproducibility

Use scatterplots and hierarchical clustering to determine reproducibility of gDNA, cDNA, and polysome libraries.

```{r COMT replicate correlation pairwise}

# Transform into IQLRs and plot examples
comt_rep_cast_mat<- as.matrix(comt_rep_cast_dt[,-c("ROI", "VAR_ID")])
rownames(comt_rep_cast_mat)<- comt_rep_cast_dt[,VAR_ID]
comt_rep_cast_mat_reclose<- apply(10^comt_rep_cast_mat, 2, reclose_coda)
comt_rep_cast_mat_iqlr<- apply(comt_rep_cast_mat_reclose, 2, iqlr)
comt_rep_cast_mat_iqlr_dt<- as.data.table(comt_rep_cast_mat_iqlr)

comt_rep_cast_mat_iqlr_dt[,ROI:=comt_rep_cast_dt[,ROI]]
comt_rep_cast_mat_iqlr_dt[,VAR_ID:=comt_rep_cast_dt[,VAR_ID]]

rm(comt_rep_cast_mat)
rm(comt_rep_cast_mat_reclose)
rm(comt_rep_cast_mat_iqlr)

### gDNA

# ROI_1 gDNA
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_1"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi1_samples[roi1_samples%in%gdna_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# ROI_2 gDNA
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_2"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi2_samples[roi2_samples%in%gdna_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()


### cDNA

# ROI_1 cDNA
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_1"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi1_samples[roi1_samples%in%cdna_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# ROI_2 cDNA
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_2"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi2_samples[roi2_samples%in%cdna_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()


### Poly

# ROI_1 poly fraction 1
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_1"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi1_samples[roi1_samples%in%poly_metafrac_1_set2_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# Very poor correlation

# ROI_1 poly fraction 2
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_1"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi1_samples[roi1_samples%in%poly_metafrac_2_set2_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# Still poor correlation

# ROI_1 poly fraction 3
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_1"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi1_samples[roi1_samples%in%poly_metafrac_3_set2_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# ROI_1 poly fraction 4
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_1"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi1_samples[roi1_samples%in%poly_metafrac_4_set2_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()


# ROI_2 poly fraction 1
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_2"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi2_samples[roi2_samples%in%poly_metafrac_1_set2_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# Very poor correlation

# ROI_2 poly fraction 2
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_2"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi2_samples[roi2_samples%in%poly_metafrac_2_set2_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# Poor correlation for bio reps. C, D from first batch
# Decent correlation for bio reps. A, B from second batch

# ROI_2 poly fraction 3
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_2"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi2_samples[roi2_samples%in%poly_metafrac_3_set2_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# Same case: poor correlation for C, D replicates
# But very good correlation for A, B replicates

# ROI_2 poly fraction 4
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_2"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi2_samples[roi2_samples%in%poly_metafrac_4_set2_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# Same as fractions 2-4

# Look at set 1 polysome samples which are held out from analysis for validation

# Set 1 ROI_1 poly fraction 1
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_1"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi1_samples[roi1_samples%in%poly_metafrac_1_set1_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,-2.5), y=c(-6,-2.5))+
  theme_cowplot()

# Set 1 ROI_1 poly fraction 2
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_1"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi1_samples[roi1_samples%in%poly_metafrac_2_set1_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# Set 1 ROI_1 poly fraction 3
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_1"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi1_samples[roi1_samples%in%poly_metafrac_3_set1_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# Set 1 ROI_1 poly fraction 4
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_1"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi1_samples[roi1_samples%in%poly_metafrac_4_set1_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# Set 1 ROI_1 poly fraction 5
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_1"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi1_samples[roi1_samples%in%poly_metafrac_5_set1_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()


# ROI_1 poly fraction 4 + cDNA
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_1"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi1_samples[
          roi1_samples%in%c(poly_metafrac_4_set2_samples, cdna_samples)]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()


### FACS

# ROI_1 FACS total input
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_1"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi1_samples[roi1_samples%in%facs_total_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# Noisy in general, consider using gDNA readouts instead for difference

# ROI_1 FACS quadrant 1
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_1"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi1_samples[roi1_samples%in%facs_q1_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# COMT_43 outlier

# ROI_1 FACS quadrant 2
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_1"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi1_samples[roi1_samples%in%facs_q2_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# COMT_44 outlier

# ROI_1 FACS quadrant 3
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_1"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi1_samples[roi1_samples%in%facs_q3_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# Noisy in general

# ROI_1 FACS quadrant 4
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_1"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi1_samples[roi1_samples%in%facs_q4_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# Noisy in general

# ROI_2 FACS total input
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_2"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi2_samples[roi2_samples%in%facs_total_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# COMT_57 outlier

# ROI_2 FACS quadrant 1
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_2"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi2_samples[roi2_samples%in%facs_q1_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# ROI_2 FACS quadrant 2
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_2"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi2_samples[roi2_samples%in%facs_q2_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# COMT_59 outlier

# ROI_2 FACS quadrant 3
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_2"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi2_samples[roi2_samples%in%facs_q3_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# COMT_50, COMT_60 outliers

# ROI_2 FACS quadrant 4
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_2"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi2_samples[roi2_samples%in%facs_q4_samples]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# Noisy in general

# Determine if FACS samples with poor correlation are due to insufficient sequencing depth
facs_outlier_samples<- paste("COMT", c(43, 44, 50, 57, 59, 60, 123, 133), "R", sep="_")

# Choose some well-behaved replicates from different quadrants
facs_correlated_samples<- paste("COMT", c(58, 121, 122, 126, 131, 132, 141), "R", sep="_")

comt_all_filt_postprocess_subtract_drop_dt[Sample%in%facs_outlier_samples, median(DP), by=.(Sample)]
comt_all_filt_postprocess_subtract_drop_dt[Sample%in%facs_correlated_samples, median(DP), by=.(Sample)]

# Depth may explain some poor correlation between FACS samples
# Consider using both gDNA and total samples for input
# and using weighted averaging based on depth

# Plot gDNA and input samples to determine if bottlenecking is primary 
# driver of poor rep-rep correlation
# ROI_1 FACS total input
ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_1"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi1_samples[roi1_samples%in%c(facs_total_samples, gdna_samples)]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# FACS samples that show decent correlation (~ >0.8) with gDNA are COMT_42, COMT_52
# COMT_120, COMT_130 show slightly worse correlation and COMT_146 is the worst

ggpairs(data=comt_rep_cast_mat_iqlr_dt[ROI=="ROI_2"], 
        columns=which(names(comt_rep_cast_mat_iqlr_dt)%in%roi2_samples[roi2_samples%in%c(facs_total_samples, gdna_samples)]))+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

# FACS input samples COMT_125, COMT_135 show decent correlation with gDNA replicates
# COMT_145 is slightly worse, and COMT_47, COMT_57 show the worst correlation with gDNA


# Plot examples for each input
# ROI_1
roi1_gdna_rep_cor_gg<- ggplot(
  data=comt_rep_cast_mat_iqlr_dt, 
  aes(x=COMT_114_R, y=COMT_115_R))

roi1_gdna_rep_cor_gg<- roi1_gdna_rep_cor_gg+
  geom_point(alpha=0.75)+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

comt_rep_cast_mat_iqlr_dt[,cor(x=COMT_114_R, y=COMT_115_R, use="pairwise.complete.obs")]

roi1_cdna_rep_cor_gg<- ggplot(
  data=comt_rep_cast_mat_iqlr_dt, 
  aes(x=COMT_96_R, y=COMT_97_R))

roi1_cdna_rep_cor_gg<- roi1_cdna_rep_cor_gg+
  geom_point(alpha=0.75)+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

comt_rep_cast_mat_iqlr_dt[,cor(x=COMT_96_R, y=COMT_97_R, use="pairwise.complete.obs")]

roi1_poly_f3_rep_cor_gg<- ggplot(
  data=comt_rep_cast_mat_iqlr_dt, 
  aes(x=COMT_151_R, y=COMT_155_R))

roi1_poly_f3_rep_cor_gg<- roi1_poly_f3_rep_cor_gg+
  geom_point(alpha=0.75)+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

comt_rep_cast_mat_iqlr_dt[,cor(x=COMT_151_R, y=COMT_155_R, use="pairwise.complete.obs")]

roi1_poly_f4_rep_cor_gg<- ggplot(
  data=comt_rep_cast_mat_iqlr_dt, 
  aes(x=COMT_152_R, y=COMT_156_R))

roi1_poly_f4_rep_cor_gg<- roi1_poly_f4_rep_cor_gg+
  geom_point(alpha=0.75)+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

comt_rep_cast_mat_iqlr_dt[,cor(x=COMT_152_R, y=COMT_156_R, use="pairwise.complete.obs")]

roi1_facs_p1_rep_cor_gg<- ggplot(
  data=comt_rep_cast_mat_iqlr_dt, 
  aes(x=COMT_121_R, y=COMT_131_R))

roi1_facs_p1_rep_cor_gg<- roi1_facs_p1_rep_cor_gg+
  geom_point(alpha=0.75)+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

comt_rep_cast_mat_iqlr_dt[,cor(x=COMT_121_R, y=COMT_131_R, use="pairwise.complete.obs")]

roi1_facs_p3_rep_cor_gg<- ggplot(
  data=comt_rep_cast_mat_iqlr_dt, 
  aes(x=COMT_123_R, y=COMT_133_R))

roi1_facs_p3_rep_cor_gg<- roi1_facs_p3_rep_cor_gg+
  geom_point(alpha=0.75)+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

comt_rep_cast_mat_iqlr_dt[,cor(x=COMT_123_R, y=COMT_133_R, use="pairwise.complete.obs")]

# ROI_2
roi2_gdna_rep_cor_gg<- ggplot(
  data=comt_rep_cast_mat_iqlr_dt, 
  aes(x=COMT_118_R, y=COMT_119_R))

roi2_gdna_rep_cor_gg<- roi2_gdna_rep_cor_gg+
  geom_point(alpha=0.75)+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

comt_rep_cast_mat_iqlr_dt[,cor(x=COMT_118_R, y=COMT_119_R, use="pairwise.complete.obs")]

roi2_cdna_rep_cor_gg<- ggplot(
  data=comt_rep_cast_mat_iqlr_dt, 
  aes(x=COMT_106_R, y=COMT_107_R))

roi2_cdna_rep_cor_gg<- roi2_cdna_rep_cor_gg+
  geom_point(alpha=0.75)+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

comt_rep_cast_mat_iqlr_dt[,cor(x=COMT_106_R, y=COMT_107_R, use="pairwise.complete.obs")]

roi2_poly_f3_rep_cor_gg<- ggplot(
  data=comt_rep_cast_mat_iqlr_dt, 
  aes(x=COMT_159_R, y=COMT_163_R))

roi2_poly_f3_rep_cor_gg<- roi2_poly_f3_rep_cor_gg+
  geom_point(alpha=0.75)+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

comt_rep_cast_mat_iqlr_dt[,cor(x=COMT_159_R, y=COMT_163_R, use="pairwise.complete.obs")]

roi2_poly_f4_rep_cor_gg<- ggplot(
  data=comt_rep_cast_mat_iqlr_dt, 
  aes(x=COMT_160_R, y=COMT_164_R))

roi2_poly_f4_rep_cor_gg<- roi2_poly_f4_rep_cor_gg+
  geom_point(alpha=0.75)+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

comt_rep_cast_mat_iqlr_dt[,cor(x=COMT_160_R, y=COMT_164_R, use="pairwise.complete.obs")]

roi2_facs_p1_rep_cor_gg<- ggplot(
  data=comt_rep_cast_mat_iqlr_dt, 
  aes(x=COMT_127_R, y=COMT_137_R))

roi2_facs_p1_rep_cor_gg<- roi2_facs_p1_rep_cor_gg+
  geom_point(alpha=0.75)+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

comt_rep_cast_mat_iqlr_dt[,cor(x=COMT_127_R, y=COMT_137_R, use="pairwise.complete.obs")]

roi2_facs_p3_rep_cor_gg<- ggplot(
  data=comt_rep_cast_mat_iqlr_dt, 
  aes(x=COMT_128_R, y=COMT_138_R))

roi2_facs_p3_rep_cor_gg<- roi2_facs_p3_rep_cor_gg+
  geom_point(alpha=0.75)+
  geom_abline(slope=1, col="grey", linetype=2)+
  coord_cartesian(xlim=c(-6,4), y=c(-6,4))+
  theme_cowplot()

comt_rep_cast_mat_iqlr_dt[,cor(x=COMT_128_R, y=COMT_138_R, use="pairwise.complete.obs")]

plot_grid(
  roi1_gdna_rep_cor_gg, roi1_cdna_rep_cor_gg, roi1_poly_f3_rep_cor_gg, roi1_poly_f4_rep_cor_gg, roi1_facs_p1_rep_cor_gg, roi1_facs_p3_rep_cor_gg,
  roi2_gdna_rep_cor_gg, roi2_cdna_rep_cor_gg, roi2_poly_f3_rep_cor_gg, roi2_poly_f4_rep_cor_gg, roi2_facs_p1_rep_cor_gg, roi2_facs_p3_rep_cor_gg,
  nrow=2, ncol=6)

roi1_f3_bio_rep_cors<- c(0.657, 0.565, 0.513, 0.615, 0.542, 0.767)
roi1_f4_bio_rep_cors<- c(0.672, 0.390, 0.534, 0.367, 0.551, 0.542)

roi2_f3_bio_rep_cors<- c(0.364, 0.540, 0.549, 0.551, 0.560, 0.582, 0.552, 0.582, 0.588,
                         0.982, 0.993, 0.987, 0.981, 0.989, 0.983)

roi2_f4_bio_rep_cors<- c(0.537, 0.677, 0.670, 0.672, 0.661, 0.717, 0.719, 0.713, 0.720,
                         0.986, 0.991, 0.984, 0.975, 0.994, 0.974)


roi1_q3_bio_rep_cors<- c(.579, .415, .440, .432, .444, .542)
roi2_q3_bio_rep_cors<- c(.055, .551, .513, .726, .186, .280, .116, .861, .732, .735)

```


```{r COMT replicate correlation pairwise after biological summarization}

comt_rep_summ_cast_dt<- dcast(
  data=comt_all_filt_postprocess_subtract_drop_cast_combat_bio_dt, 
  formula=ROI+ID_SHORT~Key, value.var="log10_NORM_CAF")

ggpairs(data=comt_rep_summ_cast_dt[ROI=="ROI_1"], 
        columns=which(names(comt_rep_summ_cast_dt)%in%c(
          "gDNA.", "cDNA.", "Poly.3", "Poly.4",
          "FACS.1", "FACS.2", "FACS.3", "FACS.4")))+
  coord_cartesian(xlim=c(-6,-2.5), y=c(-6,-2.5))+
  geom_abline(slope=1, col="grey", linetype=2)+
  theme_cowplot()

ggpairs(data=comt_rep_summ_cast_dt[ROI=="ROI_2"], 
        columns=which(names(comt_rep_summ_cast_dt)%in%c(
          "gDNA.", "cDNA.", "Poly.2", "Poly.3", "Poly.4",
          "FACS.1", "FACS.2", "FACS.3", "FACS.4")))+
  coord_cartesian(xlim=c(-6,-2.5), y=c(-6,-2.5))+
  geom_abline(slope=1, col="grey", linetype=2)+
  theme_cowplot()

# Correlations with gDNA are stronger after summarizing biological replicates

# cDNA and Poly.3, Poly.4 samples have satisfactory correlation
# ROI_2 correlation is slightly better than ROI_1, possibly because of technical replicates
# ROI_2 Poly.2 also has decent correlation with cDNA but ROI_1 Poly.2 does not

# What proportion of variants are found in all replicates and all layers?
comt_bio_drop_keys<- c("Poly.1", "Poly.2", "FACS.4")

# How many variants were found in all biological replicates for all layers?
comt_counts_by_keybio_varid_dt<- comt_all_filt_postprocess_subtract_drop_dt[
  !Key%in%c("Input.", "FACS.Input") & !Fraction%in%c("1","2") & Quadrant!="4" & !grepl(",", REF_CODON),
  .N, by=.(Key, Key_bio, VAR_ID)]

comt_counts_by_key_varid_dt<- comt_counts_by_keybio_varid_dt[,.N,by=.(Key, VAR_ID)]

comt_counts_by_varid_dt<- comt_counts_by_key_varid_dt[N==4, .N, by=.(VAR_ID)]

#comt_counts_by_varid_dt[N==7, length(unique(VAR_ID))] / comt_counts_by_varid_dt[,length(unique(VAR_ID))]
# 96% of variants found in all biological replicates and gDNA, RNA, Poly.3, Poly.4, FACS.1, FACS.2, and FACS.3

rm(list=c("comt_counts_by_keybio_varid_dt", "comt_counts_by_key_varid_dt"))

# How many variants were found in at least one biological replicate for all layers?
#comt_all_filt_postprocess_subtract_drop_cast_combat_bio_dt[,length(unique(VAR_ID))]
#table(comt_all_filt_postprocess_subtract_drop_cast_combat_bio_dt[!Key%in%comt_bio_drop_keys, .N,by=.(VAR_ID)]$N)

```


```{r Input correlation heatmap}

input_levels<- c("gDNA.", "cDNA.", "Poly.1", "Poly.2", "Poly.3", "Poly.4", "FACS.1", "FACS.2", "FACS.3", "FACS.4")

comt_rep_summ_cast_mat<- as.matrix(comt_rep_summ_cast_dt[,-c("ROI", "ID_SHORT")])

rownames(comt_rep_summ_cast_mat)<- comt_rep_summ_cast_dt[,ID_SHORT]
comt_rep_summ_cast_mat_cor<- cor(comt_rep_summ_cast_mat, use="pairwise.complete.obs")
comt_rep_summ_cast_mat_cor_dt<- as.data.table(comt_rep_summ_cast_mat_cor)
comt_rep_summ_cast_mat_cor_dt[,Sample_1:=rownames(comt_rep_summ_cast_mat_cor)]
setkey(comt_rep_summ_cast_mat_cor_dt, Sample_1)

comt_rep_summ_cast_mat_cor_melt_dt<- melt(
  data=comt_rep_summ_cast_mat_cor_dt, id.vars="Sample_1",
  variable.name="Sample_2", value.name="Pearson_cor")

comt_rep_summ_cast_mat_cor_melt_dt[,Sample_1:=factor(Sample_1, levels=input_levels)]
comt_rep_summ_cast_mat_cor_melt_dt[,Sample_2:=factor(Sample_2, levels=input_levels)]

comt_rep_summ_cast_mat_cor_melt_gg<- ggplot(
  data=comt_rep_summ_cast_mat_cor_melt_dt[Sample_1!="FACS.Input" & Sample_2!="FACS.Input"], 
  aes(x=Sample_1, y=Sample_2, fill=Pearson_cor))

comt_rep_summ_cast_mat_cor_melt_gg+
  geom_tile()+
  scale_fill_viridis(option="C")+
  theme_cowplot()+
  labs(x="", y="", fill="Pearson r")+
  theme(axis.text.x=element_text(angle=90))

rm(comt_rep_summ_cast_mat_cor_melt_gg)

```


# Variants with differential mRNA abundance

```{r COMT RNA limma-trend}

rna_fdr_thresh<- 0.01
loose_fdr_thresh<- 0.05
replicate_count_thresh<- 8
log_caf_thresh<- -10

comt_rna_limma_dt<- comt_all_filt_postprocess_subtract_drop_cast_cdna_gdna_combat_dt[
  !grepl("Poly", Input) & !grepl("Input_library", Input),]

# Use a filter on frequency; corresponds to ~10 counts
comt_rna_limma_counts_dt<- comt_rna_limma_dt[
  Median_log_CAF>log_caf_thresh, .N, by=.(VAR_ID)]

comt_rna_limma_complete_varids<- comt_rna_limma_counts_dt[
  N>=replicate_count_thresh, unique(VAR_ID)]

comt_rna_limma_input_dt<- comt_rna_limma_dt[
  VAR_ID%in%comt_rna_limma_complete_varids]

comt_rna_limma_input_cast_dt<- dcast(
  data=comt_rna_limma_input_dt, formula=VAR_ID~Key, value.var="Median_log_CAF")

comt_rna_limma_input_cast_mat<- as.matrix(
  comt_rna_limma_input_cast_dt[,-c("VAR_ID")])

rownames(comt_rna_limma_input_cast_mat)<- comt_rna_limma_input_cast_dt[,VAR_ID]

#comt_rna_limma_input_cast_mat<- normalizeQuantiles(comt_rna_limma_input_cast_mat)

comt_rna_limma_input_cast_eset<- Biobase::ExpressionSet(
  assayData=comt_rna_limma_input_cast_mat)

comt_rna_limma_input_design_dt<- 
  comt_rna_limma_input_dt[,as.factor(unique(Input)), keyby=.(Key)]

comt_rna_limma_input_design_mat<- 
  as.matrix(comt_rna_limma_input_design_dt[,-c("Key")])

rownames(comt_rna_limma_input_design_mat)<- comt_rna_limma_input_design_dt[,Key]

comt_rna_limma_input_design<- 
  model.matrix(~V1, data=as.data.frame(comt_rna_limma_input_design_mat), 
               list(V1=contr.treatment(n=2, base=2)))

comt_rna_limma_input_cast_eset_fit<- lmFit(
  comt_rna_limma_input_cast_eset, comt_rna_limma_input_design)

comt_rna_limma_input_cast_eset_fit_eb<- eBayes(
  comt_rna_limma_input_cast_eset_fit, trend=TRUE, proportion=0.03)

comt_rna_limma_input_cast_eset_fit_eb_all<- topTable(
  comt_rna_limma_input_cast_eset_fit_eb, coef=ncol(
    comt_rna_limma_input_design), number=nrow(
      comt_rna_limma_input_cast_eset_fit_eb))

comt_rna_limma_input_cast_eset_fit_eb_all_dt<- as.data.table(
  comt_rna_limma_input_cast_eset_fit_eb_all)

comt_rna_limma_input_cast_eset_fit_eb_all_dt[
  ,VAR_ID:=rownames(comt_rna_limma_input_cast_eset_fit_eb_all)]

comt_rna_limma_input_cast_eset_fit_eb_all_dt[logFC>0, Direction:="RNA up"]
comt_rna_limma_input_cast_eset_fit_eb_all_dt[logFC<0, Direction:="RNA down"]
comt_rna_limma_input_cast_eset_fit_eb_all_dt[,Direction:=as.factor(Direction)]

# Look at p-value distr
cdna_pval_distr_gg<- ggplot(
  data=comt_rna_limma_input_cast_eset_fit_eb_all_dt, 
  aes(x=P.Value))

cdna_pval_distr_gg+
  geom_histogram()+
  theme_cowplot()

# Volcano plot
cDNA_gDNA_volcano_gg<- ggplot(
  data=comt_rna_limma_input_cast_eset_fit_eb_all_dt, 
  aes(x=logFC, y=-1*log10(adj.P.Val)))

cDNA_gDNA_volcano_gg+
  geom_point()+
  theme_cowplot()

comt_rna_limma_input_cast_eset_fit_eb_top_dt<- 
  comt_rna_limma_input_cast_eset_fit_eb_all_dt[adj.P.Val < rna_fdr_thresh]

comt_rna_limma_input_cast_eset_fit_eb_top_varids<- 
  comt_rna_limma_input_cast_eset_fit_eb_top_dt[,VAR_ID]

comt_rna_limma_input_cast_eset_fit_eb_top_up_varids<- 
  comt_rna_limma_input_cast_eset_fit_eb_top_dt[logFC>0,VAR_ID]

comt_rna_limma_input_cast_eset_fit_eb_top_down_varids<- 
  comt_rna_limma_input_cast_eset_fit_eb_top_dt[logFC<0,VAR_ID]

comt_rna_limma_dt[
  ,ID_SHORT:=paste(AA_CHANGE, paste(REF, ALT, sep=">"), sep=", ")]

comt_rna_limma_dt[
  ,ID_LONG:=paste(AA_CHANGE, paste(POS, REF, ALT, sep=":"), sep=", ")]

if("Direction"%in%names(comt_rna_limma_dt)){
  comt_rna_limma_dt[,Direction:=NULL]
}

comt_rna_limma_dt[
  VAR_ID%in%comt_rna_limma_input_cast_eset_fit_eb_top_up_varids,
  Direction:="RNA up"]

comt_rna_limma_dt[
  VAR_ID%in%comt_rna_limma_input_cast_eset_fit_eb_top_down_varids,
  Direction:="RNA down"]

comt_rna_limma_dt[,Direction:=as.factor(Direction)]

amplicon_gDNA_med_log_caf<- comt_rna_limma_dt[
  Input=="gDNA", median(Median_log_CAF)]

amplicon_cDNA_med_log_caf<- comt_rna_limma_dt[
  Input=="cDNA", median(Median_log_CAF)]

aaas_cols<- pal_aaas()(2)

comt_rna_limma_sig_dt<- comt_rna_limma_dt[
    VAR_ID%in%comt_rna_limma_input_cast_eset_fit_eb_top_varids]

comt_rna_limma_sig_dt[,Input:=factor(Input, levels=c("cDNA", "gDNA"))]

# Visualize the frequencies
comt_rna_limma_sig_gg<- ggplot(
  data=comt_rna_limma_sig_dt,
  aes(x=Median_log_CAF, y=reorder(ID_SHORT, POS), col=Input))

comt_rna_limma_sig_gg+
  geom_vline(xintercept=amplicon_gDNA_med_log_caf, col=aaas_cols[1], linetype=2)+
  geom_vline(xintercept=amplicon_cDNA_med_log_caf, col=aaas_cols[2], linetype=2)+
  facet_wrap(~Direction, scales="free_y")+
  geom_point(position=position_jitterdodge())+
  labs(y="Variant ID", x="log frequency")+
  scale_x_continuous(breaks=seq(-11,-6))+
  coord_cartesian(xlim=c(-11,-6))+
  scale_color_aaas()+
  theme_cowplot()+
  theme(axis.text.y=element_text(size=7))

# Do not include nonsense variants
comt_rna_limma_sig_gg<- ggplot(
  data=comt_rna_limma_sig_dt[VAR_TYPE!="Nonsense"],
  aes(x=Median_log_CAF, y=reorder(ID_SHORT, POS), col=Input))

comt_rna_limma_sig_gg+
  geom_vline(xintercept=amplicon_gDNA_med_log_caf, col=aaas_cols[1], linetype=2)+
  geom_vline(xintercept=amplicon_cDNA_med_log_caf, col=aaas_cols[2], linetype=2)+
  facet_wrap(~Direction, scales="free_y")+
  geom_point(position=position_jitterdodge())+
  labs(y="Variant ID", x="log frequency")+
  scale_x_continuous(breaks=seq(-11,-6))+
  coord_cartesian(xlim=c(-11,-6))+
  scale_color_aaas()+
  theme_cowplot()+
  theme(axis.text.y=element_text(size=7))

# Split by type
comt_rna_limma_sig_gg<- ggplot(
  data=comt_rna_limma_sig_dt,
  aes(x=Median_log_CAF, y=reorder(ID_SHORT, POS), col=Input))

comt_rna_limma_sig_gg+
  geom_vline(xintercept=amplicon_gDNA_med_log_caf, col=aaas_cols[1], linetype=2)+
  geom_vline(xintercept=amplicon_cDNA_med_log_caf, col=aaas_cols[2], linetype=2)+
  facet_wrap(~VAR_TYPE, scales="free_y")+
  geom_point(position=position_jitterdodge())+
  labs(y="Variant ID", x="log frequency")+
  scale_x_continuous(breaks=seq(-11,-6))+
  coord_cartesian(xlim=c(-11,-6))+
  scale_color_aaas()+
  theme_cowplot()+
  theme(axis.text.y=element_text(size=7))

# Nonsense variants only
comt_rna_limma_sig_gg<- ggplot(
  data=comt_rna_limma_sig_dt[VAR_TYPE=="Nonsense"],
  aes(x=Median_log_CAF, y=reorder(ID_SHORT, POS), col=Input))

comt_rna_limma_sig_gg+
  geom_vline(xintercept=amplicon_gDNA_med_log_caf, col=aaas_cols[1], linetype=2)+
  geom_vline(xintercept=amplicon_cDNA_med_log_caf, col=aaas_cols[2], linetype=2)+
  geom_point(position=position_jitterdodge())+
  labs(y="Variant ID", x="log frequency")+
  scale_x_continuous(breaks=seq(-10,-6))+
  coord_cartesian(xlim=c(-10,-6.5))+
  scale_color_aaas()+
  theme_cowplot()+
  theme(axis.text.y=element_text(size=7))

rm(comt_rna_limma_sig_gg)

# Annotate the limma dataset
comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt<- merge(
  comt_rna_limma_input_cast_eset_fit_eb_all_dt, annot_dt, by="VAR_ID")

# Validation variants
comt_cdna_val_var_ids<- c("p.M51L"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4827:ATG:CTC",
                          "p.M152T"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:5131:TG:CT",
                          "p.S74H"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4896:AG:CA",
                          "p.S74L"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4896:AGC:CTA",
                          "p.H62R"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4861:AC:GG",
                          "p.V63F"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4863:GTG:TTT",
                          "p.L64C"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4866:CTG:TGT",
                          "p.E56R"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4842:GAG:CGT",
                          "p.V63V"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4865:G:C",
                          "p.N71*"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4887:AAC:TAG")

```

```{r ALDEx2 utilities}

aitchison.mean<- function(n, log=FALSE){

    # Input is a vector of non-negative integer counts.
    # Output is a probability vector of expected frequencies.
    # If log-frequencies are requested, the uninformative subspace is removed.

    n <- round( as.vector( n, mode="numeric" ) )
    if ( any( n < 0 ) ) stop("counts cannot be negative")

    a <- n + 0.5
    sa <- sum(a)

    log.p <- digamma(a) - digamma(sa)
    log.p <- log.p - mean(log.p)

    if ( log ) return(log.p)

    p <- exp( log.p - max(log.p) )
    p <- p / sum(p)
    return(p)
}

aldex_effect<- function(clr, verbose = TRUE, include.sample.summary = FALSE, 
                        useMC = FALSE, CI = FALSE, glm.conds = NULL, paired.test = FALSE, CI_percentile = .75){
    if (is.vector(clr@conds)) {
        conditions <- clr@conds
    }
    else if (is.factor(clr@conds)) {
        if (length(levels(clr@conds) == 2)) {
            conditions <- clr@conds
        }
    }
    else if (is.matrix(clr@conds)) {
        if (is.null(glm.conds)) 
            stop("please provide a binary condition vector")
        conditions <- glm.conds
    }
    else {
        stop("please check that the conditions parameter for aldex.clr is correct.")
    }
    is.multicore = FALSE
    if ("BiocParallel" %in% rownames(installed.packages()) & 
        useMC == TRUE) {
        message("multicore environment is OK -- using the BiocParallel package")
        is.multicore = TRUE
    }
    else {
        if (verbose == TRUE) 
            message("operating in serial mode")
    }
    nr <- numFeatures(clr)
    rn <- getFeatureNames(clr)
    conditions <- as.factor(conditions)
    levels <- levels(conditions)
    if (length(conditions) != numConditions(clr)) 
        stop("mismatch btw 'length(conditions)' and 'ncol(reads)'")
    if (length(levels) != 2) 
        stop("only two condition levels are currently supported")
    levels <- vector("list", length(levels))
    names(levels) <- levels(conditions)
    for (l in levels(conditions)) {
        levels[[l]] <- which(conditions == l)
        if (length(levels[[l]]) < 2) 
            stop("condition level '", l, "' has less than two replicates")
    }
    if (verbose == TRUE) 
        message("sanity check complete")
    rab <- vector("list", 3)
    names(rab) <- c("all", "win", "spl")
    rab$win <- list()
    cl2p <- NULL
    cl2p <- do.call(cbind, getMonteCarloInstances(clr))
    rab$all <- Rfast::rowMedians(cl2p)
    names(rab$all) <- rownames(cl2p)
    rm(cl2p)
    if (verbose == TRUE) 
        message("rab.all  complete")
    for (level in levels(conditions)) {
        cl2p <- NULL
        cl2p <- do.call(cbind, getMonteCarloInstances(clr)[levels[[level]]])
        rab$win[[level]] <- Rfast::rowMedians(cl2p)
        rm(cl2p)
    }
    if (verbose == TRUE) 
        message("rab.win  complete")
    if (is.multicore == TRUE) 
        rab$spl <- bplapply(getMonteCarloInstances(clr), function(m) {
            t(apply(m, 1, median))
        })
    if (is.multicore == FALSE) 
        rab$spl <- lapply(getMonteCarloInstances(clr), function(m) {
            Rfast::rowMedians(m)
        })
    if (verbose == TRUE) 
        message("rab of samples complete")
    if (paired.test == FALSE) {
        l2d <- vector("list", 2)
        names(l2d) <- c("btw", "win")
        l2d$win <- list()
        for (level in levels(conditions)) {
            concat <- NULL
            for (l1 in sort(levels[[level]])) {
                concat <- cbind(getMonteCarloReplicate(clr, l1), 
                  concat)
            }
            if (ncol(concat) < 10000) {
                sampl1 <- t(apply(concat, 1, function(x) {
                  sample(x, ncol(concat))
                }))
                sampl2 <- t(apply(concat, 1, function(x) {
                  sample(x, ncol(concat))
                }))
            }
            else {
                sampl1 <- t(apply(concat, 1, function(x) {
                  sample(x, 10000)
                }))
                sampl2 <- t(apply(concat, 1, function(x) {
                  sample(x, 10000)
                }))
            }
            l2d$win[[level]] <- cbind(l2d$win[[level]], abs(sampl1 - 
                sampl2))
            rm(sampl1)
            rm(sampl2)
        }
        if (verbose == TRUE) 
            message("within sample difference calculated")
        ncol.wanted <- min(sapply(l2d$win, ncol))
        if (is.multicore == TRUE) 
            l2d$win <- bplapply(l2d$win, function(arg) {
                arg[, 1:ncol.wanted]
            })
        if (is.multicore == FALSE) 
            l2d$win <- lapply(l2d$win, function(arg) {
                arg[, 1:ncol.wanted]
            })
        concatl1 <- NULL
        concatl2 <- NULL
        for (l1 in levels[[1]]) concatl1 <- cbind(getMonteCarloReplicate(clr, 
            l1), concatl1)
        for (l2 in levels[[2]]) concatl2 <- cbind(getMonteCarloReplicate(clr, 
            l2), concatl2)
        sample.size <- min(ncol(concatl1), ncol(concatl2))
        if (sample.size < 10000) {
            smpl1 <- t(apply(concatl1, 1, function(x) {
                sample(x, sample.size)
            }))
            smpl2 <- t(apply(concatl2, 1, function(x) {
                sample(x, sample.size)
            }))
        }
        else {
            smpl1 <- t(apply(concatl1, 1, function(x) {
                sample(x, 10000)
            }))
            smpl2 <- t(apply(concatl2, 1, function(x) {
                sample(x, 10000)
            }))
        }
        l2d$btw <- smpl2 - smpl1
        rm(smpl1)
        rm(smpl2)
        if (verbose == TRUE) 
            message("between group difference calculated")
        win.max <- matrix(0, nrow = nr, ncol = ncol.wanted)
        l2d$effect <- matrix(0, nrow = nr, ncol = ncol(l2d$btw))
        rownames(l2d$effect) <- rn
        options(warn = -1)
        for (i in 1:nr) {
            win.max[i, ] <- pmax(l2d$win[[1]][i, ], l2d$win[[2]][i, 
                ])
            l2d$effect[i, ] <- l2d$btw[i, ]/win.max[i, ]
            l2d$effect[i, ][is.na(l2d$effect[i, ])] <- 0
        }
        options(warn = 0)
        rownames(win.max) <- rn
        attr(l2d$win, "max") <- win.max
        rm(win.max)
        l2s <- vector("list", 2)
        names(l2s) <- c("btw", "win")
        l2s$win <- list()
        l2s$btw <- Rfast::rowMedians(l2d$btw)
        l2s$win <- Rfast::rowMedians(attr(l2d$win, "max"))
        if (verbose == TRUE) 
            message("group summaries calculated")
        if (CI == FALSE) {
            effect <- Rfast::rowMedians(l2d$effect)
        }
        else {
            effectlow <- as.vector(t(apply(l2d$effect, 1, function(x) {
                x[is.na(x)] <- 0
                quantile(x, probs = 0 + ((1-CI_percentile)/2), names = FALSE)
            })))
            effecthigh <- as.vector(t(apply(l2d$effect, 1, function(x) {
                x[is.na(x)] <- 0
                quantile(x, probs = 1 - ((1-CI_percentile)/2), names = FALSE)
            })))
            effect <- Rfast::rowMedians(l2d$effect)
        }
        overlap <- apply(l2d$effect, 1, function(row) {
            if (all(is.na(row))) 
                warning("NAs in effect, ignore if using ALR")
            row[is.na(row)] <- 0
            min(aitchison.mean(c(sum(row < 0), sum(row > 0)) + 
                0.5))
        })
        if (verbose == TRUE) 
            message("unpaired effect size calculated")
    }
    else if (paired.test == TRUE) {
        l2s <- vector("list", 2)
        names(l2s) <- c("btw", "win")
        sets <- names(levels)
        setA <- which(conditions == sets[1])
        setB <- which(conditions == sets[2])
        diff <- NULL
        for (i in 1:length(setA)) {
            jnk1 <- getMonteCarloReplicate(clr, setA[i])
            jnk2 <- getMonteCarloReplicate(clr, setB[i])
            diff <- cbind(diff, jnk2 - jnk1)
        }
        overlap <- apply(diff, 1, function(row) {
            if (all(is.na(row))) 
                warning("NAs in effect, ignore if using ALR")
            row[is.na(row)] <- 0
            min(aitchison.mean(c(sum(row < 0), sum(row > 0)) + 
                0.5))
        })
        l2s$btw <- apply(diff, 1, median)
        l2s$win <- apply(diff, 1, function(x) mad(x, constant = 1.428))
        effect.low <- apply(diff, 1, function(x) quantile(x, 
            probs = 0 + ((1-CI_percentile)/2)))
        effect.high <- apply(diff, 1, function(x) quantile(x, 
            probs = 1 - ((1-CI_percentile)/2)))
        effectlow <- as.vector(effect.low/l2s$win)
        effecthigh <- as.vector(effect.high/l2s$win)
        effect <- l2s$btw/l2s$win
        if (verbose == TRUE) 
            message("paired effect size calculated")
    }
    if (CI == FALSE) {
        rv <- list(rab = rab, diff = l2s, effect = effect, overlap = overlap)
    }
    else {
        rv <- list(rab = rab, diff = l2s, effect = effect, effectlow = effectlow, 
            effecthigh = effecthigh, overlap = overlap)
    }
    if (verbose == TRUE) 
        message("summarizing output")
    y.rv <- data.frame(rv$rab$all)
    colnames(y.rv) <- c("rab.all")
    for (i in names(rv$rab$win)) {
        nm <- paste("rab.win", i, sep = ".")
        y.rv[, nm] <- data.frame(rv$rab$win[[i]])
    }
    if (include.sample.summary == TRUE) {
        for (i in names(rv$rab$spl)) {
            nm <- paste("rab.sample", i, sep = ".")
            if (is.multicore == TRUE) 
                y.rv[, nm] <- data.frame(t(rv$rab$spl[[i]]))
            if (is.multicore == FALSE) 
                y.rv[, nm] <- data.frame(rv$rab$spl[[i]])
        }
    }
    for (i in names(rv$diff)) {
        nm <- paste("diff", i, sep = ".")
        y.rv[, nm] <- data.frame(rv$diff[[i]])
    }
    if (CI == FALSE) {
        y.rv[, "effect"] <- data.frame(rv$effect)
        y.rv[, "overlap"] <- data.frame(rv$overlap)
    }
    else {
        y.rv[, "effect"] <- data.frame(rv$effect)
        y.rv[, "effect.low"] <- data.frame(rv$effectlow)
        y.rv[, "effect.high"] <- data.frame(rv$effecthigh)
        y.rv[, "overlap"] <- data.frame(rv$overlap)
    }
    return(y.rv)
}

# Function to extract ALDEx2 effect size, CIs, and p-values
run_aldex<- function(reads, conditions, denom="iqlr", CI=TRUE, paired=FALSE, mc_samples=128){
  
  aldex_clr<- aldex.clr(reads=reads, conds=conditions, denom=denom, mc.samples=mc_samples)
  
  aldex_effect<- aldex_effect(aldex_clr, paired.test=paired, CI=CI)
  
  aldex_ttest<- aldex.ttest(aldex_clr, paired.test=paired)
  
  aldex_effect_dt<- as.data.table(aldex_effect)
  aldex_ttest_dt<- as.data.table(aldex_ttest)
  
  aldex_effect_dt[,VAR_ID:=rownames(aldex_effect)]
  aldex_ttest_dt[,VAR_ID:=rownames(aldex_ttest)]
  
  aldex_res_merge_dt<- merge(aldex_effect_dt, aldex_ttest_dt, by="VAR_ID")
  
  return(aldex_res_merge_dt)
}

get_last_char<- function(x){
  nchars<- nchar(x)
  last_char<- substr(x, nchars, nchars)
}

fill_nas<- function(dt){
  
  refs<- dt[,unique(paste("p.", REF_AA, AA_POS, sep=""))]
  alts<- dt[,sort(unique(ALT_AA))]
  all_aa_changes<- unlist(lapply(refs, paste, alts, sep=""))
  obs_aa_changes<- dt[,unique(AA_CHANGE)]
  unobs_aa_changes<- all_aa_changes[!all_aa_changes%in%obs_aa_changes]
  
  aa_change_strip<- sapply(strsplit(as.character(unobs_aa_changes), ".", fixed=T), "[", 2)
  
  unobs_aa_refs<- substr(aa_change_strip, 1, 1)
  unobs_aa_pos<- gsub(pattern="[[:alpha:]]|[[:punct:]]", replacement="", aa_change_strip)
  unobs_aa_alts<- sapply(aa_change_strip, get_last_char)
  
  update_dt<- rbind(dt, data.table(
    REF_AA=unobs_aa_refs, AA_POS=unobs_aa_pos, ALT_AA=unobs_aa_alts, AA_CHANGE=unobs_aa_changes), fill=TRUE)
  
  update_dt[,VAR_TYPE:=ifelse(REF_AA==ALT_AA, "Silent", "Missense"), by=seq_len(nrow(update_dt))]
  update_dt[ALT_AA=="*", VAR_TYPE:="Nonsense"]
  update_dt[,VAR_TYPE:=factor(VAR_TYPE, levels=c("Silent", "Missense", "Nonsense"))]
  
  update_dt[AA_POS%in%seq(40,74), ROI:="ROI_1"]
  update_dt[AA_POS%in%seq(136,158), ROI:="ROI_2"]
  
  return(update_dt)
  
}

reclose_coda<- function(x){
  reclosed<- x/sum(x)
  return(reclosed)
}


```

```{r COMT RNA ALDEx2}

comt_rna_fdr_thresh<- 0.1

# First transform normalized frequencies to counts per million
comt_rna_input_cast_mat_cpm<- apply(comt_rna_limma_input_cast_mat, 2, function(x){round(exp(x)*1e6)})
comt_rna_input_cast_mat_cpm_iqlr<- apply(reclose_coda(comt_rna_input_cast_mat_cpm), 2, iqlr)
comt_rna_input_cast_mat_cpm_iqlr_dt<- as.data.table(comt_rna_input_cast_mat_cpm_iqlr)
comt_rna_input_cast_mat_cpm_iqlr_dt[,VAR_ID:=rownames(comt_rna_input_cast_mat_cpm_iqlr)]

comt_rna_input_cast_mat_cpm_iqlr_melt_dt<- melt(
  comt_rna_input_cast_mat_cpm_iqlr_dt, id.vars="VAR_ID", value.name="IQLR", variable.name="Input")

comt_rna_input_cast_mat_cpm_iqlr_melt_dt[,Input_clean:=sapply(strsplit(as.character(Input), "_"), "[", 1)]

# Need to use aldex.effect to get confidence intervals
set.seed(9)

comt_rna_input_cast_mat_cpm_aldex_dt<- run_aldex(comt_rna_input_cast_mat_cpm, c(rep("B", 4), rep("A", 4)))


# Look at the raw p-values; may not show uniform distr as these are posterior p-values
comt_rna_aldex2_pval_gg<- ggplot(
  data=comt_rna_input_cast_mat_cpm_aldex_dt, 
  aes(x=we.ep))

comt_rna_aldex2_pval_gg+
  geom_histogram()+
  theme_cowplot()

rm(comt_rna_aldex2_pval_gg)

comt_rna_input_cast_mat_cpm_aldex_dt[,VAR_ID:=rownames(comt_rna_input_cast_mat_cpm)]

comt_rna_input_cast_mat_cpm_aldex_merge_dt<- merge(
  comt_rna_input_cast_mat_cpm_aldex_dt, annot_dt, by="VAR_ID")

comt_rna_input_cast_mat_cpm_aldex_merge_dt[,Direction:=ifelse(effect>0, "RNA up", "RNA down")]
comt_rna_input_cast_mat_cpm_aldex_merge_dt[we.eBH>=comt_rna_fdr_thresh, Direction:="N.S."]
comt_rna_input_cast_mat_cpm_aldex_merge_dt[,Direction:=factor(Direction, levels=c("N.S.", "RNA up", "RNA down"))]

# Volcano plot of ALDEx2 results
comt_rna_aldex2_gg<- ggplot(
  data=comt_rna_input_cast_mat_cpm_aldex_merge_dt[we.eBH>=comt_rna_fdr_thresh],
  aes(x=effect, y=-1*log10(we.eBH)))

comt_rna_aldex2_gg+
  geom_point(alpha=0.3)+
  geom_point(data=comt_rna_input_cast_mat_cpm_aldex_merge_dt[
    we.eBH<comt_rna_fdr_thresh & VAR_TYPE=="Nonsense"], 
             aes(x=effect, y=-1*log10(we.eBH)), col="firebrick")+
  geom_point(data=comt_rna_input_cast_mat_cpm_aldex_merge_dt[
    we.eBH<comt_rna_fdr_thresh & VAR_TYPE=="Missense"], 
           aes(x=effect, y=-1*log10(we.eBH)), col="cyan3")+
  geom_point(data=comt_rna_input_cast_mat_cpm_aldex_merge_dt[
    we.eBH<comt_rna_fdr_thresh & VAR_TYPE=="Silent"], 
           aes(x=effect, y=-1*log10(we.eBH)), col="magenta")+
  theme_cowplot()+
  labs(x="Effect", y="-1*log10(FDR)")

rm(comt_rna_aldex2_gg)

# Plot IQLRs for variants with significant effects
comt_rna_input_cast_mat_cpm_aldex_merge_iqlr_dt<- merge(
  comt_rna_input_cast_mat_cpm_aldex_merge_dt, comt_rna_input_cast_mat_cpm_iqlr_melt_dt, by="VAR_ID")

comt_rna_aldex2_sig_gg<- ggplot(
  data=comt_rna_input_cast_mat_cpm_aldex_merge_iqlr_dt[we.eBH<comt_rna_fdr_thresh],
  aes(x=effect, y=reorder(ID_SHORT, effect), fill=VAR_TYPE))

comt_rna_aldex2_sig_gg+
  geom_bar(stat="identity", alpha=0.75)+
  scale_fill_manual(values=c("magenta", "cyan3", "firebrick"))+
  theme_cowplot()+
  labs(y="", x="Effect size", fill="")+
  theme(axis.text.y=element_text(size=8))

comt_rna_aldex2_sig_gg<- ggplot(
  data=comt_rna_input_cast_mat_cpm_aldex_merge_dt[we.eBH<0.05 & VAR_TYPE!="Nonsense"],
  aes(x=effect, y=reorder(ID_SHORT, POS), fill=-1*log10(we.eBH), xmax=effect.high, xmin=effect.low))

comt_rna_aldex2_sig_gg+
  geom_bar(stat="identity", alpha=0.75)+
  geom_errorbar(col="grey", width=0.3)+
  scale_fill_viridis(option="D", begin=0.05, end=0.95)+
  theme_cowplot()+
  labs(y="", x="Effect size", fill="")+
  theme(axis.text.x=element_text(angle=90))+
  coord_flip()

comt_rna_aldex2_sig_gg<- ggplot(
  data=comt_rna_input_cast_mat_cpm_aldex_merge_iqlr_dt[we.eBH<0.03 & VAR_TYPE!="Nonsense"],
  aes(x=IQLR, y=reorder(ID_SHORT, POS), col=Input_clean))

comt_rna_aldex2_sig_gg+
  geom_jitter(height=0.1)+
  scale_color_aaas()+
  theme_cowplot()+
  labs(y="", x="IQLR", col="Input")+
  theme(axis.text.y=element_text(size=8))+
  theme(legend.position="bottom")

rm(comt_rna_aldex2_sig_gg)


comt_rna_input_cast_mat_cpm_iqlr_melt_dt


# Generate a VE map using ALDEx2
comt_rna_aldex_rep_summ_ve_dt<- comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  !grepl(",", AA_CHANGE), .(Median_effect=median(effect, na.rm=T),
                            Max_effect=max(effect, na.rm=T),
                            Median_diff=median(diff.btw, na.rm=T),
                            Max_diff=max(diff.btw, na.rm=T)), 
  by=.(ROI, AA_POS, REF_AA, ALT_AA, AA_CHANGE, VAR_TYPE)]

comt_rna_aldex_rep_summ_ve_dt<- fill_nas(comt_rna_aldex_rep_summ_ve_dt)

comt_rna_aldex_rep_summ_ve_dt[,AA_POS:=as.integer(AA_POS)]
comt_rna_aldex_rep_summ_ve_dt[,ALT_AA:=factor(ALT_AA, levels=unlist(aa_class_list))]
comt_rna_aldex_rep_summ_ve_dt[,ALT_AA_level:=as.integer(ALT_AA)]

comt_rna_aldex_rep_summ_ve_dt[,Is_REF:=ifelse(REF_AA==ALT_AA, TRUE, FALSE)]

comt_rna_aldex_ve_gg<- ggplot(
  data=comt_rna_aldex_rep_summ_ve_dt[Is_REF==FALSE], 
  aes(x=AA_POS, y=reorder(ALT_AA, ALT_AA_level), fill=Median_effect))

comt_rna_aldex_ve_gg+
  facet_wrap(~ROI, scales="free")+
  geom_tile()+
  geom_tile(data=comt_rna_aldex_rep_summ_ve_dt[Is_REF==TRUE], 
            aes(x=AA_POS, y=reorder(ALT_AA, ALT_AA_level, fill=Median_effect)), col="black", linewidth=0.25)+
  scale_fill_gradient2(na.value="grey")+
  labs(x="Amino acid position", y="Alternate", 
       fill="Median effect")+
  theme_cowplot()

rm(comt_rna_aldex_ve_gg)

```

```{r COMT RNA limma-trend VE maps}

comt_rep_summ_ve_dt<- comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  !grepl(",", AA_CHANGE), .(Median_log_FC=median(logFC, na.rm=T)), 
  by=.(ROI, AA_POS, REF_AA, ALT_AA, AA_CHANGE, VAR_TYPE)]

comt_rep_summ_ve_dt[,AA_POS:=as.integer(AA_POS)]
comt_rep_summ_ve_dt[,ALT_AA:=factor(ALT_AA, levels=unlist(aa_class_list))]
comt_rep_summ_ve_dt[,ALT_AA_level:=as.integer(ALT_AA)]

comt_ve_heatmap_gg<- ggplot(
  data=comt_rep_summ_ve_dt, 
  aes(x=AA_POS, y=reorder(ALT_AA, ALT_AA_level), fill=Median_log_FC))

comt_ve_heatmap_gg+
  facet_wrap(~ROI, scales="free")+
  geom_tile()+
  scale_fill_gradient2()+
  labs(x="Amino acid position", y="Alternate", 
       fill="Median logFC")+
  theme_cowplot()

rm(comt_ve_heatmap_gg)

comt_ve_heatmap_roi1_gg<- ggplot(
  data=comt_rep_summ_ve_dt[ROI=="ROI_1"], 
  aes(x=AA_POS, y=reorder(ALT_AA, ALT_AA_level), fill=Median_log_FC))

comt_ve_heatmap_roi1_gg+
  geom_tile()+
  scale_fill_gradient2()+
  labs(x="Amino acid position", y="Alternate", 
       fill="Median logFC")+
  theme_cowplot()

rm(comt_ve_heatmap_roi1_gg)

comt_ve_heatmap_roi2_gg<- ggplot(
  data=comt_rep_summ_ve_dt[ROI=="ROI_2"], 
  aes(x=AA_POS, y=reorder(ALT_AA, ALT_AA_level), fill=Median_log_FC))

comt_ve_heatmap_roi2_gg+
  geom_tile()+
  scale_fill_gradient2()+
  labs(x="Amino acid position", y="Alternate", 
       fill="Median logFC")+
  theme_cowplot()

rm(comt_ve_heatmap_roi2_gg)

```

```{r Silent versus nonsense and missense mutations}

comt_roi_silent_missense_nonsense_gg<- ggplot(
  data=comt_rep_summ_ve_dt, 
  aes(x=VAR_TYPE, y=Median_log_FC))

comt_roi_silent_missense_nonsense_gg+
  facet_wrap(~ROI)+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot(fill="steelblue", alpha=0.75, width=0.5, outlier.shape=NA)+
  labs(x="Mutation type", y="Median log FC")+
  theme_cowplot()

t.test(Median_log_FC~ROI, data=comt_rep_summ_ve_dt[VAR_TYPE=="Silent"])
t.test(Median_log_FC~ROI, data=comt_rep_summ_ve_dt[VAR_TYPE=="Missense"])

# ALDEx2
comt_roi_silent_missense_nonsense_gg<- ggplot(
  data=comt_rna_aldex_rep_summ_ve_dt, 
  aes(x=VAR_TYPE, y=Median_diff, fill=ROI))

comt_roi_silent_missense_nonsense_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot(alpha=0.75, width=0.5, outlier.shape=NA)+
  scale_fill_manual(values=roi_cols)+
  labs(x="Mutation type", y="Median difference")+
  theme_cowplot()

t.test(Median_diff~ROI, data=comt_rna_aldex_rep_summ_ve_dt[VAR_TYPE=="Silent"])
t.test(Median_diff~ROI, data=comt_rna_aldex_rep_summ_ve_dt[VAR_TYPE=="Missense"])
t.test(Median_diff~ROI, data=comt_rna_aldex_rep_summ_ve_dt[VAR_TYPE=="Nonsense"])

```


## H62-V63 hotspot 

CAC-GUG-CUG-CAG-CAU
H-V-L-Q-H
N-O-O-M-N (N:nonoptimal, O:optimal, M:mid)

While the histidine codons are not optimal (especially CAU), the GUG and CUG codons for V, L are highly optimal. Thus, mutations away from these optimal codons are unlikely to lead to higher mRNA abundance through codon optimality.

Looking at where these residues reside on the cytosolic protein isoform, they are within the first alpha helix, distant from the AdoMet and catecholamine binding sites:
https://www.rcsb.org/3d-view/3BWM

CAC-GUG is not an inhibitory dicodon in yeast. GUG-CUG is not either, but these codons are found among several inhibitory dicodon pairs. 


## DeMaSk comparison

See https://demask.princeton.edu/about/

```{r DeMaSk comparison}

demask_file<- paste(demask_dir, "dc59eaf2256a3e939635.txt", sep="/")
demask_dt<- fread(demask_file, sep="\t")
demask_dt[,AA_CHANGE:=paste("p.", WT, pos, var, sep="")]

# Merge with RNA abundance data
comt_rna_input_cast_mat_cpm_aldex_merge_demask_dt<- merge(
  comt_rna_input_cast_mat_cpm_aldex_merge_dt, 
  demask_dt[,.(AA_CHANGE, score, entropy, log2f_var, matrix)], by="AA_CHANGE")

# Determine if variants that are down show lesser scores than RNA up variants
demask_gg<- ggplot(
  data=comt_rna_input_cast_mat_cpm_aldex_merge_demask_dt,
  aes(x=Direction, y=score))

demask_gg+
  geom_boxplot(width=0.5, fill="steelblue", alpha=0.75)+
  labs(x="", y="DeMaSk score")+
  theme_cowplot()

rm(demask_gg)

#comt_rna_input_cast_mat_cpm_aldex_merge_demask_dt[Direction!="N.S.", t.test(score~Direction)]

# The opposite of expectation is observed; would this hold if variants at the hotspot are removed? Yes
#comt_rna_input_cast_mat_cpm_aldex_merge_demask_dt[Direction!="N.S." & !AA_POS2%in%c(62,63), t.test(score~Direction)]

#quantile(demask_dt[,unique(entropy), by=.(pos)]$V1, probs=seq(0,1,0.1))
# Use bottom 20th percentile to identify constrained positions
demask_entropy_thresh<- 1.0779

demask_low_entropy_pos<- demask_dt[entropy<demask_entropy_thresh, unique(pos)]

# ROI_1 is codons 40-74 
# ROI_2 is codons 136-158

# Thus targeted codons with low entropy are 63, 67, 70-71, 140, 146, 156-157

```

There is no overall association between directionality of mRNA abundance effect and predicted impact on protein function using DeMaSk. However, V63 showed low entropy among homologs, suggesting it is an important conserved residue.


## Density of variants by position

Plot the density of variants by COMT transcript position to identify local hotspots of mRNA abundance dysregulation.

```{r Variants by position}

# Additionally plot changes to methionine, which may create
# new alternative TIS start codons

if("Label"%in%names(comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt)){
  comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,Label:=NULL]
}

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[abs(logFC)>0.35, Label:=AA_CHANGE]
comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,Label:=as.factor(Label)]

mut_density_gg<- ggplot(
  data=comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
    adj.P.Val < loose_fdr_thresh & VAR_TYPE!="Nonsense" & ROI=="ROI_1",], 
  aes(x=AA_POS2, y=logFC, col=adj.P.Val, label=Label))

mut_density_gg+
  scale_y_continuous(breaks=seq(-0.5, 1, 0.5))+
  scale_x_continuous(breaks=seq(40,157,5))+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  scale_color_viridis(option="A", begin=0.9, end=0.1)+
  geom_jitter(size=1.5, width=0.5)+
  geom_text_repel(col="black")+
  labs(x="Amino acid position", y="log FC, cDNA to gDNA", col="FDR")+
  theme_cowplot()+
  expand_limits(y=c(-.5,1))

mut_density_gg<- ggplot(
  data=comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
    adj.P.Val < loose_fdr_thresh & VAR_TYPE!="Nonsense",], 
  aes(x=AA_POS2, y=logFC, col=adj.P.Val, label=Label))

mut_density_gg+
  facet_wrap(~ROI, scales="free_x")+
  scale_y_continuous(breaks=seq(-0.5, 1, 0.5))+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  scale_color_viridis(option="A", begin=0.9, end=0.1)+
  geom_jitter(size=1.5, width=0.5)+
  geom_text_repel(col="black")+
  labs(x="Amino acid position", y="log FC, cDNA to gDNA", col="FDR")+
  theme_cowplot()+
  expand_limits(y=c(-.5,1))

rm(mut_density_gg)

```


## Clinically relevant SNPs

Determine if annotated clinically-relevant SNPs show differential mRNA abundance.

ClinVar mutations are:
K55K (G>A) is rs1601527185
H62H (C>T) is rs4633
A72S (G>T) is rs6267
Q73Q (C>A) is rs740602
E77E (G>A) is rs1601527383
T81T (C>A) is rs1601527404

```{r Clinically relevant SNPs}

clinrel_var_ids<- c("rs4633", "rs4680", "rs4818", "rs6267", "rs74745580")

# ROI_1
#rs74745580: TGGAA[C>T]GAG; p.N39N, 499:C:T; not in mutagenesis region
#rs4633: AACCA[C>T]GTG; p.H62H, 568:C:T
#rs6267: AAC[G>T]CACAG; p.A72S, 596:G:T

# ROI_2
#rs4818: AGGCT[C>G]ATC; p.L136L, 5084:G:C
#rs4680: GGC[G>A]TGAAG; p.V158M, 5148:G:A; forms a new TIS?; GGCGTGAAG > GGCATGAAG

```


## Dicodons and dipeptides

Determine if there are inhibitory or optimal codon pairs or dipeptides in the COMT ROIs.

```{r Dicodon pairs ALDEx2}

comt_rna_input_cast_mat_cpm_aldex_merge_sig_var_ids<- 
  comt_rna_input_cast_mat_cpm_aldex_merge_dt[we.eBH < comt_rna_fdr_thresh, VAR_ID]

gamble_grayhack_inhibitory_codon_pairs_list<- list(
  c("AGG", "CGA"), c("AGG", "CGG"), c("AUA", "CGA"), c("AUA", "CGG"),
  c("CGA", "AUA"), c("CGA", "CCG"), c("CGA", "CGA"), c("CGA", "CGG"), 
  c("CGA", "CUG"), c("CGA", "GCG"), c("CUC", "AUA"), c("CUC", "CCG"),
  c("CUG", "AUA"), c("CUG", "CCG"), c("CUG", "CGA"), c("CUG", "CUG"), 
  c("CUU", "CUG"), c("GUA", "CCG"), c("GUA", "CGA"), c("GUG", "CGA"))

gamble_grayhack_optimal_codon_pairs_list<- list(
  c("AGA", "CCA"), c("AGA", "AUU"), c("AGA", "AGA"))

gamble_grayhack_inhibitory_codon_pairs<- sapply(gamble_grayhack_inhibitory_codon_pairs_list, paste0, collapse="")
gamble_grayhack_inhibitory_codon_pairs_dna<- gsub("U", "T", gamble_grayhack_inhibitory_codon_pairs)

gamble_grayhack_optimal_codon_pairs<- sapply(gamble_grayhack_optimal_codon_pairs_list, paste0, collapse="")
gamble_grayhack_optimal_codon_pairs_dna<- gsub("U", "T", gamble_grayhack_optimal_codon_pairs)

# Determine if variants create optimal or inhibitory dicodons
check_dicodon_pair<- function(var_id, input_dt, codon_pairs){
  
  var_id_aa_pos<- input_dt[VAR_ID==var_id, unique(as.integer(AA_POS2))]
  var_id_alt_codon<- input_dt[VAR_ID==var_id, unique(ALT_CODON)]
  upstream_codon<- input_dt[AA_POS2==(var_id_aa_pos-1), unique(REF_CODON)]
  downstream_codon<- input_dt[AA_POS2==(var_id_aa_pos+1), unique(REF_CODON)]
  upstream_dicodon<- paste(c(upstream_codon, var_id_alt_codon), collapse="")
  downstream_dicodon<- paste(c(var_id_alt_codon, downstream_codon), collapse="")
  
  is_dicodon<- FALSE
  if(upstream_dicodon%in%codon_pairs || downstream_dicodon%in%codon_pairs){
    is_dicodon<- TRUE
  }
  return(is_dicodon)
}

comt_rna_input_cast_mat_cpm_aldex_merge_sig_var_ids_is_inhib<- 
  sapply(comt_rna_input_cast_mat_cpm_aldex_merge_sig_var_ids, check_dicodon_pair,
         comt_rna_input_cast_mat_cpm_aldex_merge_dt, gamble_grayhack_inhibitory_codon_pairs_dna)

# table(comt_rna_input_cast_mat_cpm_aldex_merge_sig_var_ids_is_inhib)
# One inhibitory dicodon found
comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  VAR_ID%in%comt_rna_input_cast_mat_cpm_aldex_merge_sig_var_ids[
    comt_rna_input_cast_mat_cpm_aldex_merge_sig_var_ids_is_inhib]]

comt_rna_input_cast_mat_cpm_aldex_merge_sig_var_ids_is_optimal<- 
  sapply(comt_rna_input_cast_mat_cpm_aldex_merge_sig_var_ids, check_dicodon_pair,
         comt_rna_input_cast_mat_cpm_aldex_merge_dt, gamble_grayhack_optimal_codon_pairs_dna)

# table(comt_rna_input_cast_mat_cpm_aldex_merge_sig_var_ids_is_optimal)
# No optimal dicodons found

```

```{r Dipeptide pairs ALDEx2}

comt_rna_input_cast_mat_cpm_aldex_merge_sig_var_ids<- 
  comt_rna_input_cast_mat_cpm_aldex_merge_dt[we.eBH < comt_rna_fdr_thresh, VAR_ID]

# Candidate dipeptides were identified from a reporter assay with 8x repeats

burke_subramaniam_nonoptimal_dipeptides<- c(
  "RH", "HR", "II", "KI", "KY", "FR", "FG", "FK", "FS", "KV", "YK", "VK", "LG", "SF", "WW")

burke_subramaniam_optimal_dipeptides<- c(
  "AM", "MM", "LF", "CL", "CF", "FC", "WM", "WF", "VM", "VF")

check_dipeptides<- function(var_id, input_dt, aa_pairs){
  
  var_id_aa_pos<- input_dt[VAR_ID==var_id, unique(as.integer(AA_POS2))]
  var_id_alt_aa<- input_dt[VAR_ID==var_id, unique(ALT_AA)]
  upstream_aa<- input_dt[AA_POS2==(var_id_aa_pos-1), unique(REF_AA)]
  downstream_aa<- input_dt[AA_POS2==(var_id_aa_pos+1), unique(REF_AA)]
  upstream_pair<- paste(c(upstream_aa, var_id_alt_aa), collapse="")
  downstream_pair<- paste(c(var_id_alt_aa, downstream_aa), collapse="")
  
  is_dipeptide<- FALSE
  if(upstream_pair%in%aa_pairs || downstream_pair%in%aa_pairs){
    is_dipeptide<- TRUE
  }
  return(is_dipeptide)
}

comt_rna_input_cast_mat_cpm_aldex_merge_sig_var_ids_aa_is_inhib<- 
  sapply(comt_rna_input_cast_mat_cpm_aldex_merge_sig_var_ids, check_dipeptides,
         comt_rna_input_cast_mat_cpm_aldex_merge_dt, burke_subramaniam_nonoptimal_dipeptides)

# table(comt_rna_input_cast_mat_cpm_aldex_merge_sig_var_ids_aa_is_inhib)
# One inhibitory dipeptide found among differential variants
comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  VAR_ID%in%comt_rna_input_cast_mat_cpm_aldex_merge_sig_var_ids[
    comt_rna_input_cast_mat_cpm_aldex_merge_sig_var_ids_aa_is_inhib], 
  .(VAR_ID, REF_CODON, ALT_CODON, AA_CHANGE, effect)]

# V63R is an inhibitory dipeptide but has a positive effect; 
# although it is at the H62-V63 hotspot

comt_rna_input_cast_mat_cpm_aldex_merge_sig_var_ids_aa_is_optimal<- 
  sapply(comt_rna_input_cast_mat_cpm_aldex_merge_sig_var_ids, check_dipeptides,
         comt_rna_input_cast_mat_cpm_aldex_merge_dt, burke_subramaniam_optimal_dipeptides)

# table(comt_rna_input_cast_mat_cpm_aldex_merge_sig_var_ids_aa_is_optimal)
# One optimal dipeptide found
comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  VAR_ID%in%comt_rna_input_cast_mat_cpm_aldex_merge_sig_var_ids[
    comt_rna_input_cast_mat_cpm_aldex_merge_sig_var_ids_aa_is_optimal],
  .(VAR_ID, REF_CODON, ALT_CODON, AA_CHANGE, effect)]

# V63C is an optimal dipeptide, and has a positive effect; 
# although it is at the H62-V63 hotspot

```


## Variants causing Kozak-like motifs

```{r Variants that create Kozak-like sequences}

kozak_canonical_file<- paste(kozak_dir, "Noderer_Wang_2014_FACS_seq_TIS_efficiency.txt", sep="/")

kozak_noncanonical_file<- paste(kozak_dir, "nar-01911-h-2017-File009.txt", sep="/")

tis_efficiency_dt<- fread(kozak_canonical_file, skip=1)
noncanon_tis_efficiency_dt<- fread(kozak_noncanonical_file, skip=1)
names(noncanon_tis_efficiency_dt)<- c("sequence", "efficiency")

endogenous_mem_tis<- "UUGAAGAUGCC"
flag_mem_tis<- "UUGAAGAUGGA"
endogenous_cyto_tis<- "CUGCUCAUGGG"
comt_tis_seqs<- c(endogenous_mem_tis, flag_mem_tis, endogenous_cyto_tis)

# tis_efficiency_dt[sequence%in%comt_tis_seqs]
# Membrane TIS has efficiency 81-83; Flag tag does not strongly alter efficiency
# Cytoplasmic TIS has slightly higher efficiency at 86

endogenous_cyto_tis_flush<- "CUGCUCAUGGGU"
endogenous_cyto_tis_flush_split<- unlist(strsplit(endogenous_cyto_tis_flush, ""))

# First look at the variants surrounding the alternative start codon for 
# the cytoplasmic COMT isoform

# For Noderer et al. 2014 exclude changes at the cyto TIS start codon
comt_rna_input_cast_mat_cpm_aldex_merge_tis_dt<- comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  AA_POS2%in%c(49,50,52,150,151,153) & VAR_TYPE!="Nonsense"]

get_kozak_efficiency<- function(var_id, input_dt, endo_tis, tis_efficiency_dt){
  
  endo_tis_copy<- copy(endo_tis)
  var_id_aa_pos<- input_dt[VAR_ID==var_id, unique(as.integer(AA_POS2))]
  var_id_alt_codon<- input_dt[VAR_ID==var_id, unique(ALT_CODON)]
  var_id_alt_codon_split<- unlist(strsplit(var_id_alt_codon, ""))
  
  if(var_id_aa_pos == 49){
    endo_tis_copy[1:3]<- var_id_alt_codon_split
  } else if(var_id_aa_pos == 50){
    endo_tis_copy[4:6]<- var_id_alt_codon_split
  } else if(var_id_aa_pos == 51){
    endo_tis_copy[7:9]<- var_id_alt_codon_split
  } else if(var_id_aa_pos == 52){
    endo_tis_copy[10:12]<- var_id_alt_codon_split
  }
  
  endo_tis_copy_paste<- paste(endo_tis_copy[1:11], collapse="")
  endo_tis_copy_paste<- gsub("T", "U", endo_tis_copy_paste)
  tis_efficiency<- tis_efficiency_dt[sequence==endo_tis_copy_paste, efficiency]
  return(tis_efficiency)
}

comt_rna_input_cast_mat_cpm_aldex_merge_tis_varids<- 
  comt_rna_input_cast_mat_cpm_aldex_merge_tis_dt[,as.character(unique(VAR_ID))]

comt_rna_input_cast_mat_cpm_aldex_merge_tis_efficiencies<- 
  sapply(comt_rna_input_cast_mat_cpm_aldex_merge_tis_varids, get_kozak_efficiency,
         comt_rna_input_cast_mat_cpm_aldex_merge_tis_dt, endogenous_cyto_tis_flush_split, 
         tis_efficiency_dt)

comt_rna_input_cast_mat_cpm_aldex_merge_tis_efficiencies_dt<- 
  data.table(VAR_ID=comt_rna_input_cast_mat_cpm_aldex_merge_tis_varids, 
             Kozak_efficiency=comt_rna_input_cast_mat_cpm_aldex_merge_tis_efficiencies)

# Annotate the limma dataset with canonicalk Kozak efficiency
comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  ,Kozak_efficiency:=comt_rna_input_cast_mat_cpm_aldex_merge_tis_efficiencies_dt[
    match(comt_rna_input_cast_mat_cpm_aldex_merge_dt[,VAR_ID], VAR_ID), Kozak_efficiency]]

kozak_efficiency_gg<- ggplot(
  data=comt_rna_input_cast_mat_cpm_aldex_merge_dt[!is.na(Kozak_efficiency)], 
  aes(x=effect, Kozak_efficiency))

kozak_efficiency_gg+
  geom_point(size=2)+
  theme_cowplot()+
  labs(y="Kozak efficiency")

rm(kozak_efficiency_gg)

# No clear relationship between cDNA abundance effect and TIS strength


# Get noncanonical TIS efficiencies for changes at mutagenized methionine codons
get_noncanon_kozak_efficiency<- function(var_id, input_dt, endo_tis, tis_efficiency_dt){
  
  endo_tis_copy<- copy(endo_tis)
  var_id_aa_pos<- input_dt[VAR_ID==var_id, unique(as.integer(AA_POS2))]
  var_id_alt_codon<- input_dt[VAR_ID==var_id, unique(ALT_CODON)]
  var_id_alt_codon_split<- unlist(strsplit(var_id_alt_codon, ""))
  endo_tis_copy[7:9]<- var_id_alt_codon_split
  
  endo_tis_copy_paste<- paste(endo_tis_copy[3:10], collapse="")
  endo_tis_copy_paste<- gsub("T", "U", endo_tis_copy_paste)
  tis_efficiency<- tis_efficiency_dt[sequence==endo_tis_copy_paste, efficiency]
  return(tis_efficiency)
}

comt_rna_input_cast_mat_cpm_aldex_merge_tis_noncanon_dt<- comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  AA_POS2%in%c(49,50,52,150,151,153) & VAR_TYPE!="Nonsense"]

comt_rna_input_cast_mat_cpm_aldex_merge_tis_noncanon_varids<- 
  comt_rna_input_cast_mat_cpm_aldex_merge_tis_noncanon_dt[,as.character(unique(VAR_ID))]

comt_rna_input_cast_mat_cpm_aldex_merge_tis_noncanon_efficiencies<- 
  sapply(comt_rna_input_cast_mat_cpm_aldex_merge_tis_noncanon_varids, get_noncanon_kozak_efficiency,
         comt_rna_input_cast_mat_cpm_aldex_merge_tis_noncanon_dt, endogenous_cyto_tis_flush_split, 
         noncanon_tis_efficiency_dt)

# Fill NAs for those variants not found in the noncanonical set
is.na(comt_rna_input_cast_mat_cpm_aldex_merge_tis_noncanon_efficiencies)<- lengths(comt_rna_input_cast_mat_cpm_aldex_merge_tis_noncanon_efficiencies) == 0

comt_rna_input_cast_mat_cpm_aldex_merge_tis_noncanon_efficiencies_dt<- 
  data.table(VAR_ID=comt_rna_input_cast_mat_cpm_aldex_merge_tis_noncanon_varids, 
             Kozak_efficiency=unlist(comt_rna_input_cast_mat_cpm_aldex_merge_tis_noncanon_efficiencies))

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  ,Noncanonical_Kozak_efficiency:=comt_rna_input_cast_mat_cpm_aldex_merge_tis_noncanon_efficiencies_dt[
    match(comt_rna_input_cast_mat_cpm_aldex_merge_dt[,VAR_ID], VAR_ID), Kozak_efficiency]]

kozak_efficiency_gg<- ggplot(
  data=comt_rna_input_cast_mat_cpm_aldex_merge_dt[!is.na(Noncanonical_Kozak_efficiency)], 
  aes(x=effect, Kozak_efficiency))

kozak_efficiency_gg+
  geom_point(size=2)+
  theme_cowplot()+
  labs(y="Noncanonical Kozak efficiency")

rm(kozak_efficiency_gg)


```


## tRNA abundance 

Determine if there is a relationship between direction or magnitude of differential mRNA abundance and tRNA abundance. Further determine if tRNA abundance is more variable for RNA down versus RNA up variants.

```{r tRNA abundance by hydro-tRNAseq and mim-tRNAseq RNA limma}

trna_fdr_thresh<- loose_fdr_thresh

hydro_tRNAseq_raw_dt<- fread(paste(codon_data_dir, "hydrotrnaseq_raw.txt", sep="/"), sep="\t", header=T)

mim_tRNAseq_raw_dt<- as.data.table(
  read_xlsx(paste(codon_data_dir, "1-s2.0-S1097276521000484-mmc2.xlsx", sep="/"), skip=3, na="NA"))

# Hydro-tRNAseq summarization
hydro_tRNAseq_raw_agg_dt<- hydro_tRNAseq_raw_dt[
  ,.(Sum_count=sum(`Read count`), log2_sum_count=log2(sum(`Read count`))), 
  by=.(Isotype, Anticodon, Codon)]

#' Generates expression entries for wobble codons for Hydro-tRNAseq dataset
#'
#' @param hydrotrnaseq_summarized_dt data.table containing summarized expression values for tRNAs
#' @return data.table original data.table with entries for wobble codons
expand_wobble_pairs<- function(hydrotrnaseq_summarized_dt){
  
  wobble_anticodons<- hydrotrnaseq_summarized_dt[substr(Anticodon, 1, 1)%in%c("G", "T"), unique(Anticodon)]
  
  # For each anticodon with a G or T at the 5' end, there can be one other codon that can pair with it
  n_wobble_anticodons<- length(wobble_anticodons)

  # Initialize the data.table
  wobble_codon_dt<- data.table(Isotype=vector(mode="character", length=n_wobble_anticodons),
                               Anticodon=vector(mode="character", length=n_wobble_anticodons),
                               Codon=vector(mode="character", length=n_wobble_anticodons), 
                               Sum_count=vector(mode="numeric", length=n_wobble_anticodons),
                               log2_sum_count=vector(mode="numeric", length=n_wobble_anticodons))
  
  # Iterate over the anticodons that can form wobble pairing, and create entries for wobble codons
  for(i in 1:length(wobble_anticodons)){
    anticodon<- wobble_anticodons[i]
    wobble_bp<- substr(anticodon, 1, 1)
    
    # These wobble pairing rules are the original described by Crick, which are actually conservative
    if(wobble_bp=="G"){
      wobble_anticodon<- paste0("A", substr(anticodon, 2, 3))
    } else if(wobble_bp=="T"){
      wobble_anticodon<- paste0("C", substr(anticodon, 2, 3))
    }
    
    wobble_codon<- revcomp(wobble_anticodon)
    wobble_aa<- names(AA_CODONS_LIST)[sapply(AA_CODONS_LIST, is_in, wobble_codon)]
    
    # Use the tRNA expression values of the canonical anticodon for the wobble codon
    wobble_codon_dt[i,]<- data.table(Isotype=wobble_aa, Anticodon=anticodon, Codon=wobble_codon, 
                                     Sum_count=hydrotrnaseq_summarized_dt[Anticodon==anticodon, Sum_count],
                                     log2_sum_count=hydrotrnaseq_summarized_dt[Anticodon==anticodon, log2_sum_count])
  }
  
  complete_dt<- rbind(hydrotrnaseq_summarized_dt, wobble_codon_dt, fill=T)
  setkey(complete_dt, Isotype, Anticodon, Codon)
  return(complete_dt)
}

# We accrue 23 codons by taking into account wobble pairs
hydro_tRNAseq_raw_agg_wobble_dt<- expand_wobble_pairs(hydro_tRNAseq_raw_agg_dt)

# Summarize at the level of codon after adding wobble pairs
hydro_tRNAseq_raw_agg_wobble_summ_dt<- hydro_tRNAseq_raw_agg_wobble_dt[
  ,.(Sum_count=sum(Sum_count), log2_sum_count=log2(sum(Sum_count))), 
  by=.(Isotype, Codon)]

# mim-tRNAseq summarization
mim_tRNAseq_raw_dt[,Anticodon:=sapply(strsplit(`Unique tRNA`, "-"), "[", 2)]
mim_tRNAseq_raw_dt[,Isotype:=sapply(strsplit(`Unique tRNA`, "-"), "[", 1)]
mim_tRNAseq_raw_dt[,HEK293T_mean:=(`HEK293T rep 1` + `HEK293T rep 2`)/2]
mim_tRNAseq_raw_dt[,Codon:=revcomp(Anticodon), by=seq_len(nrow(mim_tRNAseq_raw_dt))]
mim_tRNAseq_raw_dt[,Isotype:=names(aa_map)[sapply(aa_map, is_in, Isotype)], 
                   by=seq_len(nrow(mim_tRNAseq_raw_dt))]

mim_tRNAseq_raw_filt_dt<- mim_tRNAseq_raw_dt[!grepl("iMet|eColiLys|SeC|Sup", `Unique tRNA`)]

mim_tRNAseq_raw_filt_agg_dt<- mim_tRNAseq_raw_filt_dt[
  ,.(Sum_count=sum(HEK293T_mean), 
     log2_sum_count=log2(sum(HEK293T_mean)+1), 
     Sum_copy_number=sum(`Gene copy number`)),
  by=.(Isotype, Codon, Anticodon)]


#' Generates expression entries for wobble codons for mim-tRNAseq dataset
#'
#' @param mimtrnaseq_summarized_dt data.table containing summarized expression values for tRNAs
#' @return data.table original data.table with entries for wobble codons
expand_wobble_pairs<- function(mimtrnaseq_summarized_dt){
  
  wobble_anticodons<- mimtrnaseq_summarized_dt[substr(Anticodon, 1, 1)%in%c("G", "T"), unique(Anticodon)]
  
  # For each anticodon with a G or T at the 5' end, there can be one other codon that can pair with it
  n_wobble_anticodons<- length(wobble_anticodons)

  # Initialize the data.table
  wobble_codon_dt<- data.table(Isotype=vector(mode="character", length=n_wobble_anticodons),
                               Anticodon=vector(mode="character", length=n_wobble_anticodons),
                               Codon=vector(mode="character", length=n_wobble_anticodons), 
                               Sum_count=vector(mode="numeric", length=n_wobble_anticodons),
                               log2_sum_count=vector(mode="numeric", length=n_wobble_anticodons),
                               Sum_copy_number=vector(mode="integer", length=n_wobble_anticodons))
  
  # Iterate over the anticodons that can form wobble pairing, and create entries for wobble codons
  for(i in 1:length(wobble_anticodons)){
    anticodon<- wobble_anticodons[i]
    wobble_bp<- substr(anticodon, 1, 1)
    
    # These wobble pairing rules are the original described by Crick, which are actually conservative
    if(wobble_bp=="G"){
      wobble_anticodon<- paste0("A", substr(anticodon, 2, 3))
    } else if(wobble_bp=="T"){
      wobble_anticodon<- paste0("C", substr(anticodon, 2, 3))
    }
    
    wobble_codon<- revcomp(wobble_anticodon)
    wobble_aa<- names(AA_CODONS_LIST)[sapply(AA_CODONS_LIST, is_in, wobble_codon)]
    
    # Use the tRNA expression values of the canonical anticodon for the wobble codon
    wobble_codon_dt[i,]<- data.table(Isotype=wobble_aa, Anticodon=anticodon, Codon=wobble_codon, 
                                     Sum_count=mimtrnaseq_summarized_dt[Anticodon==anticodon, Sum_count],
                                     log2_sum_count=mimtrnaseq_summarized_dt[Anticodon==anticodon, log2_sum_count],
                                     Sum_copy_number=mimtrnaseq_summarized_dt[Anticodon==anticodon, Sum_copy_number])
  }
  
  complete_dt<- rbind(mimtrnaseq_summarized_dt, wobble_codon_dt, fill=T)
  setkey(complete_dt, Isotype, Anticodon, Codon)
  return(complete_dt)
}

mim_tRNAseq_raw_filt_agg_wobble_dt<- expand_wobble_pairs(mim_tRNAseq_raw_filt_agg_dt)

mim_tRNAseq_raw_filt_agg_wobble_nostop_dt<- mim_tRNAseq_raw_filt_agg_wobble_dt[
  !Codon%in%c("TAG", "TAA", "TGA")]

# Finally aggregate at the level of codon 
mim_tRNAseq_raw_filt_agg_wobble_nostop_agg_summ_dt<-
  mim_tRNAseq_raw_filt_agg_wobble_nostop_dt[
    ,.(Sum_count=sum(Sum_count), log2_sum_count=log2(sum(Sum_count)+1), 
       Sum_copy_number=sum(Sum_copy_number)), 
    by=.(Isotype, Codon)]


# Annotate Hydro-tRNAseq abundance
comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  ,REF_CODON_hydro_log2_sum:=hydro_tRNAseq_raw_agg_wobble_summ_dt[
    match(comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,REF_CODON], Codon), log2_sum_count]]

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  ,ALT_CODON_hydro_log2_sum:=hydro_tRNAseq_raw_agg_wobble_summ_dt[
    match(comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,ALT_CODON], Codon), log2_sum_count]]

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  ,hydro_log2_sum_diff:=ALT_CODON_hydro_log2_sum-REF_CODON_hydro_log2_sum]

# Annotate mim-tRNAseq abundance
comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  ,REF_CODON_mim_log2_sum:=mim_tRNAseq_raw_filt_agg_wobble_nostop_agg_summ_dt[
    match(comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,REF_CODON], Codon), log2_sum_count]]

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  ,ALT_CODON_mim_log2_sum:=mim_tRNAseq_raw_filt_agg_wobble_nostop_agg_summ_dt[
    match(comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,ALT_CODON], Codon), log2_sum_count]]

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  ,mim_log2_sum_diff:=ALT_CODON_mim_log2_sum-REF_CODON_mim_log2_sum]

# Plot magnitude of variant effect against tRNA abundance diff.
hydro_trnaseq_dt<- comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
    !is.na(hydro_log2_sum_diff) & adj.P.Val < trna_fdr_thresh]
    
hydro_trnaseq_cor_gg<- ggplot(
  data=hydro_trnaseq_dt, 
  aes(x=hydro_log2_sum_diff, y=logFC, col=adj.P.Val))
    
hydro_trnaseq_cor_gg+
  geom_vline(xintercept=0, col="grey", linetype=2)+
  geom_point()+
  scale_color_viridis(option="A", begin=0.9, end=0.1)+
  labs(x="Hydro-tRNAseq log2(sum),
ALT to REF codon", 
       y="log FC, cDNA to gDNA", 
col="FDR")+
  theme_cowplot()

rm(hydro_trnaseq_cor_gg)

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  !is.na(hydro_log2_sum_diff) & adj.P.Val < trna_fdr_thresh, 
  cor(hydro_log2_sum_diff, logFC, method="spearman")]

# No relationship found by rank-based correlation

# Is there a difference in the variance? Use Levene's test
levene.test(
  y=comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
    !is.na(hydro_log2_sum_diff) & adj.P.Val < trna_fdr_thresh, hydro_log2_sum_diff], 
  group=comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
    !is.na(hydro_log2_sum_diff) & adj.P.Val < trna_fdr_thresh, Direction], 
  kruskal.test=TRUE)

# No significant difference in variance using Hydro-tRNA-seq data


# Verify with mim-tRNAseq data

mim_trna_dt<- comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
    !is.na(mim_log2_sum_diff) & adj.P.Val < trna_fdr_thresh]

mim_trnaseq_cor_gg<- ggplot(
  data=mim_trna_dt, 
  aes(x=mim_log2_sum_diff, y=logFC, col=adj.P.Val))

mim_trnaseq_cor_gg+
  geom_vline(xintercept=0, col="grey", linetype=2)+
  geom_point()+
  scale_color_viridis(option="A", begin=0.9, end=0.1)+
  labs(x="mim-tRNAseq log2(sum),
ALT to REF codon", 
       y="log FC, cDNA to gDNA", 
col="FDR")+
  theme_cowplot()

rm(mim_trnaseq_cor_gg)

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  !is.na(mim_log2_sum_diff) & adj.P.Val < trna_fdr_thresh, 
  cor(mim_log2_sum_diff, logFC, method="spearman")]

# Similarly, no difference in tRNA abundance diff with mim-tRNAseq
levene.test(
  y=comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
    !is.na(mim_log2_sum_diff) & adj.P.Val < trna_fdr_thresh, mim_log2_sum_diff], 
  group=comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
    !is.na(mim_log2_sum_diff) & adj.P.Val < trna_fdr_thresh, Direction], 
  kruskal.test=TRUE)

# However, there was a significant difference in variance for mim-tRNA seq
# between groups. Uncertain if this is real as Hydro-tRNA-seq did not show
# a significant difference


```

```{r tRNA abundance by hydro-tRNAseq and mim-tRNAseq RNA ALDEx2}

trna_fdr_thresh<- comt_rna_fdr_thresh

hydro_tRNAseq_raw_dt<- fread(paste(codon_data_dir, "hydrotrnaseq_raw.txt", sep="/"), sep="\t", header=T)

mim_tRNAseq_raw_dt<- as.data.table(
  read_xlsx(paste(codon_data_dir, "1-s2.0-S1097276521000484-mmc2.xlsx", sep="/"), skip=3, na="NA"))

# Hydro-tRNAseq summarization
hydro_tRNAseq_raw_agg_dt<- hydro_tRNAseq_raw_dt[
  ,.(Sum_count=sum(`Read count`), log2_sum_count=log2(sum(`Read count`))), 
  by=.(Isotype, Anticodon, Codon)]

#' Generates expression entries for wobble codons for Hydro-tRNAseq dataset
#'
#' @param hydrotrnaseq_summarized_dt data.table containing summarized expression values for tRNAs
#' @return data.table original data.table with entries for wobble codons
expand_wobble_pairs<- function(hydrotrnaseq_summarized_dt){
  
  wobble_anticodons<- hydrotrnaseq_summarized_dt[substr(Anticodon, 1, 1)%in%c("G", "T"), unique(Anticodon)]
  
  # For each anticodon with a G or T at the 5' end, there can be one other codon that can pair with it
  n_wobble_anticodons<- length(wobble_anticodons)

  # Initialize the data.table
  wobble_codon_dt<- data.table(Isotype=vector(mode="character", length=n_wobble_anticodons),
                               Anticodon=vector(mode="character", length=n_wobble_anticodons),
                               Codon=vector(mode="character", length=n_wobble_anticodons), 
                               Sum_count=vector(mode="numeric", length=n_wobble_anticodons),
                               log2_sum_count=vector(mode="numeric", length=n_wobble_anticodons))
  
  # Iterate over the anticodons that can form wobble pairing, and create entries for wobble codons
  for(i in 1:length(wobble_anticodons)){
    anticodon<- wobble_anticodons[i]
    wobble_bp<- substr(anticodon, 1, 1)
    
    # These wobble pairing rules are the original described by Crick, which are actually conservative
    if(wobble_bp=="G"){
      wobble_anticodon<- paste0("A", substr(anticodon, 2, 3))
    } else if(wobble_bp=="T"){
      wobble_anticodon<- paste0("C", substr(anticodon, 2, 3))
    }
    
    wobble_codon<- revcomp(wobble_anticodon)
    wobble_aa<- names(AA_CODONS_LIST)[sapply(AA_CODONS_LIST, is_in, wobble_codon)]
    
    # Use the tRNA expression values of the canonical anticodon for the wobble codon
    wobble_codon_dt[i,]<- data.table(Isotype=wobble_aa, Anticodon=anticodon, Codon=wobble_codon, 
                                     Sum_count=hydrotrnaseq_summarized_dt[Anticodon==anticodon, Sum_count],
                                     log2_sum_count=hydrotrnaseq_summarized_dt[Anticodon==anticodon, log2_sum_count])
  }
  
  complete_dt<- rbind(hydrotrnaseq_summarized_dt, wobble_codon_dt, fill=T)
  setkey(complete_dt, Isotype, Anticodon, Codon)
  return(complete_dt)
}

# We accrue 23 codons by taking into account wobble pairs
hydro_tRNAseq_raw_agg_wobble_dt<- expand_wobble_pairs(hydro_tRNAseq_raw_agg_dt)

# Summarize at the level of codon after adding wobble pairs
hydro_tRNAseq_raw_agg_wobble_summ_dt<- hydro_tRNAseq_raw_agg_wobble_dt[
  ,.(Sum_count=sum(Sum_count), log2_sum_count=log2(sum(Sum_count))), 
  by=.(Isotype, Codon)]

# mim-tRNAseq summarization
mim_tRNAseq_raw_dt[,Anticodon:=sapply(strsplit(`Unique tRNA`, "-"), "[", 2)]
mim_tRNAseq_raw_dt[,Isotype:=sapply(strsplit(`Unique tRNA`, "-"), "[", 1)]
mim_tRNAseq_raw_dt[,HEK293T_mean:=(`HEK293T rep 1` + `HEK293T rep 2`)/2]
mim_tRNAseq_raw_dt[,Codon:=revcomp(Anticodon), by=seq_len(nrow(mim_tRNAseq_raw_dt))]
mim_tRNAseq_raw_dt[,Isotype:=names(aa_map)[sapply(aa_map, is_in, Isotype)], 
                   by=seq_len(nrow(mim_tRNAseq_raw_dt))]

mim_tRNAseq_raw_filt_dt<- mim_tRNAseq_raw_dt[!grepl("iMet|eColiLys|SeC|Sup", `Unique tRNA`)]

mim_tRNAseq_raw_filt_agg_dt<- mim_tRNAseq_raw_filt_dt[
  ,.(Sum_count=sum(HEK293T_mean), 
     log2_sum_count=log2(sum(HEK293T_mean)+1), 
     Sum_copy_number=sum(`Gene copy number`)),
  by=.(Isotype, Codon, Anticodon)]


#' Generates expression entries for wobble codons for mim-tRNAseq dataset
#'
#' @param mimtrnaseq_summarized_dt data.table containing summarized expression values for tRNAs
#' @return data.table original data.table with entries for wobble codons
expand_wobble_pairs<- function(mimtrnaseq_summarized_dt){
  
  wobble_anticodons<- mimtrnaseq_summarized_dt[substr(Anticodon, 1, 1)%in%c("G", "T"), unique(Anticodon)]
  
  # For each anticodon with a G or T at the 5' end, there can be one other codon that can pair with it
  n_wobble_anticodons<- length(wobble_anticodons)

  # Initialize the data.table
  wobble_codon_dt<- data.table(Isotype=vector(mode="character", length=n_wobble_anticodons),
                               Anticodon=vector(mode="character", length=n_wobble_anticodons),
                               Codon=vector(mode="character", length=n_wobble_anticodons), 
                               Sum_count=vector(mode="numeric", length=n_wobble_anticodons),
                               log2_sum_count=vector(mode="numeric", length=n_wobble_anticodons),
                               Sum_copy_number=vector(mode="integer", length=n_wobble_anticodons))
  
  # Iterate over the anticodons that can form wobble pairing, and create entries for wobble codons
  for(i in 1:length(wobble_anticodons)){
    anticodon<- wobble_anticodons[i]
    wobble_bp<- substr(anticodon, 1, 1)
    
    # These wobble pairing rules are the original described by Crick, which are actually conservative
    if(wobble_bp=="G"){
      wobble_anticodon<- paste0("A", substr(anticodon, 2, 3))
    } else if(wobble_bp=="T"){
      wobble_anticodon<- paste0("C", substr(anticodon, 2, 3))
    }
    
    wobble_codon<- revcomp(wobble_anticodon)
    wobble_aa<- names(AA_CODONS_LIST)[sapply(AA_CODONS_LIST, is_in, wobble_codon)]
    
    # Use the tRNA expression values of the canonical anticodon for the wobble codon
    wobble_codon_dt[i,]<- data.table(Isotype=wobble_aa, Anticodon=anticodon, Codon=wobble_codon, 
                                     Sum_count=mimtrnaseq_summarized_dt[Anticodon==anticodon, Sum_count],
                                     log2_sum_count=mimtrnaseq_summarized_dt[Anticodon==anticodon, log2_sum_count],
                                     Sum_copy_number=mimtrnaseq_summarized_dt[Anticodon==anticodon, Sum_copy_number])
  }
  
  complete_dt<- rbind(mimtrnaseq_summarized_dt, wobble_codon_dt, fill=T)
  setkey(complete_dt, Isotype, Anticodon, Codon)
  return(complete_dt)
}

mim_tRNAseq_raw_filt_agg_wobble_dt<- expand_wobble_pairs(mim_tRNAseq_raw_filt_agg_dt)

mim_tRNAseq_raw_filt_agg_wobble_nostop_dt<- mim_tRNAseq_raw_filt_agg_wobble_dt[
  !Codon%in%c("TAG", "TAA", "TGA")]

# Finally aggregate at the level of codon 
mim_tRNAseq_raw_filt_agg_wobble_nostop_agg_summ_dt<-
  mim_tRNAseq_raw_filt_agg_wobble_nostop_dt[
    ,.(Sum_count=sum(Sum_count), log2_sum_count=log2(sum(Sum_count)+1), 
       Sum_copy_number=sum(Sum_copy_number)), 
    by=.(Isotype, Codon)]


# Annotate Hydro-tRNAseq abundance
comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  ,REF_CODON_hydro_log2_sum:=hydro_tRNAseq_raw_agg_wobble_summ_dt[
    match(comt_rna_input_cast_mat_cpm_aldex_merge_dt[,REF_CODON], Codon), log2_sum_count]]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  ,ALT_CODON_hydro_log2_sum:=hydro_tRNAseq_raw_agg_wobble_summ_dt[
    match(comt_rna_input_cast_mat_cpm_aldex_merge_dt[,ALT_CODON], Codon), log2_sum_count]]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  ,hydro_log2_sum_diff:=ALT_CODON_hydro_log2_sum-REF_CODON_hydro_log2_sum]

# Annotate mim-tRNAseq abundance
comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  ,REF_CODON_mim_log2_sum:=mim_tRNAseq_raw_filt_agg_wobble_nostop_agg_summ_dt[
    match(comt_rna_input_cast_mat_cpm_aldex_merge_dt[,REF_CODON], Codon), log2_sum_count]]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  ,ALT_CODON_mim_log2_sum:=mim_tRNAseq_raw_filt_agg_wobble_nostop_agg_summ_dt[
    match(comt_rna_input_cast_mat_cpm_aldex_merge_dt[,ALT_CODON], Codon), log2_sum_count]]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  ,mim_log2_sum_diff:=ALT_CODON_mim_log2_sum-REF_CODON_mim_log2_sum]

# Plot magnitude of variant effect against tRNA abundance diff.
hydro_trnaseq_dt<- comt_rna_input_cast_mat_cpm_aldex_merge_dt[
    !is.na(hydro_log2_sum_diff) & we.eBH < trna_fdr_thresh]
    
hydro_trnaseq_cor_gg<- ggplot(
  data=hydro_trnaseq_dt, 
  aes(x=hydro_log2_sum_diff, y=effect, col=we.eBH))
    
hydro_trnaseq_cor_gg+
  geom_vline(xintercept=0, col="grey", linetype=2)+
  geom_point()+
  scale_color_viridis(option="A", begin=0.9, end=0.1)+
  labs(x="Hydro-tRNAseq log2(sum),
ALT to REF codon", 
       y="Effect size, cDNA to gDNA", 
col="FDR")+
  theme_cowplot()

rm(hydro_trnaseq_cor_gg)

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  !is.na(hydro_log2_sum_diff) & we.eBH < trna_fdr_thresh, 
  cor(hydro_log2_sum_diff, effect, method="spearman")]

# No relationship found by rank-based correlation

# Is there a difference in the variance? Use Levene's test
levene.test(
  y=comt_rna_input_cast_mat_cpm_aldex_merge_dt[
    !is.na(hydro_log2_sum_diff) & we.eBH < trna_fdr_thresh, hydro_log2_sum_diff], 
  group=comt_rna_input_cast_mat_cpm_aldex_merge_dt[
    !is.na(hydro_log2_sum_diff) & we.eBH < trna_fdr_thresh, Direction], 
  kruskal.test=TRUE)

# No significant difference in variance using Hydro-tRNA-seq data


# Verify with mim-tRNAseq data

mim_trna_dt<- comt_rna_input_cast_mat_cpm_aldex_merge_dt[
    !is.na(mim_log2_sum_diff) & we.eBH < trna_fdr_thresh]

mim_trnaseq_cor_gg<- ggplot(
  data=mim_trna_dt, 
  aes(x=mim_log2_sum_diff, y=effect, col=we.eBH))

mim_trnaseq_cor_gg+
  geom_vline(xintercept=0, col="grey", linetype=2)+
  geom_point()+
  scale_color_viridis(option="A", begin=0.9, end=0.1)+
  labs(x="mim-tRNAseq log2(sum),
ALT to REF codon", 
       y="Effect size, cDNA to gDNA", 
col="FDR")+
  theme_cowplot()

rm(mim_trnaseq_cor_gg)

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  !is.na(mim_log2_sum_diff) & we.eBH < trna_fdr_thresh, 
  cor(mim_log2_sum_diff, effect, method="spearman")]

# Similarly, no difference in tRNA abundance diff with mim-tRNAseq
levene.test(
  y=comt_rna_input_cast_mat_cpm_aldex_merge_dt[
    !is.na(mim_log2_sum_diff) & we.eBH < trna_fdr_thresh, mim_log2_sum_diff], 
  group=comt_rna_input_cast_mat_cpm_aldex_merge_dt[
    !is.na(mim_log2_sum_diff) & we.eBH < trna_fdr_thresh, Direction], 
  kruskal.test=TRUE)

# However, there was a significant difference in variance for mim-tRNA seq
# between groups. Uncertain if this is real as Hydro-tRNA-seq did not show
# a significant difference


```


## Codon stability coefficient

Determine if there is a relationship between direction or magnitude of differential mRNA abundance and codon stability coeffienct (CSC). Compare codon stability for RNA up and RNA down groups using Wu et al. 2019 (Bazzini) and Narula et al. 2019 (Rissland) CSC data for HEK293T cells.

```{r CSC RNA limma}

comt_all_limma_input_cast_eset_fit_eb_top_loose_varids<- 
  comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[adj.P.Val < loose_fdr_thresh, VAR_ID]

# Annotate the CSC scores for REF and ALT codons for both Wu and Narula et al. 2019 datasets

# Wu et al. 2019
csc_wu_2019<- paste(codon_data_dir, "elife-45396-fig1-data2-v2.csv", sep="/")
csc_wu_2019_dt<- fread(csc_wu_2019, sep=",", header=TRUE)
csc_wu_2019_dt[,Codon:=codon]

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  ,Wu_REF_CSC:=csc_wu_2019_dt[
    match(comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,REF_CODON], codon), `293T_ORFome`]]

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  ,Wu_ALT_CSC:=csc_wu_2019_dt[
    match(comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,ALT_CODON], codon), `293T_ORFome`]]

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  ,Wu_CSC_diff:=Wu_ALT_CSC-Wu_REF_CSC]


# Narula et al. 2019
csc_narula_2019<- paste(codon_data_dir, "Supplemental_Table_S5.xlsx", sep="/")
csc_narula_2019_dt<- as.data.table(read_xlsx(csc_narula_2019, skip=5))
csc_narula_2019_dt[,Codon:=toupper(Codon)]
csc_narula_2019_dt[,ORFome_CSC_mean:=(
  `ORF14 CSC` + `ORF DMSO CSC` + `ORF23 CSC` + `ORF56 CSC`)/4]

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  ,Narula_REF_CSC:=csc_narula_2019_dt[
    match(comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,REF_CODON], Codon), ORFome_CSC_mean]]

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  ,Narula_ALT_CSC:=csc_narula_2019_dt[
    match(comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,ALT_CODON], Codon), ORFome_CSC_mean]]

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  ,Narula_CSC_diff:=Narula_ALT_CSC-Narula_REF_CSC]

# Compare CSC between groups
if("Significant"%in%names(comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt)){
  comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,Significant:=NULL]
}

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  VAR_ID%in%comt_all_limma_input_cast_eset_fit_eb_top_loose_varids,
  Significant:="True"]

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  is.na(Significant), Significant:="False"]

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  logFC>0 & Significant=="True", Direction:="RNA up"]

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  logFC<0 & Significant=="True", Direction:="RNA down"]

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  Significant=="False", Direction:="N.S."]

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  ,Direction:=factor(Direction, levels=c("N.S.", "RNA down", "RNA up"))]

# Test if there are differences by one-sided Wilcoxon rank sum tests
# RNA down should come before RNA up for alternative=less

# Wu et al. 2019 data
# RNA up versus RNA down
comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  !is.na(Wu_CSC_diff) & Significant=="True", wilcox.test(
    formula=Wu_CSC_diff~Direction, data=.SD, alternative="less")]

# RNA down versus N.S.
comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  !is.na(Wu_CSC_diff) & Direction!="RNA up", wilcox.test(
    formula=Wu_CSC_diff~Direction, data=.SD, alternative="greater")]

# RNA up versus N.S.
comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  !is.na(Wu_CSC_diff) & Direction!="RNA down", wilcox.test(
    formula=Wu_CSC_diff~Direction, data=.SD, alternative="less")]

# Narula et al. 2019 data
# RNA up versus RNA down
comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  !is.na(Narula_CSC_diff) & Significant=="True", wilcox.test(
    formula=Narula_CSC_diff~Direction, data=.SD, alternative="less")]

# RNA down versus N.S.
comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  !is.na(Narula_CSC_diff) & Direction!="RNA up", wilcox.test(
    formula=Narula_CSC_diff~Direction, data=.SD, alternative="greater")]

# RNA up versus N.S.
comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  !is.na(Narula_CSC_diff) & Direction!="RNA down", wilcox.test(
    formula=Narula_CSC_diff~Direction, data=.SD, alternative="less")]

# Tests are significant with Wu et al. 2019 at p < 0.05
# At this same alpha threshold no tests are significant with Narula et al. 2019

# Is the same correlation observed between Wu and Narula datasets?
csc_merge_dt<- merge(csc_wu_2019_dt, csc_narula_2019_dt, by="Codon")

csc_cor_gg<- ggplot(
  data=csc_merge_dt, aes(x=`293T_ORFome`, y=ORFome_CSC_mean))

csc_cor_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point()+
  geom_smooth(method="lm")+
  labs(x="Wu et al. 2019 CSC", y="Narula et al. 2019 CSC")+
  theme_cowplot()

rm(csc_cor_gg)

#csc_merge_dt[,cor(x=`293T_ORFome`, y=ORFome_CSC_mean, method="pearson")]
# 0.8397968

#csc_merge_dt[,cor(x=`293T_ORFome`, y=ORFome_CSC_mean, method="spearman")]
# 0.8138022


# Plot CSC by cDNA effect direction
jitter_pos<- position_jitter(width=0.25, seed=9)

# Label CSC diff outlier points
if("Label"%in%names(comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt)){
  comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,Label:=NULL]
}

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  Direction!="N.S." & (Wu_CSC_diff>0.12 | Wu_CSC_diff<(-0.12)), 
  Label:=gsub("p.", "", AA_CHANGE)]

# Wu et al. 2019
# !AA_POS%in%seq(62,63)
csc_gg<- ggplot(
  data=comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt,
  aes(x=Direction, y=Wu_CSC_diff, label=Label))

# title="Wu et al. 2019, 293T ORFome"

csc_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot(width=0.5, outlier.shape=NA)+
  geom_jitter(data=comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
    Direction!="N.S."], aes(x=Direction, y=Wu_CSC_diff), position=jitter_pos)+
  geom_text_repel(data=comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
    Direction!="N.S."], aes(x=Direction, y=Wu_CSC_diff), box.padding=0.7, position=jitter_pos)+
  labs(x="", y="Alt. - Ref. CSC", col="FDR")+
  scale_color_viridis(option="A", begin=0.9, end=0.1)+
  theme_cowplot()


# Narula et al. 2019
if("Label"%in%names(comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt)){
  comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,Label:=NULL]
}

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  Direction!="N.S." & (Narula_CSC_diff>0.18 | Narula_CSC_diff<(-0.22)), 
  Label:=gsub("p.", "", AA_CHANGE)]

csc_gg<- ggplot(
  data=comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt,
  aes(x=Direction, y=Narula_CSC_diff, label=Label))

# title="Narula et al. 2019, 293T ORFome"
csc_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot(width=0.5, fill="steelblue", alpha=0.75, outlier.shape=NA)+
  geom_jitter(data=comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
    Direction!="N.S."], aes(x=Direction, y=Narula_CSC_diff), position=jitter_pos)+
  geom_text_repel(data=comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
    Direction!="N.S."], aes(x=Direction, y=Narula_CSC_diff), box.padding=0.7, position=jitter_pos)+
  labs(x="", y="Alt. - Ref. CSC", col="FDR")+
  scale_color_viridis(option="A", begin=0.9, end=0.1)+
  theme_cowplot()

rm(csc_gg)


```

```{r CSC RNA ALDEx2}

comt_rna_input_cast_mat_cpm_aldex_var_ids<- comt_rna_input_cast_mat_cpm_aldex_merge_dt[we.eBH < comt_rna_fdr_thresh, VAR_ID]

# Annotate the CSC scores for REF and ALT codons for both Wu and Narula et al. 2019 datasets

# Wu et al. 2019
csc_wu_2019<- paste(codon_data_dir, "elife-45396-fig1-data2-v2.csv", sep="/")
csc_wu_2019_dt<- fread(csc_wu_2019, sep=",", header=TRUE)
csc_wu_2019_dt[,Codon:=codon]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  ,Wu_REF_CSC:=csc_wu_2019_dt[
    match(comt_rna_input_cast_mat_cpm_aldex_merge_dt[,REF_CODON], codon), `293T_ORFome`]]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  ,Wu_ALT_CSC:=csc_wu_2019_dt[
    match(comt_rna_input_cast_mat_cpm_aldex_merge_dt[,ALT_CODON], codon), `293T_ORFome`]]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  ,Wu_CSC_diff:=Wu_ALT_CSC-Wu_REF_CSC]


# Narula et al. 2019
csc_narula_2019<- paste(codon_data_dir, "Supplemental_Table_S5.xlsx", sep="/")
csc_narula_2019_dt<- as.data.table(read_xlsx(csc_narula_2019, skip=5))
csc_narula_2019_dt[,Codon:=toupper(Codon)]
csc_narula_2019_dt[,ORFome_CSC_mean:=(
  `ORF14 CSC` + `ORF DMSO CSC` + `ORF23 CSC` + `ORF56 CSC`)/4]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  ,Narula_REF_CSC:=csc_narula_2019_dt[
    match(comt_rna_input_cast_mat_cpm_aldex_merge_dt[,REF_CODON], Codon), ORFome_CSC_mean]]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  ,Narula_ALT_CSC:=csc_narula_2019_dt[
    match(comt_rna_input_cast_mat_cpm_aldex_merge_dt[,ALT_CODON], Codon), ORFome_CSC_mean]]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  ,Narula_CSC_diff:=Narula_ALT_CSC-Narula_REF_CSC]

# Compare CSC between groups
if("Significant"%in%names(comt_rna_input_cast_mat_cpm_aldex_merge_dt)){
  comt_rna_input_cast_mat_cpm_aldex_merge_dt[,Significant:=NULL]
}

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  VAR_ID%in%comt_rna_input_cast_mat_cpm_aldex_var_ids,
  Significant:="True"]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  is.na(Significant), Significant:="False"]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  effect>0 & Significant=="True", Direction:="RNA up"]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  effect<0 & Significant=="True", Direction:="RNA down"]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  Significant=="False", Direction:="N.S."]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  ,Direction:=factor(Direction, levels=c("N.S.", "RNA down", "RNA up"))]

# Test if there are differences by one-sided Wilcoxon rank sum tests
# RNA down should come before RNA up for alternative=less

# Wu et al. 2019 data
# RNA up versus RNA down
comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  !is.na(Wu_CSC_diff) & Significant=="True", wilcox.test(
    formula=Wu_CSC_diff~Direction, data=.SD, alternative="less")]

# RNA down versus N.S.
comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  !is.na(Wu_CSC_diff) & Direction!="RNA up", wilcox.test(
    formula=Wu_CSC_diff~Direction, data=.SD, alternative="greater")]

# RNA up versus N.S.
comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  !is.na(Wu_CSC_diff) & Direction!="RNA down", wilcox.test(
    formula=Wu_CSC_diff~Direction, data=.SD, alternative="less")]

# Narula et al. 2019 data
# RNA up versus RNA down
comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  !is.na(Narula_CSC_diff) & Significant=="True", wilcox.test(
    formula=Narula_CSC_diff~Direction, data=.SD, alternative="less")]

# RNA down versus N.S.
comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  !is.na(Narula_CSC_diff) & Direction!="RNA up", wilcox.test(
    formula=Narula_CSC_diff~Direction, data=.SD, alternative="greater")]

# RNA up versus N.S.
comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  !is.na(Narula_CSC_diff) & Direction!="RNA down", wilcox.test(
    formula=Narula_CSC_diff~Direction, data=.SD, alternative="less")]

# Tests are significant with Wu et al. 2019 at p < 0.05
# At this same alpha threshold no tests are significant with Narula et al. 2019

# Is the same correlation observed between Wu and Narula datasets?
csc_merge_dt<- merge(csc_wu_2019_dt, csc_narula_2019_dt, by="Codon")

csc_cor_gg<- ggplot(
  data=csc_merge_dt, aes(x=`293T_ORFome`, y=ORFome_CSC_mean))

csc_cor_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point()+
  geom_smooth(method="lm")+
  labs(x="Wu et al. 2019 CSC", y="Narula et al. 2019 CSC")+
  theme_cowplot()

#csc_merge_dt[,cor(x=`293T_ORFome`, y=ORFome_CSC_mean, method="pearson")]
# 0.8397968

#csc_merge_dt[,cor(x=`293T_ORFome`, y=ORFome_CSC_mean, method="spearman")]
# 0.8138022


# Plot CSC by cDNA effect direction
jitter_pos<- position_jitter(width=0.25, seed=9)

# Label CSC diff outlier points
if("Label"%in%names(comt_rna_input_cast_mat_cpm_aldex_merge_dt)){
  comt_rna_input_cast_mat_cpm_aldex_merge_dt[,Label:=NULL]
}

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  Direction!="N.S." & (Wu_CSC_diff>0.12 | Wu_CSC_diff<(-0.12)), 
  Label:=gsub("p.", "", AA_CHANGE)]

# Wu et al. 2019
# !AA_POS%in%seq(62,63)
csc_gg<- ggplot(
  data=comt_rna_input_cast_mat_cpm_aldex_merge_dt,
  aes(x=Direction, y=Wu_CSC_diff, label=Label))

# title="Wu et al. 2019, 293T ORFome"

csc_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot(width=0.5, fill="steelblue", alpha=0.75, outlier.shape=NA)+
  geom_jitter(data=comt_rna_input_cast_mat_cpm_aldex_merge_dt[
    Direction!="N.S."], aes(x=Direction, y=Wu_CSC_diff), position=jitter_pos)+
  geom_text_repel(data=comt_rna_input_cast_mat_cpm_aldex_merge_dt[
    Direction!="N.S."], aes(x=Direction, y=Wu_CSC_diff), box.padding=0.7, position=jitter_pos)+
  labs(x="", y="Alt. - Ref. CSC", col="FDR")+
  scale_color_viridis(option="A", begin=0.9, end=0.1)+
  theme_cowplot()


# Narula et al. 2019
if("Label"%in%names(comt_rna_input_cast_mat_cpm_aldex_merge_dt)){
  comt_rna_input_cast_mat_cpm_aldex_merge_dt[,Label:=NULL]
}

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  Direction!="N.S." & (Narula_CSC_diff>0.18 | Narula_CSC_diff<(-0.22)), 
  Label:=gsub("p.", "", AA_CHANGE)]

csc_gg<- ggplot(
  data=comt_rna_input_cast_mat_cpm_aldex_merge_dt,
  aes(x=Direction, y=Narula_CSC_diff, label=Label))

# title="Narula et al. 2019, 293T ORFome"
csc_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot(width=0.5, fill="steelblue", alpha=0.75, outlier.shape=NA)+
  geom_jitter(data=comt_rna_input_cast_mat_cpm_aldex_merge_dt[
    Direction!="N.S."], aes(x=Direction, y=Narula_CSC_diff), position=jitter_pos)+
  geom_text_repel(data=comt_rna_input_cast_mat_cpm_aldex_merge_dt[
    Direction!="N.S."], aes(x=Direction, y=Narula_CSC_diff), box.padding=0.7, position=jitter_pos)+
  labs(x="", y="Alt. - Ref. CSC", col="FDR")+
  scale_color_viridis(option="A", begin=0.9, end=0.1)+
  theme_cowplot()

```


## Ribosome dwell time

```{r Eichhorn 2014 Ribo-seq dwell time limma}

# Determine if there associations with ribosome dwell time
# using Narula et al. 2019 (from Eichhorn et al. 2014 Ribo-seq data)
# Wu and Green

# Add the A- and P-site dwell time data from Narula et al. 2019
# derived from Eichhorn et al. 2014
comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  ,Narula_ALT_Asite:=csc_narula_2019_dt[
    match(comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,ALT_CODON], Codon), HEK293T]]

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  ,Narula_REF_Asite:=csc_narula_2019_dt[
    match(comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,REF_CODON], Codon), HEK293T]]

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  ,Narula_ALT_Psite:=csc_narula_2019_dt[
    match(comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,ALT_CODON], Codon), `HEK293T P site`]]

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  ,Narula_REF_Psite:=csc_narula_2019_dt[
    match(comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,REF_CODON], Codon), `HEK293T P site`]]

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,Narula_Asite_diff:=Narula_ALT_Asite-Narula_REF_Asite]
comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,Narula_Psite_diff:=Narula_ALT_Psite-Narula_REF_Psite]

# A site dwell time
if("Label"%in%names(comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt)){
  comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,Label:=NULL]
}

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  Direction!="N.S." & (Narula_Asite_diff>0.68 | Narula_Asite_diff<(-0.75)), 
  Label:=gsub("p.", "", AA_CHANGE)]

asite_dwell_gg<- ggplot(
  data=comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt,
  aes(x=Direction, y=Narula_Asite_diff, label=Label))

# title="Narula et al. 2019, 293T A-site dwell time"
asite_dwell_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot(width=0.5, fill="steelblue", alpha=0.75, outlier.shape=NA)+
  geom_jitter(data=comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
    Direction!="N.S."], aes(x=Direction, y=Narula_Asite_diff), position=jitter_pos)+
  geom_text_repel(data=comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
    Direction!="N.S."], aes(x=Direction, y=Narula_Asite_diff), box.padding=0.7, position=jitter_pos)+
  labs(x="", y="Alt. - Ref. RPF norm. score", col="FDR")+
  scale_color_viridis(option="A", begin=0.9, end=0.1)+
  theme_cowplot()

# RNA up versus RNA down
comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  !is.na(Narula_Asite_diff) & Significant=="True", wilcox.test(
    formula=Narula_Asite_diff~Direction, data=.SD, alternative="greater")]

# RNA down versus N.S.
comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  !is.na(Narula_Asite_diff) & Direction!="RNA up", wilcox.test(
    formula=Narula_Asite_diff~Direction, data=.SD, alternative="less")]

# RNA up versus N.S.
comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  !is.na(Narula_Asite_diff) & Direction!="RNA down", wilcox.test(
    formula=Narula_Asite_diff~Direction, data=.SD, alternative="greater")]

# RNA up versus RNA down as well as RNA down versus N.S. were 
# significant at p < 0.05


# P site dwell time
comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,Label:=NULL]

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  Direction!="N.S." & (Narula_Psite_diff>0.4 | Narula_Psite_diff<(-0.4)), 
  Label:=gsub("p.", "", AA_CHANGE)]

psite_dwell_gg<- ggplot(
  data=comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt,
  aes(x=Direction, y=Narula_Psite_diff, label=Label))

psite_dwell_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot(width=0.5,  fill="steelblue", alpha=0.75, outlier.shape=NA)+
  geom_jitter(data=comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
    Direction!="N.S."], aes(x=Direction, y=Narula_Psite_diff), position=jitter_pos)+
  geom_text_repel(data=comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
    Direction!="N.S."], aes(x=Direction, y=Narula_Psite_diff), box.padding=0.7, position=jitter_pos)+
  labs(x="", y="Alt. - Ref. norm. score", col="FDR", title="Narula et al. 2019, 293T P-site dwell time")+
  scale_color_viridis(option="A", begin=0.9, end=0.1)+
  theme_cowplot()

# No relationship with P site dwell time 

```

```{r Eichhorn 2014 Ribo-seq dwell time ALDEx2}

# Determine if there associations with ribosome dwell time
# using Narula et al. 2019 (from Eichhorn et al. 2014 Ribo-seq data)
# Wu and Green

# Add the A- and P-site dwell time data from Narula et al. 2019
# derived from Eichhorn et al. 2014
comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  ,Narula_ALT_Asite:=csc_narula_2019_dt[
    match(comt_rna_input_cast_mat_cpm_aldex_merge_dt[,ALT_CODON], Codon), HEK293T]]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  ,Narula_REF_Asite:=csc_narula_2019_dt[
    match(comt_rna_input_cast_mat_cpm_aldex_merge_dt[,REF_CODON], Codon), HEK293T]]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  ,Narula_ALT_Psite:=csc_narula_2019_dt[
    match(comt_rna_input_cast_mat_cpm_aldex_merge_dt[,ALT_CODON], Codon), `HEK293T P site`]]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  ,Narula_REF_Psite:=csc_narula_2019_dt[
    match(comt_rna_input_cast_mat_cpm_aldex_merge_dt[,REF_CODON], Codon), `HEK293T P site`]]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[,Narula_Asite_diff:=Narula_ALT_Asite-Narula_REF_Asite]
comt_rna_input_cast_mat_cpm_aldex_merge_dt[,Narula_Psite_diff:=Narula_ALT_Psite-Narula_REF_Psite]

# A site dwell time
if("Label"%in%names(comt_rna_input_cast_mat_cpm_aldex_merge_dt)){
  comt_rna_input_cast_mat_cpm_aldex_merge_dt[,Label:=NULL]
}

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  Direction!="N.S." & (Narula_Asite_diff>0.68 | Narula_Asite_diff<(-0.75)), 
  Label:=gsub("p.", "", AA_CHANGE)]

asite_dwell_gg<- ggplot(
  data=comt_rna_input_cast_mat_cpm_aldex_merge_dt,
  aes(x=Direction, y=Narula_Asite_diff, label=Label))

asite_dwell_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot(width=0.5, fill="steelblue", alpha=0.75, outlier.shape=NA)+
  geom_jitter(data=comt_rna_input_cast_mat_cpm_aldex_merge_dt[
    Direction!="N.S."], aes(x=Direction, y=Narula_Asite_diff), position=jitter_pos)+
  geom_text_repel(data=comt_rna_input_cast_mat_cpm_aldex_merge_dt[
    Direction!="N.S."], aes(x=Direction, y=Narula_Asite_diff), box.padding=0.7, position=jitter_pos)+
  labs(x="", y="Alt. - Ref. RPF norm. score", col="FDR", title="Narula et al. 2019, 293T A-site dwell time")+
  scale_color_viridis(option="A", begin=0.9, end=0.1)+
  theme_cowplot()

rm(asite_dwell_gg)

# RNA up versus RNA down
comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  !is.na(Narula_Asite_diff) & Significant=="True", wilcox.test(
    formula=Narula_Asite_diff~Direction, data=.SD, alternative="greater")]

# RNA down versus N.S.
comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  !is.na(Narula_Asite_diff) & Direction!="RNA up", wilcox.test(
    formula=Narula_Asite_diff~Direction, data=.SD, alternative="less")]

# RNA up versus N.S.
comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  !is.na(Narula_Asite_diff) & Direction!="RNA down", wilcox.test(
    formula=Narula_Asite_diff~Direction, data=.SD, alternative="greater")]

# RNA up versus RNA down as well as RNA down versus N.S. were 
# significant at p < 0.05


# P site dwell time
comt_rna_input_cast_mat_cpm_aldex_merge_dt[,Label:=NULL]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  Direction!="N.S." & (Narula_Psite_diff>0.4 | Narula_Psite_diff<(-0.4)), 
  Label:=gsub("p.", "", AA_CHANGE)]

psite_dwell_gg<- ggplot(
  data=comt_rna_input_cast_mat_cpm_aldex_merge_dt,
  aes(x=Direction, y=Narula_Psite_diff, label=Label))

psite_dwell_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot(width=0.5,  fill="steelblue", alpha=0.75, outlier.shape=NA)+
  geom_jitter(data=comt_rna_input_cast_mat_cpm_aldex_merge_dt[
    Direction!="N.S."], aes(x=Direction, y=Narula_Psite_diff), position=jitter_pos)+
  geom_text_repel(data=comt_rna_input_cast_mat_cpm_aldex_merge_dt[
    Direction!="N.S."], aes(x=Direction, y=Narula_Psite_diff), box.padding=0.7, position=jitter_pos)+
  labs(x="", y="Alt. - Ref. norm. score", col="FDR", title="Narula et al. 2019, 293T P-site dwell time")+
  scale_color_viridis(option="A", begin=0.9, end=0.1)+
  theme_cowplot()

rm(psite_dwell_gg)

# No relationship with P site dwell time 

```

```{r Wu 2019 Ribo-seq dwell time ALDEx2}

# Determine if there associations with ribosome dwell time
# Wu and Green
wu_dwell_time_21nt_dt<- fread(paste(
  codon_data_dir, "21ntRPF_occu.csv", sep="/"), sep=",", header=T)

wu_dwell_time_21nt_dt[,RPF_length:="21"]

wu_dwell_time_28nt_dt<- fread(paste(
  codon_data_dir, "28ntRPF_occu.csv", sep="/"), sep=",", header=T)

wu_dwell_time_28nt_dt[,RPF_length:="28"]

wu_dwell_time_dt<- rbind(wu_dwell_time_21nt_dt, wu_dwell_time_28nt_dt)
wu_dwell_time_dt[,V9:=NULL]
wu_dwell_time_dt[,RPF_length:=as.factor(RPF_length)]

rm(list=c("wu_dwell_time_21nt_dt", "wu_dwell_time_28nt_dt"))

# Annotate the differential cDNA:gDNA dataset
wu_dwell_time_dt[,CHXTIG:=(CHXTIG1+CHXTIG2+CHXTIG3)/3]
wu_dwell_time_dt[,CHXANS:=(CHXANS1+CHXANS2)/2]
wu_dwell_time_dt[,CHX:=(CHX1+CHX2)/2]

wu_dwell_time_21nt<- wu_dwell_time_dt[RPF_length=="21"]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  ,Wu_REF_21nt_CHXTIG:=wu_dwell_time_21nt[
    match(comt_rna_input_cast_mat_cpm_aldex_merge_dt[,REF_CODON], codon), CHXTIG]]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  ,Wu_ALT_21nt_CHXTIG:=wu_dwell_time_21nt[
    match(comt_rna_input_cast_mat_cpm_aldex_merge_dt[,ALT_CODON], codon), CHXTIG]]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  ,Wu_21nt_CHXTIG_diff:=Wu_ALT_21nt_CHXTIG-Wu_REF_21nt_CHXTIG]

# RNA up versus RNA down
comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  !is.na(Wu_21nt_CHXTIG_diff) & Significant=="True", wilcox.test(
    formula=Wu_21nt_CHXTIG_diff~Direction, data=.SD, alternative="greater")]

# A site dwell time with 21 nt RPFs, CHX+TIG
if("Label"%in%names(comt_rna_input_cast_mat_cpm_aldex_merge_dt)){
  comt_rna_input_cast_mat_cpm_aldex_merge_dt[,Label:=NULL]
}

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  Direction!="N.S." & (Wu_21nt_CHXTIG_diff>2 | Wu_21nt_CHXTIG_diff<(-2)), 
  Label:=gsub("p.", "", AA_CHANGE)]

asite_dwell_gg<- ggplot(
  data=comt_rna_input_cast_mat_cpm_aldex_merge_dt,
  aes(x=Direction, y=Wu_21nt_CHXTIG_diff, label=Label))

# title="Wu et al. 2019, 21 nt CHX+TIG"
asite_dwell_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot(width=0.5, outlier.shape=NA)+
  geom_jitter(data=comt_rna_input_cast_mat_cpm_aldex_merge_dt[
    Direction!="N.S."], aes(x=Direction, y=Wu_21nt_CHXTIG_diff, col=we.eBH), position=jitter_pos)+
  geom_text_repel(data=comt_rna_input_cast_mat_cpm_aldex_merge_dt[
    Direction!="N.S."], aes(x=Direction, y=Wu_21nt_CHXTIG_diff), box.padding=0.7, position=jitter_pos)+
  labs(x="", y="Alt. - Ref. RPF norm. score", col="FDR")+
  scale_color_viridis(option="A", begin=0.9, end=0.1)+
  theme_cowplot()

rm(asite_dwell_gg)

```

Comparison to Wu 2019 CHX+TIG 21 nt RFP normalized densities indicates an unexpected reverse trend to the Narula 2019 analysis (data from Eichhorn 2014).

See also:

Gobet et al. 2020: Robust landscapes of ribosome dwell times and aminoacyl-tRNAs in response to nutrient stress in liver

Ingolia et al. 2011: Ribosome profiling of mouse embryonic stem cells reveals the complexity and dynamics of mammalian proteomes

```{r CSC of native COMT coding sequence}

flag_comt_moxgfp_cds<- "ATGGACTACAAAGACGATGACGACAAGccggaggccccgcctctgctgttggcagctgtgttgctgggcctggtgctgctggtggtgctgctgctgcttctgaggcactggggctggggcctgtgccttatcggctggaacgagttcatcctgcagcccatccacaacctgctcatgggtgacaccaaggagcagcgcatcctgaaccacgtgctgcagcatgcggagcccgggaacgcacagagcgtgctggaggccattgacacctactgcgagcagaaggagtgggccatgaacgtgggcgacaagaaaggcaagatcgtggacgccgtgattcaggagcaccagccctccgtgctgctggagctgggggcctactgtggctactcagctgtgcgcatggcccgcctgctgtcaccaggggcgaggctgatcaccatcgagatcaaccccgactgtgccgccatcacccagcggatggtggatttcgctggcgtgaaggacaaggtcacccttgtggttggagcgtcccaggacatcatcccccagctgaagaagaagtatgatgtggacacactggacatggtcttcctcgaccactggaaggaccggtacctgccggacacgcttctcttggaggaatgtggcctgctgcggaaggggacagtgctactggctgacaacgtgatctgcccaggtgcgccagacttcctagcacacgtgcgcgggagcagctgctttgagtgcacacactaccaatcgttcctggaatacagggaggtggtggacggcctggagaaggccatctacaagggcccaggcagcgaagcagggcccGTGTCCAAGGGCGAGGAGCTGTTCACCGGGGTGGTGCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCTCCGTGCGGGGCGAGGGCGAGGGCGATGCCACCAACGGCAAGCTGACCCTGAAGTTCATCAGCACCACCGGCAAGCTGCCCGTGCCCTGGCCCACCCTCGTGACCACCCTGACCTACGGCGTGCAGAGCTTCTCCCGCTACCCCGACCACATGAAGCGCCACGACTTCTTCAAGAGCGCCATGCCCGAAGGCTACGTCCAGGAGCGCACCATCTCCTTCAAGGACGACGGCACCTACAAGACCCGCGCCGAGGTGAAGTTCGAGGGCGACACCCTGGTGAACCGCATCGAGCTGAAGGGCATCGACTTCAAGGAGGACGGCAACATCCTGGGGCACAAGCTGGAGTACAACTTCAACTCCCACAACGTCTATATCACCGCCGACAAGCAGAAGAACGGCATCAAGGCCAACTTCAAGATCCGCCACAACGTGGAGGACGGCTCCGTGCAGCTCGCCGACCACTACCAGCAGAACACCCCCATCGGCGACGGCCCCGTGCTGCTGCCCGACAACCACTACCTGTCCACCCAGTCCAAGCTGTCCAAAGACCCCAACGAGAAGCGCGATCACATGGTCCTTCTGGAATTCGTGACCGCCGCCGGGATCACTCACGGCATGGACGAGCTGTACAAGTGA"

comt_cds<- "ATGccggaggccccgcctctgctgttggcagctgtgttgctgggcctggtgctgctggtggtgctgctgctgcttctgaggcactggggctggggcctgtgccttatcggctggaacgagttcatcctgcagcccatccacaacctgctcatgggtgacaccaaggagcagcgcatcctgaaccacgtgctgcagcatgcggagcccgggaacgcacagagcgtgctggaggccattgacacctactgcgagcagaaggagtgggccatgaacgtgggcgacaagaaaggcaagatcgtggacgccgtgattcaggagcaccagccctccgtgctgctggagctgggggcctactgtggctactcagctgtgcgcatggcccgcctgctgtcaccaggggcgaggctgatcaccatcgagatcaaccccgactgtgccgccatcacccagcggatggtggatttcgctggcgtgaaggacaaggtcacccttgtggttggagcgtcccaggacatcatcccccagctgaagaagaagtatgatgtggacacactggacatggtcttcctcgaccactggaaggaccggtacctgccggacacgcttctcttggaggaatgtggcctgctgcggaaggggacagtgctactggctgacaacgtgatctgcccaggtgcgccagacttcctagcacacgtgcgcgggagcagctgctttgagtgcacacactaccaatcgttcctggaatacagggaggtggtggacggcctggagaaggccatctacaagggcccaggcagcgaagcagggcccTGA"

# ROI_1 is 40-74
# ROI_2 is 136-158 
roi1_aas<- seq(40,74)
roi2_aas<- seq(136,158)

comt_cds_split<- unlist(strsplit(toupper(comt_cds), ""))

get_per_codon_data<- function(cds_seq, measure_dt, measure_name){
  
  cds_seq_split<- unlist(strsplit(toupper(cds_seq), ""))
  codon_indices<- seq(1, length(cds_seq_split) - 3, 3)
  
  codon_vec<- NULL
  res_vec<- NULL
  for(i in codon_indices){
    cds_codon<- paste(cds_seq_split[i:(i+2)], collapse="")
    cds_codon_measure<- unlist(measure_dt[codon==cds_codon, measure_name, with=FALSE])
    codon_vec<- append(codon_vec, cds_codon)
    res_vec<- append(res_vec, cds_codon_measure)
  }
  res_dt<- data.table(Codon=codon_vec, Measure=res_vec)
  return(res_dt)
}

comt_cds_csc_dt<- get_per_codon_data(comt_cds, csc_wu_2019_dt, "293T_ORFome")

comt_cds_csc_dt[,Codon_number:=seq(1:.N)]

comt_cds_csc_gg<- ggplot(
  data=comt_cds_csc_dt, 
  aes(x=Codon_number, y=Measure))

comt_cds_csc_gg+
  geom_line(lwd=0.7)+
  theme_cowplot()+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_vline(xintercept=c(62,66), col="limegreen", linetype=2)+
  labs(x="Codon index", y="293T ORFome CSC")

rm(comt_cds_csc_gg)

comt_cds_csc_dt[Codon_number%in%roi1_aas, .(Mean_CSC=mean(Measure), SD_CSC=sd(Measure))]
comt_cds_csc_dt[Codon_number%in%roi2_aas, .(Mean_CSC=mean(Measure), SD_CSC=sd(Measure))]

# No large difference in mean or SD

```


## Prediction of variant impact on mRNA folding

Compare differential variants with exhaustive mRNA folding effects of SNPs.

```{r Gaither et al. 2021 exhaustive RNA stability of synonymous mutations}

gaither_2021_dt<- fread("/Users/ianhoskins/synon_rna_stability/RNAStability_v10.5.1_hg38_distinct_SURF_SPI_Phred_GitHub_Export_COMT_rehead.tsv", sep="\t", header=T)

gaither_2021_dt[,TYPE:=as.factor(TYPE)]

# Custom convert from NM_000754.3 to pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS, which was used as reference
# 19961199 corresponds to first base of exon 2, which starts at POS 4562 of pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS
# 19962530 corresponds to first base of second codon of membrane isoform
gaither_2021_comt_select_dt<- gaither_2021_dt[Selected_Trans=="NM_000754.3"]
gaither_2021_comt_select_dt[,Genomic_POS:=POS]
gaither_2021_comt_trx_idx_dt<- gaither_2021_comt_select_dt[,.(Genomic_POS=unique(POS))]

# POS will be the pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS transcript position
# Need to account for Flag in the reference
# The first seq() maps the 5' UTR and start codon; the second seq() maps the rest of 
# the coding sequence (omitting the Flag tag)
# Because 3' UTR was mapped by Gaither, generate placeholder positions that do
# not exist in pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS
gaither_2021_comt_trx_idx_dt[,POS:=c(seq(4562,4562+93), seq(4680,4680+1954))]

gaither_2021_comt_select_dt[,POS:=gaither_2021_comt_trx_idx_dt[
  match(gaither_2021_comt_select_dt[,POS], Genomic_POS), POS]]

# Look at global folding statistics across the trx
comt_dMFE_gg<- ggplot(
  data=gaither_2021_comt_select_dt[TYPE!="other"], 
  aes(x=POS, y=deltaMFE))

comt_dMFE_gg+
  facet_wrap(~TYPE, scales="free_x")+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_smooth()+
  theme_cowplot()+
  labs(x="COMT transcript position", y="delta MFE")

rm(comt_dMFE_gg)

# Look at synonymous mutations
comt_dMFE_synon_gg<- ggplot(
  data=gaither_2021_comt_select_dt[TYPE=="synonymous"], 
  aes(x=POS, y=deltaMFE))

comt_dMFE_synon_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_point()+
  geom_smooth()+
  theme_cowplot()+
  labs(x="COMT transcript position", y="delta MFE")

rm(comt_dMFE_synon_gg)

# Missense
comt_dMFE_missense_gg<- ggplot(
  data=gaither_2021_comt_select_dt[TYPE=="missense"], 
  aes(x=POS, y=deltaMFE))

comt_dMFE_missense_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_point()+
  geom_smooth()+
  theme_cowplot()+
  labs(x="COMT transcript position", y="delta MFE")

rm(comt_dMFE_missense_gg)

# Plot SPI norm across the COMT transcript
comt_spi_norm_gg<- ggplot(
  data=gaither_2021_comt_select_dt[TYPE=="synonymous"], 
  aes(x=POS, y=SPI_Norm))

comt_spi_norm_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_point()+
  geom_smooth(method="lm")+
  scale_color_npg()+
  theme_cowplot()+
  labs(x="COMT transcript position", y="Norm. SPI")

rm(comt_spi_norm_gg)

# Now merge stats with the differential variants and test if variants with differential abundance
# show a trend in global folding statistics
comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,SURF_key:=paste(POS, REF, ALT, sep=":")]
gaither_2021_comt_select_dt[,SURF_key:=paste(POS, REF, ALT, sep=":")]

comt_all_limma_input_cast_eset_fit_eb_all_merge_surf_dt<- 
  merge(comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt, 
        gaither_2021_comt_select_dt, by="SURF_key")

# Compare differential to non-differential variants
comt_all_limma_input_cast_eset_fit_eb_all_merge_surf_dt[,Direction:="N.S."]

comt_all_limma_input_cast_eset_fit_eb_all_merge_surf_dt[
  adj.P.Val < loose_fdr_thresh & logFC > 0, Direction:="RNA up"]

comt_all_limma_input_cast_eset_fit_eb_all_merge_surf_dt[
  adj.P.Val < loose_fdr_thresh & logFC < 0, Direction:="RNA down"]

comt_all_limma_input_cast_eset_fit_eb_all_merge_surf_dt[
  ,Direction:=factor(Direction, levels=c("N.S.", "RNA down", "RNA up"))]

# deltaMFE, deltaCFE, deltaCD, SURF, SPI_Norm
limma_vs_surf_gg<- ggplot(
  data=comt_all_limma_input_cast_eset_fit_eb_all_merge_surf_dt, 
  aes(x=Direction, y=SURF))

limma_vs_surf_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot(outlier.shape=NA)+
  geom_jitter(width=0.2)+
  theme_cowplot()+
  labs(x="Direction")

rm(limma_vs_surf_gg)

# Look specifically at the hotspot
limma_vs_surf_hotspot_gg<- ggplot(
  data=comt_all_limma_input_cast_eset_fit_eb_all_merge_surf_dt[
    AA_POS2%in%seq(62,66) & Direction!="RNA down"], 
  aes(x=Direction, y=deltaCD))

limma_vs_surf_hotspot_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot(outlier.shape=NA)+
  geom_jitter(width=0.2)+
  theme_cowplot()+
  labs(x="Direction")

rm(limma_vs_surf_hotspot_gg)

# deltaCFE and deltaCD appear different between N.S. and RNA_up classes

# RNA up versus N.S. for delta MFE at the hotspot
comt_all_limma_input_cast_eset_fit_eb_all_merge_surf_dt[
  AA_POS2%in%seq(62,66) & !is.na(deltaMFE) & Direction!="RNA down", t.test(
    formula=deltaMFE~Direction, data=.SD)]

# RNA up versus N.S. for delta CFE at the hotspot
comt_all_limma_input_cast_eset_fit_eb_all_merge_surf_dt[
  AA_POS2%in%seq(62,66) & !is.na(deltaCFE) & Direction!="RNA down", t.test(
    formula=deltaCFE~Direction, data=.SD)]

# RNA up versus N.S. for delta CD at the hotspot
comt_all_limma_input_cast_eset_fit_eb_all_merge_surf_dt[
  AA_POS2%in%seq(62,66) & !is.na(deltaCD) & Direction!="RNA down", t.test(
    formula=deltaCD~Direction, data=.SD)]

# Compare logFC for variants between regions with altered dMFE
comt_all_limma_input_cast_eset_fit_eb_all_merge_surf_all_dt<- 
  merge(comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt, 
        gaither_2021_comt_select_dt, by="SURF_key", all.y=TRUE)

# ROI_1: 4794, 4898
# ROI_2: 5082, 5150

comt_roi_map_gg<- ggplot(
  data=comt_all_limma_input_cast_eset_fit_eb_all_merge_surf_all_dt[
    TYPE.y%in%c("synonymous", "missense")], 
  aes(x=POS.y, y=deltaMFE))

comt_roi_map_gg+
  geom_rect(aes(xmin=4794, xmax=4898, ymin=-Inf, ymax=Inf), fill="pink")+
  geom_rect(aes(xmin=5082, xmax=5150, ymin=-Inf, ymax=Inf), fill="pink")+
  geom_vline(xintercept=5115, col="yellow")+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_point()+
  geom_smooth(method="loess", span=0.2)+
  theme_cowplot()+
  labs(x="COMT transcript position", y="delta MFE")

rm(comt_roi_map_gg)

# Compare logFC for variants after 5115 to all other variants
comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  ,dMFE_cutoff:=ifelse(POS>5115, TRUE, FALSE)]

# adj.P.Val<fdr_thresh
dmfe_cutoff_gg<- ggplot(
  data=comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[ROI=="ROI_2" & VAR_TYPE!="Nonsense"],
  aes(x=dMFE_cutoff, y=logFC))

dmfe_cutoff_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot()+
  labs(x="POS > 5115", y="log ratio, cDNA to gDNA")+
  theme_cowplot()

rm(dmfe_cutoff_gg)

comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  ROI=="ROI_2" & VAR_TYPE!="Nonsense", t.test(formula=logFC~dMFE_cutoff, data=.SD)]

```

```{r RNAsnp prediction for RNA up variants limma}

comt_insert_offset<- 4561

varid_to_rnasnp<- function(pos, ref, alt, ref_offset=comt_insert_offset){
  
  pos_offset<- pos - ref_offset
  refs<- unlist(strsplit(ref, ""))
  alts<- unlist(strsplit(alt, ""))
  
  if(length(refs) == 1){
    res<- paste(refs, pos_offset, alts, sep="")
    return(res)
  } else {
    
    res<- NULL
    for(i in 1:length(refs)){
      ref_i<- refs[i]
      alt_i<- alts[i]
      
      if(ref_i!=alt_i){
        subst<- paste(ref_i, pos_offset+(i-1), alt_i, sep="")
        res<- c(res, subst)
      }
    }
    final_res<- paste(res, collapse="-")
    return(final_res)
  }
}

# Generate the RNAsnp IDs
comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  ,RNAsnp_ID:=varid_to_rnasnp(POS, REF, ALT), by=seq_len(nrow(
    comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt))]

rnasnp_ids_file<- paste(comt_base_dir, "RNAsnp", "COMT_RNAsnp_IDs.txt", sep="/")

write.table(comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,RNAsnp_ID], 
            rnasnp_ids_file, quote=FALSE, row.names=FALSE, col.names=FALSE)

# Read in RNAsnp results
rnasnp_res_file<- paste(comt_base_dir, "RNAsnp", "COMT_RNAsnp_results.txt", sep="/")
rnasnp_res_dt<- fread(rnasnp_res_file, sep="\t")
names(rnasnp_res_dt)[5]<- "d_max_interval"
names(rnasnp_res_dt)[8]<- "r_min_interval"
names(rnasnp_res_dt)[7]<- "d_max_pval"
names(rnasnp_res_dt)[10]<- "r_min_pval"

# Merge with mRNA abundance annotations
comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,SNP:=RNAsnp_ID]

rnasnp_res_merge_dt<- merge(
  comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt,
  rnasnp_res_dt, by="SNP")

# Look at min correlation between WT and mutant across target regions
rnasnp_cor_by_pos_gg<- ggplot(
  data=rnasnp_res_merge_dt, 
  aes(x=POS, y=r_min, col=-1*log10(r_min_pval)))

rnasnp_cor_by_pos_gg+
  facet_wrap(~ROI, scales="free_x")+
  geom_point()+
  scale_color_viridis(option="C")+
  labs(x="Target position", y="Min correlation", col="-log10(p-value)")+
  theme_cowplot()+
  theme(legend.position="bottom")

rm(rnasnp_cor_by_pos_gg)
  
# Map ROI_1 and its mRNA abundance hotspot
rnasnp_cor_by_pos_roi1_gg<- ggplot(
  data=rnasnp_res_merge_dt[ROI=="ROI_1"], 
  aes(x=POS, y=r_min, col=-1*log10(r_min_pval)))

rnasnp_cor_by_pos_roi1_gg+
  geom_rect(aes(xmin=4860, xmax=4874, ymin=-Inf, ymax=Inf), fill="lightblue")+
  geom_point()+
  scale_color_viridis(option="C")+
  labs(x="Target position", y="Min correlation", col="-log10(p-value)")+
  theme_cowplot()+
  theme(legend.position="bottom")

rm(rnasnp_cor_by_pos_roi1_gg)

# Map ROI_2 and its mRNA abundance hotspot
rnasnp_cor_by_pos_roi2_gg<- ggplot(
  data=rnasnp_res_merge_dt[ROI=="ROI_2"], 
  aes(x=POS, y=r_min, col=-1*log10(r_min_pval)))

rnasnp_cor_by_pos_roi2_gg+
  geom_rect(aes(xmin=5115, xmax=5150, ymin=-Inf, ymax=Inf), fill="lightblue")+
  geom_point()+
  scale_color_viridis(option="C")+
  labs(x="Target position", y="Min correlation", col="-log10(p-value)")+
  theme_cowplot()+
  theme(legend.position="bottom")

rm(rnasnp_cor_by_pos_roi2_gg)

# So RNAsnp and Gaither et al. 2021 SNP analysis is consistent with greater effects after 
# target position 5115

# Compare logFC in mRNA abundance with folding correlation
rnasnp_logfc_gg<- ggplot(
  data=rnasnp_res_merge_dt[adj.P.Val < loose_fdr_thresh], 
  aes(x=logFC, y=r_min, col=-1*log10(r_min_pval)))

rnasnp_logfc_gg+
  facet_wrap(~ROI)+
  geom_point()+
  scale_color_viridis(option="C")+
  labs(x="log FC", y="Min correlation", col="-log10(p-value)")+
  theme_cowplot()+
  theme(legend.position="bottom")

rm(rnasnp_logfc_gg)

# No linear association between RNAsnp folding perturbation and log FC

# Is the magnitude of mRNA abundance effect in either direction 
# associated with disrupted folding?
rnasnp_abs_logfc_gg<- ggplot(
  data=rnasnp_res_merge_dt[adj.P.Val < loose_fdr_thresh], 
  aes(x=abs(logFC), y=r_min, col=-1*log10(r_min_pval)))

rnasnp_abs_logfc_gg+
  geom_point()+
  scale_color_viridis(option="C")+
  labs(x="abs(log FC)", y="Min correlation", col="-log10(p-value)")+
  theme_cowplot()+
  theme(legend.position="bottom")

rm(rnasnp_abs_logfc_gg)

# No association

# Finally compare variants after position 5115 to all others
rnasnp_logfc_roi2_gg<- ggplot(
  data=rnasnp_res_merge_dt[adj.P.Val < loose_fdr_thresh & r_min_pval < 0.15], 
  aes(x=dMFE_cutoff, y=logFC))

rnasnp_logfc_roi2_gg+
  geom_jitter(width=0.1)+
  labs(x="POS > 5115", y="log FC")+
  theme_cowplot()+
  theme(legend.position="bottom")

rm(rnasnp_logfc_roi2_gg)

# Choose the most interesting variants
rna_snp_aa_positions<- c(51, 56, 59, 62, 63, 74, 152)
rna_snp_aa_changes<- c("p.M51L", "p.E56R", "p.I59G", "p.H62S", "p.H62R", "p.V63L", "p.V63F", "p.S74L", "p.S74H", "p.M152T")
rna_snp_var_ids<- c("p.M51L"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4827:ATG:CTC",
                    "p.E56R"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4842:GAG:CGT",
                    "p.I59G"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4851:ATC:GGG",
                    "p.H62S"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4860:CAC:TCT", 
                    "p.H62R"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4861:AC:GG", 
                    "p.V63L"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4863:GTG:TTA", 
                    "p.V63F"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4863:GTG:TTT", 
                    "p.S74L"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4896:AGC:CTA", 
                    "p.S74H"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4896:AG:CA", 
                    "p.M152T"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:5131:TG:CT")

rna_snp_silent_var_ids<- c(
  "p.H62H"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4862:C:T",
  "p.V63V"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4865:G:A",
  "p.V63V"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4865:G:C")



# comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[VAR_ID%in%rna_snp_var_ids, POS-comt_insert_offset, keyby="VAR_ID"]
# comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[VAR_ID%in%rna_snp_silent_var_ids, POS-comt_insert_offset, keyby="VAR_ID"]

rna_snp_var_ids_lessthan_pval_0.15<- comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  VAR_ID%in%rna_snp_var_ids, POS-comt_insert_offset, keyby="VAR_ID"][5:9, VAR_ID]

# 386 is second coding codon within ENST00000361682.10 (the native reference used)
#comt_cds_offset<- 386 - 1

#comt_up_positions<- c(536, 602, 575, 578, 603, 567, 567, 596, 602, 579, 593, 533, 605)
#comt_up_positions_offset<- comt_up_positions - comt_cds_offset

# Add the UTR5 ex2 and Flag to the position offsets to get coo rdinates 
# for the expressed COMT sequence
#comt_utr5_ex2_flag_add<- 118
#comt_up_positions_RNAsnp<- comt_up_positions_offset + comt_utr5_ex2_flag_add

# Variants up at FDR < 0.01

# 269 G269C-G270A
# 335 A335C-G336T-C337A   X
# 308 C308A-G310T
# 311 C311T-A312C-T313G
# 336 G336C-C337G   X
# 300 A300G-C301G   X
# 300 A300C-C301T
# 329 G329T-A331T   X
# 335 A335C-C337A   X
# 312 A312T-T313C
# 326 A326C-A327T
# 266 A266C-G268C
# 338 G338A

```

```{r RNAsnp prediction for RNA up variants ALDEx2}

comt_insert_offset<- 4561

varid_to_rnasnp<- function(pos, ref, alt, ref_offset=comt_insert_offset){
  
  pos_offset<- pos - ref_offset
  refs<- unlist(strsplit(ref, ""))
  alts<- unlist(strsplit(alt, ""))
  
  if(length(refs) == 1){
    res<- paste(refs, pos_offset, alts, sep="")
    return(res)
  } else {
    
    res<- NULL
    for(i in 1:length(refs)){
      ref_i<- refs[i]
      alt_i<- alts[i]
      
      if(ref_i!=alt_i){
        subst<- paste(ref_i, pos_offset+(i-1), alt_i, sep="")
        res<- c(res, subst)
      }
    }
    final_res<- paste(res, collapse="-")
    return(final_res)
  }
}

# Generate the RNAsnp IDs
comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  ,RNAsnp_ID:=varid_to_rnasnp(POS, REF, ALT), by=seq_len(nrow(
    comt_rna_input_cast_mat_cpm_aldex_merge_dt))]

rnasnp_ids_file<- paste(comt_base_dir, "RNAsnp", "COMT_RNAsnp_IDs.txt", sep="/")

write.table(comt_rna_input_cast_mat_cpm_aldex_merge_dt[,RNAsnp_ID], 
            rnasnp_ids_file, quote=FALSE, row.names=FALSE, col.names=FALSE)

# Read in RNAsnp results
rnasnp_res_file<- paste(comt_base_dir, "RNAsnp", "COMT_RNAsnp_results.txt", sep="/")
rnasnp_res_dt<- fread(rnasnp_res_file, sep="\t")
names(rnasnp_res_dt)[5]<- "d_max_interval"
names(rnasnp_res_dt)[8]<- "r_min_interval"
names(rnasnp_res_dt)[7]<- "d_max_pval"
names(rnasnp_res_dt)[10]<- "r_min_pval"

# Merge with mRNA abundance annotations
comt_rna_input_cast_mat_cpm_aldex_merge_dt[,SNP:=RNAsnp_ID]

comt_rna_input_cast_mat_cpm_aldex_merge_dt<- merge(
  comt_rna_input_cast_mat_cpm_aldex_merge_dt,
  rnasnp_res_dt, by="SNP")

# Look at min correlation between WT and mutant across target regions
rnasnp_cor_by_pos_gg<- ggplot(
  data=comt_rna_input_cast_mat_cpm_aldex_merge_dt, 
  aes(x=POS, y=r_min, col=-1*log10(r_min_pval)))

rnasnp_cor_by_pos_gg+
  facet_wrap(~ROI, scales="free_x")+
  geom_point()+
  scale_color_viridis(option="C")+
  labs(x="Target position", y="Min correlation", col="-log10(p-value)")+
  theme_cowplot()+
  theme(legend.position="bottom")

rm(rnasnp_cor_by_pos_gg)
  
# Map ROI_1 and its mRNA abundance hotspot
rnasnp_cor_by_pos_roi1_gg<- ggplot(
  data=comt_rna_input_cast_mat_cpm_aldex_merge_dt[ROI=="ROI_1"], 
  aes(x=POS, y=r_min, col=-1*log10(r_min_pval)))

rnasnp_cor_by_pos_roi1_gg+
  geom_rect(aes(xmin=4860, xmax=4874, ymin=-Inf, ymax=Inf), fill="lightblue")+
  geom_point()+
  scale_color_viridis(option="C")+
  labs(x="Target position", y="Min correlation", col="-log10(p-value)")+
  theme_cowplot()+
  theme(legend.position="bottom")

rm(rnasnp_cor_by_pos_roi1_gg)

# Map ROI_2 and its mRNA abundance hotspot
rnasnp_cor_by_pos_roi2_gg<- ggplot(
  data=comt_rna_input_cast_mat_cpm_aldex_merge_dt[ROI=="ROI_2"], 
  aes(x=POS, y=r_min, col=-1*log10(r_min_pval)))

rnasnp_cor_by_pos_roi2_gg+
  geom_rect(aes(xmin=5115, xmax=5150, ymin=-Inf, ymax=Inf), fill="lightblue")+
  geom_point()+
  scale_color_viridis(option="C")+
  labs(x="Target position", y="Min correlation", col="-log10(p-value)")+
  theme_cowplot()+
  theme(legend.position="bottom")

rm(rnasnp_cor_by_pos_roi2_gg)

# So RNAsnp and Gaither et al. 2021 SNP analysis is consistent with greater effects after 
# target position 5115

# Compare effect size in mRNA abundance with folding correlation
rnasnp_effect_gg<- ggplot(
  data=comt_rna_input_cast_mat_cpm_aldex_merge_dt[we.eBH < comt_rna_fdr_thresh & VAR_TYPE!="Nonsense"], 
  aes(x=effect, y=r_min, col=-1*log10(r_min_pval)))

rnasnp_effect_gg+
  facet_wrap(~ROI)+
  geom_point()+
  scale_color_viridis(option="C")+
  labs(x="Effect size", y="Min correlation", col="-log10(p-value)")+
  theme_cowplot()+
  theme(legend.position="bottom")

rm(rnasnp_effect_gg)

# No linear association between RNAsnp folding perturbation and effect size

# Which variants are expected to alter folding?
#comt_rnasnp_sig_var_ids<- rnasnp_res_merge_dt[we.eBH < comt_rna_fdr_thresh & VAR_TYPE!="Nonsense" & r_min_pval < comt_rna_fdr_thresh, VAR_ID]

comt_rna_input_cast_mat_cpm_aldex_merge_dt[
  r_min<0.1 & Direction!="N.S." & VAR_TYPE!="Nonsense", .(ID_SHORT, effect, we.eBH)]

# What about the silent variants at H62, V63?
rna_snp_silent_var_ids<- c(
  "p.H62H"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4862:C:T",
  "p.V63V"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4865:G:A",
  "p.V63V"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4865:G:C")

comt_rna_input_cast_mat_cpm_aldex_merge_dt[VAR_ID%in%rna_snp_silent_var_ids, .(ID_SHORT, effect, we.eBH, r_min, r_min_pval)]

# Compare results for H62-V63 hotspot to all other variants
comt_rna_input_cast_mat_cpm_aldex_merge_dt[,Hotspot:=ifelse(AA_POS%in%c("62", "63"), "True", "False")]
comt_rna_input_cast_mat_cpm_aldex_merge_dt[,Hotspot:=factor(Hotspot, levels=c("True", "False"))]

rnasnp_hotspot_gg<- ggplot(
  data=comt_rna_input_cast_mat_cpm_aldex_merge_dt[r_min_pval < comt_rna_fdr_thresh], 
  aes(x=Hotspot, y=r_min))

rnasnp_hotspot_gg+
  geom_boxplot()+
  theme_cowplot()

rm(rnasnp_hotspot_gg)

# No difference

# Is the magnitude of mRNA abundance effect in either direction 
# associated with disrupted folding?
rnasnp_abs_effect_gg<- ggplot(
  data=comt_rna_input_cast_mat_cpm_aldex_merge_dt[we.eBH < comt_rna_fdr_thresh], 
  aes(x=abs(effect), y=r_min, col=-1*log10(r_min_pval)))

rnasnp_abs_effect_gg+
  geom_point()+
  scale_color_viridis(option="C")+
  labs(x="abs(effect size)", y="Min correlation", col="-log10(p-value)")+
  theme_cowplot()

rm(rnasnp_abs_effect_gg)

# No association

# Finally compare variants after position 5115 to all others
comt_rna_input_cast_mat_cpm_aldex_merge_dt[,RNAsnp_ROI:=as.factor(ifelse(POS>5115, "True", "False"))]

rnasnp_effect_roi2_gg<- ggplot(
  data=comt_rna_input_cast_mat_cpm_aldex_merge_dt[we.eBH < comt_rna_fdr_thresh], 
  aes(x=RNAsnp_ROI, y=effect))

rnasnp_effect_roi2_gg+
  geom_jitter(width=0.1)+
  labs(x="POS > 5115", y="Effect size")+
  theme_cowplot()+
  theme(legend.position="bottom")

rm(rnasnp_effect_roi2_gg)

# Choose the most interesting variants
rna_snp_aa_positions<- c(51, 56, 59, 62, 63, 74, 152)
rna_snp_aa_changes<- c("p.M51L", "p.E56R", "p.I59G", "p.H62S", "p.H62R", "p.V63L", "p.V63F", "p.S74L", "p.S74H", "p.M152T")
rna_snp_var_ids<- c("p.M51L"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4827:ATG:CTC",
                    "p.E56R"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4842:GAG:CGT",
                    "p.I59G"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4851:ATC:GGG",
                    "p.H62S"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4860:CAC:TCT", 
                    "p.H62R"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4861:AC:GG", 
                    "p.V63L"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4863:GTG:TTA", 
                    "p.V63F"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4863:GTG:TTT", 
                    "p.S74L"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4896:AGC:CTA", 
                    "p.S74H"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:4896:AG:CA", 
                    "p.M152T"="pDEST_HC_Rec_Bxb_v2_UTR5_Flag_COMT_moxGFP_LPS:5131:TG:CT")


```


# Variants with differential ribosome load

anota2seq requires at least 2 sample classes (e.g. control, treatment) to perform analysis
Use limma and ALDEx2 to compare biological replicates between total RNA and polysome metafractions 3-4

```{r COMT Poly F2 limma}

poly_f2_fdr_thresh<- 0.01
replicate_count_thresh<- 8
log_caf_thresh<- -10

comt_poly_limma_f2_dt<- comt_all_filt_postprocess_subtract_drop_cast_combat_dt[
  Input=="cDNA" | (Input=="Poly" & Fraction==2)]

# Use a filter on frequency; corresponds to ~10 counts
comt_poly_limma_f2_counts_dt<- comt_poly_limma_f2_dt[
  Median_log_CAF>log_caf_thresh, .N, by=.(VAR_ID)]

comt_poly_limma_complete_varids<- comt_poly_limma_f2_counts_dt[
  N>=replicate_count_thresh, unique(VAR_ID)]

comt_poly_limma_f2_input_dt<- comt_poly_limma_f2_dt[
  VAR_ID%in%comt_poly_limma_complete_varids]

comt_poly_limma_f2_input_cast_dt<- dcast(
  data=comt_poly_limma_f2_input_dt, formula=VAR_ID~Key, value.var="Median_log_CAF")

comt_poly_limma_f2_input_cast_mat<- as.matrix(
  comt_poly_limma_f2_input_cast_dt[,-c("VAR_ID")])

rownames(comt_poly_limma_f2_input_cast_mat)<- comt_poly_limma_f2_input_cast_dt[,VAR_ID]

#comt_poly_limma_f2_input_cast_mat<- normalizeQuantiles(comt_poly_limma_f2_input_cast_mat)

comt_poly_limma_f2_input_cast_mat_eset<- Biobase::ExpressionSet(
  assayData=comt_poly_limma_f2_input_cast_mat)

comt_poly_limma_f2_input_design_dt<- 
  comt_poly_limma_f2_input_dt[,as.factor(unique(Input)), keyby=.(Key)]

comt_poly_limma_f2_input_design_mat<- 
  as.matrix(comt_poly_limma_f2_input_design_dt[,-c("Key")])

rownames(comt_poly_limma_f2_input_design_mat)<- comt_poly_limma_f2_input_design_dt[,Key]

comt_poly_limma_f2_input_design<- 
  model.matrix(~V1, data=as.data.frame(comt_poly_limma_f2_input_design_mat), 
               list(V1=contr.treatment(n=2, base=1)))

comt_poly_limma_f2_input_cast_mat_eset_fit<- lmFit(
  comt_poly_limma_f2_input_cast_mat_eset, comt_poly_limma_f2_input_design)

comt_poly_limma_f2_input_cast_mat_eset_fit_eb<- eBayes(
  comt_poly_limma_f2_input_cast_mat_eset_fit, trend=TRUE, proportion=0.03)

comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all<- topTable(
  comt_poly_limma_f2_input_cast_mat_eset_fit_eb, coef=ncol(
    comt_poly_limma_f2_input_design), number=nrow(
      comt_poly_limma_f2_input_cast_mat_eset_fit_eb))

comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_dt<- as.data.table(
  comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all)

comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_dt[
  ,VAR_ID:=rownames(comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all)]

comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_dt[logFC>0, Direction:="Poly up"]
comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_dt[logFC<0, Direction:="Poly down"]
comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_dt[,Direction:=as.factor(Direction)]

# Look at p-value distr
poly_f2_pval_distr_gg<- ggplot(
  data=comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_dt, 
  aes(x=P.Value))

poly_f2_pval_distr_gg+
  geom_histogram()+
  theme_cowplot()

rm(poly_f2_pval_distr_gg)

# Volcano plot
poly_f2_volcano_gg<- ggplot(
  data=comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_dt, 
  aes(x=logFC, y=-1*log10(adj.P.Val)))

poly_f2_volcano_gg+
  geom_point()+
  theme_cowplot()

rm(poly_f2_volcano_gg)

# MA plot
comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_dt[adj.P.Val < poly_f2_fdr_thresh, Is_significant:="FDR < 0.01"]

ma_gg<- ggplot(
  data=comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_dt, 
  aes(x=AveExpr, y=logFC))

ma_gg+
  geom_point(alpha=0.8)+
  geom_smooth(method="lm")+
  geom_point(data=comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_dt[
    !is.na(Is_significant)], 
  aes(x=AveExpr, y=logFC), col="gold")+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  labs(x="Average expression", y="log FC, poly F2 to cDNA")+
  theme_cowplot()

rm(ma_gg)

comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_top_dt<- 
  comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_dt[adj.P.Val < poly_f2_fdr_thresh]

comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_top_varids<- 
  comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_top_dt[,VAR_ID]

comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_top_up_varids<- 
  comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_top_dt[logFC>0, VAR_ID]

comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_top_down_varids<- 
  comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_top_dt[logFC<0, VAR_ID]

comt_poly_limma_f2_dt[
  ,ID_SHORT:=paste(AA_CHANGE, paste(REF, ALT, sep=">"), sep=", ")]

comt_poly_limma_f2_dt[
  ,ID_LONG:=paste(AA_CHANGE, paste(POS, REF, ALT, sep=":"), sep=", ")]

if("Direction"%in%names(comt_poly_limma_f2_dt)){
  comt_poly_limma_f2_dt[,Direction:=NULL]
}

comt_poly_limma_f2_dt[
  VAR_ID%in%comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_top_up_varids,
  Direction:="Poly up"]

comt_poly_limma_f2_dt[
  VAR_ID%in%comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_top_down_varids,
  Direction:="Poly down"]

comt_poly_limma_f2_dt[,Direction:=as.factor(Direction)]

amplicon_gDNA_med_log_caf<- comt_poly_limma_f2_dt[
  Input=="cDNA", median(Median_log_CAF)]

amplicon_poly_med_log_caf<- comt_poly_limma_f2_dt[
  Input=="Poly", median(Median_log_CAF)]

aaas_cols<- pal_aaas()(2)

comt_poly_limma_f2_sig_dt<- comt_poly_limma_f2_dt[
    VAR_ID%in%comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_top_varids]

comt_poly_limma_f2_sig_dt[,Input:=factor(Input, levels=c("Poly", "cDNA"))]

# Visualize the frequencies
comt_poly_limma_f2_sig_gg<- ggplot(
  data=comt_poly_limma_f2_sig_dt,
  aes(x=Median_log_CAF, y=reorder(ID_SHORT, AA_POS2), col=Input))

comt_poly_limma_f2_sig_gg+
  geom_vline(xintercept=amplicon_gDNA_med_log_caf, col=aaas_cols[2], linetype=2)+
  geom_vline(xintercept=amplicon_poly_med_log_caf, col=aaas_cols[1], linetype=2)+
  facet_wrap(~Direction, scales="free_y")+
  geom_point(position=position_jitterdodge())+
  labs(y="Variant ID", x="log frequency")+
  scale_x_continuous(breaks=seq(-10,-6))+
  coord_cartesian(xlim=c(-10,-6.5))+
  scale_color_aaas()+
  theme_cowplot()+
  theme(axis.text.y=element_text(size=7))

rm(comt_poly_limma_f2_sig_gg)

# Make heatmap of the effects
comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_merge_dt<- merge(
  comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_dt, annot_dt, by="VAR_ID")

comt_poly_f2_ve_dt<- comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_merge_dt[
  !grepl(",", AA_CHANGE), .(Median_log_FC=median(logFC, na.rm=T)), 
  by=.(ROI, AA_POS, REF_AA, ALT_AA, AA_CHANGE, VAR_TYPE)]

comt_poly_f2_ve_dt[,AA_POS:=as.integer(AA_POS)]
comt_poly_f2_ve_dt[,ALT_AA:=factor(ALT_AA, levels=unlist(aa_class_list))]
comt_poly_f2_ve_dt[,ALT_AA_level:=as.integer(ALT_AA)]

comt_poly_f2_heatmap_gg<- ggplot(
  data=comt_poly_f2_ve_dt, 
  aes(x=AA_POS, y=reorder(ALT_AA, ALT_AA_level), fill=Median_log_FC))

comt_poly_f2_heatmap_gg+
  facet_wrap(~ROI, scales="free")+
  geom_tile()+
  scale_fill_gradient2()+
  labs(x="Amino acid position", y="Alternate", 
       fill="Median logFC,
poly F2 : cDNA")+
  theme_cowplot()

rm(comt_poly_f2_heatmap_gg)

# Plot fold changes by variant type
comt_poly_f2_vartype_gg<- ggplot(
  data=comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_merge_dt,
  aes(x=VAR_TYPE, y=logFC))

comt_poly_f2_vartype_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot()+
  labs(x="", y="log FC, poly F2 : RNA")+
  theme_cowplot()

rm(comt_poly_f2_vartype_gg)

```

```{r COMT Poly F2 ALDEx2}

comt_poly_f2_fdr_thresh<- 0.05

# First transform normalized frequencies to counts per million
comt_poly_f2_input_cast_mat_cpm<- apply(comt_poly_limma_f2_input_cast_mat, 2, function(x){round(exp(x)*1e6)})

set.seed(9)

comt_poly_f2_input_cast_mat_cpm_aldex_dt<- run_aldex(comt_poly_f2_input_cast_mat_cpm, c(rep("A", 4), rep("B", 4)))

# Look at the raw p-values to ensure testing procedure is correct
comt_poly_f2_aldex2_pval_gg<- ggplot(
  data=comt_poly_f2_input_cast_mat_cpm_aldex_dt, 
  aes(x=we.ep))

comt_poly_f2_aldex2_pval_gg+
  geom_histogram()+
  theme_cowplot()

rm(comt_poly_f2_aldex2_pval_gg)

# Volcano plot of ALDEx2 results
comt_poly_f2_aldex2_gg<- ggplot(
  data=comt_poly_f2_input_cast_mat_cpm_aldex_dt,
  aes(x=effect, y=-1*log10(we.eBH)))

comt_poly_f2_aldex2_gg+
  geom_point()+
  geom_point(data=comt_poly_f2_input_cast_mat_cpm_aldex_dt[we.eBH<comt_poly_f2_fdr_thresh], 
             aes(x=effect, y=-1*log10(we.eBH)), col="firebrick")+
  theme_cowplot()

rm(comt_poly_f2_aldex2_gg)

comt_poly_f2_input_cast_mat_cpm_aldex_dt[,VAR_ID:=rownames(comt_poly_f2_input_cast_mat_cpm)]

comt_poly_f2_input_cast_mat_cpm_aldex_merge_dt<- merge(
  comt_poly_f2_input_cast_mat_cpm_aldex_dt, annot_dt, by="VAR_ID")

comt_poly_f2_aldex2_sig_gg<- ggplot(
  data=comt_poly_f2_input_cast_mat_cpm_aldex_merge_dt[
    we.eBH<comt_poly_f2_fdr_thresh],
  aes(x=effect, y=reorder(ID_SHORT, POS), fill=VAR_TYPE))

comt_poly_f2_aldex2_sig_gg+
  geom_bar(stat="identity", alpha=0.75)+
  scale_fill_manual(values=c("darkviolet", "firebrick"))+
  theme_cowplot()+
  labs(y="", x="Effect size", fill="")+
  theme(axis.text.y=element_text(size=8))

rm(comt_poly_f2_aldex2_sig_gg)

# Generate a VE map using ALDEx2
comt_poly_f2_rep_summ_ve_dt<- comt_poly_f2_input_cast_mat_cpm_aldex_merge_dt[
  !grepl(",", AA_CHANGE), .(Median_effect=median(effect, na.rm=T),
                            Max_effect=max(effect, na.rm=T)), 
  by=.(ROI, AA_POS, REF_AA, ALT_AA, AA_CHANGE, VAR_TYPE)]

comt_poly_f2_rep_summ_ve_dt[,AA_POS:=as.integer(AA_POS)]
comt_poly_f2_rep_summ_ve_dt[,ALT_AA:=factor(ALT_AA, levels=unlist(aa_class_list))]
comt_poly_f2_rep_summ_ve_dt[,ALT_AA_level:=as.integer(ALT_AA)]

comt_poly_f2_aldex_ve_gg<- ggplot(
  data=comt_poly_f2_rep_summ_ve_dt, 
  aes(x=AA_POS, y=reorder(ALT_AA, ALT_AA_level), fill=Median_effect))

comt_poly_f2_aldex_ve_gg+
  facet_wrap(~ROI, scales="free")+
  geom_tile()+
  scale_fill_gradient2(low=scales::muted("blue"), high=scales::muted("red"))+
  labs(x="Amino acid position", y="Alternate", 
       fill="Median effect")+
  theme_cowplot()

rm(comt_poly_f2_aldex_ve_gg)

# Test if nonsense variants enriched in F2 compared to silent variants
f2_by_type_gg<- ggplot(
  data=comt_poly_f2_input_cast_mat_cpm_aldex_merge_dt, 
  aes(x=VAR_TYPE, y=effect))

f2_by_type_gg+
  geom_boxplot()+
  theme_cowplot()

rm(f2_by_type_gg)

comt_poly_f2_input_cast_mat_cpm_aldex_merge_dt[
  VAR_TYPE!="Missense", wilcox.test(formula=effect~VAR_TYPE, data=.SD)]

comt_poly_f2_input_cast_mat_cpm_aldex_merge_dt[
  VAR_TYPE!="Silent", wilcox.test(formula=effect~VAR_TYPE, data=.SD)]


```

```{r COMT Poly F3 limma}

poly_f3_fdr_thresh<- 0.01
replicate_count_thresh<- 8
log_caf_thresh<- -10

comt_poly_limma_f3_dt<- comt_all_filt_postprocess_subtract_drop_cast_combat_dt[
  Input=="cDNA" | (Input=="Poly" & Fraction==3)]

# Use a filter on frequency; corresponds to ~10 counts
comt_poly_limma_f3_counts_dt<- comt_poly_limma_f3_dt[
  Median_log_CAF>log_caf_thresh, .N, by=.(VAR_ID)]

comt_poly_limma_complete_varids<- comt_poly_limma_f3_counts_dt[
  N>=replicate_count_thresh, unique(VAR_ID)]

comt_poly_limma_f3_input_dt<- comt_poly_limma_f3_dt[
  VAR_ID%in%comt_poly_limma_complete_varids]

comt_poly_limma_f3_input_cast_dt<- dcast(
  data=comt_poly_limma_f3_input_dt, formula=VAR_ID~Key, value.var="Median_log_CAF")

comt_poly_limma_f3_input_cast_mat<- as.matrix(
  comt_poly_limma_f3_input_cast_dt[,-c("VAR_ID")])

rownames(comt_poly_limma_f3_input_cast_mat)<- comt_poly_limma_f3_input_cast_dt[,VAR_ID]

#comt_poly_limma_f3_input_cast_mat<- normalizeQuantiles(comt_poly_limma_f3_input_cast_mat)

comt_poly_limma_f3_input_cast_mat_eset<- Biobase::ExpressionSet(
  assayData=comt_poly_limma_f3_input_cast_mat)

comt_poly_limma_f3_input_design_dt<- 
  comt_poly_limma_f3_input_dt[,as.factor(unique(Input)), keyby=.(Key)]

comt_poly_limma_f3_input_design_mat<- 
  as.matrix(comt_poly_limma_f3_input_design_dt[,-c("Key")])

rownames(comt_poly_limma_f3_input_design_mat)<- comt_poly_limma_f3_input_design_dt[,Key]

comt_poly_limma_f3_input_design<- 
  model.matrix(~V1, data=as.data.frame(comt_poly_limma_f3_input_design_mat), 
               list(V1=contr.treatment(n=2, base=1)))

comt_poly_limma_f3_input_cast_mat_eset_fit<- lmFit(
  comt_poly_limma_f3_input_cast_mat_eset, comt_poly_limma_f3_input_design)

comt_poly_limma_f3_input_cast_mat_eset_fit_eb<- eBayes(
  comt_poly_limma_f3_input_cast_mat_eset_fit, trend=TRUE, proportion=0.03)

comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all<- topTable(
  comt_poly_limma_f3_input_cast_mat_eset_fit_eb, coef=ncol(
    comt_poly_limma_f3_input_design), number=nrow(
      comt_poly_limma_f3_input_cast_mat_eset_fit_eb))

comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_dt<- as.data.table(
  comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all)

comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_dt[
  ,VAR_ID:=rownames(comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all)]

comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_dt[logFC>0, Direction:="Poly up"]
comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_dt[logFC<0, Direction:="Poly down"]
comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_dt[,Direction:=as.factor(Direction)]

# Look at p-value distr
poly_f3_pval_distr_gg<- ggplot(
  data=comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_dt, 
  aes(x=P.Value))

poly_f3_pval_distr_gg+
  geom_histogram()+
  theme_cowplot()

rm(poly_f3_pval_distr_gg)

# Volcano plot
poly_f3_volcano_gg<- ggplot(
  data=comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_dt, 
  aes(x=logFC, y=-1*log10(adj.P.Val)))

poly_f3_volcano_gg+
  geom_point()+
  theme_cowplot()

rm(poly_f3_volcano_gg)

# MA plot
comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_dt[adj.P.Val < poly_f3_fdr_thresh, Is_significant:="FDR < 0.01"]

ma_gg<- ggplot(
  data=comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_dt, 
  aes(x=AveExpr, y=logFC))

ma_gg+
  geom_point(alpha=0.8)+
  geom_smooth(method="lm")+
  geom_point(data=comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_dt[
    !is.na(Is_significant)], 
  aes(x=AveExpr, y=logFC), col="gold")+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  labs(x="Average expression", y="log FC, poly F3 to cDNA")+
  theme_cowplot()

comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_top_dt<- 
  comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_dt[adj.P.Val < poly_f3_fdr_thresh]

comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_top_varids<- 
  comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_top_dt[,VAR_ID]

comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_top_up_varids<- 
  comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_top_dt[logFC>0, VAR_ID]

comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_top_down_varids<- 
  comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_top_dt[logFC<0, VAR_ID]

comt_poly_limma_f3_dt[
  ,ID_SHORT:=paste(AA_CHANGE, paste(REF, ALT, sep=">"), sep=", ")]

comt_poly_limma_f3_dt[
  ,ID_LONG:=paste(AA_CHANGE, paste(POS, REF, ALT, sep=":"), sep=", ")]

if("Direction"%in%names(comt_poly_limma_f3_dt)){
  comt_poly_limma_f3_dt[,Direction:=NULL]
}

comt_poly_limma_f3_dt[
  VAR_ID%in%comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_top_up_varids,
  Direction:="Poly up"]

comt_poly_limma_f3_dt[
  VAR_ID%in%comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_top_down_varids,
  Direction:="Poly down"]

comt_poly_limma_f3_dt[,Direction:=as.factor(Direction)]

amplicon_gDNA_med_log_caf<- comt_poly_limma_f3_dt[
  Input=="cDNA", median(Median_log_CAF)]

amplicon_poly_med_log_caf<- comt_poly_limma_f3_dt[
  Input=="Poly", median(Median_log_CAF)]

aaas_cols<- pal_aaas()(2)

comt_poly_limma_f3_sig_dt<- comt_poly_limma_f3_dt[
    VAR_ID%in%comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_top_varids]

comt_poly_limma_f3_sig_dt[,Input:=factor(Input, levels=c("Poly", "cDNA"))]

# Visualize the frequencies
comt_poly_limma_f3_sig_gg<- ggplot(
  data=comt_poly_limma_f3_sig_dt,
  aes(x=Median_log_CAF, y=reorder(ID_SHORT, AA_POS2), col=Input))

comt_poly_limma_f3_sig_gg+
  geom_vline(xintercept=amplicon_gDNA_med_log_caf, col=aaas_cols[2], linetype=2)+
  geom_vline(xintercept=amplicon_poly_med_log_caf, col=aaas_cols[1], linetype=2)+
  facet_wrap(~Direction, scales="free_y")+
  geom_point(position=position_jitterdodge())+
  labs(y="Variant ID", x="log frequency")+
  scale_x_continuous(breaks=seq(-10,-6))+
  coord_cartesian(xlim=c(-10,-5.5))+
  scale_color_aaas()+
  theme_cowplot()+
  theme(axis.text.y=element_text(size=7))

rm(comt_poly_limma_f3_sig_gg)

# Make heatmap of the effects
comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_merge_dt<- merge(
  comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_dt, annot_dt, by="VAR_ID")

comt_poly_f3_ve_dt<- comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_merge_dt[
  !grepl(",", AA_CHANGE), .(Median_log_FC=median(logFC, na.rm=T)), 
  by=.(ROI, AA_POS, REF_AA, ALT_AA, AA_CHANGE, VAR_TYPE)]

comt_poly_f3_ve_dt[,AA_POS:=as.integer(AA_POS)]
comt_poly_f3_ve_dt[,ALT_AA:=factor(ALT_AA, levels=unlist(aa_class_list))]
comt_poly_f3_ve_dt[,ALT_AA_level:=as.integer(ALT_AA)]

comt_poly_f3_heatmap_gg<- ggplot(
  data=comt_poly_f3_ve_dt, 
  aes(x=AA_POS, y=reorder(ALT_AA, ALT_AA_level), fill=Median_log_FC))

comt_poly_f3_heatmap_gg+
  facet_wrap(~ROI, scales="free")+
  geom_tile()+
  scale_fill_gradient2()+
  labs(x="Amino acid position", y="Alternate", 
       fill="Median logFC,
poly F3 : cDNA")+
  theme_cowplot()

rm(comt_poly_f3_heatmap_gg)

# Plot fold changes by variant type
comt_poly_f3_vartype_gg<- ggplot(
  data=comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_merge_dt,
  aes(x=VAR_TYPE, y=logFC))

comt_poly_f3_vartype_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot()+
  labs(x="", y="log FC, poly F3 : RNA")+
  theme_cowplot()

rm(comt_poly_f3_vartype_gg)

# Look at the top 3 and bottom 3 silent variants for F3
comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_merge_dt[abs(logFC)>0.8 & VAR_TYPE=="Silent"]

# Both T54T silent mutations are up
# One up/one down silent mutation at S74S!

```

```{r COMT Poly F3 ALDEx2}

comt_poly_f3_f4_fdr_thresh<- 0.05

# First transform normalized frequencies to counts per million
comt_poly_f3_input_cast_mat_cpm<- apply(comt_poly_limma_f3_input_cast_mat, 2, function(x){round(exp(x)*1e6)})

set.seed(9)

comt_poly_f3_input_cast_mat_cpm_aldex_dt<- run_aldex(comt_poly_f3_input_cast_mat_cpm, c(rep("A", 4), rep("B", 4)))

# Look at the raw p-values to ensure testing procedure is correct
comt_poly_f3_aldex2_pval_gg<- ggplot(
  data=comt_poly_f3_input_cast_mat_cpm_aldex_dt, 
  aes(x=we.ep))

comt_poly_f3_aldex2_pval_gg+
  geom_histogram()+
  theme_cowplot()

rm(comt_poly_f3_aldex2_pval_gg)

# Volcano plot of ALDEx2 results
comt_poly_f3_aldex2_gg<- ggplot(
  data=comt_poly_f3_input_cast_mat_cpm_aldex_dt,
  aes(x=effect, y=-1*log10(we.eBH)))

comt_poly_f3_aldex2_gg+
  geom_point()+
  geom_point(data=comt_poly_f3_input_cast_mat_cpm_aldex_dt[we.eBH<comt_poly_f3_f4_fdr_thresh], 
             aes(x=effect, y=-1*log10(we.eBH)), col="firebrick")+
  theme_cowplot()

rm(comt_poly_f3_aldex2_gg)

comt_poly_f3_input_cast_mat_cpm_aldex_dt[,VAR_ID:=rownames(comt_poly_f3_input_cast_mat_cpm)]

comt_poly_f3_input_cast_mat_cpm_aldex_merge_dt<- merge(
  comt_poly_f3_input_cast_mat_cpm_aldex_dt, annot_dt, by="VAR_ID")

# Generate a VE map using ALDEx2
comt_poly_f3_rep_summ_ve_dt<- comt_poly_f3_input_cast_mat_cpm_aldex_merge_dt[
  !grepl(",", AA_CHANGE), .(Median_effect=median(effect, na.rm=T),
                            Max_effect=max(effect, na.rm=T)), 
  by=.(ROI, AA_POS, REF_AA, ALT_AA, AA_CHANGE, VAR_TYPE)]

comt_poly_f3_rep_summ_ve_dt[,AA_POS:=as.integer(AA_POS)]
comt_poly_f3_rep_summ_ve_dt[,ALT_AA:=factor(ALT_AA, levels=unlist(aa_class_list))]
comt_poly_f3_rep_summ_ve_dt[,ALT_AA_level:=as.integer(ALT_AA)]

comt_poly_f3_aldex_ve_gg<- ggplot(
  data=comt_poly_f3_rep_summ_ve_dt, 
  aes(x=AA_POS, y=reorder(ALT_AA, ALT_AA_level), fill=Median_effect))

comt_poly_f3_aldex_ve_gg+
  facet_wrap(~ROI, scales="free")+
  geom_tile()+
  scale_fill_gradient2(high=scales::muted("blue"), low=scales::muted("red"))+
  labs(x="Amino acid position", y="Alternate", 
       fill="Median effect")+
  theme_cowplot()

rm(comt_poly_f3_aldex_ve_gg)

```

```{r COMT Poly F4 limma}

comt_poly_f4_fdr_thresh<- 0.01
replicate_count_thresh<- 8
log_caf_thresh<- -10

comt_poly_limma_f4_dt<- comt_all_filt_postprocess_subtract_drop_cast_combat_dt[
  Input=="cDNA" | (Input=="Poly" & Fraction==4)]

# Use a filter on frequency; corresponds to ~10 counts
comt_poly_limma_f4_counts_dt<- comt_poly_limma_f4_dt[
  Median_log_CAF>log_caf_thresh, .N, by=.(VAR_ID)]

comt_poly_limma_complete_varids<- comt_poly_limma_f4_counts_dt[
  N>=replicate_count_thresh, unique(VAR_ID)]

comt_poly_limma_f4_input_dt<- comt_poly_limma_f4_dt[
  VAR_ID%in%comt_poly_limma_complete_varids]

comt_poly_limma_f4_input_cast_dt<- dcast(
  data=comt_poly_limma_f4_input_dt, formula=VAR_ID~Key, value.var="Median_log_CAF")

comt_poly_limma_f4_input_cast_mat<- as.matrix(
  comt_poly_limma_f4_input_cast_dt[,-c("VAR_ID")])

rownames(comt_poly_limma_f4_input_cast_mat)<- comt_poly_limma_f4_input_cast_dt[,VAR_ID]

#comt_poly_limma_f4_input_cast_mat<- normalizeQuantiles(comt_poly_limma_f4_input_cast_mat)

comt_poly_limma_f4_input_cast_mat_eset<- Biobase::ExpressionSet(
  assayData=comt_poly_limma_f4_input_cast_mat)

comt_poly_limma_f4_input_design_dt<- 
  comt_poly_limma_f4_input_dt[,as.factor(unique(Input)), keyby=.(Key)]

comt_poly_limma_f4_input_design_mat<- 
  as.matrix(comt_poly_limma_f4_input_design_dt[,-c("Key")])

rownames(comt_poly_limma_f4_input_design_mat)<- comt_poly_limma_f4_input_design_dt[,Key]

comt_poly_limma_f4_input_design<- 
  model.matrix(~V1, data=as.data.frame(comt_poly_limma_f4_input_design_mat), 
               list(V1=contr.treatment(n=2, base=1)))

comt_poly_limma_f4_input_cast_mat_eset_fit<- lmFit(
  comt_poly_limma_f4_input_cast_mat_eset, comt_poly_limma_f4_input_design)

comt_poly_limma_f4_input_cast_mat_eset_fit_eb<- eBayes(
  comt_poly_limma_f4_input_cast_mat_eset_fit, trend=TRUE, proportion=0.03)

comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all<- topTable(
  comt_poly_limma_f4_input_cast_mat_eset_fit_eb, coef=ncol(
    comt_poly_limma_f4_input_design), number=nrow(
      comt_poly_limma_f4_input_cast_mat_eset_fit_eb))

comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_dt<- as.data.table(
  comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all)

comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_dt[
  ,VAR_ID:=rownames(comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all)]

comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_dt[logFC>0, Direction:="Poly up"]
comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_dt[logFC<0, Direction:="Poly down"]
comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_dt[,Direction:=as.factor(Direction)]

# Look at p-value distr
poly_f4_pval_distr_gg<- ggplot(
  data=comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_dt, 
  aes(x=P.Value))

poly_f4_pval_distr_gg+
  geom_histogram()+
  theme_cowplot()

rm(poly_f4_pval_distr_gg)

# Volcano plot
poly_f4_volcano_gg<- ggplot(
  data=comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_dt, 
  aes(x=logFC, y=-1*log10(adj.P.Val)))

poly_f4_volcano_gg+
  geom_point()+
  theme_cowplot()

rm(poly_f4_volcano_gg)

# MA plot
comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_dt[adj.P.Val < comt_poly_f4_fdr_thresh, Is_significant:="FDR < 0.01"]

ma_gg<- ggplot(
  data=comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_dt, 
  aes(x=AveExpr, y=logFC))

ma_gg+
  geom_point(alpha=0.8)+
  geom_smooth(method="lm")+
  geom_point(data=comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_dt[
    !is.na(Is_significant)], 
  aes(x=AveExpr, y=logFC), col="gold")+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  labs(x="Average expression", y="log FC, poly F4 to cDNA")+
  theme_cowplot()

rm(ma_gg)

comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_top_dt<- 
  comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_dt[adj.P.Val < comt_poly_f4_fdr_thresh]

comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_top_varids<- 
  comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_top_dt[,VAR_ID]

comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_top_up_varids<- 
  comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_top_dt[logFC>0, VAR_ID]

comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_top_down_varids<- 
  comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_top_dt[logFC<0, VAR_ID]

comt_poly_limma_f4_dt[
  ,ID_SHORT:=paste(AA_CHANGE, paste(REF, ALT, sep=">"), sep=", ")]

comt_poly_limma_f4_dt[
  ,ID_LONG:=paste(AA_CHANGE, paste(POS, REF, ALT, sep=":"), sep=", ")]

if("Direction"%in%names(comt_poly_limma_f4_dt)){
  comt_poly_limma_f4_dt[,Direction:=NULL]
}

comt_poly_limma_f4_dt[
  VAR_ID%in%comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_top_up_varids,
  Direction:="Poly up"]

comt_poly_limma_f4_dt[
  VAR_ID%in%comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_top_down_varids,
  Direction:="Poly down"]

comt_poly_limma_f4_dt[,Direction:=as.factor(Direction)]

amplicon_gDNA_med_log_caf<- comt_poly_limma_f4_dt[
  Input=="cDNA", median(Median_log_CAF)]

amplicon_poly_med_log_caf<- comt_poly_limma_f4_dt[
  Input=="Poly", median(Median_log_CAF)]

aaas_cols<- pal_aaas()(2)

comt_poly_limma_f4_sig_dt<- comt_poly_limma_f4_dt[
    VAR_ID%in%comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_top_varids]

comt_poly_limma_f4_sig_dt[,Input:=factor(Input, levels=c("Poly", "cDNA"))]

# Visualize the frequencies
comt_poly_limma_f4_sig_gg<- ggplot(
  data=comt_poly_limma_f4_sig_dt,
  aes(x=Median_log_CAF, y=reorder(ID_SHORT, AA_POS2), col=Input))

comt_poly_limma_f4_sig_gg+
  geom_vline(xintercept=amplicon_gDNA_med_log_caf, col=aaas_cols[2], linetype=2)+
  geom_vline(xintercept=amplicon_poly_med_log_caf, col=aaas_cols[1], linetype=2)+
  facet_wrap(~Direction, scales="free_y")+
  geom_point(position=position_jitterdodge())+
  labs(y="Variant ID", x="log frequency")+
  scale_x_continuous(breaks=seq(-10,-5))+
  coord_cartesian(xlim=c(-10,-5))+
  scale_color_aaas()+
  theme_cowplot()+
  theme(axis.text.y=element_text(size=7))

rm(comt_poly_limma_f4_sig_gg)

# Make heatmap of the effects
comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_merge_dt<- merge(
  comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_dt, annot_dt, by="VAR_ID")

comt_poly_f4_ve_dt<- comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_merge_dt[
  !grepl(",", AA_CHANGE), .(Median_log_FC=median(logFC, na.rm=T)), 
  by=.(ROI, AA_POS, REF_AA, ALT_AA, AA_CHANGE, VAR_TYPE)]

comt_poly_f4_ve_dt[,AA_POS:=as.integer(AA_POS)]
comt_poly_f4_ve_dt[,ALT_AA:=factor(ALT_AA, levels=unlist(aa_class_list))]
comt_poly_f4_ve_dt[,ALT_AA_level:=as.integer(ALT_AA)]

comt_poly_f4_heatmap_gg<- ggplot(
  data=comt_poly_f4_ve_dt, 
  aes(x=AA_POS, y=reorder(ALT_AA, ALT_AA_level), fill=Median_log_FC))

comt_poly_f4_heatmap_gg+
  facet_wrap(~ROI, scales="free")+
  geom_tile()+
  scale_fill_gradient2()+
  labs(x="Amino acid position", y="Alternate", 
       fill="Median logFC,
poly F4 : cDNA")+
  theme_cowplot()

rm(comt_poly_f4_heatmap_gg)

comt_poly_f4_vartype_gg<- ggplot(
  data=comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_merge_dt,
  aes(x=VAR_TYPE, y=logFC))

comt_poly_f4_vartype_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_boxplot()+
  labs(x="", y="log FC, poly F4 : RNA")+
  theme_cowplot()

rm(comt_poly_f4_vartype_gg)

```

```{r COMT Poly F4 ALDEx2}

# First transform normalized frequencies to counts per million
comt_poly_f4_input_cast_mat_cpm<- apply(comt_poly_limma_f4_input_cast_mat, 2, function(x){round(exp(x)*1e6)})

set.seed(9)

comt_poly_f4_input_cast_mat_cpm_aldex_dt<- run_aldex(comt_poly_f4_input_cast_mat_cpm, c(rep("A", 4), rep("B", 4)))

# Look at the raw p-values to ensure testing procedure is correct
comt_poly_f4_aldex2_pval_gg<- ggplot(
  data=comt_poly_f4_input_cast_mat_cpm_aldex_dt, 
  aes(x=we.ep))

comt_poly_f4_aldex2_pval_gg+
  geom_histogram()+
  theme_cowplot()

rm(comt_poly_f4_aldex2_pval_gg)

# Volcano plot of ALDEx2 results
comt_poly_f4_aldex2_gg<- ggplot(
  data=comt_poly_f4_input_cast_mat_cpm_aldex_dt,
  aes(x=effect, y=-1*log10(we.eBH)))

comt_poly_f4_aldex2_gg+
  geom_point()+
  geom_point(data=comt_poly_f4_input_cast_mat_cpm_aldex_dt[we.eBH<comt_poly_f3_f4_fdr_thresh], 
             aes(x=effect, y=-1*log10(we.eBH)), col="firebrick")+
  theme_cowplot()

rm(comt_poly_f4_aldex2_gg)

comt_poly_f4_input_cast_mat_cpm_aldex_dt[,VAR_ID:=rownames(comt_poly_f4_input_cast_mat_cpm)]

comt_poly_f4_input_cast_mat_cpm_aldex_merge_dt<- merge(
  comt_poly_f4_input_cast_mat_cpm_aldex_dt, annot_dt, by="VAR_ID")

# Combine F3 and F4 and plot together
comt_poly_f3_f4_input_cast_mat_cpm_aldex_merge_dt<- rbind(
  comt_poly_f3_input_cast_mat_cpm_aldex_merge_dt,
  comt_poly_f4_input_cast_mat_cpm_aldex_merge_dt)

comt_poly_f3_f4_input_cast_mat_cpm_aldex_merge_dt[
  ,Fraction:=as.factor(rep(c("F3", "F4"), times=c(
    nrow(comt_poly_f3_input_cast_mat_cpm_aldex_merge_dt), 
    nrow(comt_poly_f4_input_cast_mat_cpm_aldex_merge_dt))))]

comt_poly_f3_f4_aldex2_sig_gg<- ggplot(
  data=comt_poly_f3_f4_input_cast_mat_cpm_aldex_merge_dt[
    we.eBH<comt_poly_f3_f4_fdr_thresh],
  aes(x=effect, y=reorder(ID_SHORT, POS), fill=we.eBH, xmin=effect.low, xmax=effect.high))

comt_poly_f3_f4_aldex2_sig_gg+
  facet_wrap(~Fraction, scales="free_y")+
  geom_bar(stat="identity", alpha=0.75)+
  geom_errorbar(col="grey", width=0.3)+
  scale_fill_viridis(option="D", begin=0.95, end=0.05)+
  theme_cowplot()+
  labs(y="", x="Effect size", fill="")+
  theme(axis.text.y=element_text(size=8))

rm(comt_poly_f3_f4_aldex2_sig_gg)

# Generate a VE map using ALDEx2
comt_poly_f4_rep_summ_ve_dt<- comt_poly_f4_input_cast_mat_cpm_aldex_merge_dt[
  !grepl(",", AA_CHANGE), .(Median_effect=median(effect, na.rm=T),
                            Max_effect=max(effect, na.rm=T)), 
  by=.(ROI, AA_POS, REF_AA, ALT_AA, AA_CHANGE, VAR_TYPE)]

comt_poly_f4_rep_summ_ve_dt[,AA_POS:=as.integer(AA_POS)]
comt_poly_f4_rep_summ_ve_dt[,ALT_AA:=factor(ALT_AA, levels=unlist(aa_class_list))]
comt_poly_f4_rep_summ_ve_dt[,ALT_AA_level:=as.integer(ALT_AA)]

comt_poly_f4_aldex_ve_gg<- ggplot(
  data=comt_poly_f4_rep_summ_ve_dt, 
  aes(x=AA_POS, y=reorder(ALT_AA, ALT_AA_level), fill=Median_effect))

comt_poly_f4_aldex_ve_gg+
  facet_wrap(~ROI, scales="free")+
  geom_tile()+
  scale_fill_gradient2(high=scales::muted("blue"), low=scales::muted("red"))+
  labs(x="Amino acid position", y="Alternate", 
       fill="Median effect")+
  theme_cowplot()

rm(comt_poly_f4_aldex_ve_gg)

```

```{r COMT polysome differential variants analysis limma}

# First combine all datasets
comt_poly_limma_input_cast_mat_eset_fit_eb_all_dt<- rbind(
  comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_dt, 
  comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_dt, 
  comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_dt)

comt_poly_limma_input_cast_mat_eset_fit_eb_all_dt[,Fraction:=as.factor(rep(
  c("F2", "F3", "F4"), times=c(
    nrow(comt_poly_limma_f2_input_cast_mat_eset_fit_eb_all_dt), 
    nrow(comt_poly_limma_f3_input_cast_mat_eset_fit_eb_all_dt), 
    nrow(comt_poly_limma_f4_input_cast_mat_eset_fit_eb_all_dt))))]

comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt<- merge(
  comt_poly_limma_input_cast_mat_eset_fit_eb_all_dt, annot_dt, by="VAR_ID")

# Compare to CSC and ribosome A-site dwell score

# CSC from Wu et al. 2019
comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[
  ,Wu_REF_CSC:=csc_wu_2019_dt[
    match(comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[,REF_CODON], codon), `293T_ORFome`]]

comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[
  ,Wu_ALT_CSC:=csc_wu_2019_dt[
    match(comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[,ALT_CODON], codon), `293T_ORFome`]]

comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[
  ,Wu_CSC_diff:=Wu_ALT_CSC-Wu_REF_CSC]

# CSC from Narula et al. 2019
comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[
  ,Narula_REF_CSC:=csc_narula_2019_dt[
    match(comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[,REF_CODON], Codon), ORFome_CSC_mean]]

comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[
  ,Narula_ALT_CSC:=csc_narula_2019_dt[
    match(comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[,ALT_CODON], Codon), ORFome_CSC_mean]]

comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[
  ,Narula_CSC_diff:=Narula_ALT_CSC-Narula_REF_CSC]

# Add the A- and P-site dwell time data from Narula et al. 2019
# derived from Eichhorn et al. 2014
comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[
  ,Narula_ALT_Asite:=csc_narula_2019_dt[
    match(comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[,ALT_CODON], Codon), HEK293T]]

comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[
  ,Narula_REF_Asite:=csc_narula_2019_dt[
    match(comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[,REF_CODON], Codon), HEK293T]]

comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[
  ,Narula_ALT_Psite:=csc_narula_2019_dt[
    match(comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[,ALT_CODON], Codon), `HEK293T P site`]]

comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[
  ,Narula_REF_Psite:=csc_narula_2019_dt[
    match(comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[,REF_CODON], Codon), `HEK293T P site`]]

comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[,Narula_Asite_diff:=Narula_ALT_Asite-Narula_REF_Asite]
comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[,Narula_Psite_diff:=Narula_ALT_Psite-Narula_REF_Psite]

# Add tRNA abundance measurements
# Annotate Hydro-tRNAseq abundance
comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[
  ,REF_CODON_hydro_log2_sum:=hydro_tRNAseq_raw_agg_wobble_summ_dt[
    match(comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[,REF_CODON], Codon), log2_sum_count]]

comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[
  ,ALT_CODON_hydro_log2_sum:=hydro_tRNAseq_raw_agg_wobble_summ_dt[
    match(comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[,ALT_CODON], Codon), log2_sum_count]]

comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[
  ,hydro_log2_sum_diff:=ALT_CODON_hydro_log2_sum-REF_CODON_hydro_log2_sum]

# Annotate mim-tRNAseq abundance
comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[
  ,REF_CODON_mim_log2_sum:=mim_tRNAseq_raw_filt_agg_wobble_nostop_agg_summ_dt[
    match(comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[,REF_CODON], Codon), log2_sum_count]]

comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[
  ,ALT_CODON_mim_log2_sum:=mim_tRNAseq_raw_filt_agg_wobble_nostop_agg_summ_dt[
    match(comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[,ALT_CODON], Codon), log2_sum_count]]

comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[
  ,mim_log2_sum_diff:=ALT_CODON_mim_log2_sum-REF_CODON_mim_log2_sum]


# Call any variant < 0.05 as significant
comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[,Significant:=Direction]
comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[adj.P.Val>=0.05, Significant:="N.S."]
comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[,Significant:=factor(Significant, levels=c("N.S.", "Poly down", "Poly up"))]

# Is there an association between ribosome load and CSC? Use Wu et al 2019 data
poly_csc_gg<- ggplot(
  data=comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt,
  aes(x=Significant, y=Wu_CSC_diff))

poly_csc_gg+
  facet_wrap(~Fraction)+
  geom_boxplot()+
  theme_cowplot()+
  labs(x="", y="Wu Alt. - Ref. CSC")

poly_csc_gg<- ggplot(
  data=comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt,
  aes(x=Significant, y=Narula_CSC_diff))

poly_csc_gg+
  facet_wrap(~Fraction)+
  geom_boxplot()+
  theme_cowplot()+
  labs(x="", y="Narula Alt. - Ref. CSC")

# 
poly_asite_gg<- ggplot(
  data=comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt,
  aes(x=Significant, y=Narula_Asite_diff))

poly_asite_gg+
  facet_wrap(~Fraction)+
  geom_boxplot()+
  theme_cowplot()+
  labs(x="", y="Narula A-site dwell score")

poly_psite_gg<- ggplot(
  data=comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt,
  aes(x=Significant, y=Narula_Psite_diff))

poly_psite_gg+
  facet_wrap(~Fraction)+
  geom_boxplot()+
  theme_cowplot()+
  labs(x="", y="Narula P-site dwell score")

# Hydroseq
poly_trna_gg<- ggplot(
  data=comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt,
  aes(x=Significant, y=hydro_log2_sum_diff))

poly_trna_gg+
  facet_wrap(~Fraction)+
  geom_boxplot()+
  theme_cowplot()+
  labs(x="", y="Hydro-tRNAseq Alt. - Ref.")

# mim-tRNAseq
poly_trna_gg<- ggplot(
  data=comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt,
  aes(x=Significant, y=mim_log2_sum_diff))

poly_trna_gg+
  facet_wrap(~Fraction)+
  geom_boxplot()+
  theme_cowplot()+
  labs(x="", y="mim-tRNAseq Alt. - Ref.")

# No association with tRNA abundance

# Check dicodon and dipeptide pairs

# Inhibitory dicodons
comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_codon_inhib<- 
  sapply(comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[,VAR_ID], check_dicodon_pair,
         comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt, gamble_grayhack_inhibitory_codon_pairs_dna)

comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[
  comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_codon_inhib, table(Significant)]

comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[
  comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_codon_inhib & Significant=="Poly down", ID_SHORT]

comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[
  comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_codon_inhib & Significant=="Poly up", ID_SHORT]

# Of variants generating inhibitory dicodons, eleven were down in polysomes compared to 1 up in polysomes

# Optimal dicodons
comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_codon_optimal<- 
  sapply(comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[,VAR_ID], check_dicodon_pair,
         comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt, gamble_grayhack_optimal_codon_pairs_dna)

comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[
  comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_codon_optimal, table(Significant)]

# No variants generating optimal dicodons

# Inhibitory dipeptides
comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_aa_inhib<- 
  sapply(comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[,VAR_ID], check_dipeptides,
         comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt, burke_subramaniam_nonoptimal_dipeptides)

comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[
  comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_aa_inhib, table(Significant)]

# 69 variants leading to inhibitory dipeptides found in poly down compared
# to 1 in poly up!

# Optimal dipeptides
comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_aa_optimal<- 
  sapply(comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[,VAR_ID], check_dipeptides,
         comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt, burke_subramaniam_optimal_dipeptides)

comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[
  comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_aa_optimal, table(Significant)]

# 29 poly down variants leading to optimal dipeptides

# Investigate if variants altering 
comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[REF_AA=="M", M_variant:="From"]
comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[ALT_AA=="M", M_variant:="To"]
comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt[,M_variant:=as.factor(M_variant)]

comt_poly_methionine_gg<- ggplot(
  data=comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt,
  aes(x=M_variant, y=logFC))

comt_poly_methionine_gg+
  facet_wrap(~Fraction)+
  geom_boxplot()+
  theme_cowplot()+
  labs(x="Methionine variant", y="log FC")

# No global effect for fractions 3,4
# Variants to M seem to increase ribosome load in fraction 2

# What about silent variants in the wobble position?


```

```{r COMT polysome differential variants analysis ALDEx2}

comt_poly_loose_fdr_thresh<- 0.1

# First combine all datasets
comt_poly_input_cast_mat_cpm_aldex_all_dt<- rbind(
  comt_poly_f2_input_cast_mat_cpm_aldex_dt, 
  comt_poly_f3_input_cast_mat_cpm_aldex_dt, 
  comt_poly_f4_input_cast_mat_cpm_aldex_dt)

comt_poly_input_cast_mat_cpm_aldex_all_dt[,Fraction:=as.factor(rep(
  c("F2", "F3", "F4"), times=c(
    nrow(comt_poly_f2_input_cast_mat_cpm_aldex_dt), 
    nrow(comt_poly_f3_input_cast_mat_cpm_aldex_dt), 
    nrow(comt_poly_f4_input_cast_mat_cpm_aldex_dt))))]

comt_poly_input_cast_mat_cpm_aldex_all_merge_dt<- merge(
  comt_poly_input_cast_mat_cpm_aldex_all_dt, annot_dt, by="VAR_ID")

# Compare to CSC and ribosome A-site dwell score

# CSC from Wu et al. 2019
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  ,Wu_REF_CSC:=csc_wu_2019_dt[
    match(comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[,REF_CODON], codon), `293T_ORFome`]]

comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  ,Wu_ALT_CSC:=csc_wu_2019_dt[
    match(comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[,ALT_CODON], codon), `293T_ORFome`]]

comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  ,Wu_CSC_diff:=Wu_ALT_CSC-Wu_REF_CSC]

# CSC from Narula et al. 2019
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  ,Narula_REF_CSC:=csc_narula_2019_dt[
    match(comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[,REF_CODON], Codon), ORFome_CSC_mean]]

comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  ,Narula_ALT_CSC:=csc_narula_2019_dt[
    match(comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[,ALT_CODON], Codon), ORFome_CSC_mean]]

comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  ,Narula_CSC_diff:=Narula_ALT_CSC-Narula_REF_CSC]

# Add the A- and P-site dwell time data from Narula et al. 2019
# derived from Eichhorn et al. 2014
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  ,Narula_ALT_Asite:=csc_narula_2019_dt[
    match(comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[,ALT_CODON], Codon), HEK293T]]

comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  ,Narula_REF_Asite:=csc_narula_2019_dt[
    match(comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[,REF_CODON], Codon), HEK293T]]

comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  ,Narula_ALT_Psite:=csc_narula_2019_dt[
    match(comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[,ALT_CODON], Codon), `HEK293T P site`]]

comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  ,Narula_REF_Psite:=csc_narula_2019_dt[
    match(comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[,REF_CODON], Codon), `HEK293T P site`]]

comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[,Narula_Asite_diff:=Narula_ALT_Asite-Narula_REF_Asite]
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[,Narula_Psite_diff:=Narula_ALT_Psite-Narula_REF_Psite]

# Add tRNA abundance measurements
# Annotate Hydro-tRNAseq abundance
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  ,REF_CODON_hydro_log2_sum:=hydro_tRNAseq_raw_agg_wobble_summ_dt[
    match(comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[,REF_CODON], Codon), log2_sum_count]]

comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  ,ALT_CODON_hydro_log2_sum:=hydro_tRNAseq_raw_agg_wobble_summ_dt[
    match(comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[,ALT_CODON], Codon), log2_sum_count]]

comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  ,hydro_log2_sum_diff:=ALT_CODON_hydro_log2_sum-REF_CODON_hydro_log2_sum]

# Annotate mim-tRNAseq abundance
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  ,REF_CODON_mim_log2_sum:=mim_tRNAseq_raw_filt_agg_wobble_nostop_agg_summ_dt[
    match(comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[,REF_CODON], Codon), log2_sum_count]]

comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  ,ALT_CODON_mim_log2_sum:=mim_tRNAseq_raw_filt_agg_wobble_nostop_agg_summ_dt[
    match(comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[,ALT_CODON], Codon), log2_sum_count]]

comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  ,mim_log2_sum_diff:=ALT_CODON_mim_log2_sum-REF_CODON_mim_log2_sum]


# Call any variant < 0.05 as significant
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[,Direction:=ifelse(effect>0, "Poly up", "Poly down")]
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[we.eBH>=comt_poly_loose_fdr_thresh, Direction:="N.S."]

# Is there an association between ribosome load and CSC? Use Wu et al 2019 data
poly_csc_gg<- ggplot(
  data=comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[Fraction!="F2"],
  aes(x=Direction, y=Wu_CSC_diff))

poly_csc_gg+
  facet_wrap(~Fraction)+
  geom_boxplot(width=0.5, fill="steelblue", alpha=0.75)+
  theme_cowplot()+
  labs(x="", y="Wu Alt. - Ref. CSC")

poly_csc_gg<- ggplot(
  data=comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[Fraction!="F2"],
  aes(x=Direction, y=Narula_CSC_diff))

poly_csc_gg+
  facet_wrap(~Fraction)+
  geom_boxplot(width=0.5, fill="steelblue", alpha=0.75)+
  theme_cowplot()+
  labs(x="", y="Narula Alt. - Ref. CSC")

poly_csc_gg<- ggplot(
  data=comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[Fraction=="F3"],
  aes(x=Direction, y=Narula_CSC_diff))

poly_csc_gg+
  geom_boxplot(width=0.5, fill="steelblue", alpha=0.75)+
  theme_cowplot()+
  labs(x="", y="Narula Alt. - Ref. CSC")

rm(poly_csc_gg)

comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[,Significant:="False"]
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[we.eBH<comt_poly_loose_fdr_thresh, Significant:="True"]

# Fraction 3
# Poly up versus poly down
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  !is.na(Narula_CSC_diff) & Significant=="True" & Fraction=="F3", wilcox.test(
    formula=Narula_CSC_diff~Direction, data=.SD, alternative="less")]

# Poly down versus N.S.
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  !is.na(Narula_CSC_diff) & Direction!="Poly up" & Fraction=="F3", wilcox.test(
    formula=Narula_CSC_diff~Direction, data=.SD, alternative="greater")]

# Poly up versus N.S.
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  !is.na(Narula_CSC_diff) & Direction!="Poly down" & Fraction=="F3", wilcox.test(
    formula=Narula_CSC_diff~Direction, data=.SD, alternative="less")]

# Is a similar result seen with the Wu CSC scores?
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  !is.na(Narula_CSC_diff) & Significant=="True" & Fraction=="F3", wilcox.test(
    formula=Wu_CSC_diff~Direction, data=.SD, alternative="less")]

# Fraction 4
# Poly up versus poly down
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  !is.na(Narula_CSC_diff) & Significant=="True" & Fraction=="F4", wilcox.test(
    formula=Narula_CSC_diff~Direction, data=.SD, alternative="less", exact=FALSE)]

# Poly down versus N.S.
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  !is.na(Narula_CSC_diff) & Direction!="Poly up" & Fraction=="F4", wilcox.test(
    formula=Narula_CSC_diff~Direction, data=.SD, alternative="greater")]

# Poly up versus N.S.
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  !is.na(Narula_CSC_diff) & Direction!="Poly down" & Fraction=="F4", wilcox.test(
    formula=Narula_CSC_diff~Direction, data=.SD, alternative="less")]

# CSC only significant for F3

# Compare variants leading to the same amino acid change but different codons, one significant and the other(s) N.S.
# comt_poly_loose_fdr_thresh
# comt_poly_f3_f4_fdr_thresh

# Remove p.G157T as it has two significant poly down variants; both of which have more negative CSC diff than the N.S. variants
poly_sig_aas<- comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[Fraction=="F4" & we.eBH < comt_poly_loose_fdr_thresh & AA_CHANGE!="p.G157T", unique(AA_CHANGE)]

comt_poly_input_cast_mat_cpm_aldex_all_merge_sig_dt<- comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[Fraction=="F4" & AA_CHANGE%in%poly_sig_aas]

# Plot only those variants with 1 other N.S. variant?
comt_poly_input_cast_mat_cpm_aldex_all_merge_sig_counts_dt<- comt_poly_input_cast_mat_cpm_aldex_all_merge_sig_dt[
  Fraction=="F4",.N, by=.(AA_CHANGE)]

poly_sig_aas_filt<- poly_sig_aas[!poly_sig_aas%in%comt_poly_input_cast_mat_cpm_aldex_all_merge_sig_counts_dt[N==1, AA_CHANGE]]

# Determine if the significant variant has the max/min CSC compared to N.S. variants
sig_is_max_csc<- function(aa_change, dt){
  
  dt_slice<- dt[AA_CHANGE==aa_change]
  sig_csc<- dt_slice[Significant=="True", Narula_CSC_diff]
  nonsig_csc<- dt_slice[Significant=="False", Narula_CSC_diff]
  sig_direction<- dt_slice[Significant=="True", Direction]
  
  if(length(sig_direction)==0){
    print(dt_slice)
  }
  
  if(sig_direction=="Poly up"){
    return(sig_csc > max(nonsig_csc))
  } else {
    return(sig_csc < min(nonsig_csc))
  }
}

# Add TRUE for this variant
sig_is_max_res<- sapply(poly_sig_aas_filt, sig_is_max_csc, dt=comt_poly_input_cast_mat_cpm_aldex_all_merge_sig_dt)
# table(sig_is_max_res)
# Not significant overall

comt_poly_csc_comp_aa_changes_selected<- comt_poly_input_cast_mat_cpm_aldex_all_merge_sig_counts_dt[N==2, AA_CHANGE]

# Add in two more variants with two N.S. variants
comt_poly_csc_comp_aa_changes_selected<- c(comt_poly_csc_comp_aa_changes_selected, c("p.P45V", "p.V63V"))

comt_poly_input_cast_mat_cpm_aldex_all_merge_sig_select_dt<- comt_poly_input_cast_mat_cpm_aldex_all_merge_sig_dt[
  AA_CHANGE%in%comt_poly_csc_comp_aa_changes_selected]

comt_poly_input_cast_mat_cpm_aldex_all_merge_sig_select_dt[,Group:=as.integer(as.factor(AA_CHANGE))]

# Determine the null distribution of CSC diff scores
csc_diff_dist<- comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  Fraction=="F4", as.numeric(dist(Narula_CSC_diff, method="manhattan"))]

csc_diff_dist_nonas<- csc_diff_dist[!is.na(csc_diff_dist)]

csc_diff_cast_dt<- dcast(
  data=comt_poly_input_cast_mat_cpm_aldex_all_merge_sig_select_dt, 
  AA_CHANGE~VAR_ID, value.var="Narula_CSC_diff", fun.aggregate=median)

get_diff<- function(x){
  x_vec<- as.vector(x)
  x_no_na<- sort(x_vec[!is.na(x_vec)])
  x_diff<- abs(x_no_na[length(x_no_na)] - x_no_na[1])
  return(x_diff)
}

csc_diff_cast_dt$csc_diff<- apply(csc_diff_cast_dt[,-c("AA_CHANGE")], 1, get_diff)
#csc_diff_cast_dt[,.(AA_CHANGE, csc_diff), keyby=.(csc_diff)]

# None of the changes are in the upper quartile of the null distribution
# Plot the top four AA changes

codon_opti_cols<- c("black", "darkorange", "dodgerblue")

comt_poly_csc_comp_aa_changes_final<- c("p.E68N", "p.P45V", "p.V63V", "p.N71F", "p.P143F", "p.L136F", "p.I141F", "p.I148C", "p.A156Y")

comt_poly_csc_comp_gg<- ggplot(
  data=comt_poly_input_cast_mat_cpm_aldex_all_merge_sig_select_dt[
    Fraction=="F4" & AA_CHANGE%in%comt_poly_csc_comp_aa_changes_final], 
  aes(x=Narula_CSC_diff, y=reorder(AA_CHANGE, AA_POS2)))

comt_poly_csc_comp_gg+
  geom_vline(xintercept=0, col="grey", linetype=2)+
  scale_color_manual(values=codon_opti_cols)+
  geom_jitter(aes(col=Direction), size=4, height=0.05)+
  labs(y="Amino acid", x="Alt. - Ref. CSC", col="Direction")+
  theme_cowplot()

# Among AA changes with one significant and one N.S. variants
# 10 showed CSC in expected direction and 4 showed CSC in unexpected direction
#pbinom(10,14,0.5,lower.tail=FALSE)

# Determine the identity of the changes to V and F
#comt_poly_input_cast_mat_cpm_aldex_all_merge_sig_select_dt[Fraction=="F4" & AA_CHANGE%in%comt_poly_csc_comp_aa_changes_final & Significant=="True", .(AA_CHANGE, REF_CODON, ALT_CODON)]

# V=GTC, F=TTC, N=AAC
# All poly up variants have a C in the wobble position
# A site dwell
poly_asite_gg<- ggplot(
  data=comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[Fraction!="F2"],
  aes(x=Direction, y=Narula_Asite_diff))

poly_asite_gg+
  facet_wrap(~Fraction)+
  geom_boxplot(width=0.5, fill="steelblue", alpha=0.75)+
  theme_cowplot()+
  labs(x="", y="Narula A-site dwell score")

rm(poly_asite_gg)

# Poly down versus poly up
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  !is.na(Narula_Asite_diff) & Significant=="True" & Fraction=="F4", wilcox.test(
    formula=Narula_Asite_diff~Direction, data=.SD, exact=FALSE)]

# Poly down versus N.S.
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  !is.na(Narula_Asite_diff) & Direction!="Poly up" & Fraction=="F4", wilcox.test(
    formula=Narula_Asite_diff~Direction, data=.SD)]

# Poly up versus N.S.
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  !is.na(Narula_Asite_diff) & Direction!="Poly down" & Fraction=="F4", wilcox.test(
    formula=Narula_Asite_diff~Direction, data=.SD)]

# A-site not significant for F4

# P-site dwell
poly_psite_gg<- ggplot(
  data=comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[Fraction!="F2"],
  aes(x=Direction, y=Narula_Psite_diff))

poly_psite_gg+
  facet_wrap(~Fraction)+
  geom_boxplot(width=0.5, fill="steelblue", alpha=0.75)+
  theme_cowplot()+
  labs(x="", y="Narula P-site dwell score")

poly_psite_gg<- ggplot(
  data=comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[Fraction=="F3"],
  aes(x=Direction, y=Narula_Psite_diff))

poly_psite_gg+
  geom_boxplot(width=0.5, fill="steelblue", alpha=0.75)+
  theme_cowplot()+
  labs(x="", y="Narula P-site dwell score")

rm(poly_psite_gg)

# Poly down versus poly up
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  !is.na(Narula_Psite_diff) & Significant=="True" & Fraction=="F3", wilcox.test(
    formula=Narula_Psite_diff~Direction, data=.SD)]

# Poly down versus N.S.
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  !is.na(Narula_Psite_diff) & Direction!="Poly up" & Fraction=="F3", wilcox.test(
    formula=Narula_Psite_diff~Direction, data=.SD)]

# Poly up versus N.S.
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  !is.na(Narula_Psite_diff) & Direction!="Poly down" & Fraction=="F3", wilcox.test(
    formula=Narula_Psite_diff~Direction, data=.SD)]

# P-site dwell significantly up in poly down compared to poly up

# Hydroseq
poly_trna_gg<- ggplot(
  data=comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[Fraction!="F2"],
  aes(x=Direction, y=hydro_log2_sum_diff))

poly_trna_gg+
  facet_wrap(~Fraction)+
  geom_boxplot(width=0.5, fill="steelblue", alpha=0.75)+
  theme_cowplot()+
  labs(x="", y="Hydro-tRNAseq Alt. - Ref.")

poly_trna_gg<- ggplot(
  data=comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[Fraction=="F4"],
  aes(x=Direction, y=hydro_log2_sum_diff))

poly_trna_gg+
  geom_boxplot(width=0.5, fill="steelblue", alpha=0.75)+
  theme_cowplot()+
  labs(x="", y="Hydro-tRNAseq Alt. - Ref.")

# Poly down versus poly up
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  !is.na(hydro_log2_sum_diff) & Significant=="True" & Fraction=="F4", wilcox.test(
    formula=hydro_log2_sum_diff~Direction, data=.SD)]

# Poly down versus N.S.
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  !is.na(hydro_log2_sum_diff) & Direction!="Poly up" & Fraction=="F4", wilcox.test(
    formula=hydro_log2_sum_diff~Direction, data=.SD)]

# Poly up versus N.S.
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  !is.na(hydro_log2_sum_diff) & Direction!="Poly down" & Fraction=="F4", wilcox.test(
    formula=hydro_log2_sum_diff~Direction, data=.SD)]

# mim-tRNAseq
poly_trna_gg<- ggplot(
  data=comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[Fraction!="F2"],
  aes(x=Direction, y=mim_log2_sum_diff))

poly_trna_gg+
  facet_wrap(~Fraction)+
  geom_boxplot()+
  theme_cowplot()+
  labs(x="", y="mim-tRNAseq Alt. - Ref.")

rm(poly_trna_gg)

# tRNA not significant

# Check dicodon and dipeptide pairs

# Inhibitory dicodons
comt_poly_input_cast_mat_cpm_aldex_all_merge_codon_inhib<- 
  sapply(comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[,VAR_ID], check_dicodon_pair,
         comt_poly_input_cast_mat_cpm_aldex_all_merge_dt, gamble_grayhack_inhibitory_codon_pairs_dna)

#comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[comt_poly_input_cast_mat_cpm_aldex_all_merge_codon_inhib, table(Direction)]

comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  comt_poly_input_cast_mat_cpm_aldex_all_merge_codon_inhib & Direction=="Poly down", .(Fraction, ID_SHORT, effect, we.eBH)]

# Of variants generating inhibitory dicodons, 1 was down in polysome F3

# Optimal dicodons
comt_poly_input_cast_mat_cpm_aldex_all_merge_codon_optimal<- 
  sapply(comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[,VAR_ID], check_dicodon_pair,
         comt_poly_input_cast_mat_cpm_aldex_all_merge_dt, gamble_grayhack_optimal_codon_pairs_dna)

#comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[comt_poly_input_cast_mat_cpm_aldex_all_merge_codon_optimal, table(Direction)]

# No optimal dicodons found

# Inhibitory dipeptides
comt_poly_input_cast_mat_cpm_aldex_all_merge_aa_inhib<- 
  sapply(comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[,VAR_ID], check_dipeptides,
         comt_poly_input_cast_mat_cpm_aldex_all_merge_dt, burke_subramaniam_nonoptimal_dipeptides)

#comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[comt_poly_input_cast_mat_cpm_aldex_all_merge_aa_inhib, table(Direction)]

# 6 variants to inhibitory AAs were down in polysome F3, 2 up in polysomes (F3, F4)

comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  comt_poly_input_cast_mat_cpm_aldex_all_merge_aa_inhib & Direction=="Poly down", .(Fraction, ID_SHORT, effect, we.eBH)]

comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[
  comt_poly_input_cast_mat_cpm_aldex_all_merge_aa_inhib & Direction=="Poly up", .(Fraction, ID_SHORT, effect, we.eBH)]


# Optimal dipeptides
comt_poly_input_cast_mat_cpm_aldex_all_merge_aa_optimal<- 
  sapply(comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[,VAR_ID], check_dipeptides,
         comt_poly_input_cast_mat_cpm_aldex_all_merge_dt, burke_subramaniam_optimal_dipeptides)

#comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[comt_poly_input_cast_mat_cpm_aldex_all_merge_aa_optimal, table(Direction)]

# No variants leading to optimal dipeptides

# Investigate if variants altering methionine have any properties
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[REF_AA=="M", M_variant:="From"]
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[ALT_AA=="M", M_variant:="To"]
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[,M_variant:=as.factor(M_variant)]

comt_poly_methionine_gg<- ggplot(
  data=comt_poly_input_cast_mat_cpm_aldex_all_merge_dt,
  aes(x=M_variant, y=effect))

comt_poly_methionine_gg+
  facet_wrap(~Fraction)+
  geom_boxplot()+
  theme_cowplot()+
  labs(x="Methionine variant", y="Effect size")

rm(comt_poly_methionine_gg)

# No global effect for fractions 3,4
# Variants to M seem to increase ribosome load in fraction 2

# How are silent variants distributed?
# comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[VAR_TYPE=="Silent", table(Direction, Fraction)]

comt_poly_silent_gg<- ggplot(
  data=comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[VAR_TYPE=="Silent"], 
  aes(x=Direction, y=effect))

comt_poly_silent_gg+
  facet_wrap(~Fraction)+
  geom_boxplot()+
  geom_jitter(width=0.2)+
  theme_cowplot()+
  labs(x="Direction", y="Effect size")

comt_poly_silent_gg<- ggplot(
  data=comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[VAR_TYPE=="Silent"], 
  aes(x=Narula_CSC_diff, y=effect))

comt_poly_silent_gg+
  facet_wrap(~Fraction)+
  geom_point()+
  theme_cowplot()+
  labs(x="CSC diff", y="Effect size")

# Possible subtle correlation for F3
comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[VAR_TYPE=="Silent" & Fraction=="F3", 
                                                cor.test(Narula_CSC_diff, effect, method="spearman")]

comt_poly_silent_gg<- ggplot(
  data=comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[VAR_TYPE=="Silent"], 
  aes(x=Narula_Asite_diff, y=effect))

comt_poly_silent_gg+
  facet_wrap(~Fraction)+
  geom_point()+
  theme_cowplot()+
  labs(x="A-site diff", y="Effect size")

comt_poly_silent_gg<- ggplot(
  data=comt_poly_input_cast_mat_cpm_aldex_all_merge_dt[VAR_TYPE=="Silent"], 
  aes(x=Narula_Psite_diff, y=effect))

comt_poly_silent_gg+
  facet_wrap(~Fraction)+
  geom_point()+
  theme_cowplot()+
  labs(x="P-site diff", y="Effect size")

# No association with A- and P-site dwell scores

# What about silent variants in the wobble position?


```

```{r Poly versus RNA abundance limma}

comt_rna_vs_protein_dt<- merge(
  comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,.(VAR_ID, logFC, adj.P.Val)],
  comt_poly_limma_input_cast_mat_eset_fit_eb_all_merge_dt, 
  by="VAR_ID")

comt_rna_vs_protein_dt[
  (logFC.x < -0.4 & logFC.y > 0.3) |
    (logFC.x > 0.5 & logFC.y < -0.5), Label:=AA_CHANGE]

comt_rna_vs_protein_gg<- ggplot(
  data=comt_rna_vs_protein_dt[Fraction%in%c("F3","F4")], 
  aes(x=logFC.x, y=logFC.y, label=Label))

comt_rna_vs_protein_gg+
  geom_point(alpha=0.5)+
  facet_wrap(~Fraction)+
  geom_point(data=comt_rna_vs_protein_dt[Fraction%in%c("F3","F4") & VAR_TYPE=="Silent"], 
             aes(x=logFC.x, y=logFC.y), col="dodgerblue")+
  geom_point(data=comt_rna_vs_protein_dt[Fraction%in%c("F3","F4") & VAR_TYPE=="Nonsense"], 
             aes(x=logFC.x, y=logFC.y), col="firebrick")+
  geom_text_repel()+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_vline(xintercept=0, col="grey", linetype=2)+
  theme_cowplot()+
  labs(x="RNA log FC", y="Poly log FC")

rm(comt_rna_vs_protein_gg)

```

```{r Poly versus RNA abundance ALDEx2}

comt_rna_vs_protein_aldex_dt<- merge(
  comt_rna_input_cast_mat_cpm_aldex_merge_dt[,.(VAR_ID, diff.btw, effect, we.eBH)],
  comt_poly_input_cast_mat_cpm_aldex_all_merge_dt, 
  by="VAR_ID")

comt_rna_vs_protein_aldex_gg<- ggplot(
  data=comt_rna_vs_protein_aldex_dt[Fraction%in%c("F3","F4") & VAR_TYPE=="Missense"], 
  aes(x=effect.x, y=effect.y))

comt_rna_vs_protein_aldex_gg+
  geom_point(alpha=0.5)+
  facet_wrap(~Fraction)+
  geom_point(data=comt_rna_vs_protein_aldex_dt[Fraction%in%c("F3","F4") & VAR_TYPE=="Silent"], 
             aes(x=effect.x, y=effect.y), col="dodgerblue")+
  geom_point(data=comt_rna_vs_protein_aldex_dt[Fraction%in%c("F3","F4") & VAR_TYPE=="Nonsense"], 
             aes(x=effect.x, y=effect.y), col="firebrick")+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_vline(xintercept=0, col="grey", linetype=2)+
  theme_cowplot()+
  labs(x="RNA effect", y="Poly effect")

rm(comt_rna_vs_protein_aldex_gg)

comt_rna_silent_vs_nonsense_aldex_gg<- ggplot(
  data=comt_rna_vs_protein_aldex_dt[Fraction%in%c("F3","F4") & VAR_TYPE!="Missense"],
  aes(x=Fraction, y=effect.y, col=VAR_TYPE))

comt_rna_silent_vs_nonsense_aldex_gg+
  geom_point(position=position_jitterdodge(), alpha=0.75)+
  scale_color_manual(values=c("dodgerblue", "firebrick"))+
  stat_summary(fun=median, fun.min=median, fun.max=median,
               geom="crossbar", width=0.3, position=position_dodge(0.75))+
  theme_cowplot()+
  labs(x="", y="Polysome effect", col="")

rm(comt_rna_silent_vs_nonsense_aldex_gg)

# Test silent versus nonsense variants in each fraction
comt_rna_vs_protein_aldex_dt[
  VAR_TYPE!="Missense" & Fraction=="F3", wilcox.test(
    formula=effect.y~VAR_TYPE, data=.SD)]

comt_rna_vs_protein_aldex_dt[
  VAR_TYPE!="Missense" & Fraction=="F4", wilcox.test(
    formula=effect.y~VAR_TYPE, data=.SD)]

```


# Variants with differential protein abundance

Analyze variant frequencies in each population/quadrant relative to gDNA frequency to determine effects on protein abundance.

```{r COMT FACS plots}

comt_facs_trial2_dir<- paste(comt_facs_dir, "Trial2", sep="/")
comt_facs_trial2_files<- list.files(comt_facs_trial2_dir, full.names=TRUE)
comt_facs_trial2_list<- lapply(comt_facs_trial2_files, fread, sep=",", header=TRUE)
comt_facs_trial2_list_nrows<- sapply(comt_facs_trial2_list, nrow)

comt_facs_trial2_all_dt<- rbindlist(comt_facs_trial2_list)

comt_facs_trial2_all_dt[,Sample:=factor(
  rep(c("ROI_1", "ROI_2", "WT no Dox", "WT"), times=comt_facs_trial2_list_nrows), 
  levels=c("WT no Dox", "WT", "ROI_1", "ROI_2"))]

rm(comt_facs_trial2_list)

comt_facs_trial2_all_dt[,`:=`(log10_GFP=log10(`GFP-A`), log10_mCherry=log10(`mCherry-A`))]

facs_three_cols<- c("#374E55FF", roi1_rgb, roi2_rgb)

comt_facs_trial2_fluor_gg<- ggplot(
  data=comt_facs_trial2_all_dt[Sample!="WT no Dox"], 
  aes(x=log10_mCherry, y=log10_GFP, col=Sample))

comt_facs_trial2_fluor_gg+
  facet_wrap(~Sample)+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.3)+
  scale_color_manual(values=facs_three_cols)+
  labs(x="mCherry (mRNA proxy)", y="moxGFP (protein)")+
  theme_cowplot()

comt_facs_trial2_all_dt[!is.na(log10_mCherry) & !is.na(log10_GFP), cor(log10_mCherry, log10_GFP), by=.(Sample)]

```

```{r COMT FACS analysis}

# comt_all_filt_postprocess_subtract_drop_bio_dt
# comt_all_filt_postprocess_subtract_drop_cast_combat_dt

comt_all_filt_postprocess_subtract_drop_cast_combat_bio_dt[,AA_POS:=as.integer(AA_POS)]
comt_all_filt_postprocess_subtract_drop_cast_combat_bio_dt[,ALT_AA:=factor(ALT_AA, levels=unlist(aa_class_list))]
comt_all_filt_postprocess_subtract_drop_cast_combat_bio_dt[,ALT_AA_level:=as.integer(ALT_AA)]

# First use summarized data to count variant types in each quadrant
comt_all_filt_postprocess_subtract_drop_bio_facs_quad_counts_dt<- 
  comt_all_filt_postprocess_subtract_drop_cast_combat_bio_dt[Input=="FACS", .N, by=.(ROI, VAR_TYPE, Quadrant)]

facs_quad_counts_gg<- ggplot(
  data=comt_all_filt_postprocess_subtract_drop_bio_facs_quad_counts_dt[VAR_TYPE!="Missense"], 
  aes(x=VAR_TYPE, y=N, fill=Quadrant))

facs_quad_counts_gg+
  facet_wrap(~ROI)+
  geom_bar(stat="identity", position="dodge")+
  scale_fill_aaas()+
  labs(x="", y="Number variants")+
  theme_cowplot()

rm(facs_quad_counts_gg)

# No obvious count difference between FACS quadrants; this could 
# simply be due to equivalent detection in quadrants

# Determine the quadrant of max frequency and then count
# Do not consider frequency in the input for now
comt_all_filt_postprocess_subtract_drop_bio_facs_max_dt<- 
  comt_all_filt_postprocess_subtract_drop_cast_combat_bio_dt[
    Input=="FACS" & Quadrant!="Input", 
    .(Max_pop=.SD[which.max(log10_NORM_CAF), Quadrant]),
    by=.(ROI, VAR_ID, POS, REF, ALT, REF_CODON, ALT_CODON, 
         AA_POS, REF_AA, ALT_AA, ALT_AA_level, AA_CHANGE, TYPE, VAR_TYPE), 
    .SDcols=c("Quadrant", "log10_NORM_CAF")]

comt_all_filt_postprocess_subtract_drop_bio_facs_max_quad_counts_dt<- 
  comt_all_filt_postprocess_subtract_drop_bio_facs_max_dt[,.N,by=.(ROI, VAR_TYPE, Max_pop)]

facs_maxpop_quad_counts_gg<- ggplot(
  data=comt_all_filt_postprocess_subtract_drop_bio_facs_max_quad_counts_dt[VAR_TYPE!="Missense"], 
  aes(x=VAR_TYPE, y=N, fill=Max_pop))

facs_maxpop_quad_counts_gg+
  facet_wrap(~ROI)+
  geom_bar(stat="identity", position="dodge", alpha=0.75)+
  scale_fill_aaas()+
  labs(x="", y="Number variants")+
  theme_cowplot()

rm(facs_maxpop_quad_counts_gg)

# Now counts make more sense: silent variants are found more often in the 1st and 4th quadrants
# and nonsense are found more in the 2nd and 3rd quadrants

# For simplicity, just plot counts for low versus high protein
comt_all_filt_postprocess_subtract_drop_bio_facs_max_dt[
  Max_pop%in%c("2", "3"), Pop_class:="Low protein"]

comt_all_filt_postprocess_subtract_drop_bio_facs_max_dt[
  Max_pop%in%c("1", "4"), Pop_class:="High protein"]

comt_all_filt_postprocess_subtract_drop_bio_facs_max_dt[,Pop_class:=as.factor(Pop_class)]

comt_all_filt_postprocess_subtract_drop_bio_facs_max_quad_counts_dt<- 
  comt_all_filt_postprocess_subtract_drop_bio_facs_max_dt[,.N,by=.(ROI, VAR_TYPE, Pop_class)]

facs_maxpop_quad_counts_gg<- ggplot(
  data=comt_all_filt_postprocess_subtract_drop_bio_facs_max_quad_counts_dt[VAR_TYPE!="Missense"], 
  aes(x=VAR_TYPE, y=N, fill=Pop_class))

facs_maxpop_quad_counts_gg+
  facet_wrap(~ROI)+
  geom_bar(stat="identity", position="dodge", alpha=0.8)+
  scale_fill_aaas()+
  labs(x="", y="Number variants", fill="")+
  theme_cowplot()

# Plot both ROIs together
facs_maxpop_quad_counts_gg<- ggplot(
  data=comt_all_filt_postprocess_subtract_drop_bio_facs_max_quad_counts_dt[VAR_TYPE!="Missense"], 
  aes(x=VAR_TYPE, y=N, fill=Pop_class))

facs_maxpop_quad_counts_gg+
  geom_bar(stat="identity", position="dodge", alpha=0.8)+
  scale_fill_aaas()+
  labs(x="", y="Number variants", fill="")+
  theme_cowplot()

rm(facs_maxpop_quad_counts_gg)

# Make a max-pop variant effect heatmap
# Aggregate variants for each amino acid change
comt_all_filt_postprocess_subtract_drop_bio_facs_max_summ_dt<- 
  comt_all_filt_postprocess_subtract_drop_bio_facs_max_dt[
    , .N, by=.(ROI, AA_POS, REF_AA, ALT_AA, AA_CHANGE, ALT_AA_level, Pop_class)]

comt_all_filt_postprocess_subtract_drop_bio_facs_max_summ_dt<- 
  comt_all_filt_postprocess_subtract_drop_bio_facs_max_summ_dt[
    ,.(Max_pop_class=.SD[which.max(N), Pop_class]), 
    by=.(ROI, AA_POS, REF_AA, ALT_AA, AA_CHANGE, ALT_AA_level), .SDcols=c("N", "Pop_class")]

comt_facs_ve_heatmap_gg<- ggplot(
  data=comt_all_filt_postprocess_subtract_drop_bio_facs_max_summ_dt[!is.na(AA_POS)], 
  aes(x=AA_POS, y=reorder(ALT_AA, ALT_AA_level), fill=Max_pop_class))

comt_facs_ve_heatmap_gg+
  facet_wrap(~ROI, scales="free")+
  geom_tile(alpha=0.8)+
  scale_fill_aaas()+
  labs(x="Amino acid position", y="Alternate", 
       fill="Max population")+
  theme_cowplot()

rm(comt_facs_ve_heatmap_gg)

# As another method of analysis, take the frequency difference between
# each quadrant and gDNA and take the upper tail of the absolute differences 
# to identify variants enriched in any one quadrant

comt_all_filt_postprocess_subtract_drop_bio_merge_dt<- 
  merge(comt_all_filt_postprocess_subtract_drop_cast_combat_bio_dt[
    Input=="gDNA", .(log10_NORM_CAF, log_NORM_CAF, VAR_ID)],
    comt_all_filt_postprocess_subtract_drop_cast_combat_bio_dt[
      Input=="FACS"], by="VAR_ID")

comt_all_filt_postprocess_subtract_drop_bio_merge_dt[,log10_NORM_CAF_diff:=log10_NORM_CAF.y-log10_NORM_CAF.x]
comt_all_filt_postprocess_subtract_drop_bio_merge_dt[,log10_NORM_CAF_diff_abs:=abs(log10_NORM_CAF_diff)]

comt_all_filt_postprocess_subtract_drop_bio_merge_dt[,log_NORM_CAF_diff:=log_NORM_CAF.y-log_NORM_CAF.x]
comt_all_filt_postprocess_subtract_drop_bio_merge_dt[,log_NORM_CAF_diff_abs:=abs(log_NORM_CAF_diff)]

# Plot distribution of the difference for each quadrant
comt_facs_freq_diff_distr_gg<- ggplot(
  data=comt_all_filt_postprocess_subtract_drop_bio_merge_dt[Quadrant!="Input"], 
  aes(x=log10_NORM_CAF_diff, col=Quadrant))

comt_facs_freq_diff_distr_gg+
  facet_wrap(~Quadrant)+
  geom_density(linewidth=1.2)+
  geom_vline(xintercept=0, col="grey", linetype=2)+
  labs(x="log10 frequency difference to gDNA", col="Quadrant")+
  theme_cowplot()

rm(comt_facs_freq_diff_distr_gg)

comt_all_filt_postprocess_subtract_drop_bio_merge_dt[
  Quadrant%in%c("2", "3"), Pop_class:="Low protein"]

comt_all_filt_postprocess_subtract_drop_bio_merge_dt[
  Quadrant%in%c("1", "4"), Pop_class:="High protein"]

comt_all_filt_postprocess_subtract_drop_bio_merge_dt[,Pop_class:=as.factor(Pop_class)]

# As a sanity check, determine the difference for all nonsense variants
comt_facs_freq_diff_nonsense_distr_gg<- ggplot(
  data=comt_all_filt_postprocess_subtract_drop_bio_merge_dt[
    Quadrant!="Input" & VAR_TYPE=="Nonsense"], 
  aes(x=log10_NORM_CAF_diff, col=Quadrant))

comt_facs_freq_diff_nonsense_distr_gg+
  facet_wrap(~Quadrant)+
  geom_density(size=1.2)+
  geom_vline(xintercept=0, col="grey", linetype=2)+
  scale_color_nejm()+
  labs(x="log10 frequency difference to gDNA", col="Quadrant")+
  theme_cowplot()

comt_facs_freq_diff_nonsense_distr_gg<- ggplot(
  data=comt_all_filt_postprocess_subtract_drop_bio_merge_dt[
    Quadrant%in%c("1", "3") & VAR_TYPE!="Missense"], 
  aes(x=log10_NORM_CAF_diff, fill=Quadrant))

comt_facs_freq_diff_nonsense_distr_gg+
  facet_wrap(VAR_TYPE~Pop_class)+
  geom_histogram(alpha=0.75)+
  geom_vline(xintercept=0, col="grey", linetype=2)+
  scale_fill_aaas()+
  labs(x="log10 frequency difference to gDNA", col="Quadrant")+
  theme_cowplot()

rm(comt_facs_freq_diff_nonsense_distr_gg)

```

```{r COMT FACS Q1 limma}

facs_fdr_thresh<- 0.01
replicate_count_thresh<- 8
log_caf_thresh<- -10

# As a final method to identify variants with differential protein
# use limma to compare gDNA frequencies to frequencies in each quadrant

# Discard the FACS input samples and use gDNA in place
comt_facs_limma_dt<- comt_all_filt_postprocess_subtract_drop_cast_combat_dt[
  Input%in%c("gDNA", "FACS") & !grepl("Input", Key_bio)]

comt_facs_limma_dt[,Quadrant:=sapply(strsplit(as.character(Key_bio), "_"), "[", 3)]

comt_facs_limma_q1_dt<- comt_facs_limma_dt[is.na(Quadrant) | Quadrant==1]

# Use a filter on frequency; corresponds to ~10 counts
comt_facs_limma_q1_counts_dt<- comt_facs_limma_q1_dt[Median_log_CAF>log_caf_thresh, .N, by=.(VAR_ID)]

comt_facs_limma_complete_varids<- comt_facs_limma_q1_counts_dt[
  N>=replicate_count_thresh, unique(VAR_ID)]

comt_facs_limma_q1_input_dt<- comt_facs_limma_q1_dt[VAR_ID%in%comt_facs_limma_complete_varids]

comt_facs_limma_q1_input_cast_dt<- dcast(
  data=comt_facs_limma_q1_input_dt, formula=VAR_ID~Key, value.var="Median_log_CAF")

comt_facs_limma_q1_input_cast_mat<- as.matrix(
  comt_facs_limma_q1_input_cast_dt[,-c("VAR_ID")])

rownames(comt_facs_limma_q1_input_cast_mat)<- comt_facs_limma_q1_input_cast_dt[,VAR_ID]

#comt_facs_limma_q1_input_cast_mat<- normalizeQuantiles(comt_facs_limma_q1_input_cast_mat)

comt_facs_limma_q1_input_cast_mat_eset<- Biobase::ExpressionSet(
  assayData=comt_facs_limma_q1_input_cast_mat)

comt_facs_limma_q1_input_design_dt<- 
  comt_facs_limma_q1_input_dt[,as.factor(unique(Input)), keyby=.(Key)]

comt_facs_limma_q1_input_design_mat<- 
  as.matrix(comt_facs_limma_q1_input_design_dt[,-c("Key")])

rownames(comt_facs_limma_q1_input_design_mat)<- comt_facs_limma_q1_input_design_dt[,Key]

comt_facs_limma_q1_input_design<- 
  model.matrix(~V1, data=as.data.frame(comt_facs_limma_q1_input_design_mat), 
               list(V1=contr.treatment(n=2, base=2)))

comt_facs_limma_q1_input_cast_mat_eset_fit<- lmFit(
  comt_facs_limma_q1_input_cast_mat_eset, comt_facs_limma_q1_input_design)

comt_facs_limma_q1_input_cast_mat_eset_fit_eb<- eBayes(
  comt_facs_limma_q1_input_cast_mat_eset_fit, trend=TRUE, proportion=0.03)

comt_facs_limma_q1_input_cast_mat_eset_fit_eb_all<- topTable(
  comt_facs_limma_q1_input_cast_mat_eset_fit_eb, coef=ncol(
    comt_facs_limma_q1_input_design), number=nrow(
      comt_facs_limma_q1_input_cast_mat_eset_fit_eb))

comt_facs_limma_q1_input_cast_mat_eset_fit_eb_all_dt<- as.data.table(
  comt_facs_limma_q1_input_cast_mat_eset_fit_eb_all)

comt_facs_limma_q1_input_cast_mat_eset_fit_eb_all_dt[
  ,VAR_ID:=rownames(comt_facs_limma_q1_input_cast_mat_eset_fit_eb_all)]

rm(comt_facs_limma_q1_input_cast_mat_eset_fit_eb_all)

comt_facs_limma_q1_input_cast_mat_eset_fit_eb_all_dt[logFC>0, Direction:="Quadrant up"]
comt_facs_limma_q1_input_cast_mat_eset_fit_eb_all_dt[logFC<0, Direction:="Quadrant down"]
comt_facs_limma_q1_input_cast_mat_eset_fit_eb_all_dt[,Direction:=as.factor(Direction)]

# Look at p-value distr
facs_q1_distr_gg<- ggplot(
  data=comt_facs_limma_q1_input_cast_mat_eset_fit_eb_all_dt, 
  aes(x=P.Value))

facs_q1_distr_gg+
  geom_histogram()+
  theme_cowplot()

rm(facs_q1_distr_gg)

# Volcano plot
facs_q1_volcano_gg<- ggplot(
  data=comt_facs_limma_q1_input_cast_mat_eset_fit_eb_all_dt, 
  aes(x=logFC, y=-1*log10(adj.P.Val)))

facs_q1_volcano_gg+
  geom_point()+
  theme_cowplot()

rm(facs_q1_volcano_gg)

# MA plot
comt_facs_limma_q1_input_cast_mat_eset_fit_eb_all_dt[adj.P.Val < facs_fdr_thresh, Is_significant:="FDR < 0.01"]

ma_gg<- ggplot(
  data=comt_facs_limma_q1_input_cast_mat_eset_fit_eb_all_dt[is.na(Is_significant)], 
  aes(x=AveExpr, y=logFC))

ma_gg+
  geom_point(alpha=0.8)+
  geom_smooth(method="lm")+
  geom_point(data=comt_facs_limma_q1_input_cast_mat_eset_fit_eb_all_dt[
    !is.na(Is_significant)], 
  aes(x=AveExpr, y=logFC), col="gold")+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  labs(x="Average expression", y="log FC, FACS P1 to cDNA")+
  theme_cowplot()

rm(ma_gg)

comt_facs_limma_q1_input_cast_mat_eset_fit_eb_top_dt<- 
  comt_facs_limma_q1_input_cast_mat_eset_fit_eb_all_dt[adj.P.Val < facs_fdr_thresh]

comt_facs_limma_q1_input_cast_mat_eset_fit_eb_top_varids<- 
  comt_facs_limma_q1_input_cast_mat_eset_fit_eb_top_dt[,VAR_ID]

comt_facs_limma_q1_input_cast_mat_eset_fit_eb_top_up_varids<- 
  comt_facs_limma_q1_input_cast_mat_eset_fit_eb_top_dt[logFC>0,VAR_ID]

comt_facs_limma_q1_input_cast_mat_eset_fit_eb_top_down_varids<- 
  comt_facs_limma_q1_input_cast_mat_eset_fit_eb_top_dt[logFC<0,VAR_ID]

comt_facs_limma_q1_dt[,ID_SHORT:=paste(AA_CHANGE, paste(REF, ALT, sep=">"), sep=", ")]

comt_facs_limma_q1_dt[,ID_LONG:=paste(AA_CHANGE, paste(POS, REF, ALT, sep=":"), sep=", ")]

if("Direction"%in%names(comt_facs_limma_q1_dt)){
  comt_facs_limma_q1_dt[,Direction:=NULL]
}

comt_facs_limma_q1_dt[
  VAR_ID%in%comt_facs_limma_q1_input_cast_mat_eset_fit_eb_top_up_varids,
  Direction:="Quadrant up"]

comt_facs_limma_q1_dt[
  VAR_ID%in%comt_facs_limma_q1_input_cast_mat_eset_fit_eb_top_down_varids,
  Direction:="Quadrant down"]

comt_facs_limma_q1_dt[,Direction:=as.factor(Direction)]

gDNA_med_log_caf<- comt_facs_limma_q1_dt[Input=="gDNA", median(Median_log_CAF)]

facs_med_log_caf<- comt_facs_limma_q1_dt[Quadrant=="1", median(Median_log_CAF)]

aaas_cols<- pal_aaas()(2)

comt_facs_limma_q1_sig_dt<- comt_facs_limma_q1_dt[
    VAR_ID%in%comt_facs_limma_q1_input_cast_mat_eset_fit_eb_top_varids]

comt_facs_limma_q1_sig_dt[,Input:=factor(Input, levels=c("FACS", "gDNA"))]

# Visualize the frequencies
# reorder(ID_LONG, Median_log_CAF)

comt_facs_limma_q1_sig_gg<- ggplot(
  data=comt_facs_limma_q1_sig_dt,
  aes(x=Median_log_CAF, y=reorder(ID_SHORT, POS), col=Input))

comt_facs_limma_q1_sig_gg+
  geom_vline(xintercept=facs_med_log_caf, col=aaas_cols[1], linetype=2)+
  geom_vline(xintercept=gDNA_med_log_caf, col=aaas_cols[2], linetype=2)+
  facet_wrap(~Direction, scales="free_y")+
  geom_point(position=position_jitterdodge())+
  labs(y="Variant ID", x="log frequency")+
  scale_x_continuous(breaks=seq(-10,-5))+
  coord_cartesian(xlim=c(-10,-5))+
  scale_color_aaas()+
  theme_cowplot()+
  theme(axis.text.y=element_text(size=7))

rm(comt_facs_limma_q1_sig_gg)

```

```{r COMT FACS Q1 ALDEx2}

comt_facs_q1_fdr_thresh<- 0.01

# First transform normalized frequencies to counts per million
comt_facs_limma_q1_input_cast_mat_cpm<- apply(comt_facs_limma_q1_input_cast_mat, 2, function(x){round(exp(x)*1e6)})

set.seed(9)

comt_facs_q1_limma_input_cast_mat_cpm_aldex_dt<- run_aldex(comt_facs_limma_q1_input_cast_mat_cpm, c(rep("A", 4), rep("B", 4)))

# Look at the raw p-values to ensure testing procedure is correct
comt_facs_q1_aldex2_pval_gg<- ggplot(
  data=comt_facs_q1_limma_input_cast_mat_cpm_aldex_dt, 
  aes(x=we.ep))

comt_facs_q1_aldex2_pval_gg+
  geom_histogram()+
  theme_cowplot()

rm(comt_facs_q1_aldex2_pval_gg)

# Volcano plot of ALDEx2 results
comt_facs_q1_aldex2_gg<- ggplot(
  data=comt_facs_q1_limma_input_cast_mat_cpm_aldex_dt,
  aes(x=effect, y=-1*log10(we.eBH)))

comt_facs_q1_aldex2_gg+
  geom_point()+
  geom_point(data=comt_facs_q1_limma_input_cast_mat_cpm_aldex_dt[we.eBH<comt_facs_q1_fdr_thresh], 
             aes(x=effect, y=-1*log10(we.eBH)), col="firebrick")+
  theme_cowplot()

rm(comt_facs_q1_aldex2_gg)

comt_facs_q1_limma_input_cast_mat_cpm_aldex_dt[,VAR_ID:=rownames(comt_facs_limma_q1_input_cast_mat_cpm)]

comt_facs_q1_limma_input_cast_mat_cpm_aldex_merge_dt<- merge(
  comt_facs_q1_limma_input_cast_mat_cpm_aldex_dt, annot_dt, by="VAR_ID")

comt_facs_q1_aldex2_sig_gg<- ggplot(
  data=comt_facs_q1_limma_input_cast_mat_cpm_aldex_merge_dt[
    we.eBH<comt_facs_q1_fdr_thresh & effect<0],
  aes(x=effect, y=reorder(ID_SHORT, POS), fill=we.eBH, xmin=effect.low, xmax=effect.high))

comt_facs_q1_aldex2_sig_gg+
  geom_bar(stat="identity", alpha=0.75)+
  geom_errorbar(col="grey", width=0.3)+
  scale_fill_viridis(option="D", begin=0.95, end=0.05)+
  theme_cowplot()+
  labs(y="", x="Effect size", fill="")+
  theme(axis.text.y=element_text(size=8))

rm(comt_facs_q1_aldex2_sig_gg)

# Generate a VE map using ALDEx2
comt_facs_q1_rep_summ_ve_dt<- comt_facs_q1_limma_input_cast_mat_cpm_aldex_merge_dt[
  !grepl(",", AA_CHANGE), .(Median_effect=median(effect, na.rm=T),
                            Max_effect=max(effect, na.rm=T)), 
  by=.(ROI, AA_POS, REF_AA, ALT_AA, AA_CHANGE, VAR_TYPE)]

comt_facs_q1_rep_summ_ve_dt[,AA_POS:=as.integer(AA_POS)]
comt_facs_q1_rep_summ_ve_dt[,ALT_AA:=factor(ALT_AA, levels=unlist(aa_class_list))]
comt_facs_q1_rep_summ_ve_dt[,ALT_AA_level:=as.integer(ALT_AA)]

comt_facs_q1_aldex_ve_gg<- ggplot(
  data=comt_facs_q1_rep_summ_ve_dt, 
  aes(x=AA_POS, y=reorder(ALT_AA, ALT_AA_level), fill=Median_effect))

comt_facs_q1_aldex_ve_gg+
  facet_wrap(~ROI, scales="free")+
  geom_tile()+
  scale_fill_gradient2(high=scales::muted("blue"), low=scales::muted("red"))+
  labs(x="Amino acid position", y="Alternate", 
       fill="Median effect")+
  theme_cowplot()

# Max effect
comt_facs_q1_aldex_ve_gg<- ggplot(
  data=comt_facs_q1_rep_summ_ve_dt, 
  aes(x=AA_POS, y=reorder(ALT_AA, ALT_AA_level), fill=Max_effect))

comt_facs_q1_aldex_ve_gg+
  facet_wrap(~ROI, scales="free")+
  geom_tile()+
  scale_fill_gradient2(high=scales::muted("blue"), low=scales::muted("red"))+
  labs(x="Amino acid position", y="Alternate", 
       fill="Max effect")+
  theme_cowplot()

rm(comt_facs_q1_aldex_ve_gg)

```

```{r COMT FACS Q2 limma}

facs_fdr_thresh<- 0.1
replicate_count_thresh<- 8
log_caf_thresh<- -10

# As a final method to identify variants with differential protein
# use limma to compare gDNA frequencies to frequencies in each quadrant

# Discard the FACS input samples and use gDNA in place
comt_facs_limma_dt<- comt_all_filt_postprocess_subtract_drop_cast_combat_dt[
  Input%in%c("gDNA", "FACS") & !grepl("Input", Key_bio)]

comt_facs_limma_dt[,Quadrant:=sapply(strsplit(as.character(Key_bio), "_"), "[", 3)]

comt_facs_limma_q2_dt<- comt_facs_limma_dt[is.na(Quadrant) | Quadrant==2]

# Use a filter on frequency; corresponds to ~10 counts
comt_facs_limma_q2_counts_dt<- comt_facs_limma_q2_dt[Median_log_CAF>log_caf_thresh, .N, by=.(VAR_ID)]

comt_facs_limma_complete_varids<- comt_facs_limma_q2_counts_dt[
  N>=replicate_count_thresh, unique(VAR_ID)]

comt_facs_limma_q2_input_dt<- comt_facs_limma_q2_dt[VAR_ID%in%comt_facs_limma_complete_varids]

comt_facs_limma_q2_input_cast_dt<- dcast(
  data=comt_facs_limma_q2_input_dt, formula=VAR_ID~Key, value.var="Median_log_CAF")

comt_facs_limma_q2_input_cast_mat<- as.matrix(
  comt_facs_limma_q2_input_cast_dt[,-c("VAR_ID")])

rownames(comt_facs_limma_q2_input_cast_mat)<- comt_facs_limma_q2_input_cast_dt[,VAR_ID]

#comt_facs_limma_q2_input_cast_mat<- normalizeQuantiles(comt_facs_limma_q2_input_cast_mat)

comt_facs_limma_q2_input_cast_mat_eset<- Biobase::ExpressionSet(
  assayData=comt_facs_limma_q2_input_cast_mat)

comt_facs_limma_q2_input_design_dt<- 
  comt_facs_limma_q2_input_dt[,as.factor(unique(Input)), keyby=.(Key)]

comt_facs_limma_q2_input_design_mat<- 
  as.matrix(comt_facs_limma_q2_input_design_dt[,-c("Key")])

rownames(comt_facs_limma_q2_input_design_mat)<- comt_facs_limma_q2_input_design_dt[,Key]

comt_facs_limma_q2_input_design<- 
  model.matrix(~V1, data=as.data.frame(comt_facs_limma_q2_input_design_mat), 
               list(V1=contr.treatment(n=2, base=2)))

comt_facs_limma_q2_input_cast_mat_eset_fit<- lmFit(
  comt_facs_limma_q2_input_cast_mat_eset, comt_facs_limma_q2_input_design)

comt_facs_limma_q2_input_cast_mat_eset_fit_eb<- eBayes(
  comt_facs_limma_q2_input_cast_mat_eset_fit, trend=TRUE, proportion=0.03)

comt_facs_limma_q2_input_cast_mat_eset_fit_eb_all<- topTable(
  comt_facs_limma_q2_input_cast_mat_eset_fit_eb, coef=ncol(
    comt_facs_limma_q2_input_design), number=nrow(
      comt_facs_limma_q2_input_cast_mat_eset_fit_eb))

comt_facs_limma_q2_input_cast_mat_eset_fit_eb_all_dt<- as.data.table(
  comt_facs_limma_q2_input_cast_mat_eset_fit_eb_all)

comt_facs_limma_q2_input_cast_mat_eset_fit_eb_all_dt[
  ,VAR_ID:=rownames(comt_facs_limma_q2_input_cast_mat_eset_fit_eb_all)]

rm(comt_facs_limma_q2_input_cast_mat_eset_fit_eb_all)

comt_facs_limma_q2_input_cast_mat_eset_fit_eb_all_dt[logFC>0, Direction:="Quadrant up"]
comt_facs_limma_q2_input_cast_mat_eset_fit_eb_all_dt[logFC<0, Direction:="Quadrant down"]
comt_facs_limma_q2_input_cast_mat_eset_fit_eb_all_dt[,Direction:=as.factor(Direction)]

# Look at p-value distr
facs_q2_distr_gg<- ggplot(
  data=comt_facs_limma_q2_input_cast_mat_eset_fit_eb_all_dt, 
  aes(x=P.Value))

facs_q2_distr_gg+
  geom_histogram()+
  theme_cowplot()

rm(facs_q2_distr_gg)

# Volcano plot
facs_q2_volcano_gg<- ggplot(
  data=comt_facs_limma_q2_input_cast_mat_eset_fit_eb_all_dt, 
  aes(x=logFC, y=-1*log10(adj.P.Val)))

facs_q2_volcano_gg+
  geom_point()+
  theme_cowplot()

rm(facs_q2_volcano_gg)

# MA plot
comt_facs_limma_q2_input_cast_mat_eset_fit_eb_all_dt[adj.P.Val < facs_fdr_thresh, Is_significant:="FDR < 0.01"]

ma_gg<- ggplot(
  data=comt_facs_limma_q2_input_cast_mat_eset_fit_eb_all_dt[is.na(Is_significant)], 
  aes(x=AveExpr, y=logFC))

ma_gg+
  geom_point(alpha=0.8)+
  geom_smooth(method="lm")+
  geom_point(data=comt_facs_limma_q2_input_cast_mat_eset_fit_eb_all_dt[
    !is.na(Is_significant)], 
  aes(x=AveExpr, y=logFC), col="gold")+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  labs(x="Average expression", y="log FC, FACS P2 to cDNA")+
  theme_cowplot()

rm(ma_gg)

comt_facs_limma_q2_input_cast_mat_eset_fit_eb_top_dt<- 
  comt_facs_limma_q2_input_cast_mat_eset_fit_eb_all_dt[adj.P.Val < facs_fdr_thresh]

comt_facs_limma_q2_input_cast_mat_eset_fit_eb_top_varids<- 
  comt_facs_limma_q2_input_cast_mat_eset_fit_eb_top_dt[,VAR_ID]

comt_facs_limma_q2_input_cast_mat_eset_fit_eb_top_up_varids<- 
  comt_facs_limma_q2_input_cast_mat_eset_fit_eb_top_dt[logFC>0,VAR_ID]

comt_facs_limma_q2_input_cast_mat_eset_fit_eb_top_down_varids<- 
  comt_facs_limma_q2_input_cast_mat_eset_fit_eb_top_dt[logFC<0,VAR_ID]

comt_facs_limma_q2_dt[,ID_SHORT:=paste(AA_CHANGE, paste(REF, ALT, sep=">"), sep=", ")]

comt_facs_limma_q2_dt[,ID_LONG:=paste(AA_CHANGE, paste(POS, REF, ALT, sep=":"), sep=", ")]

if("Direction"%in%names(comt_facs_limma_q2_dt)){
  comt_facs_limma_q2_dt[,Direction:=NULL]
}

comt_facs_limma_q2_dt[
  VAR_ID%in%comt_facs_limma_q2_input_cast_mat_eset_fit_eb_top_up_varids,
  Direction:="Quadrant up"]

comt_facs_limma_q2_dt[
  VAR_ID%in%comt_facs_limma_q2_input_cast_mat_eset_fit_eb_top_down_varids,
  Direction:="Quadrant down"]

comt_facs_limma_q2_dt[,Direction:=as.factor(Direction)]

gDNA_med_log_caf<- comt_facs_limma_q2_dt[Input=="gDNA", median(Median_log_CAF)]

facs_med_log_caf<- comt_facs_limma_q2_dt[Quadrant=="2", median(Median_log_CAF)]

aaas_cols<- pal_aaas()(2)

comt_facs_limma_q2_sig_dt<- comt_facs_limma_q2_dt[
    VAR_ID%in%comt_facs_limma_q2_input_cast_mat_eset_fit_eb_top_varids]

comt_facs_limma_q2_sig_dt[,Input:=factor(Input, levels=c("FACS", "gDNA"))]

# Visualize the frequencies
# reorder(ID_LONG, Median_log_CAF)

comt_facs_limma_q2_sig_gg<- ggplot(
  data=comt_facs_limma_q2_sig_dt,
  aes(x=Median_log_CAF, y=reorder(ID_SHORT, POS), col=Input))

comt_facs_limma_q2_sig_gg+
  geom_vline(xintercept=facs_med_log_caf, col=aaas_cols[1], linetype=2)+
  geom_vline(xintercept=gDNA_med_log_caf, col=aaas_cols[2], linetype=2)+
  facet_wrap(~Direction, scales="free_y")+
  geom_point(position=position_jitterdodge())+
  labs(y="Variant ID", x="log frequency")+
  scale_x_continuous(breaks=seq(-10,-5))+
  coord_cartesian(xlim=c(-10,-5))+
  scale_color_aaas()+
  theme_cowplot()+
  theme(axis.text.y=element_text(size=7))

rm(comt_facs_limma_q2_sig_gg)


```

```{r COMT FACS Q2 ALDEx2}

comt_facs_q2_fdr_thresh<- 0.2

# First transform normalized frequencies to counts per million
comt_facs_limma_q2_input_cast_mat_cpm<- apply(comt_facs_limma_q2_input_cast_mat, 2, function(x){round(exp(x)*1e6)})

set.seed(9)

comt_facs_q2_limma_input_cast_mat_cpm_aldex_dt<- run_aldex(comt_facs_limma_q2_input_cast_mat_cpm, c(rep("A", 4), rep("B", 4)))

# Look at the raw p-values to ensure testing procedure is correct
comt_facs_q2_aldex2_pval_gg<- ggplot(
  data=comt_facs_q2_limma_input_cast_mat_cpm_aldex_dt, 
  aes(x=we.ep))

comt_facs_q2_aldex2_pval_gg+
  geom_histogram()+
  theme_cowplot()

rm(comt_facs_q2_aldex2_pval_gg)

# Volcano plot of ALDEx2 results
comt_facs_q2_aldex2_gg<- ggplot(
  data=comt_facs_q2_limma_input_cast_mat_cpm_aldex_dt,
  aes(x=effect, y=-1*log10(we.eBH)))

comt_facs_q2_aldex2_gg+
  geom_point()+
  geom_point(data=comt_facs_q2_limma_input_cast_mat_cpm_aldex_dt[we.eBH<comt_facs_q2_fdr_thresh], 
             aes(x=effect, y=-1*log10(we.eBH)), col="firebrick")+
  theme_cowplot()

rm(comt_facs_q2_aldex2_gg)

comt_facs_q2_limma_input_cast_mat_cpm_aldex_dt[,VAR_ID:=rownames(comt_facs_limma_q2_input_cast_mat_cpm)]

comt_facs_q2_limma_input_cast_mat_cpm_aldex_merge_dt<- merge(
  comt_facs_q2_limma_input_cast_mat_cpm_aldex_dt, annot_dt, by="VAR_ID")

comt_facs_q2_aldex2_sig_gg<- ggplot(
  data=comt_facs_q2_limma_input_cast_mat_cpm_aldex_merge_dt[
    we.eBH<comt_facs_q2_fdr_thresh],
  aes(x=effect, y=reorder(ID_SHORT, POS), fill=we.eBH, xmin=effect.low, xmax=effect.high))

comt_facs_q2_aldex2_sig_gg+
  geom_bar(stat="identity", alpha=0.75)+
  geom_errorbar(col="grey", width=0.3)+
  scale_fill_viridis(option="D", begin=0.95, end=0.05)+
  theme_cowplot()+
  labs(y="", x="Effect size", fill="")+
  theme(axis.text.y=element_text(size=8))

rm(comt_facs_q2_aldex2_sig_gg)

# Generate a VE map using ALDEx2
comt_facs_q2_rep_summ_ve_dt<- comt_facs_q2_limma_input_cast_mat_cpm_aldex_merge_dt[
  !grepl(",", AA_CHANGE), .(Median_effect=median(effect, na.rm=T),
                            Max_effect=max(effect, na.rm=T),
                            Median_log2FC=median(diff.btw, na.rm=T),
                            Max_log2FC=max(diff.btw, na.rm=T)), 
  by=.(ROI, AA_POS, REF_AA, ALT_AA, AA_CHANGE, VAR_TYPE)]

comt_facs_q2_rep_summ_ve_dt[,AA_POS:=as.integer(AA_POS)]
comt_facs_q2_rep_summ_ve_dt[,ALT_AA:=factor(ALT_AA, levels=unlist(aa_class_list))]
comt_facs_q2_rep_summ_ve_dt[,ALT_AA_level:=as.integer(ALT_AA)]

comt_facs_q2_aldex_ve_gg<- ggplot(
  data=comt_facs_q2_rep_summ_ve_dt, 
  aes(x=AA_POS, y=reorder(ALT_AA, ALT_AA_level), fill=Median_effect))

comt_facs_q2_aldex_ve_gg+
  facet_wrap(~ROI, scales="free")+
  geom_tile()+
  scale_fill_gradient2(low=scales::muted("blue"), high=scales::muted("red"))+
  labs(x="Amino acid position", y="Alternate", 
       fill="Median effect")+
  theme_cowplot()

# Max effect
comt_facs_q2_aldex_ve_gg<- ggplot(
  data=comt_facs_q2_rep_summ_ve_dt, 
  aes(x=AA_POS, y=reorder(ALT_AA, ALT_AA_level), fill=Max_effect))

comt_facs_q2_aldex_ve_gg+
  facet_wrap(~ROI, scales="free")+
  geom_tile()+
  scale_fill_gradient2(low=scales::muted("blue"), high=scales::muted("red"))+
  labs(x="Amino acid position", y="Alternate", 
       fill="Max effect")+
  theme_cowplot()

```

```{r COMT FACS Q3 limma}

facs_fdr_thresh<- 0.004
replicate_count_thresh<- 8
log_caf_thresh<- -10

# As a final method to identify variants with differential protein
# use limma to compare gDNA frequencies to frequencies in each quadrant

# Discard the FACS input samples and use gDNA in place
comt_facs_limma_dt<- comt_all_filt_postprocess_subtract_drop_cast_combat_dt[
  Input%in%c("gDNA", "FACS") & !grepl("Input", Key_bio)]

comt_facs_limma_dt[,Quadrant:=sapply(strsplit(as.character(Key_bio), "_"), "[", 3)]

# Combine quadrants 2, 3 (low protein) to test against gDNA?
comt_facs_limma_q3_dt<- comt_facs_limma_dt[is.na(Quadrant) | Quadrant==3]

# Use a filter on frequency; corresponds to ~10 counts
comt_facs_limma_q3_counts_dt<- comt_facs_limma_q3_dt[Median_log_CAF>log_caf_thresh, .N, by=.(VAR_ID)]

comt_facs_limma_complete_varids<- comt_facs_limma_q3_counts_dt[
  N>=replicate_count_thresh, unique(VAR_ID)]

comt_facs_limma_q3_input_dt<- comt_facs_limma_q3_dt[VAR_ID%in%comt_facs_limma_complete_varids]

comt_facs_limma_q3_input_cast_dt<- dcast(
  data=comt_facs_limma_q3_input_dt, formula=VAR_ID~Key, value.var="Median_log_CAF")

comt_facs_limma_q3_input_cast_mat<- as.matrix(
  comt_facs_limma_q3_input_cast_dt[,-c("VAR_ID")])

rownames(comt_facs_limma_q3_input_cast_mat)<- comt_facs_limma_q3_input_cast_dt[,VAR_ID]

#comt_facs_limma_q3_input_cast_mat<- normalizeQuantiles(comt_facs_limma_q3_input_cast_mat)

comt_facs_limma_q3_input_cast_mat_eset<- Biobase::ExpressionSet(
  assayData=comt_facs_limma_q3_input_cast_mat)

comt_facs_limma_q3_input_design_dt<- 
  comt_facs_limma_q3_input_dt[,as.factor(unique(Input)), keyby=.(Key)]

comt_facs_limma_q3_input_design_mat<- 
  as.matrix(comt_facs_limma_q3_input_design_dt[,-c("Key")])

rownames(comt_facs_limma_q3_input_design_mat)<- comt_facs_limma_q3_input_design_dt[,Key]

comt_facs_limma_q3_input_design<- 
  model.matrix(~V1, data=as.data.frame(comt_facs_limma_q3_input_design_mat), 
               list(V1=contr.treatment(n=2, base=2)))

comt_facs_limma_q3_input_cast_mat_eset_fit<- lmFit(
  comt_facs_limma_q3_input_cast_mat_eset, comt_facs_limma_q3_input_design)

comt_facs_limma_q3_input_cast_mat_eset_fit_eb<- eBayes(
  comt_facs_limma_q3_input_cast_mat_eset_fit, trend=TRUE, proportion=0.03)

comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all<- topTable(
  comt_facs_limma_q3_input_cast_mat_eset_fit_eb, coef=ncol(
    comt_facs_limma_q3_input_design), number=nrow(
      comt_facs_limma_q3_input_cast_mat_eset_fit_eb))

comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_dt<- as.data.table(
  comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all)

comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_dt[
  ,VAR_ID:=rownames(comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all)]

rm(comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all)

comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_dt[logFC>0, Direction:="Quadrant up"]
comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_dt[logFC<0, Direction:="Quadrant down"]
comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_dt[,Direction:=as.factor(Direction)]

# Look at p-value distr
facs_q3_distr_gg<- ggplot(
  data=comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_dt, 
  aes(x=P.Value))

facs_q3_distr_gg+
  geom_histogram()+
  theme_cowplot()

rm(facs_q3_distr_gg)

# Volcano plot
facs_q3_volcano_gg<- ggplot(
  data=comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_dt, 
  aes(x=logFC, y=-1*log10(adj.P.Val)))

facs_q3_volcano_gg+
  geom_point()+
  theme_cowplot()

rm(facs_q3_volcano_gg)

# MA plot
comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_dt[adj.P.Val < facs_fdr_thresh, Is_significant:="FDR < 0.01"]

ma_gg<- ggplot(
  data=comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_dt[is.na(Is_significant)], 
  aes(x=AveExpr, y=logFC))

ma_gg+
  geom_point(alpha=0.8)+
  geom_smooth(method="loess", span=1)+
  geom_point(data=comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_dt[
    !is.na(Is_significant)], 
  aes(x=AveExpr, y=logFC), col="gold")+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  labs(x="Average expression", y="log FC, FACS P3 to cDNA")+
  theme_cowplot()

rm(ma_gg)

comt_facs_limma_q3_input_cast_mat_eset_fit_eb_top_dt<- 
  comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_dt[adj.P.Val < facs_fdr_thresh]

comt_facs_limma_q3_input_cast_mat_eset_fit_eb_top_varids<- 
  comt_facs_limma_q3_input_cast_mat_eset_fit_eb_top_dt[,VAR_ID]

comt_facs_limma_q3_input_cast_mat_eset_fit_eb_top_up_varids<- 
  comt_facs_limma_q3_input_cast_mat_eset_fit_eb_top_dt[logFC>0,VAR_ID]

comt_facs_limma_q3_input_cast_mat_eset_fit_eb_top_down_varids<- 
  comt_facs_limma_q3_input_cast_mat_eset_fit_eb_top_dt[logFC<0,VAR_ID]

comt_facs_limma_q3_dt[,ID_SHORT:=paste(AA_CHANGE, paste(REF, ALT, sep=">"), sep=", ")]

comt_facs_limma_q3_dt[,ID_LONG:=paste(AA_CHANGE, paste(POS, REF, ALT, sep=":"), sep=", ")]

if("Direction"%in%names(comt_facs_limma_q3_dt)){
  comt_facs_limma_q3_dt[,Direction:=NULL]
}

comt_facs_limma_q3_dt[
  VAR_ID%in%comt_facs_limma_q3_input_cast_mat_eset_fit_eb_top_up_varids,
  Direction:="Quadrant up"]

comt_facs_limma_q3_dt[
  VAR_ID%in%comt_facs_limma_q3_input_cast_mat_eset_fit_eb_top_down_varids,
  Direction:="Quadrant down"]

comt_facs_limma_q3_dt[,Direction:=as.factor(Direction)]

gDNA_med_log_caf<- comt_facs_limma_q3_dt[Input=="gDNA", median(Median_log_CAF)]

facs_med_log_caf<- comt_facs_limma_q3_dt[Quadrant=="3", median(Median_log_CAF)]

aaas_cols<- pal_aaas()(2)

comt_facs_limma_q3_sig_dt<- comt_facs_limma_q3_dt[
    VAR_ID%in%comt_facs_limma_q3_input_cast_mat_eset_fit_eb_top_varids]

comt_facs_limma_q3_sig_dt[,Input:=factor(Input, levels=c("FACS", "gDNA"))]

# Visualize the frequencies
# reorder(ID_LONG, Median_log_CAF)

# All variants
comt_facs_limma_q3_sig_gg<- ggplot(
  data=comt_facs_limma_q3_sig_dt,
  aes(x=Median_log_CAF, y=reorder(ID_SHORT, POS), col=Input))

comt_facs_limma_q3_sig_gg+
  geom_vline(xintercept=facs_med_log_caf, col=aaas_cols[1], linetype=2)+
  geom_vline(xintercept=gDNA_med_log_caf, col=aaas_cols[2], linetype=2)+
  facet_wrap(~Direction, scales="free_y")+
  geom_point(position=position_jitterdodge())+
  labs(y="Variant ID", x="log frequency")+
  scale_x_continuous(breaks=seq(-10,-5))+
  coord_cartesian(xlim=c(-9.5,-4.5))+
  scale_color_aaas()+
  theme_cowplot()+
  theme(axis.text.y=element_text(size=7))

# Only variant enriched in Q3
comt_facs_limma_q3_sig_gg<- ggplot(
  data=comt_facs_limma_q3_sig_dt[Direction=="Quadrant up"],
  aes(x=Median_log_CAF, y=reorder(ID_SHORT, POS), col=Input))

comt_facs_limma_q3_sig_gg+
  geom_vline(xintercept=facs_med_log_caf, col=aaas_cols[1], linetype=2)+
  geom_vline(xintercept=gDNA_med_log_caf, col=aaas_cols[2], linetype=2)+
  geom_point(position=position_jitterdodge())+
  labs(y="Variant ID", x="log frequency")+
  scale_x_continuous(breaks=seq(-10,-5))+
  scale_color_aaas()+
  theme_cowplot()

# Aggregate at level of AA change
comt_facs_limma_q3_sig_gg<- ggplot(
  data=comt_facs_limma_q3_sig_dt[Direction=="Quadrant up"],
  aes(x=Median_log_CAF, y=reorder(AA_CHANGE, POS), col=Input))

comt_facs_limma_q3_sig_gg+
  geom_vline(xintercept=facs_med_log_caf, col=aaas_cols[1], linetype=2)+
  geom_vline(xintercept=gDNA_med_log_caf, col=aaas_cols[2], linetype=2)+
  geom_point(position=position_jitterdodge())+
  labs(y="Variant ID", x="log frequency")+
  scale_x_continuous(breaks=seq(-10,-5))+
  scale_color_aaas()+
  theme_cowplot()

rm(comt_facs_limma_q3_sig_gg)

# Merge with DeMask data and determine if there is a correlation with sites of conservation
comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_annot_dt<- merge(
  comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_dt, annot_dt, by="VAR_ID")

comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_annot_dt[,AA_POS:=as.integer(AA_POS)]
comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_annot_dt[,ALT_AA:=factor(ALT_AA, levels=unlist(aa_class_list))]
comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_annot_dt[,ALT_AA_level:=as.integer(ALT_AA)]

comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_annot_loose_dt<- comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_annot_dt[adj.P.Val < 0.01]

comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_annot_loose_demask_dt<- 
  merge(comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_annot_loose_dt, demask_dt, by="AA_CHANGE")

# Plot DeMask score against direction of effect
comt_facs_limma_q3_sig_demask_gg<- ggplot(
  data=comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_annot_loose_demask_dt,
  aes(col=ROI, x=score, linetype=Direction))

comt_facs_limma_q3_sig_demask_gg+
  geom_density()+
  scale_color_jama()+
  labs(x="DeMask score")+
  theme_cowplot()

rm(comt_facs_limma_q3_sig_demask_gg)

# Results track with the DeMask score; test with Wilcoxon rank sum 

comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_annot_loose_demask_dt[,t.test(score~Direction)]

# What about entropy?
comt_facs_limma_q3_sig_demask_gg<- ggplot(
  data=comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_annot_loose_demask_dt,
  aes(x=Direction, y=entropy))

comt_facs_limma_q3_sig_demask_gg+
  geom_boxplot(fill="steelblue", alpha=0.75, width=0.5)+
  labs(x="", y="DeMask entropy")+
  theme_cowplot()

rm(comt_facs_limma_q3_sig_demask_gg)

# Less separation when looking at entropy

# Plot results as a heatmap
comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_annot_agg_dt<- 
  comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_annot_dt[
  !grepl(",", AA_CHANGE), .(Median_log_FC=median(logFC, na.rm=T)), 
  by=.(ROI, AA_POS, REF_AA, ALT_AA, AA_CHANGE, ALT_AA_level, VAR_TYPE)]

comt_facs_q3_ve_heatmap_gg<- ggplot(
  data=comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_annot_agg_dt, 
  aes(x=AA_POS, y=reorder(ALT_AA, ALT_AA_level), fill=Median_log_FC))

comt_facs_q3_ve_heatmap_gg+
  facet_wrap(~ROI, scales="free")+
  geom_tile()+
  scale_fill_gradient2(high=scales::muted("red"), low=scales::muted("blue"))+
  labs(x="Amino acid position", y="Alternate", 
       fill="Median log FC,
Q3 to gDNA")+
  theme_cowplot()

rm(comt_facs_q3_ve_heatmap_gg)

# comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_annot_agg_dt[AA_POS%in%c(59, 63, 67, 136, 137, 152, 153, 156, 158), sort(unique(AA_CHANGE))]

```

```{r COMT FACS Q3 ALDEx2}

comt_facs_q3_fdr_thresh<- 0.05
comt_facs_q3_fdr_loose_thresh<- 0.15

# First transform normalized frequencies to counts per million
comt_facs_limma_q3_input_cast_mat_cpm<- apply(comt_facs_limma_q3_input_cast_mat, 2, function(x){round(exp(x)*1e6)})

set.seed(9)

comt_facs_q3_limma_input_cast_mat_cpm_aldex_dt<- run_aldex(comt_facs_limma_q3_input_cast_mat_cpm, c(rep("A", 4), rep("B", 4)))

# Look at the raw p-values to ensure testing procedure is correct
comt_facs_q3_aldex2_pval_gg<- ggplot(
  data=comt_facs_q3_limma_input_cast_mat_cpm_aldex_dt, 
  aes(x=we.ep))

comt_facs_q3_aldex2_pval_gg+
  geom_histogram()+
  theme_cowplot()

rm(comt_facs_q3_aldex2_pval_gg)

# Volcano plot of ALDEx2 results
comt_facs_q3_aldex2_gg<- ggplot(
  data=comt_facs_q3_limma_input_cast_mat_cpm_aldex_dt,
  aes(x=effect, y=-1*log10(we.eBH)))

comt_facs_q3_aldex2_gg+
  geom_point()+
  geom_point(data=comt_facs_q3_limma_input_cast_mat_cpm_aldex_dt[we.eBH<comt_facs_q3_fdr_thresh], 
             aes(x=effect, y=-1*log10(we.eBH)), col="firebrick")+
  theme_cowplot()

rm(comt_facs_q3_aldex2_gg)

comt_facs_q3_limma_input_cast_mat_cpm_aldex_dt[,VAR_ID:=rownames(comt_facs_limma_q3_input_cast_mat_cpm)]

comt_facs_q3_limma_input_cast_mat_cpm_aldex_merge_dt<- merge(
  comt_facs_q3_limma_input_cast_mat_cpm_aldex_dt, annot_dt, by="VAR_ID")

comt_facs_q3_aldex2_sig_gg<- ggplot(
  data=comt_facs_q3_limma_input_cast_mat_cpm_aldex_merge_dt[
    we.eBH<comt_facs_q3_fdr_thresh & effect>0],
  aes(x=effect, y=reorder(ID_SHORT, POS), fill=we.eBH, xmin=effect.low, xmax=effect.high))

comt_facs_q3_aldex2_sig_gg+
  geom_bar(stat="identity", alpha=0.75)+
  geom_errorbar(col="grey", width=0.3)+
  scale_fill_viridis(option="D", begin=0.95, end=0.05)+
  theme_cowplot()+
  labs(y="", x="Effect size", fill="FDR")+
  theme(axis.text.y=element_text(size=8))

rm(comt_facs_q3_aldex2_sig_gg)

# Generate a VE map using ALDEx2
comt_facs_q3_rep_summ_ve_dt<- comt_facs_q3_limma_input_cast_mat_cpm_aldex_merge_dt[
  !grepl(",", AA_CHANGE), .(Median_effect=median(effect, na.rm=T),
                            Max_effect=max(effect, na.rm=T)), 
  by=.(ROI, AA_POS, REF_AA, ALT_AA, AA_CHANGE, VAR_TYPE)]

comt_facs_q3_rep_summ_ve_dt<- fill_nas(comt_facs_q3_rep_summ_ve_dt)

comt_facs_q3_rep_summ_ve_dt[,AA_POS:=as.integer(AA_POS)]
comt_facs_q3_rep_summ_ve_dt[,ALT_AA:=factor(ALT_AA, levels=unlist(aa_class_list))]
comt_facs_q3_rep_summ_ve_dt[,ALT_AA_level:=as.integer(ALT_AA)]

comt_facs_q3_rep_summ_ve_dt[,Is_REF:=ifelse(REF_AA==ALT_AA, TRUE, FALSE)]

comt_facs_q3_aldex_ve_gg<- ggplot(
  data=comt_facs_q3_rep_summ_ve_dt[Is_REF==FALSE], 
  aes(x=AA_POS, y=reorder(ALT_AA, ALT_AA_level), fill=Median_effect))

comt_facs_q3_aldex_ve_gg+
  facet_wrap(~ROI, scales="free")+
  geom_tile()+
  geom_tile(data=comt_facs_q3_rep_summ_ve_dt[Is_REF==TRUE], 
            aes(x=AA_POS, y=reorder(ALT_AA, ALT_AA_level, fill=Median_effect)), col="black", linewidth=0.25)+
  scale_fill_gradient2(low=scales::muted("blue"), high=scales::muted("red"), na.value="grey")+
  labs(x="Amino acid position", y="Alternate", 
       fill="Median effect")+
  theme_cowplot()

rm(comt_facs_q3_aldex_ve_gg)

# Merge with DeMask data and determine if there is a correlation with sites of conservation
comt_facs_q3_limma_input_cast_mat_cpm_aldex_merge_dt[,Direction:="Pop. down"]
comt_facs_q3_limma_input_cast_mat_cpm_aldex_merge_dt[effect>0, Direction:="Pop. up"]
comt_facs_q3_limma_input_cast_mat_cpm_aldex_merge_dt[,Direction:=as.factor(Direction)]

comt_facs_q3_limma_input_cast_mat_cpm_aldex_merge_demask_dt<- 
  merge(comt_facs_q3_limma_input_cast_mat_cpm_aldex_merge_dt, demask_dt, by="AA_CHANGE")

# Plot DeMask score against direction of effect
comt_facs_limma_q3_sig_demask_gg<- ggplot(
  data=comt_facs_q3_limma_input_cast_mat_cpm_aldex_merge_demask_dt[we.eBH<0.15],
  aes(col=ROI, x=score, linetype=Direction))

comt_facs_limma_q3_sig_demask_gg+
  geom_density()+
  scale_color_manual(values=roi_cols)+
  labs(x="DeMask score")+
  theme_cowplot()

rm(comt_facs_limma_q3_sig_demask_gg)

# Results track with the DeMask score; test with Wilcoxon rank sum 
comt_facs_q3_limma_input_cast_mat_cpm_aldex_merge_demask_dt[
  ROI=="ROI_1" & we.eBH<0.2, wilcox.test(score~Direction, data=.SD), .SDcols=c("score", "Direction")]

comt_facs_q3_limma_input_cast_mat_cpm_aldex_merge_demask_dt[
  ROI=="ROI_2" & we.eBH<0.2, wilcox.test(score~Direction, data=.SD), .SDcols=c("score", "Direction")]

```

```{r COMT FACS Q4 limma}

facs_fdr_thresh<- 0.004
replicate_count_thresh<- 8
log_caf_thresh<- -10

# As a final method to identify variants with differential protein
# use limma to compare gDNA frequencies to frequencies in each quadrant

# Discard the FACS input samples and use gDNA in place
comt_facs_limma_dt<- comt_all_filt_postprocess_subtract_drop_cast_combat_dt[
  Input%in%c("gDNA", "FACS") & !grepl("Input", Key_bio)]

comt_facs_limma_dt[,Quadrant:=sapply(strsplit(as.character(Key_bio), "_"), "[", 3)]

# Combine quadrants 2, 3 (low protein) to test against gDNA?
comt_facs_limma_q4_dt<- comt_facs_limma_dt[is.na(Quadrant) | Quadrant==4]

# Use a filter on frequency; corresponds to ~10 counts
comt_facs_limma_q4_counts_dt<- comt_facs_limma_q4_dt[Median_log_CAF>log_caf_thresh, .N, by=.(VAR_ID)]

comt_facs_limma_complete_varids<- comt_facs_limma_q4_counts_dt[
  N>=replicate_count_thresh, unique(VAR_ID)]

comt_facs_limma_q4_input_dt<- comt_facs_limma_q4_dt[VAR_ID%in%comt_facs_limma_complete_varids]

comt_facs_limma_q4_input_cast_dt<- dcast(
  data=comt_facs_limma_q4_input_dt, formula=VAR_ID~Key, value.var="Median_log_CAF")

comt_facs_limma_q4_input_cast_mat<- as.matrix(
  comt_facs_limma_q4_input_cast_dt[,-c("VAR_ID")])

rownames(comt_facs_limma_q4_input_cast_mat)<- comt_facs_limma_q4_input_cast_dt[,VAR_ID]

#comt_facs_limma_q4_input_cast_mat<- normalizeQuantiles(comt_facs_limma_q4_input_cast_mat)

comt_facs_limma_q4_input_cast_mat_eset<- Biobase::ExpressionSet(
  assayData=comt_facs_limma_q4_input_cast_mat)

comt_facs_limma_q4_input_design_dt<- 
  comt_facs_limma_q4_input_dt[,as.factor(unique(Input)), keyby=.(Key)]

comt_facs_limma_q4_input_design_mat<- 
  as.matrix(comt_facs_limma_q4_input_design_dt[,-c("Key")])

rownames(comt_facs_limma_q4_input_design_mat)<- comt_facs_limma_q4_input_design_dt[,Key]

comt_facs_limma_q4_input_design<- 
  model.matrix(~V1, data=as.data.frame(comt_facs_limma_q4_input_design_mat), 
               list(V1=contr.treatment(n=2, base=2)))

comt_facs_limma_q4_input_cast_mat_eset_fit<- lmFit(
  comt_facs_limma_q4_input_cast_mat_eset, comt_facs_limma_q4_input_design)

comt_facs_limma_q4_input_cast_mat_eset_fit_eb<- eBayes(
  comt_facs_limma_q4_input_cast_mat_eset_fit, trend=TRUE, proportion=0.03)

comt_facs_limma_q4_input_cast_mat_eset_fit_eb_all<- topTable(
  comt_facs_limma_q4_input_cast_mat_eset_fit_eb, coef=ncol(
    comt_facs_limma_q4_input_design), number=nrow(
      comt_facs_limma_q4_input_cast_mat_eset_fit_eb))

comt_facs_limma_q4_input_cast_mat_eset_fit_eb_all_dt<- as.data.table(
  comt_facs_limma_q4_input_cast_mat_eset_fit_eb_all)

comt_facs_limma_q4_input_cast_mat_eset_fit_eb_all_dt[
  ,VAR_ID:=rownames(comt_facs_limma_q4_input_cast_mat_eset_fit_eb_all)]

rm(comt_facs_limma_q4_input_cast_mat_eset_fit_eb_all)

comt_facs_limma_q4_input_cast_mat_eset_fit_eb_all_dt[logFC>0, Direction:="Quadrant up"]
comt_facs_limma_q4_input_cast_mat_eset_fit_eb_all_dt[logFC<0, Direction:="Quadrant down"]
comt_facs_limma_q4_input_cast_mat_eset_fit_eb_all_dt[,Direction:=as.factor(Direction)]

# Look at p-value distr
facs_q4_distr_gg<- ggplot(
  data=comt_facs_limma_q4_input_cast_mat_eset_fit_eb_all_dt, 
  aes(x=P.Value))

facs_q4_distr_gg+
  geom_histogram()+
  theme_cowplot()

rm(facs_q4_distr_gg)

# Volcano plot
facs_q4_volcano_gg<- ggplot(
  data=comt_facs_limma_q4_input_cast_mat_eset_fit_eb_all_dt, 
  aes(x=logFC, y=-1*log10(adj.P.Val)))

facs_q4_volcano_gg+
  geom_point()+
  theme_cowplot()

rm(facs_q4_volcano_gg)

# MA plot
comt_facs_limma_q4_input_cast_mat_eset_fit_eb_all_dt[adj.P.Val < facs_fdr_thresh, Is_significant:="FDR < 0.004"]

ma_gg<- ggplot(
  data=comt_facs_limma_q4_input_cast_mat_eset_fit_eb_all_dt[is.na(Is_significant)], 
  aes(x=AveExpr, y=logFC))

ma_gg+
  geom_point(alpha=0.8)+
  geom_smooth(method="lm")+
  geom_point(data=comt_facs_limma_q4_input_cast_mat_eset_fit_eb_all_dt[
    !is.na(Is_significant)], 
  aes(x=AveExpr, y=logFC), col="gold")+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  labs(x="Average expression", y="log FC, FACS P3 to cDNA")+
  theme_cowplot()

rm(ma_gg)

comt_facs_limma_q4_input_cast_mat_eset_fit_eb_top_dt<- 
  comt_facs_limma_q4_input_cast_mat_eset_fit_eb_all_dt[adj.P.Val < facs_fdr_thresh]

comt_facs_limma_q4_input_cast_mat_eset_fit_eb_top_varids<- 
  comt_facs_limma_q4_input_cast_mat_eset_fit_eb_top_dt[,VAR_ID]

comt_facs_limma_q4_input_cast_mat_eset_fit_eb_top_up_varids<- 
  comt_facs_limma_q4_input_cast_mat_eset_fit_eb_top_dt[logFC>0,VAR_ID]

comt_facs_limma_q4_input_cast_mat_eset_fit_eb_top_down_varids<- 
  comt_facs_limma_q4_input_cast_mat_eset_fit_eb_top_dt[logFC<0,VAR_ID]

comt_facs_limma_q4_dt[,ID_SHORT:=paste(AA_CHANGE, paste(REF, ALT, sep=">"), sep=", ")]

comt_facs_limma_q4_dt[,ID_LONG:=paste(AA_CHANGE, paste(POS, REF, ALT, sep=":"), sep=", ")]

if("Direction"%in%names(comt_facs_limma_q4_dt)){
  comt_facs_limma_q4_dt[,Direction:=NULL]
}

comt_facs_limma_q4_dt[
  VAR_ID%in%comt_facs_limma_q4_input_cast_mat_eset_fit_eb_top_up_varids,
  Direction:="Quadrant up"]

comt_facs_limma_q4_dt[
  VAR_ID%in%comt_facs_limma_q4_input_cast_mat_eset_fit_eb_top_down_varids,
  Direction:="Quadrant down"]

comt_facs_limma_q4_dt[,Direction:=as.factor(Direction)]

gDNA_med_log_caf<- comt_facs_limma_q4_dt[Input=="gDNA", median(Median_log_CAF)]

facs_med_log_caf<- comt_facs_limma_q4_dt[Quadrant=="4", median(Median_log_CAF)]

aaas_cols<- pal_aaas()(2)

comt_facs_limma_q4_sig_dt<- comt_facs_limma_q4_dt[
    VAR_ID%in%comt_facs_limma_q4_input_cast_mat_eset_fit_eb_top_varids]

comt_facs_limma_q4_sig_dt[,Input:=factor(Input, levels=c("FACS", "gDNA"))]

# Visualize the frequencies
# reorder(ID_LONG, Median_log_CAF)

comt_facs_limma_q4_sig_gg<- ggplot(
  data=comt_facs_limma_q4_sig_dt,
  aes(x=Median_log_CAF, y=reorder(ID_SHORT, POS), col=Input))

comt_facs_limma_q4_sig_gg+
  geom_vline(xintercept=facs_med_log_caf, col=aaas_cols[1], linetype=2)+
  geom_vline(xintercept=gDNA_med_log_caf, col=aaas_cols[2], linetype=2)+
  facet_wrap(~Direction, scales="free_y")+
  geom_point(position=position_jitterdodge())+
  labs(y="Variant ID", x="log frequency")+
  scale_x_continuous(breaks=seq(-10,-5))+
  coord_cartesian(xlim=c(-10,-5))+
  scale_color_aaas()+
  theme_cowplot()+
  theme(axis.text.y=element_text(size=7))

rm(comt_facs_limma_q4_sig_gg)

```

```{r COMT FACS Q4 ALDEx2}

comt_facs_q4_fdr_thresh<- 0.01

# First transform normalized frequencies to counts per million
comt_facs_limma_q4_input_cast_mat_cpm<- apply(comt_facs_limma_q4_input_cast_mat, 2, function(x){round(exp(x)*1e6)})

set.seed(9)

comt_facs_q4_limma_input_cast_mat_cpm_aldex_dt<- run_aldex(comt_facs_limma_q4_input_cast_mat_cpm, c(rep("A", 4), rep("B", 4)))

comt_facs_q4_aldex2_pval_gg<- ggplot(
  data=comt_facs_q4_limma_input_cast_mat_cpm_aldex_dt, 
  aes(x=we.ep))

comt_facs_q4_aldex2_pval_gg+
  geom_histogram()+
  theme_cowplot()

rm(comt_facs_q4_aldex2_pval_gg)

# Volcano plot of ALDEx2 results
comt_facs_q4_aldex2_gg<- ggplot(
  data=comt_facs_q4_limma_input_cast_mat_cpm_aldex_dt,
  aes(x=effect, y=-1*log10(we.eBH)))

comt_facs_q4_aldex2_gg+
  geom_point()+
  geom_point(data=comt_facs_q4_limma_input_cast_mat_cpm_aldex_dt[we.eBH<comt_facs_q4_fdr_thresh], 
             aes(x=effect, y=-1*log10(we.eBH)), col="firebrick")+
  theme_cowplot()

rm(comt_facs_q4_aldex2_gg)

comt_facs_q4_limma_input_cast_mat_cpm_aldex_dt[,VAR_ID:=rownames(comt_facs_limma_q4_input_cast_mat_cpm)]

comt_facs_q4_limma_input_cast_mat_cpm_aldex_merge_dt<- merge(
  comt_facs_q4_limma_input_cast_mat_cpm_aldex_dt, annot_dt, by="VAR_ID")

comt_facs_q4_aldex2_sig_gg<- ggplot(
  data=comt_facs_q4_limma_input_cast_mat_cpm_aldex_merge_dt[
    we.eBH<comt_facs_q4_fdr_thresh],
  aes(x=effect, y=reorder(ID_SHORT, POS), fill=we.eBH, xmin=effect.low, xmax=effect.high))

comt_facs_q4_aldex2_sig_gg+
  geom_bar(stat="identity", alpha=0.75)+
  geom_errorbar(col="grey", width=0.3)+
  scale_fill_viridis(option="D", begin=0.95, end=0.05)+
  theme_cowplot()+
  labs(y="", x="Effect size", fill="FDR")+
  theme(axis.text.y=element_text(size=8))

rm(comt_facs_q4_aldex2_sig_gg)

# Generate a VE map using ALDEx2
comt_facs_q4_rep_summ_ve_dt<- comt_facs_q4_limma_input_cast_mat_cpm_aldex_merge_dt[
  !grepl(",", AA_CHANGE), .(Median_effect=median(effect, na.rm=T),
                            Max_effect=max(effect, na.rm=T)), 
  by=.(ROI, AA_POS, REF_AA, ALT_AA, AA_CHANGE, VAR_TYPE)]

comt_facs_q4_rep_summ_ve_dt[,AA_POS:=as.integer(AA_POS)]
comt_facs_q4_rep_summ_ve_dt[,ALT_AA:=factor(ALT_AA, levels=unlist(aa_class_list))]
comt_facs_q4_rep_summ_ve_dt[,ALT_AA_level:=as.integer(ALT_AA)]

comt_facs_q4_aldex_ve_gg<- ggplot(
  data=comt_facs_q4_rep_summ_ve_dt, 
  aes(x=AA_POS, y=reorder(ALT_AA, ALT_AA_level), fill=Median_effect))

comt_facs_q4_aldex_ve_gg+
  facet_wrap(~ROI, scales="free")+
  geom_tile()+
  scale_fill_gradient2(high=scales::muted("blue"), low=scales::muted("red"))+
  labs(x="Amino acid position", y="Alternate", 
       fill="Median effect")+
  theme_cowplot()

# Max effect
comt_facs_q4_aldex_ve_gg<- ggplot(
  data=comt_facs_q4_rep_summ_ve_dt, 
  aes(x=AA_POS, y=reorder(ALT_AA, ALT_AA_level), fill=Max_effect))

comt_facs_q4_aldex_ve_gg+
  facet_wrap(~ROI, scales="free")+
  geom_tile()+
  scale_fill_gradient2(high=scales::muted("blue"), low=scales::muted("red"))+
  labs(x="Amino acid position", y="Alternate", 
       fill="Max effect")+
  theme_cowplot()

rm(comt_facs_q4_aldex_ve_gg)

```

```{r COMT protein versus mRNA abundance limma}

comt_facs_limma_input_cast_mat_eset_fit_eb_all_dt<- rbind(
  comt_facs_limma_q1_input_cast_mat_eset_fit_eb_all_dt, 
  comt_facs_limma_q2_input_cast_mat_eset_fit_eb_all_dt, 
  comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_dt, 
  comt_facs_limma_q4_input_cast_mat_eset_fit_eb_all_dt)

comt_facs_limma_input_cast_mat_eset_fit_eb_all_dt[
  ,Quadrant:=rep(c("Q1", "Q2", "Q3", "Q4"), times=c(
    nrow(comt_facs_limma_q1_input_cast_mat_eset_fit_eb_all_dt), 
    nrow(comt_facs_limma_q2_input_cast_mat_eset_fit_eb_all_dt), 
    nrow(comt_facs_limma_q3_input_cast_mat_eset_fit_eb_all_dt), 
    nrow(comt_facs_limma_q4_input_cast_mat_eset_fit_eb_all_dt)))]

comt_facs_limma_input_cast_mat_eset_fit_eb_all_merge_dt<- merge(
  comt_facs_limma_input_cast_mat_eset_fit_eb_all_dt, annot_dt, by="VAR_ID")


comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[,ID_SHORT:=paste(POS, REF, ALT, AA_CHANGE, sep=":")]

comt_rna_vs_protein_key_dt<- comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[
  ,.(ID_SHORT, VAR_ID, REF_AA, ALT_AA, AA_CHANGE, AA_POS2, Direction)]

comt_rep_summ_cast_rna_dir_dt<- merge(comt_rep_summ_cast_dt, comt_rna_vs_protein_key_dt, by="ID_SHORT")

comt_rep_summ_cast_rna_dir_dt[,FACS.3_gDNA:=FACS.3-gDNA.]

rna_sig_short_ids<- comt_rna_limma_input_cast_eset_fit_eb_all_merge_dt[adj.P.Val < rna_fdr_thresh, unique(ID_SHORT)]

# Plot N.S. variants as well

rna_vs_facs1_gg<- ggplot(
  data=comt_rep_summ_cast_rna_dir_dt, 
  aes(x=Direction, y=FACS.1, fill=Direction))

rna_vs_facs1_gg+
  geom_boxplot(width=0.5, alpha=0.75, outlier.shape=NA)+
  scale_fill_manual(values=c("grey", "firebrick", "deepskyblue4"))+
  labs(x="cDNA:gDNA direction", y="Quadrant 1")+
  theme_cowplot()

rna_vs_facs2_gg<- ggplot(
  data=comt_rep_summ_cast_rna_dir_dt, 
  aes(x=Direction, y=FACS.2, fill=Direction))

rna_vs_facs2_gg+
  geom_boxplot(width=0.5, alpha=0.75, outlier.shape=NA)+
  scale_fill_manual(values=c("grey", "firebrick", "deepskyblue4"))+
  labs(x="cDNA:gDNA direction", y="Quadrant 2")+
  theme_cowplot()

rna_vs_facs3_gg<- ggplot(
  data=comt_rep_summ_cast_rna_dir_dt, 
  aes(x=Direction, y=FACS.3, fill=Direction))

rna_vs_facs3_gg+
  geom_boxplot(width=0.5, alpha=0.75, outlier.shape=NA)+
  scale_fill_manual(values=c("grey", "firebrick", "deepskyblue4"))+
  labs(x="cDNA:gDNA direction", y="Quadrant 3")+
  theme_cowplot()

# Influence of H62-V63?
rna_vs_facs3_no_hotspot_gg<- ggplot(
  data=comt_rep_summ_cast_rna_dir_dt[!grepl("H62|V63", ID_SHORT)], 
  aes(x=Direction, y=FACS.3, fill=Direction))

rna_vs_facs3_no_hotspot_gg+
  geom_boxplot(width=0.5, alpha=0.75, outlier.shape=NA)+
  scale_fill_manual(values=c("grey", "firebrick", "deepskyblue4"))+
  labs(x="cDNA:gDNA direction", y="Quadrant 3")+
  theme_cowplot()

rna_vs_facs4_gg<- ggplot(
  data=comt_rep_summ_cast_rna_dir_dt, 
  aes(x=Direction, y=FACS.4, fill=Direction))

rna_vs_facs4_gg+
  geom_boxplot(width=0.5, alpha=0.75, outlier.shape=NA)+
  scale_fill_manual(values=c("grey", "firebrick", "deepskyblue4"))+
  labs(x="cDNA:gDNA direction", y="Quadrant 4")+
  theme_cowplot()

rm(list=c("rna_vs_facs1_gg", "rna_vs_facs2_gg", "rna_vs_facs3_gg", "rna_vs_facs4_gg", 
          "rna_vs_facs3_no_hotspot_gg"))

```

```{r COMT protein versus mRNA abundance ALDEx2}

comt_facs_all_limma_input_cast_mat_cpm_aldex_dt<- rbind(
  comt_facs_q1_limma_input_cast_mat_cpm_aldex_dt, 
  comt_facs_q2_limma_input_cast_mat_cpm_aldex_dt, 
  comt_facs_q3_limma_input_cast_mat_cpm_aldex_dt, 
  comt_facs_q4_limma_input_cast_mat_cpm_aldex_dt)

comt_facs_all_limma_input_cast_mat_cpm_aldex_dt[
  ,Quadrant:=rep(c("Q1", "Q2", "Q3", "Q4"), times=c(
    nrow(comt_facs_q1_limma_input_cast_mat_cpm_aldex_dt), 
    nrow(comt_facs_q2_limma_input_cast_mat_cpm_aldex_dt), 
    nrow(comt_facs_q3_limma_input_cast_mat_cpm_aldex_dt), 
    nrow(comt_facs_q4_limma_input_cast_mat_cpm_aldex_dt)))]

comt_facs_all_limma_input_cast_mat_cpm_aldex_merge_dt<- merge(
  comt_facs_all_limma_input_cast_mat_cpm_aldex_dt, annot_dt, by="VAR_ID")

comt_facs_all_limma_input_cast_mat_cpm_aldex_merge_dt[,Direction:=ifelse(effect>0, "Pop. up", "Pop. down")]
comt_facs_all_limma_input_cast_mat_cpm_aldex_merge_dt[we.eBH>=comt_facs_q3_fdr_thresh, Direction:="N.S."]

#rm(comt_facs_all_limma_input_cast_mat_cpm_aldex_dt)

# Merge with the RNA abundance dataset
comt_rna_input_cast_mat_cpm_aldex_merge_dt[,Direction:=ifelse(effect>0, "RNA up", "RNA down")]
comt_rna_input_cast_mat_cpm_aldex_merge_dt[we.eBH>=0.15, Direction:="N.S."]
comt_rna_input_cast_mat_cpm_aldex_merge_dt[,Direction:=factor(Direction, levels=c("N.S.", "RNA down", "RNA up"))]

comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt<- merge(
  comt_rna_input_cast_mat_cpm_aldex_merge_dt[,.(diff.btw, effect, we.eBH, Direction, VAR_ID)],
  comt_facs_all_limma_input_cast_mat_cpm_aldex_merge_dt, by="VAR_ID")

# Plot N.S. variants as well

rna_vs_facs1_gg<- ggplot(
  data=comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[Quadrant=="Q1"], 
  aes(x=Direction.x, y=effect.y, fill=Direction.x))

rna_vs_facs1_gg+
  geom_boxplot(width=0.5, alpha=0.75, outlier.shape=NA)+
  scale_fill_manual(values=c("grey", "firebrick", "deepskyblue4"))+
  labs(x="cDNA:gDNA direction", y="P1 effect", fill="")+
  theme_cowplot()

rna_vs_facs2_gg<- ggplot(
  data=comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[Quadrant=="Q2"], 
  aes(x=Direction.x, y=effect.y, fill=Direction.x))

rna_vs_facs2_gg+
  geom_boxplot(width=0.5, alpha=0.75, outlier.shape=NA)+
  scale_fill_manual(values=c("grey", "firebrick", "deepskyblue4"))+
  labs(x="cDNA:gDNA direction", y="P2 effect")+
  theme_cowplot()

rna_vs_facs3_gg<- ggplot(
  data=comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[Quadrant=="Q3"], 
  aes(x=Direction.x, y=effect.y, fill=Direction.x))

rna_vs_facs3_gg+
  geom_boxplot(width=0.5, alpha=0.75, outlier.shape=NA)+
  scale_fill_manual(values=c("grey", "firebrick", "deepskyblue4"))+
  labs(x="cDNA:gDNA direction", y="P3 effect")+
  theme_cowplot()

# Influence of H62-V63?
rna_vs_facs3_no_hotspot_gg<- ggplot(
  data=comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[
    Quadrant=="Q3" & !grepl("H62|V63", ID_SHORT)], 
  aes(x=Direction.x, y=effect.y, fill=Direction.x))

rna_vs_facs3_no_hotspot_gg+
  geom_boxplot(width=0.5, alpha=0.75, outlier.shape=NA)+
  scale_fill_manual(values=c("grey", "firebrick", "deepskyblue4"))+
  labs(x="cDNA:gDNA direction", y="P3 effect, no H62, V63")+
  theme_cowplot()

rna_vs_facs4_gg<- ggplot(
  data=comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[Quadrant=="Q4"], 
  aes(x=Direction.x, y=effect.y, fill=Direction.x))

rna_vs_facs4_gg+
  geom_boxplot(width=0.5, alpha=0.75, outlier.shape=NA)+
  scale_fill_manual(values=c("grey", "firebrick", "deepskyblue4"))+
  labs(x="cDNA:gDNA direction", y="P4 effect")+
  theme_cowplot()

# Combine P1, P3 plots
rna_vs_facs1_3_gg<- ggplot(
  data=comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[Quadrant%in%c("Q1", "Q3")], 
  aes(x=Direction.x, y=effect.y, fill=Quadrant))

rna_vs_facs1_3_gg+
  geom_boxplot(width=0.5, alpha=0.75, outlier.shape=NA)+
  scale_fill_manual(values=c("deepskyblue4", "firebrick"))+
  coord_cartesian(ylim=c(-8,5))+
  labs(x="cDNA:gDNA direction", y="FACS effect size")+
  theme_cowplot()


# Scatterplot of RNA versus P3 enrichment
rna_vs_facs1_3_gg<- ggplot(
  data=comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[
    Quadrant=="Q3" & (we.eBH.x < comt_rna_fdr_thresh | we.eBH.y < comt_facs_q3_fdr_thresh)], 
  aes(x=effect.x, y=effect.y))

rna_vs_facs1_3_gg+
  geom_point()+
  coord_cartesian(ylim=c(-8,5))+
  labs(x="RNA effect size", y="FACS effect size")+
  theme_cowplot()

comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[
  Quadrant=="Q1" & Direction.x!="RNA up", wilcox.test(
    formula=effect.y~Direction.x, data=.SD)]

comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[
  Quadrant=="Q3" & Direction.x!="RNA up", wilcox.test(
    formula=effect.y~Direction.x, data=.SD)]

comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[
  Quadrant=="Q1" & Direction.x!="N.S.", wilcox.test(
    formula=effect.y~Direction.x, data=.SD)]

comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[
  Quadrant=="Q3" & Direction.x!="N.S.", wilcox.test(
    formula=effect.y~Direction.x, data=.SD)]

rm(list=c("rna_vs_facs1_gg", "rna_vs_facs2_gg", "rna_vs_facs3_gg", "rna_vs_facs4_gg", 
          "rna_vs_facs3_no_hotspot_gg", "rna_vs_facs1_3_gg"))

```

```{r COMT protein effect by amino acid property ALDEx2}

comt_facs_q3_fdr_aa_type_thresh<- 0.1

positive_charged_aas<- c("R", "H", "K")
negative_charged_aas<- c("D", "E")
polar_uncharged_aas<- c("S", "T", "N", "Q") # C
nonpolar_uncharged_aas<- c("A", "V", "I", "L", "M", "F", "Y", "W", "G")
other_aas<- c("C", "P")

# Annotate the AA type
comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[REF_AA%in%positive_charged_aas, REF_TYPE:="PC"]
comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[REF_AA%in%negative_charged_aas, REF_TYPE:="NC"]
comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[REF_AA%in%polar_uncharged_aas, REF_TYPE:="PU"]
comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[REF_AA%in%nonpolar_uncharged_aas, REF_TYPE:="NU"]
comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[REF_AA%in%other_aas, REF_TYPE:="Other"]

comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[ALT_AA%in%positive_charged_aas, ALT_TYPE:="PC"]
comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[ALT_AA%in%negative_charged_aas, ALT_TYPE:="NC"]
comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[ALT_AA%in%polar_uncharged_aas, ALT_TYPE:="PU"]
comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[ALT_AA%in%nonpolar_uncharged_aas, ALT_TYPE:="NU"]
comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[ALT_AA%in%other_aas, ALT_TYPE:="Other"]

comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[
  ,REF_TYPE:=factor(REF_TYPE, levels=c(
    "PC", "NC", "PU", "NU", "Other"))]

comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[
  ,ALT_TYPE:=factor(ALT_TYPE, levels=c(
    "PC", "NC", "PU", "NU", "Other"))]

comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[,REF_ALT_TYPE:=paste(REF_TYPE, ALT_TYPE, sep=", ")]

comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[,AA_POS_key:=paste(REF_AA, AA_POS, sep="")]
comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[,AA_POS_key2:=paste(REF_AA, AA_POS, ALT_AA, sep="")]

# ADD
comt_protein_del_positions<- c(59, 62, 63, 66, 67, 136, 137, 139, 152, 153, 156, 157, 158)
comt_protein_del_ids<- c("I59", "H62", "V63", "H66", "A67", "L136", "I137", "I139", "M152", "V153", "A156", "G157", "V158")
comt_protein_del_silent_ids<- c("I59I", "H62H", "V63V", "H66H", "A67A", "L136L", "I137I", "I139I", "M152M", "V153V", "A156A", "G157G", "V158")

comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_aa_type_dt<- 
  comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_dt[
    Quadrant=="Q3" & AA_POS_key%in%comt_protein_del_ids & 
      (we.eBH.y<comt_facs_q3_fdr_aa_type_thresh | AA_POS_key2%in%comt_protein_del_silent_ids), 
    .(Median_RNA_effect=median(effect.x), Median_FACS_effect=median(effect.y), 
      MAD_RNA_effect=mad(effect.x), MAD_FACS_effect=mad(effect.y), N=.N),
    by=.(REF_ALT_TYPE, REF_TYPE, ALT_TYPE, ROI, AA_POS2, AA_POS_key2)]

comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_aa_type_dt[,WT:="False"]

comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_aa_type_dt[
  AA_POS_key2%in%comt_protein_del_silent_ids, WT:="True"]

comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_aa_type_dt[Median_FACS_effect>5, Label:=AA_POS_key2]

comt_facs_vs_aa_type_gg<- ggplot(
  data=comt_facs_rna_limma_input_cast_mat_cpm_aldex_merge_aa_type_dt[!is.na(ALT_TYPE)], 
  aes(x=as.factor(AA_POS2), y=Median_FACS_effect, col=ALT_TYPE, shape=WT, label=Label))

comt_facs_vs_aa_type_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_jitter(width=0.1, size=6, alpha=0.75)+
  geom_text_repel(size=6)+
  scale_color_brewer(palette="Set2")+
  labs(x="Amino acid position", y="P3 effect size", col="Alternate type")+
  theme_cowplot()+
  theme(legend.position="bottom")

rm(comt_facs_vs_aa_type_gg)

```

```{r COMT protein versus polysome abundance limma}

comt_poly_f3_vs_protein_dt<- merge(
  comt_poly_limma_input_cast_mat_eset_fit_eb_all_dt[Fraction=="F3", .(VAR_ID, logFC, adj.P.Val)],
  comt_facs_limma_input_cast_mat_eset_fit_eb_all_merge_dt, by="VAR_ID")

# Highlight variants enriched in Q3 and enriched in heavy polysomes (unexpected)
comt_poly_f3_vs_protein_dt[Quadrant=="Q3" & logFC.x > 1 & logFC.y > 1, Label:=AA_CHANGE]

comt_poly_f3_vs_protein_gg<- ggplot(
  data=comt_poly_f3_vs_protein_dt[Quadrant%in%c("Q1", "Q3")], 
  aes(x=logFC.x, y=logFC.y, label=Label))

comt_poly_f3_vs_protein_gg+
  geom_point(alpha=0.3)+
  facet_wrap(~Quadrant)+
  geom_point(data=comt_poly_f3_vs_protein_dt[Quadrant%in%c("Q1", "Q3") & VAR_TYPE=="Silent"], 
             aes(x=logFC.x, y=logFC.y), col="dodgerblue")+
  geom_point(data=comt_poly_f3_vs_protein_dt[Quadrant%in%c("Q1", "Q3") & VAR_TYPE=="Nonsense"], 
             aes(x=logFC.x, y=logFC.y), col="firebrick")+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_vline(xintercept=0, col="grey", linetype=2)+
  geom_text_repel()+
  theme_cowplot()+
  labs(x="Poly F3 log FC", y="FACS log FC")

# No strong relationship with polysome F3

rm(comt_poly_f3_vs_protein_gg)

comt_poly_f4_vs_protein_dt<- merge(
  comt_poly_limma_input_cast_mat_eset_fit_eb_all_dt[Fraction=="F4", .(VAR_ID, logFC, adj.P.Val)],
  comt_facs_limma_input_cast_mat_eset_fit_eb_all_merge_dt, by="VAR_ID")

# Highlight variants enriched in Q3 and enriched in heavy polysomes (unexpected)
comt_poly_f4_vs_protein_dt[Quadrant=="Q3" & logFC.x > 1 & logFC.y > 1, Label:=AA_CHANGE]

comt_poly_f4_vs_protein_gg<- ggplot(
  data=comt_poly_f4_vs_protein_dt[Quadrant%in%c("Q1", "Q3")], 
  aes(x=logFC.x, y=logFC.y, label=Label))

comt_poly_f4_vs_protein_gg+
  geom_point(alpha=0.3)+
  facet_wrap(~Quadrant)+
  geom_point(data=comt_poly_f4_vs_protein_dt[Quadrant%in%c("Q1", "Q3") & VAR_TYPE=="Silent"], 
             aes(x=logFC.x, y=logFC.y), col="dodgerblue")+
  geom_point(data=comt_poly_f4_vs_protein_dt[Quadrant%in%c("Q1", "Q3") & VAR_TYPE=="Nonsense"], 
             aes(x=logFC.x, y=logFC.y), col="firebrick")+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_vline(xintercept=0, col="grey", linetype=2)+
  geom_text_repel()+
  theme_cowplot()+
  labs(x="Poly F4 log FC", y="FACS log FC")

rm(comt_poly_f4_vs_protein_gg)

```

```{r COMT multi-layer IQLR correlation}

#comt_rep_summ_cast_dt derives from comt_all_filt_postprocess_subtract_drop_bio_dt

comt_rep_summ_cast_mat<- as.matrix(comt_rep_summ_cast_dt[,-c("ROI", "ID_SHORT", "Poly.2")])
rownames(comt_rep_summ_cast_mat)<- comt_rep_summ_cast_dt[,ID_SHORT]
#comt_rep_summ_cast_mat<- normalizeQuantiles(comt_rep_summ_cast_mat)

# Convert to IQLR
comt_rep_summ_cast_nona_mat<- comt_rep_summ_cast_mat[apply(comt_rep_summ_cast_mat, 1, row_is_complete),]

comt_rep_summ_cast_nona_mat_freqs<- 10^comt_rep_summ_cast_nona_mat

# First reclose the dataset since WT frequency is not included in matrix
comt_rep_summ_cast_nona_mat_freqs_reclose<- apply(comt_rep_summ_cast_nona_mat_freqs, 2, reclose_coda)

comt_rep_summ_cast_nona_iqlr_mat<- apply(comt_rep_summ_cast_nona_mat_freqs_reclose, 2, iqlr)
comt_rep_summ_cast_nona_iqlr_dt<- as.data.table(comt_rep_summ_cast_nona_iqlr_mat)

# Correlation between DNA and RNA
comt_multilayer_dna_rna_gg<- ggplot(
  data=comt_rep_summ_cast_nona_iqlr_dt, 
  aes(x=gDNA., y=cDNA.))

comt_multilayer_dna_rna_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth()+
  theme_cowplot()

comt_rep_summ_cast_nona_iqlr_dt[,cor(x=gDNA., y=cDNA.)]

rm(comt_multilayer_dna_rna_gg)

# Correlation between RNA and poly fraction 3
comt_multilayer_rna_f3_gg<- ggplot(
  data=comt_rep_summ_cast_nona_iqlr_dt, 
  aes(x=cDNA., y=Poly.3))

comt_multilayer_rna_f3_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth()+
  theme_cowplot()

comt_rep_summ_cast_nona_iqlr_dt[,cor(x=cDNA., y=Poly.3)]

rm(comt_multilayer_rna_f3_gg)

# Correlation between RNA and poly fraction 4
comt_multilayer_rna_f4_gg<- ggplot(
  data=comt_rep_summ_cast_nona_iqlr_dt, 
  aes(x=cDNA., y=Poly.4))

comt_multilayer_rna_f4_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth()+
  theme_cowplot()

rm(comt_multilayer_rna_f4_gg)

comt_rep_summ_cast_nona_iqlr_dt[,cor(x=cDNA., y=Poly.4)]


# Correlation between RNA and FACS P1 (high protein); expected positive correlation
comt_multilayer_rna_p1_gg<- ggplot(
  data=comt_rep_summ_cast_nona_iqlr_dt, 
  aes(x=cDNA., y=FACS.1))

comt_multilayer_rna_p1_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth()+
  theme_cowplot()

comt_rep_summ_cast_nona_iqlr_dt[,cor(x=cDNA., y=FACS.1)]

rm(comt_multilayer_rna_p1_gg)

# Correlation between RNA and FACS P3 (low protein); expected inverse correlation
comt_multilayer_rna_p3_gg<- ggplot(
  data=comt_rep_summ_cast_nona_iqlr_dt, 
  aes(x=cDNA., y=FACS.3))

comt_multilayer_rna_p3_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth()+
  theme_cowplot()

comt_rep_summ_cast_nona_iqlr_dt[,cor(x=cDNA., y=FACS.3)]

rm(comt_multilayer_rna_p3_gg)

# Correlation between poly fraction 3 and FACS P1
comt_multilayer_f3_p1_gg<- ggplot(
  data=comt_rep_summ_cast_nona_iqlr_dt, 
  aes(x=Poly.3, y=FACS.1))

comt_multilayer_f3_p1_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth()+
  theme_cowplot()

comt_rep_summ_cast_nona_iqlr_dt[,cor(x=Poly.3, y=FACS.1)]

rm(comt_multilayer_f3_p1_gg)

# Correlation between poly fraction 4 and FACS P1
comt_multilayer_f4_p1_gg<- ggplot(
  data=comt_rep_summ_cast_nona_iqlr_dt, 
  aes(x=Poly.4, y=FACS.1))

comt_multilayer_f4_p1_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth()+
  theme_cowplot()

comt_rep_summ_cast_nona_iqlr_dt[,cor(x=Poly.4, y=FACS.1)]

rm(comt_multilayer_f4_p1_gg)

# Correlation between poly fraction 3 and FACS P3
comt_multilayer_f3_p3_gg<- ggplot(
  data=comt_rep_summ_cast_nona_iqlr_dt, 
  aes(x=Poly.3, y=FACS.3))

comt_multilayer_f3_p3_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth()+
  theme_cowplot()

comt_rep_summ_cast_nona_iqlr_dt[,cor(x=Poly.3, y=FACS.3)]

rm(comt_multilayer_f3_p3_gg)

# Correlation between poly fraction 4 and FACS P3
comt_multilayer_f4_p3_gg<- ggplot(
  data=comt_rep_summ_cast_nona_iqlr_dt, 
  aes(x=Poly.4, y=FACS.3))

comt_multilayer_f4_p3_gg+
  geom_abline(slope=1, col="grey", linetype=2)+
  geom_point(alpha=0.5)+
  geom_smooth()+
  theme_cowplot()

comt_rep_summ_cast_nona_iqlr_dt[,cor(x=Poly.4, y=FACS.3)]

rm(comt_multilayer_f4_p3_gg)

```


# Multi-layer expression analysis with unsupervised ML

```{r COMT multilayer hierarchical clustering ALDEx2}

# ALDEx2 multilayer analysis

multilayer_rna_fdr_thresh<- 0.1
multilayer_poly_protein_fdr_thresh<- 0.05

if("comt_all_expr_aldex_merge_dt"%in%ls()){
  rm(comt_all_expr_aldex_merge_dt)
}

# Combine ALDEx2 effect sizes
comt_all_expr_aldex_merge_dt<- merge(
  comt_rna_input_cast_mat_cpm_aldex_merge_dt,
  comt_poly_f3_f4_input_cast_mat_cpm_aldex_merge_dt,
  all=TRUE, by="VAR_ID")

comt_all_expr_aldex_merge_dt<- merge(
  comt_all_expr_aldex_merge_dt,
  comt_facs_q3_limma_input_cast_mat_cpm_aldex_merge_dt,
  all=TRUE, by="VAR_ID")

comt_all_expr_aldex_merge_dt[is.na(ROI), ROI:=ROI.x]
comt_all_expr_aldex_merge_dt[is.na(ROI), ROI:=ROI.y]

comt_all_expr_aldex_merge_dt[is.na(POS), POS:=POS.x]
comt_all_expr_aldex_merge_dt[is.na(POS), POS:=POS.y]

comt_all_expr_aldex_merge_dt[is.na(REF), REF:=REF.x]
comt_all_expr_aldex_merge_dt[is.na(REF), REF:=REF.y]

comt_all_expr_aldex_merge_dt[is.na(ALT), ALT:=ALT.x]
comt_all_expr_aldex_merge_dt[is.na(ALT), ALT:=ALT.y]

comt_all_expr_aldex_merge_dt[is.na(REF_CODON), REF_CODON:=REF_CODON.x]
comt_all_expr_aldex_merge_dt[is.na(REF_CODON), REF_CODON:=REF_CODON.y]

comt_all_expr_aldex_merge_dt[is.na(ALT_CODON), ALT_CODON:=ALT_CODON.x]
comt_all_expr_aldex_merge_dt[is.na(ALT_CODON), ALT_CODON:=ALT_CODON.y]

comt_all_expr_aldex_merge_dt[is.na(AA_POS), AA_POS:=AA_POS.x]
comt_all_expr_aldex_merge_dt[is.na(AA_POS), AA_POS:=AA_POS.y]

comt_all_expr_aldex_merge_dt[is.na(REF_AA), REF_AA:=REF_AA.x]
comt_all_expr_aldex_merge_dt[is.na(REF_AA), REF_AA:=REF_AA.y]

comt_all_expr_aldex_merge_dt[is.na(ALT_AA), ALT_AA:=ALT_AA.x]
comt_all_expr_aldex_merge_dt[is.na(ALT_AA), ALT_AA:=ALT_AA.y]

comt_all_expr_aldex_merge_dt[is.na(AA_CHANGE), AA_CHANGE:=AA_CHANGE.x]
comt_all_expr_aldex_merge_dt[is.na(AA_CHANGE), AA_CHANGE:=AA_CHANGE.y]

comt_all_expr_aldex_merge_dt[is.na(VAR_TYPE), VAR_TYPE:=VAR_TYPE.x]
comt_all_expr_aldex_merge_dt[is.na(VAR_TYPE), VAR_TYPE:=VAR_TYPE.y]

comt_all_expr_aldex_merge_dt[is.na(TYPE), TYPE:=TYPE.x]
comt_all_expr_aldex_merge_dt[is.na(TYPE), TYPE:=TYPE.y]

comt_all_expr_aldex_merge_dt[is.na(ID_SHORT), ID_SHORT:=ID_SHORT.x]
comt_all_expr_aldex_merge_dt[is.na(ID_SHORT), ID_SHORT:=ID_SHORT.y]

# Filter out columns not used
comt_all_expr_aldex_merge_dt<- comt_all_expr_aldex_merge_dt[
  ,.(ROI, VAR_ID, POS, REF, ALT, REF_CODON, ALT_CODON, AA_POS, AA_POS2, REF_AA, ALT_AA, AA_CHANGE, VAR_TYPE, TYPE, ID_SHORT, Fraction,
     effect, effect.x, effect.y, diff.btw, diff.btw.x, diff.btw.y, we.eBH, we.eBH.x, we.eBH.y, 
     Wu_REF_CSC, Wu_ALT_CSC , Wu_CSC_diff, Narula_REF_CSC, Narula_ALT_CSC, Narula_CSC_diff, 
     Narula_ALT_Asite, Narula_REF_Asite, Narula_ALT_Psite, Narula_REF_Psite, Narula_Asite_diff, Narula_Psite_diff,
     REF_CODON_hydro_log2_sum, ALT_CODON_hydro_log2_sum, hydro_log2_sum_diff, 
     REF_CODON_mim_log2_sum, ALT_CODON_mim_log2_sum, mim_log2_sum_diff,
     d_max, d_max_pval, r_min, r_min_pval)]

write.table(comt_all_expr_aldex_merge_dt, paste(comt_res_base_dir, "ALDEx2_results.txt", sep="/"), sep="\t", quote=FALSE, row.names=FALSE)

# effect is P3 enrichment, effect.x is RNA enrichment, effect.y is F3 or F4 enrichment

# Filter again for any significant variant for RNA, poly; and only significant, enriched nonsense variants
#comt_all_expr_aldex_merge_sig_dt<- comt_all_expr_aldex_merge_dt[
#effect > 0 & (we.eBH < multilayer_rna_fdr_thresh | we.eBH.x < multilayer_poly_protein_fdr_thresh | we.eBH.y < multilayer_poly_protein_fdr_thresh)]

comt_all_expr_aldex_merge_sig_dt<- comt_all_expr_aldex_merge_dt[
(we.eBH < multilayer_poly_protein_fdr_thresh & effect > 0) | 
  we.eBH.x < multilayer_rna_fdr_thresh | 
  we.eBH.y < multilayer_poly_protein_fdr_thresh]

comt_all_expr_aldex_merge_sig_melt_dt<- melt(
  data=comt_all_expr_aldex_merge_sig_dt, id.vars=c(
    "ROI", "VAR_ID", "Fraction", "POS", "REF", "ALT", "REF_CODON", "ALT_CODON", "AA_POS", "AA_POS2", 
    "REF_AA", "ALT_AA", "AA_CHANGE", "VAR_TYPE", "TYPE", "ID_SHORT"),
  measure.vars=c("effect", "effect.x", "effect.y"), value.name="Effect_size", variable.name="Layer")

comt_all_expr_aldex_merge_sig_melt_dt[Layer=="effect", Layer:="Protein"]
comt_all_expr_aldex_merge_sig_melt_dt[Layer=="effect.x", Layer:="RNA"]
comt_all_expr_aldex_merge_sig_melt_dt[Layer=="effect.y", Layer:="Poly"]
comt_all_expr_aldex_merge_sig_melt_dt[,Layer:=factor(Layer, levels=c("RNA", "Poly", "Protein"))]

comt_all_expr_aldex_merge_sig_melt_dt[VAR_TYPE=="Nonsense", VAR_TYPE_order:=1]
comt_all_expr_aldex_merge_sig_melt_dt[VAR_TYPE=="Missense", VAR_TYPE_order:=2]
comt_all_expr_aldex_merge_sig_melt_dt[VAR_TYPE=="Silent", VAR_TYPE_order:=3]

comt_all_expr_aldex_merge_sig_melt_dt[Layer=="Protein", Effect_size:=-1 * Effect_size]

# Finally aggregate at the level of amino acid
comt_all_expr_aldex_merge_sig_melt_summ_dt<- comt_all_expr_aldex_merge_sig_melt_dt[
  ,.(Median_effect_size=median(Effect_size, na.rm=TRUE)), by=.(
    Layer, Fraction, ROI, AA_POS2, AA_CHANGE, VAR_TYPE, VAR_TYPE_order)]

comt_all_expr_aldex_merge_sig_melt_summ_nonsense_dt<- comt_all_expr_aldex_merge_sig_melt_summ_dt[VAR_TYPE=="Nonsense"]

# Jitterplot of variants found tested in RNA, poly F4, and FC P3
comt_multilayer_nonsense_gg<- ggplot(
  data=comt_all_expr_aldex_merge_sig_melt_summ_nonsense_dt[Fraction=="F4"],
  aes(x=Layer, y=Median_effect_size))

comt_multilayer_nonsense_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_jitter(width=0.15, alpha=0.75, size=2)+
  theme_cowplot()+
  labs(x="", y="Effect size")

rm(comt_multilayer_nonsense_gg)

# Silent, missense
comt_all_expr_aldex_merge_sig_melt_summ_cast_dt<- dcast(
  data=comt_all_expr_aldex_merge_sig_melt_summ_dt[Fraction=="F4" & VAR_TYPE!="Nonsense"],
  formula=AA_CHANGE~Layer, value.var="Median_effect_size")

comt_all_expr_aldex_merge_sig_melt_summ_mat<- as.matrix(comt_all_expr_aldex_merge_sig_melt_summ_cast_dt[,-"AA_CHANGE"])
rownames(comt_all_expr_aldex_merge_sig_melt_summ_mat)<- comt_all_expr_aldex_merge_sig_melt_summ_cast_dt[,AA_CHANGE]

pheatmap(comt_all_expr_aldex_merge_sig_melt_summ_mat, cluster_cols=FALSE, clustering_method="complete",
         color=colorRampPalette(c(scales::muted("red"), "white", scales::muted("blue")))(50))

# Correlations between IQLR and predicted RNA structure?
comt_all_expr_aldex_merge_dt[Fraction=="F4", cor.test(effect.x, y=d_max, use="complete.obs", method="spearman")]
comt_all_expr_aldex_merge_dt[Fraction=="F4", cor.test(effect.y, y=d_max, use="complete.obs", method="spearman")]
comt_all_expr_aldex_merge_dt[Fraction=="F4", cor.test(effect, y=d_max, use="complete.obs", method="spearman")]

comt_all_expr_aldex_merge_dt[Fraction=="F3", cor.test(effect.x, y=d_max, use="complete.obs", method="spearman")]
comt_all_expr_aldex_merge_dt[Fraction=="F3", cor.test(effect.y, y=d_max, use="complete.obs", method="spearman")]
comt_all_expr_aldex_merge_dt[Fraction=="F3", cor.test(effect, y=d_max, use="complete.obs", method="spearman")]

# Weak but significant correlation with protein effect size
# Weak but significant correlation with poly F3 effect size
# Do these correlations hold using the min correlation metric?

comt_all_expr_aldex_merge_dt[Fraction=="F4", cor.test(effect.x, y=r_min, use="complete.obs", method="spearman")]
comt_all_expr_aldex_merge_dt[Fraction=="F4", cor.test(effect, y=r_min, use="complete.obs", method="spearman")]

# Yes

# Does filtering for significant variants in terms of effect size or predicted folding strengthen the association?
comt_all_expr_aldex_merge_dt[Fraction=="F4" & we.eBH.x < 0.2, cor.test(effect.x, y=d_max, use="complete.obs", method="spearman")]
comt_all_expr_aldex_merge_dt[Fraction=="F4" & we.eBH < 0.2, cor.test(effect, y=d_max, use="complete.obs", method="spearman")]

# After filtering on significant of the effect, only protein correlation remains
comt_all_expr_aldex_merge_dt[
  Fraction=="F4" & we.eBH < 0.1 & d_max_pval < 0.15, cor.test(effect, y=d_max, use="complete.obs", method="spearman")]

# Iterate over FDR and p-value thresholds and determine best correlation between distance and effect size
pval_thresholds<- seq(0.1, 1, 0.05)
protein_struct_mat<- matrix(data=0, nrow=length(pval_thresholds), ncol=length(pval_thresholds))

for(i in 1:length(pval_thresholds)){
  for(j in 1:length(pval_thresholds)){
    protein_struct_mat[i,j]<- comt_all_expr_aldex_merge_dt[
      Fraction=="F4" & we.eBH < pval_thresholds[i] & d_max_pval < pval_thresholds[j], cor(
        effect, y=d_max, use="complete.obs", method="spearman")]
  }
}

rna_struct_mat<- matrix(data=0, nrow=length(pval_thresholds), ncol=length(pval_thresholds))

for(i in 1:length(pval_thresholds)){
  for(j in 1:length(pval_thresholds)){
    rna_struct_mat[i,j]<- comt_all_expr_aldex_merge_dt[
      Fraction=="F4" & we.eBH.x < pval_thresholds[i] & d_max_pval < pval_thresholds[j], cor(
        effect.x, y=d_max, use="complete.obs", method="spearman")]
  }
}

comt_all_expr_aldex_merge_dt[,Protein_del:=ifelse(effect>0, "True", "False")]
comt_all_expr_aldex_merge_dt[,RNA_del:=ifelse(effect.x<0, "True", "False")]

protein_vs_struct_gg<- ggplot(
  data=comt_all_expr_aldex_merge_dt[Fraction=="F4" & we.eBH < 0.1 & d_max_pval < 0.15],
  aes(x=Protein_del, y=d_max))

protein_vs_struct_gg+
  geom_boxplot(outlier.shape=NA, fill="steelblue", alpha=0.75, width=0.5)+
  geom_jitter(width=0.1)+
  theme_cowplot()

comt_all_expr_aldex_merge_dt[Fraction=="F4" & we.eBH < 0.1 & d_max_pval < 0.15, wilcox.test(effect~Protein_del)]

```

```{r COMT multilayer SOM clustering on IQLR}

# Hierarchical clustering on frequencies
pheatmap(comt_rep_summ_cast_nona_mat, show_rownames=FALSE, clustering_method="ward.D2")

# Hierarchical clustering on IQLRs
pheatmap(comt_rep_summ_cast_nona_iqlr_mat, show_rownames=FALSE, clustering_method="ward.D2")

# Plot the most variable variants across layers
comt_rep_summ_cast_nona_iqlr_var<- apply(comt_rep_summ_cast_nona_iqlr_mat, 1, function(x){sd(x)/mean(x)})

# Select variants using percentile
comt_rep_summ_cast_nona_iqlr_var_quantile<- quantile(comt_rep_summ_cast_nona_iqlr_var, .98)

comt_rep_summ_cast_nona_iqlr_mat_high_var_mat<- comt_rep_summ_cast_nona_iqlr_mat[
  comt_rep_summ_cast_nona_iqlr_var >= comt_rep_summ_cast_nona_iqlr_var_quantile,]

pheatmap(comt_rep_summ_cast_nona_iqlr_mat_high_var_mat, clustering_method="ward.D2")


write.table(comt_rep_summ_cast_nona_iqlr_mat, paste(comt_res_base_dir, "IQLRs.txt", sep="/"), sep="\t", quote=FALSE)

# Remove P2 and P4 as they are not particularly useful
comt_rep_summ_cast_nona_iqlr_mat_filt<- comt_rep_summ_cast_nona_iqlr_mat[,
  !colnames(comt_rep_summ_cast_nona_iqlr_mat)%in%c("FACS.2", "FACS.4")]

# Plot the most variable variants across layers
comt_rep_summ_cast_nona_iqlr_mat_filt_var<- apply(comt_rep_summ_cast_nona_iqlr_mat_filt, 1, function(x){sd(x)/mean(x)})

# Select variants using percentile
comt_rep_summ_cast_nona_iqlr_mat_filt_var_quantile<- quantile(comt_rep_summ_cast_nona_iqlr_mat_filt_var, .99)

comt_rep_summ_cast_nona_iqlr_mat_filt_high_var<- comt_rep_summ_cast_nona_iqlr_mat_filt[
  comt_rep_summ_cast_nona_iqlr_mat_filt_var >= comt_rep_summ_cast_nona_iqlr_mat_filt_var_quantile,]

pheatmap(comt_rep_summ_cast_nona_iqlr_mat_filt_high_var, clustering_method="ward.D2")


# Train SOMs on centered IQLR matrices

#' Centers with the robust median
#'
#' @param x numeric or integer vector
#' @param standardize logical indicating the data should also be standardized after centering
#' @return numeric vector centered and possibly standardized
robust_center<- function(x, standardize=FALSE){
  centered_x<- x - median(x, na.rm=TRUE)
  if(standardize==TRUE){
    centered_x<- centered_x / sd(centered_x, na.rm=TRUE)
  }
  return(centered_x)
}


# Scale and standardize 
# For some reason apply with margin of 1 transposes the matrix
comt_rep_summ_cast_nona_iqlr_mat_filt_center<- t(apply(scale(comt_rep_summ_cast_nona_iqlr_mat_filt), 1, robust_center))

# comt_rep_summ_cast_nona_iqlr_mat_filt
som_input_mat<- comt_rep_summ_cast_nona_iqlr_mat_filt_center

# Determine best seed of 500 for minimizing mean distances
som_nunits<- round(5 * sqrt(nrow(som_input_mat)))

# Bump to 260 to to have more balanced factors
# 260 = 13 * 20

som_xdim<- 13
som_ydim<- 20
som_topo<- "hexagonal"
som_toroid<- TRUE

# Determine an optimal seed based on minimization of distances of each object to its respective codebook vector
# 300
determine_som_best_seed<- function(dat_mat, nseeds=500, xdim=som_xdim, ydim=som_ydim, topo=som_topo, is_toroidal=som_toroid){
  
  mean_distances<- vector(mode="numeric", length=nseeds)
  
  for(i in 1:nseeds){
    set.seed(i)
    som_grid<- kohonen::somgrid(xdim=xdim, ydim=ydim, topo=topo, toroidal=is_toroidal)
    som_res<- kohonen::som(dat_mat, grid=som_grid)
    mean_dist<- base::max(som_res$distances)
    mean_distances[i]<- mean_dist
  }
  
  names(mean_distances)<- as.character(1:nseeds)
  return(mean_distances)
}

som_mean_distances<- determine_som_best_seed(som_input_mat)
som_optimal_seed<- names(som_mean_distances[which.min(som_mean_distances)])

set.seed(as.integer(som_optimal_seed))

iqlr_multilayer_som<- kohonen::som(
  som_input_mat, 
  grid=kohonen::somgrid(xdim=som_xdim, ydim=som_ydim, topo=som_topo, toroidal=som_toroid))

plot(iqlr_multilayer_som, type="codes")

plot(iqlr_multilayer_som, type="counts")

iqlr_multilayer_som_counts_dt<- as.data.table(table(iqlr_multilayer_som$unit.classif))
names(iqlr_multilayer_som_counts_dt)<- c("SOM_neuron", "N_variants")

#plot(iqlr_multilayer_som, type="property")

# Run affinity propagation clustering
iqlr_multilayer_som_codes<- iqlr_multilayer_som$codes[[1]]

set.seed(9)

iqlr_multilayer_som_ap<- apcluster(negDistMat(r=2), iqlr_multilayer_som_codes, q=0.1)

cluster_membership<- c()
for (i in 1:length(iqlr_multilayer_som_ap@clusters)){
  cluster_membership[iqlr_multilayer_som_ap@clusters[[i]]]<- i
}

iqlr_multilayer_som_counts_dt[,AP_cluster:=cluster_membership]

iqlr_multilayer_som_counts_summ_dt<- iqlr_multilayer_som_counts_dt[,.(N_variants=sum(N_variants)), keyby=.(AP_cluster)]

# Plot kohonen map with AP clusters
som_ap_cluster_cols<- brewer.pal(n=9, name="Set3")

plot(iqlr_multilayer_som, shape="straight", type="property", property=cluster_membership, 
     palette.name=function(x){return(som_ap_cluster_cols)})

# Plot exemplars for each AP cluster
iqlr_multilayer_som_codes_dt<- as.data.table(iqlr_multilayer_som_codes[iqlr_multilayer_som_ap@exemplars,])
iqlr_multilayer_som_codes_dt[,Cluster:=as.factor(paste("C", seq(1,9), sep=""))]

iqlr_multilayer_som_codes_melt_dt<- melt(
  iqlr_multilayer_som_codes_dt, id.vars="Cluster", 
  variable.name="Layer", value.name="Scaled_IQLR")

iqlr_multilayer_som_codes_melt_gg<- ggplot(
  data=iqlr_multilayer_som_codes_melt_dt, 
  aes(x=Layer, y=Cluster, fill=Scaled_IQLR))

iqlr_multilayer_som_codes_melt_gg+
  geom_tile()+
  scale_fill_gradient2(low=scales::muted("#276419"), mid='white', high=scales::muted("#8E0152"))+
  theme_cowplot()
             
# Determine variants enriched in each cluster
iqlr_multilayer_som_var_dt<- data.table(
  "ID_SHORT"=rownames(som_input_mat),
  "SOM_neuron"=as.character(iqlr_multilayer_som$unit.classif))
                                        
iqlr_multilayer_som_var_merge_dt<- merge(iqlr_multilayer_som_var_dt, iqlr_multilayer_som_counts_dt, by="SOM_neuron")

rm(iqlr_multilayer_som_var_dt)

iqlr_multilayer_som_var_merge_annot_dt<- merge(iqlr_multilayer_som_var_merge_dt, annot_dt, by="ID_SHORT")

rm(iqlr_multilayer_som_var_merge_dt)

# Plot silent, nonsense, missense variants for each AP cluster
iqlr_multilayer_som_var_merge_annot_summ_dt<- iqlr_multilayer_som_var_merge_annot_dt[,.N,by=.(AP_cluster, VAR_TYPE)]
iqlr_multilayer_som_var_merge_annot_summ_dt[,AP_cluster:=factor(AP_cluster)]

# Cast then melt againt to include 0 counts for each cluster
iqlr_multilayer_som_var_merge_annot_summ_cast_dt<- dcast(
  data=iqlr_multilayer_som_var_merge_annot_summ_dt, formula=AP_cluster~VAR_TYPE, value.var="N")

iqlr_multilayer_som_var_merge_annot_summ_cast_dt[is.na(iqlr_multilayer_som_var_merge_annot_summ_cast_dt)]<- as.integer(0)

iqlr_multilayer_som_var_merge_annot_summ_melt_dt<- melt(
  data=iqlr_multilayer_som_var_merge_annot_summ_cast_dt, id.vars=c("AP_cluster"), variable.name="VAR_TYPE", value.name="N")

# Plot counts
jco_cols<- pal_jco()(3)[c(1,3,2)]

som_ap_cluster_vartype_counts_gg<- ggplot(
  data=iqlr_multilayer_som_var_merge_annot_summ_melt_dt, 
  aes(x=AP_cluster, y=log2(N+1), fill=VAR_TYPE))

som_ap_cluster_vartype_counts_gg+
  geom_bar(stat="identity", position="dodge")+
  scale_fill_manual(values=jco_cols)+
  scale_x_discrete(drop=FALSE)+
  labs(x="AP cluster", y="log2 variants + 1", fill="")+
  theme_cowplot()

# Plot proportions
iqlr_multilayer_som_counts_summ_dt[,AP_cluster:=factor(AP_cluster)]

iqlr_multilayer_som_var_merge_annot_summ_melt_nvars_dt<- merge(
  iqlr_multilayer_som_var_merge_annot_summ_melt_dt, 
  iqlr_multilayer_som_counts_summ_dt, by="AP_cluster")

iqlr_multilayer_som_var_merge_annot_summ_melt_nvars_dt[,Proportion:=N/N_variants]

som_ap_cluster_vartype_counts_gg<- ggplot(
  data=iqlr_multilayer_som_var_merge_annot_summ_melt_nvars_dt, 
  aes(x=AP_cluster, y=Proportion, fill=VAR_TYPE))

som_ap_cluster_vartype_counts_gg+
  geom_bar(stat="identity", position="dodge")+
  scale_fill_manual(values=jco_cols)+
  scale_x_discrete(drop=FALSE)+
  labs(x="AP cluster", y="Proportion variants", fill="")+
  theme_cowplot()
  
rm(som_ap_cluster_vartype_counts_gg)

# Determine which missense variants are in the cluster with high proportion of nonsense variants
# Amino acid positions I59 and A67 stand out

iqlr_multilayer_som_ap_clust_del_ids<- iqlr_multilayer_som_var_merge_annot_dt[
  AP_cluster=="9" & VAR_TYPE=="Missense", ID_SHORT]

# Include the silent variants in the nearby cluster 6
iqlr_multilayer_som_ap_clust_del_silent_ids<- iqlr_multilayer_som_var_merge_annot_dt[
  AP_cluster%in%c("7","8") & VAR_TYPE=="Silent", ID_SHORT]

silent_ids<- iqlr_multilayer_som_var_merge_annot_dt[VAR_TYPE=="Silent", ID_SHORT]
missense_ids<- iqlr_multilayer_som_var_merge_annot_dt[VAR_TYPE=="Missense", ID_SHORT]

iqlr_multilayer_som_ap_clust_ids<- c(iqlr_multilayer_som_ap_clust_del_ids, iqlr_multilayer_som_ap_clust_del_silent_ids)

# Plot data for these missense variants
comt_rep_summ_cast_nona_iqlr_mat_filt_center_dt<- as.data.table(comt_rep_summ_cast_nona_iqlr_mat_filt_center)
comt_rep_summ_cast_nona_iqlr_mat_filt_center_dt[,ID_SHORT:=rownames(comt_rep_summ_cast_nona_iqlr_mat_filt_center)]

comt_rep_summ_cast_nona_iqlr_mat_filt_center_melt_dt<- melt(
  data=comt_rep_summ_cast_nona_iqlr_mat_filt_center_dt, "ID_SHORT", variable.name="Layer", value.name="IQLR")

comt_rep_summ_cast_nona_iqlr_mat_filt_center_melt_dt[,POS:=as.integer(sapply(strsplit(ID_SHORT, ":"), "[", 1))]

iqlr_multilayer_som_ap_clust_del_gg<- ggplot(
  data=comt_rep_summ_cast_nona_iqlr_mat_filt_center_melt_dt[
    ID_SHORT%in%iqlr_multilayer_som_ap_clust_ids], 
  aes(x=Layer, y=reorder(ID_SHORT, POS), fill=IQLR))

iqlr_multilayer_som_ap_clust_del_gg+
  geom_tile()+
  scale_fill_gradient2(low=scales::muted("#8E0152"), high=scales::muted("#276419"))+
  labs(x="", y="", fill="IQLR")+
  theme_cowplot()

comt_rep_summ_cast_nona_iqlr_mat_filt_center_melt_dt[,REF:=sapply(strsplit(ID_SHORT, ":"), "[", 2)]
comt_rep_summ_cast_nona_iqlr_mat_filt_center_melt_dt[,ALT:=sapply(strsplit(ID_SHORT, ":"), "[", 3)]
comt_rep_summ_cast_nona_iqlr_mat_filt_center_melt_dt[,AA_CHANGE:=sapply(strsplit(ID_SHORT, ":"), "[", 4)]

comt_rep_summ_cast_nona_iqlr_mat_filt_center_melt_dt[
  ,SNP:=varid_to_rnasnp(POS, REF, ALT), by=seq_len(nrow(
    comt_rep_summ_cast_nona_iqlr_mat_filt_center_melt_dt))]

comt_rep_summ_cast_nona_iqlr_mat_filt_center_melt_rnasnp_dt<- merge(
  comt_rep_summ_cast_nona_iqlr_mat_filt_center_melt_dt, rnasnp_res_dt, by="SNP")

# Determine if any of the variants are predicted to have an impact on secondary structure
comt_rep_summ_cast_nona_iqlr_mat_filt_center_melt_rnasnp_uniq_dt<-
  unique(comt_rep_summ_cast_nona_iqlr_mat_filt_center_melt_rnasnp_dt[
  ID_SHORT%in%iqlr_multilayer_som_ap_clust_ids, 
  .(SNP, POS, REF, ALT, AA_CHANGE, ID_SHORT, d_max, d_max_pval, r_min, r_min_pval)])

# Generate a key for plotting
comt_rep_summ_cast_nona_iqlr_mat_filt_center_melt_rnasnp_uniq_dt[,Key:=paste(REF, ALT, AA_CHANGE, sep=":")]

setkey(comt_rep_summ_cast_nona_iqlr_mat_filt_center_melt_rnasnp_uniq_dt, POS)

iqlr_multilayer_som_ap_clust_del_rnasnp_gg<- ggplot(
  data=comt_rep_summ_cast_nona_iqlr_mat_filt_center_melt_rnasnp_uniq_dt, 
  aes(x=reorder(Key, POS), y=d_max))

iqlr_multilayer_som_ap_clust_del_rnasnp_gg+
  geom_segment(aes(x=reorder(Key, POS), xend=reorder(Key, POS), y=0, yend=d_max), col="grey")+
  geom_point(aes(col=d_max_pval), size=3)+
  scale_color_viridis(option="A", begin=0.95, end=0.05)+
  expand_limits(y=c(0,0.4))+
  scale_y_continuous(breaks=seq(0,0.6,0.1))+
  theme_cowplot()+
  labs(x="Amino acid position", y="RNAsnp max dist.", col="p-value")+
  coord_flip()

# What proportion of silent and missense variants alter structure?
comt_rep_summ_cast_nona_iqlr_mat_filt_center_melt_rnasnp_dt[ID_SHORT%in%silent_ids & d_max_pval < 0.1, length(unique(ID_SHORT))] / length(silent_ids)

comt_rep_summ_cast_nona_iqlr_mat_filt_center_melt_rnasnp_dt[ID_SHORT%in%missense_ids & d_max_pval < 0.1, length(unique(ID_SHORT))] / length(missense_ids)

# Z compositions library for NAs?

# For LRA uses weighted parts for each variant (spectral mapping, Greenacre 2019 review)
#comt_rep_summ_cast_nona_mat_freqs<- 10^comt_rep_summ_cast_nona_mat

# First reclose the dataset since WT frequency is not included in matrix
#comt_rep_summ_cast_nona_mat_freqs_reclose<- apply(comt_rep_summ_cast_nona_mat_freqs, 2, reclose_coda)

#comt_rep_summ_cast_nona_mat_freqs_margins<- rowSums(comt_rep_summ_cast_nona_mat_freqs_reclose)
#comt_rep_summ_cast_nona_mat_freqs_weights<- comt_rep_summ_cast_nona_mat_freqs_margins/sum(comt_rep_summ_cast_nona_mat_freqs_margins)

# LRA expects samples in the rows and parts in the columns
#comt_rep_summ_cast_nona_mat_freqs_reclose_t<- t(comt_rep_summ_cast_nona_mat_freqs_reclose)

#comt_rep_summ_cast_nona_mat_freqs_reclose_lra<- LRA(comt_rep_summ_cast_nona_mat_freqs_reclose, row.wt=comt_rep_summ_cast_nona_mat_freqs_weights)
```


## GnomAD variants and cDNA altering variants

According to gnomAD o/e value COMT is not highly constrained.

```{r gnomAD variants}
  
gnomad_dt<- fread(paste(gnomad_dir, "gnomAD_v3.1.2_ENST00000361682_2022_07_26_17_02_01.csv", sep="/"), sep=",", header=T)

gnomad_dt[,gnomAD_ID:=paste(Position, Reference, Alternate, sep=":")]

mutagenesis_coord_dt[,Position:=Genomic]
# Add in REF and ALT?

gnomad_merge_dt<- merge(x=gnomad_dt[
  nchar(Reference) == nchar(Alternate),],
  y=mutagenesis_coord_dt, by="Position")

aa_map<- list("A"="Ala", "C"="Cys", "D"="Asp", "E"="Glu", "F"="Phe", 
              "G"="Gly", "H"="His", "I"="Ile", "K"="Lys", "L"="Leu", 
              "M"="Met", "N"="Asn", "P"="Pro", "Q"="Gln", "R"="Arg", 
              "S"="Ser", "T"="Thr", "V"="Val", "W"="Trp", "Y"="Tyr",
              "*"="?")

#' Coverts long to short AA formats e.g. p.Ala41Thr to p.A41T
#'
#' @param aa_change character HGVS amino acid change, e.g. p.Ala41Thr
convert_aa_notation<- function(aa_change){
  
  aa_change_strip<- strsplit(as.character(aa_change), ".", fixed=T)
  
  if(length(aa_change_strip[[1]])==1){
    aa_change_strip<- aa_change
  } else{
    aa_change_strip<- sapply(aa_change_strip, "[", 2)
  }
  
  pos<- gsub(pattern="[[:alpha:]]|[[:punct:]]", replacement="", aa_change_strip)
  
  aa_change_no_int<- gsub(pattern="[[:digit:]]", replacement=":", aa_change_strip)
  
  aa_change_no_int_split<- unlist(strsplit(aa_change_no_int, ":+"))
  
  final_res<- c()
  for(e in 1:length(aa_change_no_int_split)){
    
    if(e==2){
      final_res<- append(final_res, pos)
      
      if(grepl("[[:punct:]]", aa_change_no_int_split[e])){
        match_res<- names(aa_map)[which(aa_change_no_int_split[1]==aa_map)]
        final_res<- append(final_res, match_res)
        next
      }
    }
    match_res<- names(aa_map)[which(aa_change_no_int_split[e]==aa_map)]
    final_res<- append(final_res, match_res)
  }
  
  final_res<- paste0(c("p.",final_res), collapse="")
  return(final_res)
}

gnomad_merge_dt[,Key:=sapply(`Protein Consequence`, convert_aa_notation)]
gnomad_merge_dt[,Key:=paste(Key, Reference, Alternate, sep=":")]

# Merge with ALDEx2 effect sizes for each layer
comt_all_expr_aldex_merge_dt[,Key:=paste(AA_CHANGE, REF, ALT, sep=":")]

comt_all_expr_aldex_merge_gnomad_dt<- merge(
  comt_all_expr_aldex_merge_dt, gnomad_merge_dt, by="Key")

# RNA and protein
comt_all_expr_aldex_merge_gnomad_gg<- ggplot(
  data=comt_all_expr_aldex_merge_gnomad_dt,
  aes(x=effect.x, y=effect, col=log10(`Allele Frequency`)))

comt_all_expr_aldex_merge_gnomad_gg+
  geom_hline(yintercept=0, col="grey", linetype=2)+
  geom_vline(xintercept=0, col="grey", linetype=2)+
  geom_point()+
  scale_color_viridis(option="C", begin=0.95, end=0.05)+
  theme_cowplot()+
  labs(x="RNA effect size", y="Protein effect size", col="log10 gnomAD frequency")

rm(comt_all_expr_aldex_merge_gnomad_gg)

# RNA versus gnomAD frequency
comt_all_expr_aldex_merge_gnomad_rna_gg<- ggplot(
  data=comt_all_expr_aldex_merge_gnomad_dt,
  aes(x=effect.x, y=log10(`Allele Frequency`)))

comt_all_expr_aldex_merge_gnomad_rna_gg+
  geom_vline(xintercept=0, col="grey", linetype=2)+
  geom_point()+
  theme_cowplot()+
  labs(x="RNA effect size", y="log10 gnomAD frequency")

rm(comt_all_expr_aldex_merge_gnomad_rna_gg)

# Protein versus gnomAD frequency
comt_all_expr_aldex_merge_gnomad_protein_gg<- ggplot(
  data=comt_all_expr_aldex_merge_gnomad_dt,
  aes(x=effect, y=log10(`Allele Frequency`)))

comt_all_expr_aldex_merge_gnomad_protein_gg+
  geom_vline(xintercept=0, col="grey", linetype=2)+
  geom_point()+
  theme_cowplot()+
  labs(x="Protein effect size", y="log10 gnomAD frequency")

rm(comt_all_expr_aldex_merge_gnomad_protein_gg)

unique(comt_all_expr_aldex_merge_gnomad_dt[effect.x>2 & we.eBH.x < 0.15, .(
  ID_SHORT, effect.x, we.eBH.x, d_max, d_max_pval, `Allele Frequency`, rsIDs)])

# Deleterious RNA
unique(comt_all_expr_aldex_merge_gnomad_dt[effect.x<(-2) & we.eBH.x < 0.15, .(
  ID_SHORT, effect.x, we.eBH.x, d_max, d_max_pval, `Allele Frequency`, rsIDs)])

# Protein up; ignore
unique(comt_all_expr_aldex_merge_gnomad_dt[effect<(-2) & we.eBH < 0.15, .(
  ID_SHORT, effect, we.eBH, d_max, d_max_pval, `Allele Frequency`, rsIDs)])

# Deleterious protein
unique(comt_all_expr_aldex_merge_gnomad_dt[we.eBH < 0.15 & effect > 0, .(
  ID_SHORT, effect, we.eBH, d_max, d_max_pval, `Allele Frequency`, rsIDs)])

unique(comt_all_expr_aldex_merge_gnomad_dt[log10(`Allele Frequency`) > -1, .(
  ID_SHORT, effect, we.eBH, d_max, d_max_pval, `Allele Frequency`, rsIDs)])

# Poly up
unique(comt_all_expr_aldex_merge_gnomad_dt[effect.y > 0 & we.eBH.y < 0.15, .(
  ID_SHORT, effect.y, we.eBH.y, d_max, d_max_pval, `Allele Frequency`, rsIDs)])

unique(comt_all_expr_aldex_merge_gnomad_dt[effect.y < 0 & we.eBH.y < 0.15, .(
  ID_SHORT, effect.y, we.eBH.y, d_max, d_max_pval, `Allele Frequency`, rsIDs)])

comt_all_expr_aldex_merge_gnomad_res_dt<- 
  unique(comt_all_expr_aldex_merge_gnomad_dt[
  (we.eBH.x < 0.15 & (effect.x>2 | effect.x<(-2))) |
    (we.eBH < 0.15 & effect > 0) | 
    (effect.y > 0 & we.eBH.y < 0.15) |
    (effect.y < 0 & we.eBH.y < 0.15), 
  .(rsIDs, POS, REF, ALT, AA_CHANGE, ID_SHORT, Fraction, 
    effect.x, we.eBH.x, effect.y, we.eBH.y, effect, we.eBH, `Allele Frequency`, 
    Transcript, `Transcript Consequence`, Chromosome, Position, 
    `ClinVar Clinical Significance`)])

write.table(comt_all_expr_aldex_merge_gnomad_res_dt, 
            paste(comt_res_base_dir, "gnomAD_results.txt", sep="/"), sep="\t", quote=FALSE, row.names=FALSE)

# Poly down

# Are variants that are significant less likely to be in gnomAD?
comt_all_expr_aldex_merge_dt[,Significant:="False"]

comt_all_expr_aldex_merge_dt[
  (we.eBH < multilayer_poly_protein_fdr_thresh & effect > 0) | 
  we.eBH.x < multilayer_rna_fdr_thresh | 
  we.eBH.y < multilayer_poly_protein_fdr_thresh, Significant:="True"]

comt_all_expr_aldex_merge_gnomad_all_dt<- merge(
  comt_all_expr_aldex_merge_dt[Fraction=="F4"], gnomad_merge_dt, by="Key", all=T)

comt_all_expr_aldex_merge_gnomad_all_dt[is.na(gnomAD_ID), In_gnomad:="False"]
comt_all_expr_aldex_merge_gnomad_all_dt[is.na(In_gnomad), In_gnomad:="True"]

gnomad_table<- comt_all_expr_aldex_merge_gnomad_all_dt[,table(Significant, In_gnomad)]
fisher.test(gnomad_table)

# Majority of significant variants are not in gnomAD, and an association test was not significant
# Could be little power

```


